// @canva/app-ui-kit/dist/cjs/ui/base/avatar/private/acronym_avatar.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "AcronymAvatar", {
    enumerable: true,
    get: function() {
        return AcronymAvatar;
    }
});
const _jsxruntime = require("react/jsx-runtime");
const _color = require('../../../../base/color/color');
const _css_colors = require('../../../../base/color/css_colors');
const _preconditions = require('../../../../base/preconditions');
const _classnames = _interop_require_default(require("classnames"));
const _react = _interop_require_wildcard(require("react"));
const _colors = require('../../colors/colors');
const _initials = require('../../initials/initials');
const _acronym_avatarcss = _interop_require_wildcard(require("./acronym_avatar.css"));
const _utils = require("./utils");
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function _getRequireWildcardCache(nodeInterop) {
    if (typeof WeakMap !== "function") return null;
    var cacheBabelInterop = new WeakMap();
    var cacheNodeInterop = new WeakMap();
    return (_getRequireWildcardCache = function(nodeInterop) {
        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
    })(nodeInterop);
}
function _interop_require_wildcard(obj, nodeInterop) {
    if (!nodeInterop && obj && obj.__esModule) {
        return obj;
    }
    if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
        return {
            default: obj
        };
    }
    var cache = _getRequireWildcardCache(nodeInterop);
    if (cache && cache.has(obj)) {
        return cache.get(obj);
    }
    var newObj = {
        __proto__: null
    };
    var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
    for(var key in obj){
        if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
            if (desc && (desc.get || desc.set)) {
                Object.defineProperty(newObj, key, desc);
            } else {
                newObj[key] = obj[key];
            }
        }
    }
    newObj.default = obj;
    if (cache) {
        cache.set(obj, newObj);
    }
    return newObj;
}
const AcronymAvatar = _react.memo((props)=>{
    const { name, borderColor, backgroundColor, backgroundSeed, role, ariaLabel, className, style, shape = 'circle', ...tooltipTriggerProps } = props;
    let normalizedColor;
    if (backgroundColor) {
        const cssColor = _css_colors.CssColors.fromString(backgroundColor);
        if (cssColor instanceof _color.RgbaColor) normalizedColor = _css_colors.CssColors.toRgbaString(cssColor);
        else normalizedColor = _css_colors.CssColors.toHexString(cssColor);
    } else if (backgroundSeed) normalizedColor = (0, _utils.generateAvatarBackgroundColor)(backgroundSeed);
    else normalizedColor = _utils.DEFAULT_BACKGROUND_COLOR;
    const clipId = _react.useId();
    let defs;
    switch(shape){
        case 'circle':
            defs = (0, _jsxruntime.jsx)("defs", {
                children: (0, _jsxruntime.jsx)("clipPath", {
                    id: clipId,
                    children: (0, _jsxruntime.jsx)("circle", {
                        id: `${clipId}-path`,
                        cx: "50%",
                        cy: "50%",
                        r: "50%"
                    })
                })
            });
            break;
        case 'square':
            defs = (0, _jsxruntime.jsx)("defs", {
                children: (0, _jsxruntime.jsx)("clipPath", {
                    id: clipId,
                    children: (0, _jsxruntime.jsx)("rect", {
                        id: `${clipId}-path`,
                        width: "100%",
                        height: "100%"
                    })
                })
            });
            break;
        default:
            throw new _preconditions.UnreachableError(shape);
    }
    let fill;
    switch(shape){
        case 'circle':
            fill = (0, _jsxruntime.jsx)("circle", {
                cx: "50%",
                cy: "50%",
                fill: normalizedColor,
                shapeRendering: "geometricPrecision",
                clipPath: `url(#${clipId})`,
                stroke: borderColor,
                className: (0, _classnames.default)(_acronym_avatarcss.default.circle, {
                    [_acronym_avatarcss.default.withStroke]: borderColor
                })
            });
            break;
        case 'square':
            fill = (0, _jsxruntime.jsx)("rect", {
                width: "100%",
                height: "100%",
                fill: normalizedColor,
                shapeRendering: "geometricPrecision",
                clipPath: `url(#${clipId})`,
                stroke: borderColor,
                className: (0, _classnames.default)(_acronym_avatarcss.default.square, {
                    [_acronym_avatarcss.default.withStroke]: borderColor
                })
            });
            break;
        default:
            throw new _preconditions.UnreachableError(shape);
    }
    const avatarTextColor = (0, _colors.getContrastingForegroundColor)(normalizedColor);
    return (0, _jsxruntime.jsx)("span", {
        role: role,
        "aria-label": ariaLabel,
        className: (0, _classnames.default)(_acronym_avatarcss.default.container, {
            [_acronym_avatarcss.default.square]: shape === 'square'
        }),
        style: {
            [_acronym_avatarcss.customProperties.avatarTextColor]: avatarTextColor
        },
        ...tooltipTriggerProps,
        children: (0, _jsxruntime.jsxs)("svg", {
            className: className,
            style: style,
            children: [
                defs,
                fill,
                (0, _jsxruntime.jsx)("text", {
                    x: "50%",
                    y: "50%",
                    "aria-hidden": "true",
                    textAnchor: "middle",
                    fill: avatarTextColor,
                    fontWeight: "500",
                    fontSize: "50%",
                    dominantBaseline: "central",
                    letterSpacing: "0.01rem",
                    dy: "-0.04em",
                    children: (0, _initials.getInitials)(name).toUpperCase()
                })
            ]
        })
    });
});
