// @canva/app-ui-kit/dist/esm/ui/base/color_picker/private/color_picker.js
import { jsx as _jsx } from "react/jsx-runtime";
import * as React from 'react';
import { HsvColor } from '../hsv_color/hsv_color';
import { InternalColorPicker } from './internal_color_picker';
export function ColorPicker(props) {
    const { alpha, sizingMode, isSheet, autoFocusHexInput, showInputSwatch, onChangeStart, onChangeComplete, onStartEyedropper, onEyedropperBackButton, onDeleteColor } = props;
    const prevColorRef = React.useRef(undefined);
    const [, forceUpdate] = React.useState(1);
    const setColor = React.useCallback((newColor)=>{
        prevColorRef.current = newColor;
        forceUpdate((n)=>-n);
    }, []);
    const color = prevColorRef.current ? prevColorRef.current.matchHexString(props.color) : HsvColor.fromHexString(props.color, alpha);
    const onInputChange = (hex)=>{
        setColor(HsvColor.fromHexString(hex));
        props.onChange(hex, 'hex');
    };
    const onSatValChange = (color)=>{
        setColor(color);
        props.onChange(color.toRgb().toHexString(), 'satVal', alpha);
    };
    const onHueChange = (hue)=>{
        const newColor = new HsvColor(hue, color.s, color.v, alpha);
        setColor(newColor);
        if (newColor.v > 0 && newColor.s > 0)
        props.onChange(newColor.toRgb().toHexString(), 'hue', alpha);
    };
    const onAlphaChange = (alpha)=>{
        const newColor = new HsvColor(color.h, color.s, color.v, alpha);
        setColor(newColor);
        props.onChange(newColor.toRgb().toHexString(), 'alpha', alpha);
    };
    return _jsx(InternalColorPicker, {
        color: color,
        sizingMode: sizingMode,
        isSheet: isSheet,
        showInputSwatch: showInputSwatch,
        autoFocusHexInput: autoFocusHexInput,
        onChangeStart: onChangeStart,
        onChangeComplete: onChangeComplete,
        onInputChange: onInputChange,
        onHueChange: onHueChange,
        onSatValChange: onSatValChange,
        onAlphaChange: alpha != null ? onAlphaChange : undefined,
        onStartEyedropper: onStartEyedropper,
        onEyedropperBackButton: onEyedropperBackButton,
        onDeleteColor: onDeleteColor
    });
}
