// @canva/app-ui-kit/dist/esm/ui/base/color_picker/private/color_picker_flyout.js
import { jsx as _jsx } from "react/jsx-runtime";
import * as React from 'react';
import { Box } from '../../box/box';
import { Flyout } from '../../surface/flyout/flyout';
import { AddColorButton } from '../../swatch/swatch';
import { ColorPicker } from './color_picker';
export const ColorPickerFlyout = ({ trigger, color, alpha, ariaLabel, ariaLabelledBy, onChange, onChangeStart, onChangeComplete, onStartEyedropper, onOpen, onClose, onDeleteColor })=>{
    const [open, setOpen] = React.useState(false);
    const toggleFlyout = React.useCallback(()=>{
        if (open)
            onClose === null || onClose === void 0 ? void 0 : onClose();
        else
            onOpen === null || onOpen === void 0 ? void 0 : onOpen();
        setOpen((prev)=>!prev);
    }, [
        open,
        onOpen,
        onClose,
        setOpen
    ]);
    const triggerId = React.useId();
    return _jsx(StatelessColorPickerFlyout, {
        trigger: trigger,
        open: open,
        id: triggerId,
        onClick: toggleFlyout,
        ariaLabel: ariaLabel,
        ariaLabelledBy: ariaLabelledBy,
        color: color,
        alpha: alpha,
        onChange: onChange,
        onChangeStart: onChangeStart,
        onChangeComplete: onChangeComplete,
        onStartEyedropper: onStartEyedropper,
        onDeleteColor: onDeleteColor
    });
};
export function StatelessColorPickerFlyout(props) {
    const { trigger = (props)=> _jsx(AddColorButton, {
            size: "small",
            ...props
        }), open, id, ariaLabel, ariaLabelledBy, color, alpha, onClick, onChange, onChangeStart, onChangeComplete, onStartEyedropper, onDeleteColor } = props;
    return _jsx(Flyout, {
        role: "dialog",
        ariaLabelledBy: id,
        placement: "bottom-start",
        open: open,
        onRequestClose: onClick,
        captureFocus: true,
        width: "40u",
        trigger: (triggerProps)=>trigger({
                id,
                ariaLabel,
                ariaLabelledBy,
                onClick,
                ...triggerProps
            }),
        children: ({ mode })=>_jsx(Box, {
                padding: mode === 'popover' ? '2u' : '0',
                children: _jsx(ColorPicker, {
                    sizingMode: mode === 'popover' ? 'fixed' : 'fill',
                    color: color,
                    alpha: alpha,
                    onChange: onChange,
                    onChangeStart: onChangeStart,
                    onChangeComplete: onChangeComplete,
                    onStartEyedropper: onStartEyedropper,
                    onDeleteColor: onDeleteColor
                })
            })
    });
}
