// @canva/app-ui-kit/dist/esm/ui/base/color_picker/private/internal_color_picker.js
import { jsx as _jsx, jsxs as _jsxs, Fragment as _Fragment } from "react/jsx-runtime";
import classNames from 'classnames';
import * as React from 'react';
import { Box } from '../../box/box';
import { Button } from '../../button/button';
import { AlphaSlider } from '../alpha_slider/alpha_slider';
import { ColorCodeInput } from '../color_code_input/color_code_input';
import { EyeDropperButton } from '../eye_dropper_button/eye_dropper_button';
import { HueSlider } from '../hue_slider/hue_slider';
import { SatValPicker } from '../sat_val_picker/sat_val_picker';
import { DraggableView } from '../../draggable/draggable_view';
import { TrashIcon } from '../../icons/trash/icon';
import { ColorPickerMessages } from './color_picker.messages';
import styles from './internal_color_picker.css';
export const InternalColorPicker = React.memo((props)=>{
    const { color, sizingMode = 'fill', isSheet = sizingMode === 'fill', autoFocusHexInput = 'on-desktop', onStartEyedropper, onEyedropperBackButton, onDeleteColor, onChangeStart, onChangeComplete, onInputChange, onHueChange, onSatValChange, onAlphaChange, showInputSwatch } = props;
    const eyeDropperButton = onStartEyedropper && _jsx(EyeDropperButton, {
        onStart: onStartEyedropper,
        onBackButton: onEyedropperBackButton,
        isSheet: isSheet
    });
    const deleteColorButton = onDeleteColor && _jsx(DeleteColorButton, {
        onDeleteColor: onDeleteColor
    });
    if (sizingMode === 'fixed') return _jsxs(_Fragment, {
        children: [
            _jsx(ColorPickerBody, {
                color: color,
                onChangeStart: onChangeStart,
                onChangeComplete: onChangeComplete,
                onHueChange: onHueChange,
                onSatValChange: onSatValChange,
                onAlphaChange: onAlphaChange
            }),
            _jsx(ColorPickerInputGroup, {
                showInputSwatch: showInputSwatch === 'auto' ? 'never' : showInputSwatch,
                autoFocusHexInput: autoFocusHexInput,
                color: color,
                hexInputSizingMode: "fill",
                onInputChange: onInputChange,
                onAlphaChange: onAlphaChange,
                eyeDropperButton: eyeDropperButton,
                deleteColorButton: deleteColorButton
            })
        ]
    });
    return _jsxs(_Fragment, {
        children: [
            _jsx("div", {
                className: styles.colorPickerHeader,
                children: _jsx(ColorPickerInputGroup, {
                    color: color,
                    hexInputSizingMode: "fixed",
                    showInputSwatch: showInputSwatch === 'auto' ? 'always' : showInputSwatch,
                    onInputChange: onInputChange,
                    onAlphaChange: onAlphaChange,
                    eyeDropperButton: eyeDropperButton,
                    deleteColorButton: deleteColorButton
                })
            }),
            _jsx(Box, {
                padding: "2u",
                children: _jsx(ColorPickerBody, {
                    color: color,
                    showDivider: true,
                    onChangeStart: onChangeStart,
                    onChangeComplete: onChangeComplete,
                    onHueChange: onHueChange,
                    onSatValChange: onSatValChange,
                    onAlphaChange: onAlphaChange
                })
            })
        ]
    });
});
const ColorPickerInputGroup = React.memo((props)=>_jsxs("div", {
        className: styles.colorInputSection,
        children: [
            props.deleteColorButton,
            _jsx("div", {
                className: classNames(styles.colorCodeInput, {
                    [styles.fixedWithoutAlpha]: props.hexInputSizingMode === 'fixed' && props.onAlphaChange == null,
                    [styles.fixedWithAlpha]: props.hexInputSizingMode === 'fixed' && props.onAlphaChange != null,
                    [styles.fill]: props.hexInputSizingMode === 'fill'
                }),
                children: _jsx(ColorCodeInput, {
                    autoFocus: props.autoFocusHexInput,
                    color: props.color.toRgb().toHexString(),
                    alpha: props.color.alpha,
                    onAlphaChange: props.onAlphaChange,
                    onChange: props.onInputChange,
                    withSwatch: props.showInputSwatch === 'always',
                    ariaLabel: ColorPickerMessages.colorCodeInputLabel()
                })
            }),
            props.eyeDropperButton
        ]
    }));
const ColorPickerBody = React.memo((props)=>_jsxs(_Fragment, {
        children: [
            _jsx(SatValPicker, {
                color: props.color,
                onChange: props.onSatValChange,
                onChangeStart: props.onChangeStart,
                onChangeComplete: props.onChangeComplete,
                Draggable: DraggableView
            }),
            props.showDivider && _jsx("div", {
                className: styles.divider
            }),
            _jsx(HueSlider, {
                hue: props.color.h,
                onChangeStart: props.onChangeStart,
                onChangeComplete: props.onChangeComplete,
                onChange: props.onHueChange
            }),
            props.onAlphaChange && _jsx(AlphaSlider, {
                alpha: props.color.alpha,
                color: props.color.toRgb().toHexString(),
                onChangeStart: props.onChangeStart,
                onChangeComplete: props.onChangeComplete,
                onChange: props.onAlphaChange
            })
        ]
    }));
const DeleteColorButton = (props)=>_jsx(Button, {
        variant: "tertiary",
        icon: TrashIcon,
        className: styles.deleteColorButton,
        tooltipLabel: ColorPickerMessages.deleteColorTooltip(),
        onClick: props.onDeleteColor
    });
