// @canva/app-ui-kit/dist/cjs/ui/base/color_picker/hue_slider/private/hue_slider.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "HueSlider", {
    enumerable: true,
    get: function() {
        return HueSlider;
    }
});
const _jsxruntime = require("react/jsx-runtime");
const _classnames = _interop_require_default(require("classnames"));
const _react = _interop_require_wildcard(require("react"));
const _slider = require('../../../form/slider/slider');
const _hue_slidercss = _interop_require_default(require("./hue_slider.css"));
const _hue_slidermessages = require("./hue_slider.messages");
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function _getRequireWildcardCache(nodeInterop) {
    if (typeof WeakMap !== "function") return null;
    var cacheBabelInterop = new WeakMap();
    var cacheNodeInterop = new WeakMap();
    return (_getRequireWildcardCache = function(nodeInterop) {
        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
    })(nodeInterop);
}
function _interop_require_wildcard(obj, nodeInterop) {
    if (!nodeInterop && obj && obj.__esModule) {
        return obj;
    }
    if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
        return {
            default: obj
        };
    }
    var cache = _getRequireWildcardCache(nodeInterop);
    if (cache && cache.has(obj)) {
        return cache.get(obj);
    }
    var newObj = {
        __proto__: null
    };
    var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
    for(var key in obj){
        if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
            if (desc && (desc.get || desc.set)) {
                Object.defineProperty(newObj, key, desc);
            } else {
                newObj[key] = obj[key];
            }
        }
    }
    newObj.default = obj;
    if (cache) {
        cache.set(obj, newObj);
    }
    return newObj;
}
const track = ()=>(0, _jsxruntime.jsx)("div", {
        className: _hue_slidercss.default.track
    });
const handle = ({ focused, trackHovering })=>(0, _jsxruntime.jsx)("div", {
        className: (0, _classnames.default)(_hue_slidercss.default.handle, {
            [_hue_slidercss.default.handleHoverOrFocus]: focused || trackHovering
        })
    });
function HueSlider({ hue, onChange: onChangeProp, onChangeStart, onChangeComplete, ariaLabel = _hue_slidermessages.HueSliderMessages.colorHueLabel() }) {
    const handleAtEndRef = _react.useRef(false);
    const onChange = _react.useCallback((value)=>{
        handleAtEndRef.current = value === 1;
        onChangeProp === null || onChangeProp === void 0 ? void 0 : onChangeProp(value);
    }, [
        onChangeProp
    ]);
    const handlePosition = handleAtEndRef.current && hue === 0 ? 1 : hue;
    return (0, _jsxruntime.jsx)("div", {
        className: _hue_slidercss.default.container,
        children: (0, _jsxruntime.jsx)("div", {
            className: _hue_slidercss.default.widener,
            children: (0, _jsxruntime.jsx)(_slider.BaseSlider, {
                value: handlePosition,
                min: 0,
                max: 1,
                step: 0.005,
                onChange: onChange,
                onDragStart: onChangeStart,
                onDragEnd: onChangeComplete,
                ariaLabel: ariaLabel,
                fillColor: "transparent",
                handle: handle,
                track: track
            })
        })
    });
}
