// @canva/app-ui-kit/dist/cjs/ui/apps/developing/ui_kit/components/card/video/video.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "VideoCard", {
    enumerable: true,
    get: function() {
        return VideoCard;
    }
});
const _jsxruntime = require("react/jsx-runtime");
const _preconditions = require('../../../../../../../base/preconditions');
const _classnames = _interop_require_default(require("classnames"));
require("react");
const _badge = require('../../../../../../base/badge/badge');
const _box = require('../../../../../../base/box/box');
const _card = require('../../../../../../base/card/card');
const _duration = require('../../../../../../base/date_time/duration/duration');
const _icon = require('../../../../../../base/icons/animated_spinner/icon');
const _draggable = require("../draggable/draggable");
const _videocss = _interop_require_default(require("./video.css"));
const _videomessages = require("./video.messages");
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
const VideoCard = (props)=>(0, _jsxruntime.jsx)(_draggable.Draggable, {
        ...props,
        children: (conf)=>(0, _jsxruntime.jsx)(InternalVideoCard, {
                ...props,
                ...conf
            })
    });
const InternalVideoCard = (props)=>{
    const { ariaLabel, thumbnailUrl, mimeType, onDragStart, onDragEnd, onClick, borderRadius = 'none', thumbnailHeight, thumbnailAspectRatio, bottomEnd, bottomEndVisibility, selectable, selected, disabled, loading } = props;
    const commonProps = {
        buttonAriaLabel: ariaLabel,
        onClick,
        draggable: onDragStart != null,
        hoverEffect: 'none',
        thumbnailHeight,
        selectableMode: selectable ? 'primary' : 'none',
        selected,
        disabled,
        onDragStart,
        onDragEnd
    };
    if (isVideoProps(props)) {
        const badgeLabel = getBadgeLabel(mimeType, props.durationInSeconds);
        return (0, _jsxruntime.jsx)(_card.Card, {
            decorators: (0, _jsxruntime.jsxs)(_jsxruntime.Fragment, {
                children: [
                    (0, _jsxruntime.jsx)(_card.CardDecorator, {
                        location: "bottom-start",
                        children: (0, _jsxruntime.jsx)(_badge.Badge, {
                            tone: "contrast",
                            text: badgeLabel
                        })
                    }),
                    bottomEnd && (0, _jsxruntime.jsx)(_card.CardDecorator, {
                        location: "bottom-end",
                        visibility: bottomEndVisibility,
                        isInteractive: true,
                        children: bottomEnd
                    })
                ]
            }),
            ...commonProps,
            thumbnail: (0, _jsxruntime.jsxs)(_card.CardThumbnailContainer, {
                background: "none",
                padding: "none",
                borderRadius: borderRadius === 'standard' ? 'element' : 'none',
                children: [
                    (0, _jsxruntime.jsx)(_card.CardVideoThumbnail, {
                        imageSrc: thumbnailUrl,
                        videoSrc: props.videoPreviewUrl,
                        whenToPlay: "on-hover",
                        borderRadius: borderRadius === 'standard' ? 'element' : 'none',
                        height: thumbnailHeight,
                        className: (0, _classnames.default)({
                            [_videocss.default.thumbnailAutoWidth]: !thumbnailAspectRatio && thumbnailHeight
                        })
                    }),
                    loading && (0, _jsxruntime.jsx)(_box.Box, {
                        className: _videocss.default.loading,
                        borderRadius: borderRadius === 'standard' ? 'legacyLarge' : 'none',
                        display: "flex",
                        justifyContent: "center",
                        alignItems: "center",
                        width: "full",
                        height: "full",
                        children: (0, _jsxruntime.jsx)(_icon.AnimatedSpinnerIcon, {
                            size: "medium",
                            className: _videocss.default.spinner
                        })
                    })
                ]
            }),
            hoverEffect: "scaleThumbnail",
            layout: "vertical",
            hoverBoundary: "thumbnail"
        });
    }
    return (0, _jsxruntime.jsx)(_card.Card, {
        decorators: (0, _jsxruntime.jsxs)(_jsxruntime.Fragment, {
            children: [
                (0, _jsxruntime.jsx)(_card.CardDecorator, {
                    location: "bottom-start",
                    children: (0, _jsxruntime.jsx)(_badge.Badge, {
                        tone: "contrast",
                        text: _videomessages.VideoCardMessages.gifLabel()
                    })
                }),
                bottomEnd && (0, _jsxruntime.jsx)(_card.CardDecorator, {
                    location: "bottom-end",
                    visibility: bottomEndVisibility,
                    isInteractive: true,
                    children: bottomEnd
                })
            ]
        }),
        ...commonProps,
        thumbnail: (0, _jsxruntime.jsxs)(_card.CardThumbnailContainer, {
            background: "none",
            padding: "none",
            borderRadius: borderRadius === 'standard' ? 'element' : 'none',
            children: [
                (0, _jsxruntime.jsx)(_card.CardImageThumbnail, {
                    src: thumbnailUrl,
                    alt: props.alt,
                    borderRadius: borderRadius === 'standard' ? 'element' : 'none',
                    height: thumbnailHeight,
                    className: (0, _classnames.default)({
                        [_videocss.default.thumbnailAutoWidth]: !thumbnailAspectRatio && thumbnailHeight
                    })
                }),
                loading && (0, _jsxruntime.jsx)(_box.Box, {
                    className: _videocss.default.loading,
                    borderRadius: borderRadius === 'standard' ? 'legacyLarge' : 'none',
                    display: "flex",
                    justifyContent: "center",
                    alignItems: "center",
                    width: "full",
                    height: "full",
                    children: (0, _jsxruntime.jsx)(_icon.AnimatedSpinnerIcon, {
                        size: "medium",
                        className: _videocss.default.spinner
                    })
                })
            ]
        }),
        hoverEffect: "scaleThumbnail",
        layout: "vertical",
        hoverBoundary: "thumbnail"
    });
};
const isVideoProps = (props)=>{
    return props.videoPreviewUrl !== undefined;
};
const getBadgeLabel = (mimeType, duration)=>{
    switch(mimeType){
        case 'video/x-msvideo':
        case 'video/avi':
        case 'video/x-m4v':
        case 'video/x-matroska':
        case 'video/quicktime':
        case 'video/mp4':
        case 'video/mpeg':
        case 'video/webm':
            return duration ? (0, _duration.formatDuration)(duration, 'short') : _videomessages.VideoCardMessages.defaultLabel();
        case 'image/gif':
            return _videomessages.VideoCardMessages.gifLabel();
        default:
            throw new _preconditions.UnreachableError(mimeType);
    }
};
