// @canva/app-ui-kit/dist/esm/ui/base/form/date_time_input/private/date_time_picker_decorator.js
import { jsx as _jsx } from "react/jsx-runtime";
import * as React from 'react';
import { Box } from '../../../box/box';
import { DateTimePicker } from '../../../date_time/date_time_picker/date_time_picker';
import { ButtonDecorator } from '../../input_decorators/input_decorators';
import { CalendarIcon } from '../../../icons/calendar/icon';
import { Popover } from '../../../surface/popover/popover';
import { DateTimeInputMessages } from './date_time_input.messages';
export const DateTimePickerDecorator = React.memo(
    function DateTimePickerDecorator({ onChangeComplete, disabled, blockOutsidePointerEvents, onChange, ...dateTimePickerProps }) {
        const [isFlyoutOpen, setFlyoutOpen] = React.useState(false);
        const toggleFlyout = ()=>{
            if (isFlyoutOpen) onChangeComplete === null || onChangeComplete === void 0 ? void 0 : onChangeComplete(dateTimePickerProps.date);
            setFlyoutOpen(!isFlyoutOpen);
        };
        const closeFlyout = ()=>{
            onChangeComplete === null || onChangeComplete === void 0 ? void 0 : onChangeComplete(dateTimePickerProps.date);
            setFlyoutOpen(false);
        };
        const handleChangeAndClose = (value)=>{
            onChange === null || onChange === void 0 ? void 0 : onChange(value);
            if (value != null) {
                onChangeComplete === null || onChangeComplete === void 0 ? void 0 : onChangeComplete(value);
                setFlyoutOpen(false);
            }
        };
        const label = dateTimePickerProps.mode === 'date' ? DateTimeInputMessages.pickDate() : DateTimeInputMessages.pickDateTime();
        return _jsx(Popover, {
            open: isFlyoutOpen,
            onRequestClose: closeFlyout,
            placement: "bottom-end",
            blockOutsidePointerEvents: blockOutsidePointerEvents,
            reference: (triggerProps)=>_jsx(ButtonDecorator, {
                    label: label,
                    icon: CalendarIcon,
                    onClick: toggleFlyout,
                    disabled: disabled,
                    ...triggerProps
                }),
            children: _jsx(Box, {
                padding: "2u",
                children: _jsx(DateTimePicker, {
                    ...dateTimePickerProps,
                    canSelectPastDates: true,
                    onChange: dateTimePickerProps.mode === 'date' ? handleChangeAndClose : onChange
                })
            })
        });
    }
);
