// @canva/app-ui-kit/dist/esm/ui/base/video/video.js
import { jsx as _jsx } from "react/jsx-runtime";
import composeRefs from '@seznam/compose-react-refs';
import { cleanupVideoSrcOnUnmount } from '../../../base/platform_quirks/video_src_cleanup';
import * as React from 'react';
import { prefersReducedMotion as prefersReducedMotionBase } from '../animation/supports_animation';
import { useAutoplayVideos, useEnableAnimations } from '../provider/provider';
function setMutedAndPlaybackRate(video, muted, playbackRate) {
    video.muted = muted;
    video.playbackRate = playbackRate;
    video.defaultPlaybackRate = playbackRate;
}
export function createVideoComponent(cleanup = cleanupVideoSrcOnUnmount) {
    return (React.forwardRef(function Video(props, ref) {
                const videoRef = React.useRef(null);
                const { muted = false, playbackRate = 1, ariaLabel, ariaHidden, ...rest } = props;
                React.useEffect(()=>{
                    const video = videoRef.current;
                    if (video) setMutedAndPlaybackRate(video, muted, playbackRate);
                }, [
                    muted,
                    playbackRate
                ]);
                const onVideoRef = React.useCallback((ref)=>{
                    const oldRef = videoRef.current;
                    videoRef.current = ref;
                    if (oldRef && ref !== oldRef) {
                        ref && setMutedAndPlaybackRate(ref, oldRef.muted, oldRef.playbackRate);
                        cleanup(oldRef);
                    }
                }, []);
                return (_jsx("video", {
                        ref: composeRefs(onVideoRef, ref),
                        "aria-label": ariaLabel,
                        "aria-hidden": ariaHidden,
                        ...rest
                    }));
            }));
}
export function createVideoA11ySafeComponent(Video, prefersReducedMotion = prefersReducedMotionBase) {
    const autoPlayingVideos = [];
    const addAutoplayingVideo = (video)=>autoPlayingVideos.push(video);
    const maybeRemoveAutoplayingVideo = (video)=>{
        const index = autoPlayingVideos.indexOf(video);
        if (index >= 0) autoPlayingVideos.splice(index, 1);
    };
    const maybeSuppressPlayError = (e)=>{
        if (e.name === 'NotAllowedError' || e.name === 'AbortError') return;
        throw e;
    };
    if (typeof window !== 'undefined') {
        window.addEventListener('blur', ()=>autoPlayingVideos.forEach((video)=>video.pause()));
        window.addEventListener('focus', ()=>autoPlayingVideos.forEach((video)=>video.src && video.play().catch(maybeSuppressPlayError)));
    }
    const VideoA11ySafe = React.forwardRef((props, ref)=>{
        const autoplayVideos = useAutoplayVideos();
        const enableAnimations = useEnableAnimations();
        const enabledAutoPlay = canA11ySafeVideoAutoplay({
            prefersReducedMotion,
            autoplayVideos,
            enableAnimations
        });
        const videoRef = React.useRef(null);
        const { autoPlay = false, ...omitAutoPlayProps } = props;
        React.useEffect(()=>{
            const video = videoRef.current;
            if (video) {
                video.autoplay = autoPlay && enabledAutoPlay;
                if (video.autoplay) addAutoplayingVideo(video);
                else maybeRemoveAutoplayingVideo(video);
                return ()=>maybeRemoveAutoplayingVideo(video);
            }
        }, [
            autoPlay,
            enabledAutoPlay
        ]);
        return _jsx(Video, {
            ref: composeRefs(videoRef, ref),
            ...omitAutoPlayProps
        });
    });
    return VideoA11ySafe;
}
export function canA11ySafeVideoAutoplay({ prefersReducedMotion = prefersReducedMotionBase, autoplayVideos, enableAnimations }) {
    const autoplayVideosSetting = autoplayVideos !== null && autoplayVideos !== void 0 ? autoplayVideos : 'ADAPTIVE';
    return enableAnimations && (autoplayVideosSetting === 'AUTOPLAY' || autoplayVideosSetting === 'ADAPTIVE' && !prefersReducedMotion());
}
export const Video = createVideoComponent();
export const VideoA11ySafe = createVideoA11ySafeComponent(Video);
