// @canva/app-ui-kit/dist/cjs/ui/base/handle/handle.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: Object.getOwnPropertyDescriptor(all, name).get
    });
}
_export(exports, {
    get createBasicHTMLHandle () {
        return _basic_html_handle.createBasicHTMLHandle;
    },
    get createDOMElementHandle () {
        return _dom_element_handle.createDOMElementHandle;
    },
    get createFocusHandle () {
        return _focus_handle.createFocusHandle;
    },
    get createGestureHandle () {
        return _gesture_handle.createGestureHandle;
    },
    get createHTMLElementHandle () {
        return _create_handle.createHTMLElementHandle;
    },
    get createLayoutHandle () {
        return _layout_handle.createLayoutHandle;
    },
    get createNodeHandle () {
        return _node_handle.createNodeHandle;
    },
    get createScrollHandle () {
        return _scroll_handle.createScrollHandle;
    },
    get createTraverseHandle () {
        return _traverse_handle.createTraverseHandle;
    },
    get dangerouslyGetElement () {
        return _dom_element_handle.dangerouslyGetElement;
    },
    get dangerouslyGetHTMLElement () {
        return _dom_element_handle.dangerouslyGetHTMLElement;
    },
    get useHandleRef () {
        return _handle.useHandleRef;
    }
});
const _basic_html_handle = require("./private/composite_handle/basic_html_handle");
const _gesture_handle = require("./private/composite_handle/gesture_handle");
const _create_handle = require("./private/html_element_handle/create_handle");
const _layout_handle = require("./private/html_element_handle/layout_handle");
const _focus_handle = require("./private/html_element_handle/focus_handle");
const _scroll_handle = require("./private/html_element_handle/scroll_handle");
const _node_handle = require("./private/html_element_handle/node_handle");
const _traverse_handle = require("./private/html_element_handle/traverse_handle");
const _dom_element_handle = require("./private/extended_functionality_handle/dom_element_handle");
const _handle = require("./private/handle");
