// @canva/app-ui-kit/dist/cjs/ui/base/a11y/announcer/private/accessibility_announcer_region.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "AccessibilityAnnouncerRegion", {
    enumerable: true,
    get: function() {
        return AccessibilityAnnouncerRegion;
    }
});
const _jsxruntime = require("react/jsx-runtime");
const _react = require("react");
const _accessibility_announcer_regioncss = _interop_require_default(require("./accessibility_announcer_region.css"));
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
const AccessibilityAnnouncerRegion = ({ onMount, onUnmount })=>{
    const assertiveRef = (0, _react.useRef)(null);
    const politeRef = (0, _react.useRef)(null);
    (0, _react.useEffect)(()=>{
        if (assertiveRef.current && politeRef.current) onMount({
            assertiveRegion: assertiveRef,
            politeRegion: politeRef
        });
        return ()=>{
            onUnmount();
        };
    }, [
        onMount,
        onUnmount
    ]);
    return (0, _jsxruntime.jsxs)(_jsxruntime.Fragment, {
        children: [
            (0, _jsxruntime.jsx)("div", {
                ref: assertiveRef,
                className: _accessibility_announcer_regioncss.default.visuallyHidden,
                "aria-atomic": "true",
                "aria-live": "assertive"
            }),
            (0, _jsxruntime.jsx)("div", {
                ref: politeRef,
                className: _accessibility_announcer_regioncss.default.visuallyHidden,
                "aria-atomic": "true",
                "aria-live": "polite"
            })
        ]
    });
};
