// @canva/app-ui-kit/dist/cjs/ui/base/a11y/announcer/private/announcer_fallback.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "useFallbackAnnouncer", {
    enumerable: true,
    get: function() {
        return useFallbackAnnouncer;
    }
});
const _react = require("react");
const _screen_reader_content = require('../../screen_reader_content/screen_reader_content');
const _accessibility_announcer_regioncss = _interop_require_default(require("./accessibility_announcer_region.css"));
const _announcer_utils = require("./announcer_utils");
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
const FALLBACK_ASSERTIVE_ID = 'canva-a11y-fallback-assertive';
const FALLBACK_POLITE_ID = 'canva-a11y-fallback-polite';
function createLiveRegion(ariaLive, id, document) {
    const element = document.createElement('div');
    element.id = id;
    element.setAttribute('aria-live', ariaLive);
    element.setAttribute('aria-atomic', 'true');
    element.className = _accessibility_announcer_regioncss.default.visuallyHidden;
    return element;
}
function ensureFallbackRegions(document) {
    let assertiveRegion = null;
    let politeRegion = null;
    assertiveRegion = document.getElementById(FALLBACK_ASSERTIVE_ID);
    if (!assertiveRegion) {
        assertiveRegion = createLiveRegion('assertive', FALLBACK_ASSERTIVE_ID, document);
        document.body.prepend(assertiveRegion);
    }
    politeRegion = document.getElementById(FALLBACK_POLITE_ID);
    if (!politeRegion) {
        politeRegion = createLiveRegion('polite', FALLBACK_POLITE_ID, document);
        document.body.prepend(politeRegion);
    }
    return {
        assertiveRegion,
        politeRegion
    };
}
async function announceFallback(message, document, regions, priority = 'medium') {
    if (!message) return;
    let { assertiveRegion, politeRegion } = regions;
    if (!assertiveRegion || !politeRegion) {
        const newRegions = ensureFallbackRegions(document);
        assertiveRegion = newRegions.assertiveRegion;
        politeRegion = newRegions.politeRegion;
    }
    if (assertiveRegion && (priority === 'critical' || priority === 'high')) await (0, _announcer_utils.updateLiveRegionContent)(assertiveRegion, message);
    else if (politeRegion) await (0, _announcer_utils.updateLiveRegionContent)(politeRegion, message);
}
function deferFallbackAnnouncement(message, document, regions, priority) {
    let announced = false;
    return ()=>{
        if (announced) return;
        announced = true;
        announceFallback(message, document, regions, priority);
    };
}
const NO_OP_CONTROLLER = {
    announce: ()=>{},
    deferAnnouncement: ()=>()=>{}
};
function useFallbackAnnouncer(shouldUseFallback = true) {
    const window = (0, _screen_reader_content.useWindow)();
    const assertiveRegionRef = (0, _react.useRef)(null);
    const politeRegionRef = (0, _react.useRef)(null);
    (0, _react.useEffect)(()=>{
        if (shouldUseFallback && (window === null || window === void 0 ? void 0 : window.document)) {
            const { assertiveRegion, politeRegion } = ensureFallbackRegions(window.document);
            assertiveRegionRef.current = assertiveRegion;
            politeRegionRef.current = politeRegion;
        }
    }, [
        shouldUseFallback,
        window
    ]);
    const announce = (0, _react.useCallback)((message, priority = 'medium')=>{
        if (shouldUseFallback && (window === null || window === void 0 ? void 0 : window.document)) {
            const regions = {
                assertiveRegion: assertiveRegionRef.current,
                politeRegion: politeRegionRef.current
            };
            announceFallback(message, window.document, regions, priority);
        }
    }, [
        shouldUseFallback,
        window
    ]);
    const deferAnnouncement = (0, _react.useCallback)((message, priority = 'medium')=>{
        if (shouldUseFallback && (window === null || window === void 0 ? void 0 : window.document)) {
            const regions = {
                assertiveRegion: assertiveRegionRef.current,
                politeRegion: politeRegionRef.current
            };
            return deferFallbackAnnouncement(message, window.document, regions, priority);
        } else
            return ()=>{};
    }, [
        shouldUseFallback,
        window
    ]);
    return (0, _react.useMemo)(()=>{
        return shouldUseFallback ? {
            announce,
            deferAnnouncement
        } : NO_OP_CONTROLLER;
    }, [
        shouldUseFallback,
        announce,
        deferAnnouncement
    ]);
}
