// @canva/app-ui-kit/dist/esm/ui/base/form/text_input/private/text_input.js
import { jsx as _jsx, jsxs as _jsxs, Fragment as _Fragment } from "react/jsx-runtime";
import * as React from 'react';
import { BaseInput } from '../../base_input/base_input';
import { ClearDecorator } from '../../input_decorators/input_decorators';
export const TextInput = React.forwardRef(function TextInput(props, ref) {
    const { type = 'text', clearable = 'never', end } = props;
    const endContent = getEndContent(end, clearable !== 'never');
    return _jsx(BaseInput, {
        ...props,
        type: type,
        end: endContent,
        ref: ref
    });
});
TextInput.displayName = 'TextInput';
function getEndContent(end, showClearable) {
    const endContent = typeof end === 'function' ? end() : end;
    if (!showClearable) return endContent;
    return _jsxs(_Fragment, {
        children: [
            _jsx(ClearDecorator, {
                showDivider: !!endContent
            }),
            React.isValidElement(endContent) && endContent.type === React.Fragment ? React.Children.toArray(endContent.props.children) : endContent
        ]
    });
}
