// @canva/app-ui-kit/dist/esm/ui/apps/developing/ui_kit/components/card/video/video.js
import { jsx as _jsx, jsxs as _jsxs, Fragment as _Fragment } from "react/jsx-runtime";
import { UnreachableError } from '../../../../../../../base/preconditions';
import classNames from 'classnames';
import 'react';
import { Badge } from '../../../../../../base/badge/badge';
import { Box } from '../../../../../../base/box/box';
import { Card, CardDecorator, CardImageThumbnail, CardThumbnailContainer, CardVideoThumbnail } from '../../../../../../base/card/card';
import { formatDuration } from '../../../../../../base/date_time/duration/duration';
import { AnimatedSpinnerIcon } from '../../../../../../base/icons/animated_spinner/icon';
import { Draggable } from '../draggable/draggable';
import styles from './video.css';
import { VideoCardMessages as Messages } from './video.messages';
export const VideoCard = (props)=>_jsx(Draggable, {
        ...props,
        children: (conf)=>_jsx(InternalVideoCard, {
                ...props,
                ...conf
            })
    });
const InternalVideoCard = (props)=>{
    const { ariaLabel, thumbnailUrl, mimeType, onDragStart, onDragEnd, onClick, borderRadius = 'none', thumbnailHeight, thumbnailAspectRatio, bottomEnd, bottomEndVisibility, selectable, selected, disabled, loading } = props;
    const commonProps = {
        buttonAriaLabel: ariaLabel,
        onClick,
        draggable: onDragStart != null,
        hoverEffect: 'none',
        thumbnailHeight,
        selectableMode: selectable ? 'primary' : 'none',
        selected,
        disabled,
        onDragStart,
        onDragEnd
    };
    if (isVideoProps(props)) {
        const badgeLabel = getBadgeLabel(mimeType, props.durationInSeconds);
        return _jsx(Card, {
            decorators: _jsxs(_Fragment, {
                children: [
                    _jsx(CardDecorator, {
                        location: "bottom-start",
                        children: _jsx(Badge, {
                            tone: "contrast",
                            text: badgeLabel
                        })
                    }),
                    bottomEnd && _jsx(CardDecorator, {
                        location: "bottom-end",
                        visibility: bottomEndVisibility,
                        isInteractive: true,
                        children: bottomEnd
                    })
                ]
            }),
            ...commonProps,
            thumbnail: _jsxs(CardThumbnailContainer, {
                background: "none",
                padding: "none",
                borderRadius: borderRadius === 'standard' ? 'element' : 'none',
                children: [
                    _jsx(CardVideoThumbnail, {
                        imageSrc: thumbnailUrl,
                        videoSrc: props.videoPreviewUrl,
                        whenToPlay: "on-hover",
                        borderRadius: borderRadius === 'standard' ? 'element' : 'none',
                        height: thumbnailHeight,
                        className: classNames({
                            [styles.thumbnailAutoWidth]: !thumbnailAspectRatio && thumbnailHeight
                        })
                    }),
                    loading && _jsx(Box, {
                        className: styles.loading,
                        borderRadius: borderRadius === 'standard' ? 'legacyLarge' : 'none',
                        display: "flex",
                        justifyContent: "center",
                        alignItems: "center",
                        width: "full",
                        height: "full",
                        children: _jsx(AnimatedSpinnerIcon, {
                            size: "medium",
                            className: styles.spinner
                        })
                    })
                ]
            }),
            hoverEffect: "scaleThumbnail",
            layout: "vertical",
            hoverBoundary: "thumbnail"
        });
    }
    return _jsx(Card, {
        decorators: _jsxs(_Fragment, {
            children: [
                _jsx(CardDecorator, {
                    location: "bottom-start",
                    children: _jsx(Badge, {
                        tone: "contrast",
                        text: Messages.gifLabel()
                    })
                }),
                bottomEnd && _jsx(CardDecorator, {
                    location: "bottom-end",
                    visibility: bottomEndVisibility,
                    isInteractive: true,
                    children: bottomEnd
                })
            ]
        }),
        ...commonProps,
        thumbnail: _jsxs(CardThumbnailContainer, {
            background: "none",
            padding: "none",
            borderRadius: borderRadius === 'standard' ? 'element' : 'none',
            children: [
                _jsx(CardImageThumbnail, {
                    src: thumbnailUrl,
                    alt: props.alt,
                    borderRadius: borderRadius === 'standard' ? 'element' : 'none',
                    height: thumbnailHeight,
                    className: classNames({
                        [styles.thumbnailAutoWidth]: !thumbnailAspectRatio && thumbnailHeight
                    })
                }),
                loading && _jsx(Box, {
                    className: styles.loading,
                    borderRadius: borderRadius === 'standard' ? 'legacyLarge' : 'none',
                    display: "flex",
                    justifyContent: "center",
                    alignItems: "center",
                    width: "full",
                    height: "full",
                    children: _jsx(AnimatedSpinnerIcon, {
                        size: "medium",
                        className: styles.spinner
                    })
                })
            ]
        }),
        hoverEffect: "scaleThumbnail",
        layout: "vertical",
        hoverBoundary: "thumbnail"
    });
};
const isVideoProps = (props)=>{
    return props.videoPreviewUrl !== undefined;
};
const getBadgeLabel = (mimeType, duration)=>{
    switch(mimeType){
        case 'video/x-msvideo':
        case 'video/avi':
        case 'video/x-m4v':
        case 'video/x-matroska':
        case 'video/quicktime':
        case 'video/mp4':
        case 'video/mpeg':
        case 'video/webm':
            return duration ? formatDuration(duration, 'short') : Messages.defaultLabel();
        case 'image/gif':
            return Messages.gifLabel();
        default:
            throw new UnreachableError(mimeType);
    }
};
