// @canva/asset/lib/esm/sdk/asset/public.js
const { canva } = window;
/**
 * @public
 * This method creates a new asset upload task and adds it to the upload queue.
 * It returns a asset reference, and a function called `whenUploaded()` that can
 * be used to await the upload task completion.
 */ export function upload(options) {
    return canva.content.upload(options);
}
/**
 * @public
 * Returns a temporary URL to an asset already present in the current design.
 */ export function getTemporaryUrl(options) {
    return canva.content.getTemporaryUrl(options);
}
/**
 * @public
 * Lists a curated selection of fonts available for use within Canva when no `options` is provided,
 * otherwise performs filtering based on the criteria specified in the `options`.
 *
 * @remarks
 * To list all available fonts, please use the `requestFontSelection` method.
 */ export function findFonts(options) {
    return canva.content.findFonts(options);
}
/**
 * @public
 * Prompts the user to choose a font that can be used within Canva.
 */ export function requestFontSelection(request = {}) {
    return canva.content.requestFontSelection(request);
}
/**
 * @public
 * Prompts the user to choose a color that can be used within Canva.
 *
 */ export function openColorSelector(anchor, options) {
    return canva.content.openColorSelector(anchor, options);
}
/**
 * @internal
 * When it comes to method overloads, API extractor is omitting the last signature, even if it has
 * the public identifier, that's why the generic upload the returns `QueuedAsset` is repeated
 */ // eslint-disable-next-line @typescript-eslint/no-unused-expressions
'comment';
