// @canva/app-ui-kit/dist/cjs/ui/base/form/internal/input.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: Object.getOwnPropertyDescriptor(all, name).get
    });
}
_export(exports, {
    get BaseInputContextProvider () {
        return BaseInputContextProvider;
    },
    get useCreateInputControls () {
        return useCreateInputControls;
    },
    get useInputControls () {
        return useInputControls;
    },
    get useInputValue () {
        return useInputValue;
    },
    get usePropOverrides () {
        return usePropOverrides;
    }
});
const _jsxruntime = require("react/jsx-runtime");
const _preconditions = require('../../../../base/preconditions');
const _react = _interop_require_wildcard(require("react"));
function _getRequireWildcardCache(nodeInterop) {
    if (typeof WeakMap !== "function") return null;
    var cacheBabelInterop = new WeakMap();
    var cacheNodeInterop = new WeakMap();
    return (_getRequireWildcardCache = function(nodeInterop) {
        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
    })(nodeInterop);
}
function _interop_require_wildcard(obj, nodeInterop) {
    if (!nodeInterop && obj && obj.__esModule) {
        return obj;
    }
    if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
        return {
            default: obj
        };
    }
    var cache = _getRequireWildcardCache(nodeInterop);
    if (cache && cache.has(obj)) {
        return cache.get(obj);
    }
    var newObj = {
        __proto__: null
    };
    var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
    for(var key in obj){
        if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
            if (desc && (desc.get || desc.set)) {
                Object.defineProperty(newObj, key, desc);
            } else {
                newObj[key] = obj[key];
            }
        }
    }
    newObj.default = obj;
    if (cache) {
        cache.set(obj, newObj);
    }
    return newObj;
}
const InputControlsContext = _react.createContext(undefined);
const InputValueContext = _react.createContext('undefined');
function useCreateInputControls(controls) {
    const innerControls = _react.useRef(controls);
    innerControls.current.focus = controls.focus;
    innerControls.current.setValue = controls.setValue;
    innerControls.current.setType = controls.setType;
    innerControls.current.setDisabled = controls.setDisabled;
    return innerControls.current;
}
function useInputControls() {
    return _preconditions.Preconditions.checkExists(_react.useContext(InputControlsContext));
}
function useInputValue() {
    return _preconditions.Preconditions.checkExists(_react.useContext(InputValueContext));
}
function BaseInputContextProvider(props) {
    return (0, _jsxruntime.jsx)(InputControlsContext.Provider, {
        value: props.controls,
        children: (0, _jsxruntime.jsx)(InputValueContext.Provider, {
            value: props.value,
            children: props.children
        })
    });
}
function usePropOverrides(props) {
    const [overrides, setPropOverrides] = _react.useState(undefined);
    return {
        props: overrides != null ? {
            ...props,
            ...overrides
        } : props,
        setPropOverrides
    };
}
