// @canva/app-ui-kit/dist/esm/ui/base/animation/supports_animation.js
import * as React from 'react';
export function prefersReducedMotion() {
    const listener = getPrefersReducedMotionListener();
    return listener.prefersReducedMotion;
}
export function usePrefersReducedMotion() {
    const listener = getPrefersReducedMotionListener();
    const [reduce, setReduce] = React.useState(listener.prefersReducedMotion);
    React.useEffect(()=>{
        if (listener != null) {
            const callback = (e)=>setReduce(e.matches);
            listener.addListener(callback);
            return ()=>{
                listener.removeListener(callback);
            };
        }
    }, [
        listener,
        setReduce
    ]);
    return reduce;
}
function getPrefersReducedMotionListener() {
    return globalListener !== null && globalListener !== void 0 ? globalListener : globalListener = new PrefersReducedMotionListener();
}
class PrefersReducedMotionListener {
    get prefersReducedMotion() {
        return this.mediaQuery != null && this.mediaQuery.matches;
    }
    addListener(listener) {
        var _this_mediaQuery;
        (_this_mediaQuery = this.mediaQuery) === null || _this_mediaQuery === void 0 ? void 0 : _this_mediaQuery.addListener(listener);
    }
    removeListener(listener) {
        var _this_mediaQuery;
        (_this_mediaQuery = this.mediaQuery) === null || _this_mediaQuery === void 0 ? void 0 : _this_mediaQuery.removeListener(listener);
    }
    constructor(){
        this.mediaQuery = typeof window === 'undefined' ? undefined : window.matchMedia && window.matchMedia('(prefers-reduced-motion: reduce)');
    }
}
let globalListener;
