// @canva/app-ui-kit/dist/esm/ui/base/mobile_event_handler/mobile_event_handler.js
import { jsx as _jsx } from "react/jsx-runtime";
import * as React from 'react';
const MobileEventHandlerContext = React.createContext(undefined);
export const MobileEvents = {
  BACK: "backbutton"
};
export const MobileEventHandler = function MobileEventHandler({
  onBackButton,
  children
}) {
  const [nodeContext] = React.useState(() => ({
    handler: onBackButton,
    childContexts: []
  }));
  React.useEffect(() => {
    nodeContext.handler = onBackButton;
  }, [nodeContext, onBackButton]);
  const parentContext = React.useContext(MobileEventHandlerContext);
  React.useEffect(() => {
    if (parentContext) {
      parentContext.childContexts.unshift(nodeContext);
      return () => {
        const {
          childContexts
        } = parentContext;
        const index = childContexts.indexOf(nodeContext);
        if (index > -1) childContexts.splice(index, 1);
      };
    } else {
      const processHandlerTree = () => processContext(nodeContext);
      return backButtonTrigger(processHandlerTree);
    }
  },
  []);
  return _jsx(MobileEventHandlerContext.Provider, {
    value: nodeContext,
    children: children
  });
};
const backButtonTrigger = onTrigger => {
  try {
    var _self_Android_getBackButtonEventName, _self_Android;
    var _self_Android_getBackButtonEventName1;
    const backButtonEventName = (_self_Android_getBackButtonEventName1 = (_self_Android = self.Android) === null || _self_Android === void 0 ? void 0 : (_self_Android_getBackButtonEventName = _self_Android.getBackButtonEventName) === null || _self_Android_getBackButtonEventName === void 0 ? void 0 : _self_Android_getBackButtonEventName.call(_self_Android)) !== null && _self_Android_getBackButtonEventName1 !== void 0 ? _self_Android_getBackButtonEventName1 : "backbutton";
    document.addEventListener(backButtonEventName, onTrigger);
    return () => document.removeEventListener(backButtonEventName, onTrigger);
  } catch (e) {
    if (window.self === window.top)
      throw e;
  }
};
const processContext = ({
  childContexts,
  handler
}) => {
  let propagationStopped = false;
  for (const ctx of childContexts) {
    propagationStopped = processContext(ctx);
    if (propagationStopped) break;
  }
  if (!propagationStopped)
    handler({
      stopPropagation: () => propagationStopped = true
    });
  return propagationStopped;
};
