// @canva/app-ui-kit/dist/esm/ui/base/carousel/private/scroll_button.js
import { jsx as _jsx } from "react/jsx-runtime";
import classNames from 'classnames';
import { observer } from 'mobx-react-lite';
import * as React from 'react';
import { accessModeState } from '../../a11y/access_mode_state/access_mode_state';
import { useIsHoveringInput } from '../../a11y/pointer_input_state/pointer_input_state';
import { ChevronLeftIcon } from '../../icons/chevron_left/icon';
import { ChevronRightIcon } from '../../icons/chevron_right/icon';
import styles, { customProperties, getStyle } from './scroll_button.css';
import { ButtonMessages as Messages } from './scroll_button.messages';
const getClassName = ({ className, active, direction, variant, hoverSupported })=>classNames(styles.base, getStyle(variant), {
        [styles.active]: active,
        [styles.hoverSupported]: hoverSupported,
        [styles.forward]: direction === 'forward',
        [styles.backward]: direction === 'backward'
    }, className);
function getButtonContent(props) {
    const { direction } = props;
    if (direction === 'forward') return _jsx(ChevronRightIcon, {
        size: "medium",
        className: styles.icon
    });
    else return _jsx(ChevronLeftIcon, {
        size: "medium",
        className: styles.icon
    });
}
const focusElement = (ref)=>{
    var _ref_current;
    ref === null || ref === void 0 ? void 0 : (_ref_current = ref.current) === null || _ref_current === void 0 ? void 0 : _ref_current.focus();
};
const Button = observer((props)=>{
    const shouldFocusAfterClick = React.useRef(false);
    React.useLayoutEffect(()=>{
        if (shouldFocusAfterClick.current) {
            if (props.scrollState.atStart) {
                if (accessModeState.isMouseMode)
                    focusElement(props.oppositeButtonRef);
                else
                    focusElement(props.startLinkRef);
            }
            if (props.scrollState.atEnd) {
                if (accessModeState.isMouseMode)
                    focusElement(props.oppositeButtonRef);
                else
                    focusElement(props.endLinkRef);
            }
        }
    }, [
        props.scrollState.atStart,
        props.scrollState.atEnd,
        props.oppositeButtonRef,
        props.direction,
        props.buttonRef,
        props.startLinkRef,
        props.endLinkRef
    ]);
    const handleClick = ()=>{
        shouldFocusAfterClick.current = true;
        props.onClick();
    };
    const hoverSupported = useIsHoveringInput();
    if (props.scrollState.atStart === false && props.direction === 'backward' || props.scrollState.atEnd === false && props.direction === 'forward') return _jsx("button", {
        onClick: handleClick,
        className: getClassName({
            active: props.active,
            variant: props.variant,
            direction: props.direction,
            hoverSupported
        }),
        "aria-controls": props.scrollableId,
        "aria-label": props.direction === 'forward' ? Messages.screenReaderNext() : Messages.screenReaderPrevious(),
        style: props.style,
        type: "button",
        ref: props.buttonRef,
        children: getButtonContent(props)
    });
    return null;
});
function getButtonStyle(verticalOffsetPx, horizontalOpts) {
    const horizontalDirection = horizontalOpts && (horizontalOpts.direction === 'backward' ? 1 : -1);
    const horizontalOffsetPx = horizontalDirection && horizontalOpts.horizontalOffsetPx * horizontalDirection;
    return {
        [customProperties.horizontalOffset]: horizontalOffsetPx ? `${horizontalOffsetPx}px` : undefined,
        [customProperties.verticalOffset]: verticalOffsetPx ? `${verticalOffsetPx}px` : undefined
    };
}
export const CircularScrollButton = (props)=>{
    return _jsx(Button, {
        ...props,
        variant: "circular",
        style: getButtonStyle(props.verticalOffsetPx, props.horizontalOffsetPx ? {
            horizontalOffsetPx: props.horizontalOffsetPx,
            direction: props.direction
        } : undefined)
    });
};
export const ChevronScrollButton = (props)=>_jsx(Button, {
        ...props,
        variant: "chevron",
        style: getButtonStyle(props.verticalOffsetPx)
    });
