// @canva/app-ui-kit/dist/cjs/ui/base/surface/popover/private/arrow.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "Arrow", {
    enumerable: true,
    get: function() {
        return Arrow;
    }
});
const _jsxruntime = require("react/jsx-runtime");
const _react = _interop_require_wildcard(require("react"));
const _arrowcss = _interop_require_default(require("./arrow.css"));
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function _getRequireWildcardCache(nodeInterop) {
    if (typeof WeakMap !== "function") return null;
    var cacheBabelInterop = new WeakMap();
    var cacheNodeInterop = new WeakMap();
    return (_getRequireWildcardCache = function(nodeInterop) {
        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
    })(nodeInterop);
}
function _interop_require_wildcard(obj, nodeInterop) {
    if (!nodeInterop && obj && obj.__esModule) {
        return obj;
    }
    if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
        return {
            default: obj
        };
    }
    var cache = _getRequireWildcardCache(nodeInterop);
    if (cache && cache.has(obj)) {
        return cache.get(obj);
    }
    var newObj = {
        __proto__: null
    };
    var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
    for(var key in obj){
        if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
            if (desc && (desc.get || desc.set)) {
                Object.defineProperty(newObj, key, desc);
            } else {
                newObj[key] = obj[key];
            }
        }
    }
    newObj.default = obj;
    if (cache) {
        cache.set(obj, newObj);
    }
    return newObj;
}
const Arrow = _react.forwardRef((_props, ref)=>(0, _jsxruntime.jsx)("div", {
        ref: ref,
        className: _arrowcss.default.wrapper,
        children: (0, _jsxruntime.jsxs)("svg", {
            width: "44",
            height: "45",
            viewBox: "-6 -27 44 45",
            fill: "none",
            xmlns: "http://www.w3.org/2000/svg",
            className: _arrowcss.default.arrow,
            children: [
                (0, _jsxruntime.jsx)("g", {
                    filter: "url(#filter0_f_3010_39642)",
                    children: (0, _jsxruntime.jsx)("rect", {
                        x: "24",
                        y: "12",
                        width: "16",
                        height: "4",
                        transform: "rotate(-180 24 12)",
                        className: _arrowcss.default.shadow
                    })
                }),
                (0, _jsxruntime.jsx)("path", {
                    d: "M23.8892 15.3409C24.9794 16.4032 26.4582 17 28 17V18H4V17C5.54184 17 7.02055 16.4032 8.11078 15.3409L11.8892 11.6592C14.1595 9.44694 17.8405 9.44694 20.1108 11.6592L23.8892 15.3409Z",
                    className: _arrowcss.default.content
                }),
                (0, _jsxruntime.jsx)("path", {
                    d: "M28 16C26.7151 16 25.4879 15.5024 24.5871 14.6246L20.8087 10.943C18.1501 8.35235 13.8499 8.35235 11.1913 10.943L7.4129 14.6246C6.51208 15.5024 5.28487 16 4 16H7.29842C7.58721 15.8061 7.85933 15.5859 8.11078 15.3409L11.8892 11.6592C14.1595 9.44694 17.8405 9.44694 20.1108 11.6592L23.8892 15.3409C24.1407 15.5859 24.4128 15.8061 24.7016 16H28Z",
                    className: _arrowcss.default.border
                }),
                (0, _jsxruntime.jsx)("defs", {
                    children: (0, _jsxruntime.jsxs)("filter", {
                        id: "filter0_f_3010_39642",
                        x: "0",
                        y: "0",
                        width: "32",
                        height: "20",
                        filterUnits: "userSpaceOnUse",
                        colorInterpolationFilters: "sRGB",
                        children: [
                            (0, _jsxruntime.jsx)("feFlood", {
                                floodOpacity: "0",
                                result: "BackgroundImageFix"
                            }),
                            (0, _jsxruntime.jsx)("feBlend", {
                                mode: "normal",
                                in: "SourceGraphic",
                                in2: "BackgroundImageFix",
                                result: "shape"
                            }),
                            (0, _jsxruntime.jsx)("feGaussianBlur", {
                                stdDeviation: "4",
                                result: "effect1_foregroundBlur_3010_39642"
                            })
                        ]
                    })
                })
            ]
        })
    }));
