// @canva/app-ui-kit/dist/cjs/ui/apps/developing/ui_kit/components/app_ui_provider/app_ui_provider.js
"use strict"
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: Object.getOwnPropertyDescriptor(all, name).get
    });
}
_export(exports, {
    get AppUiProvider () {
        return AppUiProvider;
    },
    get toEaselDirection () {
        return toEaselDirection;
    }
});
const _jsxruntime = require("react/jsx-runtime");
const _react = _interop_require_wildcard(require("react"));
const _config = require('../../config/config');
const _provider = require('../../../../../base/provider/provider');
const _theme = require('../../../../../base/theme/theme');
function _getRequireWildcardCache(nodeInterop) {
    if (typeof WeakMap !== "function") return null;
    var cacheBabelInterop = new WeakMap();
    var cacheNodeInterop = new WeakMap();
    return (_getRequireWildcardCache = function(nodeInterop) {
        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
    })(nodeInterop);
}
function _interop_require_wildcard(obj, nodeInterop) {
    if (!nodeInterop && obj && obj.__esModule) {
        return obj;
    }
    if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
        return {
            default: obj
        };
    }
    var cache = _getRequireWildcardCache(nodeInterop);
    if (cache && cache.has(obj)) {
        return cache.get(obj);
    }
    var newObj = {
        __proto__: null
    };
    var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
    for(var key in obj){
        if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
            if (desc && (desc.get || desc.set)) {
                Object.defineProperty(newObj, key, desc);
            } else {
                newObj[key] = obj[key];
            }
        }
    }
    newObj.default = obj;
    if (cache) {
        cache.set(obj, newObj);
    }
    return newObj;
}
var _internalSdkWindow___canva__;
const toEaselDirection = (direction)=>{
    if (typeof direction === 'string' && direction.toUpperCase() === 'RTL') return 'RTL';
    return 'LTR';
};
const internalSdkWindow = window;
const uiKit = internalSdkWindow === null || internalSdkWindow === void 0 ? void 0 : (_internalSdkWindow___canva__ = internalSdkWindow['__canva__']) === null || _internalSdkWindow___canva__ === void 0 ? void 0 : _internalSdkWindow___canva__['uiKit'];
uiKit === null || uiKit === void 0 ? void 0 : uiKit.registerAppUiKitVersion({
    version: _config.LATEST_APP_UI_KIT_VERSION
}).catch(()=>{});
function AppUiProvider({ children }) {
    const [theme, setTheme] = _react.useState();
    function setStateFromContext(context) {
        setTheme(context.theme);
    }
    _react.useEffect(()=>{
        if (!uiKit) throw new Error("`AppUiProvider` must be used within the context of a Canva app. If you're attempting to use this component in a test environment, please use `TestAppUiProvider` component instead. Use of App UI Kit components outside of a Canva app is forbidden and strictly prohibited as per our License.");
        uiKit.getUiContext().then(setStateFromContext);
        return uiKit.onUiContextChange(setStateFromContext);
    }, []);
    if (!theme)
        return (0, _jsxruntime.jsx)(_jsxruntime.Fragment, {});
    const directionFromDocumentElement = document.documentElement.getAttribute('dir');
    return (0, _jsxruntime.jsxs)(_provider.EaselProvider, {
        enableAnimations: true,
        direction: toEaselDirection(directionFromDocumentElement),
        disableFocusTraps: true,
        enableUserSelection: false,
        disableDialogBlur: false,
        children: [
            (0, _jsxruntime.jsx)(_theme.ApplyThemeRoot, {
                appearance: theme,
                classicLight: theme,
                classicDark: theme,
                light: theme,
                dark: theme
            }),
            children
        ]
    });
}
