// @canva/app-ui-kit/dist/cjs/ui/apps/developing/ui_kit/components/app_ui_provider/test_app_ui_provider.js
"use strict"
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "TestAppUiProvider", {
    enumerable: true,
    get: function() {
        return TestAppUiProvider;
    }
});
const _jsxruntime = require("react/jsx-runtime");
require("react");
const _provider = require('../../../../../base/provider/provider');
const _theme = require('../../../../../base/theme/theme');
const _app_ui_provider = require("./app_ui_provider");
function TestAppUiProvider({ children, theme = 'light', direction = 'LTR' }) {
    var _wrappedWindow___canva__;
    const wrappedWindow = window;
    const uiKit = wrappedWindow === null || wrappedWindow === void 0 ? void 0 : (_wrappedWindow___canva__ = wrappedWindow['__canva__']) === null || _wrappedWindow___canva__ === void 0 ? void 0 : _wrappedWindow___canva__['uiKit'];
    if (uiKit != null)
        throw new Error(`'TestAppUiProvider' detected in a non-test environment. Replace 'TestAppUiProvider' usage with 'AppUiProvider' in your app.`);
    return (0, _jsxruntime.jsxs)(_provider.EaselProvider, {
        enableAnimations: true,
        direction: (0, _app_ui_provider.toEaselDirection)(direction),
        disableFocusTraps: true,
        enableUserSelection: false,
        disableDialogBlur: false,
        children: [
            (0, _jsxruntime.jsx)(_theme.ApplyThemeRoot, {
                appearance: theme,
                classicLight: theme,
                classicDark: theme,
                light: theme,
                dark: theme
            }),
            children
        ]
    });
}
