// @canva/app-ui-kit/dist/esm/ui/base/provider/private/provider.js
import { jsx as _jsx } from "react/jsx-runtime";
import { Preconditions } from '../../../../base/preconditions';
import * as React from 'react';
import { InteractionManager } from '../../a11y/interaction_manager/interaction_manager';
const EaselContext = React.createContext(null);
export function EaselProvider({ enableAnimations = true, direction, disableFocusTraps, enableUserSelection, theme, enableHighColorContrast, autoplayVideos, disableDialogBlur, expectedPixelsPerRem, children }) {
    return _jsx(InteractionManager, {
        children: _jsx(EaselContext.Provider, {
            value: {
                enableAnimations,
                direction,
                disableFocusTraps,
                enableUserSelection,
                theme,
                enableHighColorContrast,
                autoplayVideos,
                disableDialogBlur,
                expectedPixelsPerRem
            },
            children: children
        })
    });
}
function useEaselContext() {
    return Preconditions.checkExists(React.useContext(EaselContext), 'Did you forget to render a provider?');
}
export function useEnableAnimations() {
    return useEaselContext().enableAnimations;
}
export function useDirection() {
    return useEaselContext().direction;
}
export function useDisableFocusTraps() {
    return useEaselContext().disableFocusTraps;
}
export function useEnableUserSelection() {
    return useEaselContext().enableUserSelection;
}
export function useTheme() {
    return useEaselContext().theme;
}
export function useEnableHighColorContrast() {
    return useEaselContext().enableHighColorContrast;
}
export function useAutoplayVideos() {
    return useEaselContext().autoplayVideos;
}
export function useDisableDialogBlur() {
    return useEaselContext().disableDialogBlur;
}
export function useExpectedPixelsPerRem() {
    return useEaselContext().expectedPixelsPerRem;
}
