// @canva/app-ui-kit/dist/esm/ui/apps/developing/ui_kit/components/pill/pill.js
import { jsx as _jsx } from "react/jsx-runtime";
import { UnreachableError } from '../../../../../../base/preconditions';
import { Pill as EaselPill } from '../../../../../base/pill/pill';
/** 
 * Pills show selected entities, and suggest or filter content.
 *
 * Use pills for:
 * - Presenting available filters in a view
 * - Making contextual suggestions for searching and navigating content
 */ export function Pill({ truncateText, ...props }) {
    var _props_showTooltip, _props_showTooltip1, _props_showTooltip2;
    var _props_showTooltip_label;
    return _jsx(EaselPill, {
        ...props,
        maxWidth: mapMaxWidth(truncateText),
        size: props.size ? props.size : 'medium',
        shape: "rectangle",
        tooltipLabel: typeof props.showTooltip === 'object' ? (_props_showTooltip_label = (_props_showTooltip = props.showTooltip) === null || _props_showTooltip === void 0 ? void 0 : _props_showTooltip.label) !== null && _props_showTooltip_label !== void 0 ? _props_showTooltip_label : props.text : props.showTooltip === true ? props.text : undefined,
        tooltipPlacement: typeof props.showTooltip === 'object' ? (_props_showTooltip1 = props.showTooltip) === null || _props_showTooltip1 === void 0 ? void 0 : _props_showTooltip1.placement : undefined,
        tooltipDisabled: typeof props.showTooltip === 'object' ? (_props_showTooltip2 = props.showTooltip) === null || _props_showTooltip2 === void 0 ? void 0 : _props_showTooltip2.disabled : undefined
    });
}
function mapMaxWidth(truncateText) {
    switch(truncateText){
        case undefined:
            return undefined;
        case true:
            return '25u';
        case false:
            return '100%';
        default:
            throw new UnreachableError(truncateText);
    }
}
