// @canva/app-ui-kit/dist/cjs/ui/base/surface/header/private/header.js
"use strict"
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: Object.getOwnPropertyDescriptor(all, name).get
    });
}
_export(exports, {
    get SurfaceHeader () {
        return SurfaceHeader;
    },
    get SurfaceHeaderBackButton () {
        return SurfaceHeaderBackButton;
    },
    get SurfaceHeaderCloseButton () {
        return SurfaceHeaderCloseButton;
    },
    get SurfaceHeaderDescription () {
        return SurfaceHeaderDescription;
    },
    get SurfaceHeaderIconButton () {
        return SurfaceHeaderIconButton;
    },
    get SurfaceHeaderTextButton () {
        return SurfaceHeaderTextButton;
    },
    get SurfaceHeaderTitle () {
        return SurfaceHeaderTitle;
    }
});
const _jsxruntime = require("react/jsx-runtime");
const _classnames = _interop_require_default(require("classnames"));
const _react = _interop_require_wildcard(require("react"));
const _box = require('../../../box/box');
const _button = require('../../../button/button');
const _icon = require('../../../icons/arrow_left/icon');
const _icon1 = require('../../../icons/x/icon');
const _layout = require('../../../layout/layout');
const _typography = require('../../../typography/typography');
const _headercss = _interop_require_default(require("./header.css"));
const _headermessages = require("./header.messages");
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function _getRequireWildcardCache(nodeInterop) {
    if (typeof WeakMap !== "function") return null;
    var cacheBabelInterop = new WeakMap();
    var cacheNodeInterop = new WeakMap();
    return (_getRequireWildcardCache = function(nodeInterop) {
        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
    })(nodeInterop);
}
function _interop_require_wildcard(obj, nodeInterop) {
    if (!nodeInterop && obj && obj.__esModule) {
        return obj;
    }
    if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
        return {
            default: obj
        };
    }
    var cache = _getRequireWildcardCache(nodeInterop);
    if (cache && cache.has(obj)) {
        return cache.get(obj);
    }
    var newObj = {
        __proto__: null
    };
    var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
    for(var key in obj){
        if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
            if (desc && (desc.get || desc.set)) {
                Object.defineProperty(newObj, key, desc);
            } else {
                newObj[key] = obj[key];
            }
        }
    }
    newObj.default = obj;
    if (cache) {
        cache.set(obj, newObj);
    }
    return newObj;
}
const SurfaceHeader = ({ title, titleId, description, descriptionId, alignment = 'start', start, end, divider = true, padding = '2u', background = 'surface' })=>{
    const hasStructuredPart = title != null || description != null || start != null || end != null;
    return (0, _jsxruntime.jsx)(_box.Box, {
        padding: padding,
        width: "full",
        background: background === 'none' ? undefined : background,
        borderBottom: divider ? 'low' : undefined,
        children: hasStructuredPart && (0, _jsxruntime.jsx)(_box.Box, {
            className: _headercss.default.bleed,
            display: "flex",
            children: (0, _jsxruntime.jsx)(StructuredHeaderContent, {
                title: title,
                titleId: titleId,
                description: description,
                descriptionId: descriptionId,
                alignment: alignment,
                start: start,
                end: end
            })
        })
    });
};
const TitleAndDescription = ({ title, titleId, description, descriptionId, alignment })=>{
    return (0, _jsxruntime.jsxs)(_jsxruntime.Fragment, {
        children: [
            typeof title === 'string' ? (0, _jsxruntime.jsx)(SurfaceHeaderTitle, {
                id: titleId,
                alignment: alignment,
                children: title
            }) : titleId ? (0, _jsxruntime.jsx)("div", {
                id: titleId,
                children: title
            }) : title,
            typeof description === 'string' ? (0, _jsxruntime.jsx)(SurfaceHeaderDescription, {
                id: descriptionId,
                alignment: alignment,
                children: description
            }) : descriptionId ? (0, _jsxruntime.jsx)("div", {
                id: descriptionId,
                children: description
            }) : description
        ]
    });
};
const renderDecorator = (decorator)=>
    _react.isValidElement(decorator) && decorator.type === _react.Fragment ? (0, _jsxruntime.jsx)(_jsxruntime.Fragment, {
        children: _react.Children.map(decorator.props.children, (child)=>child && renderDecorator(child))
    }) : (0, _jsxruntime.jsx)(_layout.Column, {
        width: "content",
        children: decorator
    });
const renderDecorators = (decorator)=>{
    return (0, _jsxruntime.jsx)(_box.Box, {
        className: _headercss.default.minHeight,
        display: "flex",
        alignItems: "center",
        children: (0, _jsxruntime.jsx)(_layout.Columns, {
            spacing: "0.5u",
            alignY: "center",
            children: renderDecorator(decorator)
        })
    });
};
const StructuredHeaderContent = ({ title, titleId, description, descriptionId, alignment, start, end })=>{
    const titleAndDescriptionClassName = (0, _classnames.default)(_headercss.default.minHeight, {
        [_headercss.default.widthFluid]: alignment === 'start'
    });
    return (0, _jsxruntime.jsxs)(_jsxruntime.Fragment, {
        children: [
            (0, _jsxruntime.jsx)(_box.Box, {
                justifyContent: "start",
                className: alignment === 'center' && _headercss.default.widthFluid,
                children: start != null ? (0, _jsxruntime.jsx)(_box.Box, {
                    paddingEnd: "0.5u",
                    children: renderDecorators(start)
                }) : null
            }),
            title || description ? (0, _jsxruntime.jsx)(_box.Box, {
                className: titleAndDescriptionClassName,
                minWidth: "0",
                display: "flex",
                flexDirection: "column",
                justifyContent: "center",
                paddingX: "0.5u",
                children: (0, _jsxruntime.jsx)(TitleAndDescription, {
                    title: title,
                    titleId: titleId,
                    description: description,
                    descriptionId: descriptionId,
                    alignment: alignment
                })
            }) : (0, _jsxruntime.jsx)(_box.Box, {
                className: titleAndDescriptionClassName
            }),
            (0, _jsxruntime.jsx)(_box.Box, {
                display: "flex",
                justifyContent: "end",
                alignItems: "start",
                className: alignment === 'center' && _headercss.default.widthFluid,
                children: end != null ? (0, _jsxruntime.jsx)(_box.Box, {
                    paddingStart: "0.5u",
                    children: renderDecorators(end)
                }) : null
            })
        ]
    });
};
const SurfaceHeaderTitle = ({ children, id, alignment })=>{
    return (0, _jsxruntime.jsx)(_typography.Title, {
        size: "xsmall",
        lineClamp: 1,
        alignment: alignment,
        id: id,
        children: children
    });
};
const SurfaceHeaderDescription = ({ children, alignment = 'start', id })=>{
    return (0, _jsxruntime.jsx)(_typography.Text, {
        size: "small",
        tone: "tertiary",
        alignment: alignment,
        id: id,
        children: children
    });
};
const SurfaceHeaderIconButton = (props)=>{
    return (0, _jsxruntime.jsx)(_button.Button, {
        ...props,
        variant: "tertiary",
        size: "small"
    });
};
const SurfaceHeaderBackButton = ({ ariaLabel, ...props })=>{
    return (0, _jsxruntime.jsx)(SurfaceHeaderIconButton, {
        icon: _icon.ArrowLeftIcon,
        ariaLabel: ariaLabel || _headermessages.HeaderMessages.defaultGoBack(),
        ...props
    });
};
const SurfaceHeaderCloseButton = ({ ariaLabel, ...props })=>{
    return (0, _jsxruntime.jsx)(SurfaceHeaderIconButton, {
        icon: _icon1.XIcon,
        ariaLabel: ariaLabel || _headermessages.HeaderMessages.defaultClose(),
        ...props
    });
};
const SurfaceHeaderTextButton = (props)=>{
    return (0, _jsxruntime.jsx)(_button.Button, {
        variant: "linkButton",
        size: "small",
        ...props
    });
};
