// @canva/app-ui-kit/dist/cjs/ui/base/color_picker/sat_val_picker/private/sat_val_picker.js
"use strict"
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: Object.getOwnPropertyDescriptor(all, name).get
    });
}
_export(exports, {
    get SatValPicker () {
        return SatValPicker;
    },
    get mouseToKnobCoords () {
        return mouseToKnobCoords;
    }
});
const _jsxruntime = require("react/jsx-runtime");
const _react = _interop_require_wildcard(require("react"));
const _hsv_color = require('../../hsv_color/hsv_color');
const _sat_val_pickercss = _interop_require_default(require("./sat_val_picker.css"));
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function _getRequireWildcardCache(nodeInterop) {
    if (typeof WeakMap !== "function") return null;
    var cacheBabelInterop = new WeakMap();
    var cacheNodeInterop = new WeakMap();
    return (_getRequireWildcardCache = function(nodeInterop) {
        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
    })(nodeInterop);
}
function _interop_require_wildcard(obj, nodeInterop) {
    if (!nodeInterop && obj && obj.__esModule) {
        return obj;
    }
    if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
        return {
            default: obj
        };
    }
    var cache = _getRequireWildcardCache(nodeInterop);
    if (cache && cache.has(obj)) {
        return cache.get(obj);
    }
    var newObj = {
        __proto__: null
    };
    var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
    for(var key in obj){
        if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
            if (desc && (desc.get || desc.set)) {
                Object.defineProperty(newObj, key, desc);
            } else {
                newObj[key] = obj[key];
            }
        }
    }
    newObj.default = obj;
    if (cache) {
        cache.set(obj, newObj);
    }
    return newObj;
}
function SatValPicker({ color, Draggable, onChange: onChangeProp, onChangeStart, onChangeComplete }) {
    const pickerRef = _react.useRef(null);
    const setPickerRef = _react.useCallback((el)=>{
        pickerRef.current = el;
        el === null || el === void 0 ? void 0 : el.addEventListener('touchmove', (event)=>event.stopPropagation());
    }, []);
    const [browserWindow, setBrowserWindow] = _react.useState();
    _react.useLayoutEffect(()=>{
        var _pickerRef_current;
        var _pickerRef_current_ownerDocument_defaultView;
        setBrowserWindow((_pickerRef_current_ownerDocument_defaultView = (_pickerRef_current = pickerRef.current) === null || _pickerRef_current === void 0 ? void 0 : _pickerRef_current.ownerDocument.defaultView) !== null && _pickerRef_current_ownerDocument_defaultView !== void 0 ? _pickerRef_current_ownerDocument_defaultView : window);
    }, []);
    const [changing, setChanging] = _react.useState(false);
    const onChange = ({ s, v })=>{
        if (!changing) {
            setChanging(true);
            onChangeStart && onChangeStart();
        }
        if (s !== color.s || v !== color.v) onChangeProp === null || onChangeProp === void 0 ? void 0 : onChangeProp(new _hsv_color.HsvColor(color.h, s, v, color.alpha));
    };
    const moveKnob = (knob)=>{
        onChange({
            s: knob.leftPercent,
            v: 1 - knob.topPercent
        });
    };
    const stopChanging = _react.useCallback(()=>{
        setChanging(false);
        if (onChangeComplete) onChangeComplete();
    }, [
        onChangeComplete
    ]);
    const onKeyDown = (event)=>{
        const INCREMENT = 0.01;
        let { s, v } = color;
        switch(event.key){
            case 'ArrowUp':
                v += INCREMENT;
                break;
            case 'ArrowDown':
                v -= INCREMENT;
                break;
            case 'ArrowLeft':
                s -= INCREMENT;
                break;
            case 'ArrowRight':
                s += INCREMENT;
                break;
            default:
                return;
        }
        event.preventDefault();
        const clamp = (n)=>Math.min(Math.max(0, n), 1);
        onChange({
            s: clamp(s),
            v: clamp(v)
        });
    };
    const onKeyUp = _react.useCallback(()=>{
        stopChanging();
    }, [
        stopChanging
    ]);
    const onMouseUp = _react.useCallback(()=>{
        stopChanging();
    }, [
        stopChanging
    ]);
    const onMouseDown = (event)=>{
        var _pickerRef_current;
        if (event.button !== 0)
            return;
        (_pickerRef_current = pickerRef.current) === null || _pickerRef_current === void 0 ? void 0 : _pickerRef_current.focus();
        moveKnob(mouseToKnobCoords(pickerRef.current.getBoundingClientRect(), event));
    };
    const onDragMove = (event)=>moveKnob(mouseToKnobCoords(pickerRef.current.getBoundingClientRect(), event));
    const onDragEnd = _react.useCallback((event)=>{
        stopChanging();
    }, [
        stopChanging
    ]);
    const knobPosition = {
        width: `${color.s * 100}%`,
        height: `${100 - color.v * 100}%`
    };
    return (0, _jsxruntime.jsx)(Draggable, {
        onMouseDown: onMouseDown,
        onDragMove: onDragMove,
        onDragEnd: onDragEnd,
        browserWindow: browserWindow,
        raf: browserWindow === null || browserWindow === void 0 ? void 0 : browserWindow.requestAnimationFrame,
        caf: browserWindow === null || browserWindow === void 0 ? void 0 : browserWindow.cancelAnimationFrame,
        children: ({
            onMouseDown,
            onTouchStart
        })=>
            (0, _jsxruntime.jsx)("div", {
                className: _sat_val_pickercss.default.root,
                onMouseDown: onMouseDown,
                onTouchStart: onTouchStart,
                onMouseUp: onMouseUp,
                onKeyDown: onKeyDown,
                onKeyUp: onKeyUp,
                ref: setPickerRef,
                tabIndex: 0,
                style: {
                    backgroundColor: `hsl(${color.h * 360}, 100%, 50%)`
                },
                children: (0, _jsxruntime.jsx)("div", {
                    className: _sat_val_pickercss.default.whiteGradient,
                    children: (0, _jsxruntime.jsx)("div", {
                        className: _sat_val_pickercss.default.blackGradient,
                        children: (0, _jsxruntime.jsx)("div", {
                            className: _sat_val_pickercss.default.knobPositioner,
                            style: knobPosition,
                            children: (0, _jsxruntime.jsx)("div", {
                                className: _sat_val_pickercss.default.knob,
                                style: {
                                    backgroundColor: color.toRgb().toHexString()
                                }
                            })
                        })
                    })
                })
            })
    });
}
function mouseToKnobCoords(pickerRect, { clientX, clientY }) {
    const clamp = (val)=>Math.min(1, Math.max(0, val));
    return {
        leftPercent: clamp((clientX - pickerRect.left) / pickerRect.width),
        topPercent: clamp((clientY - pickerRect.top) / pickerRect.height)
    };
}
