// @canva/app-ui-kit/dist/cjs/ui/base/card/card.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: Object.getOwnPropertyDescriptor(all, name).get
    });
}
_export(exports, {
    get Card () {
        return _card.Card;
    },
    get CardAction () {
        return _card_action.CardAction;
    },
    get CardButton () {
        return _card_button.CardButton;
    },
    get CardButtonGroup () {
        return _card_button.CardButtonGroup;
    },
    get CardCheckbox () {
        return _card_checkbox.CardCheckbox;
    },
    get CardCheckboxDecorator () {
        return _card_checkbox.CardCheckboxDecorator;
    },
    get CardDecorator () {
        return _card_decorator.CardDecorator;
    },
    get CardDescription () {
        return _card_content.CardDescription;
    },
    get CardIconThumbnail () {
        return _card_thumbnail.CardIconThumbnail;
    },
    get CardImageThumbnail () {
        return _card_thumbnail.CardImageThumbnail;
    },
    get CardMenuButton () {
        return _card_button.CardMenuButton;
    },
    get CardPlaceholder () {
        return _card_thumbnail.CardPlaceholder;
    },
    get CardPreTitle () {
        return _card_content.CardPreTitle;
    },
    get CardThumbnail () {
        return _card_thumbnail.CardThumbnail;
    },
    get CardThumbnailContainer () {
        return _card_thumbnail.CardThumbnailContainer;
    },
    get CardTitle () {
        return _card_content.CardTitle;
    },
    get CardVideoThumbnail () {
        return _card_thumbnail.CardVideoThumbnail;
    },
    get StatelessCardMenuButton () {
        return _card_button.StatelessCardMenuButton;
    },
    get useCardFocus () {
        return _card_context.useCardFocus;
    },
    get useCardHover () {
        return _card_context.useCardHover;
    },
    get useCardMouseDown () {
        return _card_context.useCardMouseDown;
    }
});
const _card = require("./private/card");
const _card_action = require("./private/card_action");
const _card_content = require("./private/card_content");
const _card_context = require("./private/card_context");
const _card_thumbnail = require("./private/card_thumbnail");
const _card_button = require("./private/decorators/card_button");
const _card_checkbox = require("./private/decorators/card_checkbox");
const _card_decorator = require("./private/decorators/card_decorator");
