// @canva/app-ui-kit/dist/cjs/ui/base/surface/pin/private/pin.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "Pin", {
    enumerable: true,
    get: function() {
        return Pin;
    }
});
const _jsxruntime = require("react/jsx-runtime");
const _preconditions = require('../../../../../base/preconditions');
const _classnames = _interop_require_default(require("classnames"));
const _react = _interop_require_wildcard(require("react"));
const _layer = require('../../../layer/layer');
const _backend = require("./backend");
const _pincss = _interop_require_default(require("./pin.css"));
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function _getRequireWildcardCache(nodeInterop) {
    if (typeof WeakMap !== "function") return null;
    var cacheBabelInterop = new WeakMap();
    var cacheNodeInterop = new WeakMap();
    return (_getRequireWildcardCache = function(nodeInterop) {
        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
    })(nodeInterop);
}
function _interop_require_wildcard(obj, nodeInterop) {
    if (!nodeInterop && obj && obj.__esModule) {
        return obj;
    }
    if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
        return {
            default: obj
        };
    }
    var cache = _getRequireWildcardCache(nodeInterop);
    if (cache && cache.has(obj)) {
        return cache.get(obj);
    }
    var newObj = {
        __proto__: null
    };
    var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
    for(var key in obj){
        if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
            if (desc && (desc.get || desc.set)) {
                Object.defineProperty(newObj, key, desc);
            } else {
                newObj[key] = obj[key];
            }
        }
    }
    newObj.default = obj;
    if (cache) {
        cache.set(obj, newObj);
    }
    return newObj;
}
const Pin = _react.forwardRef(
    function Pin({ open, onOutsidePointerDown, blockOutsidePointerEvents = false, blockInsidePointerEvents = false, onCalculateLayout, reference, placement = 'auto', enableFlip = true, offset = {
        main: 0,
        cross: 0
    }, boundary = 'viewport', boundaryPadding = '1u', level = _layer.LayerLevel.PINS, parentLayer, children }, handleRef) {
        const resolvedPlacement = useAutoPlacement(placement);
        const { placement: FixedPlacement, rtlAware } = typeof resolvedPlacement === 'string' ? {
            placement: resolvedPlacement,
            rtlAware: true
        } : resolvedPlacement;
        const { setReference: setBackendReference, contentProps: allContentProps, setArrow, update } = (0, _backend.useBackend)({
            placement: FixedPlacement,
            enableFlip,
            rtlAware,
            offset,
            boundary,
            boundaryPadding,
            onCalculateLayout
        });
        const { className: contentClassName, ...contentProps } = allContentProps;
        _react.useImperativeHandle(handleRef, ()=>({
                update
            }));
        const referenceWrapperRef = _react.useRef(null);
        const referenceRef = _react.useRef(null);
        const updateReferenceRefFromWrapperRef = _react.useCallback(()=>{
            if (referenceWrapperRef.current == null)
                return;
            let children = referenceWrapperRef.current.children;
            while (children.length === 1 && window.getComputedStyle(children[0]).display === 'contents')
                children = children[0].children;
            _preconditions.Preconditions.checkState(children.length === 1, `ReactNode references must resolve to 1 DOM element, got ${children.length}. Consider using a ReferenceObject instead.`);
            referenceRef.current = children[0];
            setBackendReference(children[0]);
        }, [
            setBackendReference
        ]);
        _react.useLayoutEffect(()=>{
            if ((0, _backend.isReferenceObject)(reference)) {
                referenceRef.current = reference;
                setBackendReference(reference);
            }
        }, [
            reference,
            setBackendReference
        ]);
        _react.useLayoutEffect(()=>{
            if (!(0, _backend.isReferenceObject)(reference))
                updateReferenceRefFromWrapperRef();
        }, [
            reference,
            updateReferenceRefFromWrapperRef
        ]);
        _react.useEffect(()=>{
            return ()=>{
                referenceRef.current = null;
                setBackendReference(null);
            };
        }, [
            setBackendReference
        ]);
        const onOutsideLayerPointerDown = _react.useCallback(({ target })=>{
            var _ref_ownerDocument;
            const ref = referenceRef.current;
            var _ref_ownerDocument_defaultView;
            const Node = ((_ref_ownerDocument_defaultView = ref === null || ref === void 0 ? void 0 : (_ref_ownerDocument = ref.ownerDocument) === null || _ref_ownerDocument === void 0 ? void 0 : _ref_ownerDocument.defaultView) !== null && _ref_ownerDocument_defaultView !== void 0 ? _ref_ownerDocument_defaultView : window).Node;
            const isPointerDownOnReference = target instanceof Node && ref instanceof Node && ref.contains(target);
            onOutsidePointerDown === null || onOutsidePointerDown === void 0 ? void 0 : onOutsidePointerDown({
                target: isPointerDownOnReference ? 'reference' : 'other'
            });
        }, [
            onOutsidePointerDown
        ]);
        const maybeOnOutsidePointerDown = open && onOutsidePointerDown != null ? onOutsideLayerPointerDown : undefined;
        const onBackdropMouseDown = _react.useCallback(()=>{
            if (!onOutsidePointerDown) return;
            window.setTimeout(()=>onOutsidePointerDown({
                    target: 'other'
                }), 0);
        }, [
            onOutsidePointerDown
        ]);
        const attachReferenceWrapperRef = _react.useCallback((ref)=>{
            referenceWrapperRef.current = ref;
            updateReferenceRefFromWrapperRef();
        }, [
            updateReferenceRefFromWrapperRef
        ]);
        return (0, _jsxruntime.jsxs)(_jsxruntime.Fragment, {
            children: [
                reference != null && !(0, _backend.isReferenceObject)(reference) && (0, _jsxruntime.jsx)("div", {
                    ref: attachReferenceWrapperRef,
                    className: _pincss.default.reference,
                    children: reference
                }),
                open && (0, _jsxruntime.jsxs)(_layer.Layer, {
                    open: true,
                    level: level,
                    parentLayer: parentLayer,
                    onOutsideLayerPointerDown: maybeOnOutsidePointerDown,
                    children: [
                        blockOutsidePointerEvents &&
                        (0, _jsxruntime.jsx)("div", {
                            className: _pincss.default.backdrop,
                            onMouseDown: onBackdropMouseDown
                        }),
                        (0, _jsxruntime.jsx)("div", {
                            className: (0, _classnames.default)(contentClassName, {
                                [_pincss.default.blockInsidePointerEvents]: blockInsidePointerEvents
                            }),
                            ...contentProps,
                            children: typeof children === 'function' ? children({
                                setArrow
                            }) : children
                        })
                    ]
                })
            ]
        });
    }
);
function useAutoPlacement(placement) {
    if (placement !== 'auto') return placement;
    return 'bottom-start';
}
