// @canva/app-ui-kit/dist/cjs/ui/base/scrollable/private/base_scrollable.js
"use strict"
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "BaseScrollable", {
    enumerable: true,
    get: function() {
        return BaseScrollable;
    }
});
const _jsxruntime = require("react/jsx-runtime");
const _preconditions = require('../../../../base/preconditions');
const _classnames = _interop_require_default(require("classnames"));
const _react = _interop_require_wildcard(require("react"));
const _provider = require('../../provider/provider');
const _normalized_scroll = require('../../scroll_controls/normalized_scroll');
const _base_scrollablecss = _interop_require_default(require("./base_scrollable.css"));
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function _getRequireWildcardCache(nodeInterop) {
    if (typeof WeakMap !== "function") return null;
    var cacheBabelInterop = new WeakMap();
    var cacheNodeInterop = new WeakMap();
    return (_getRequireWildcardCache = function(nodeInterop) {
        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
    })(nodeInterop);
}
function _interop_require_wildcard(obj, nodeInterop) {
    if (!nodeInterop && obj && obj.__esModule) {
        return obj;
    }
    if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
        return {
            default: obj
        };
    }
    var cache = _getRequireWildcardCache(nodeInterop);
    if (cache && cache.has(obj)) {
        return cache.get(obj);
    }
    var newObj = {
        __proto__: null
    };
    var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
    for(var key in obj){
        if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
            if (desc && (desc.get || desc.set)) {
                Object.defineProperty(newObj, key, desc);
            } else {
                newObj[key] = obj[key];
            }
        }
    }
    newObj.default = obj;
    if (cache) {
        cache.set(obj, newObj);
    }
    return newObj;
}
const ROUNDING_ERROR = 1;
const BaseScrollable = _react.forwardRef((props, forwardedRef)=>{
    const { children, direction, onScroll, className, ariaLabel, role = 'region' } = props;
    const isRtl = (0, _provider.useDirection)() === 'RTL';
    const ref = _react.useRef(null);
    _react.useImperativeHandle(forwardedRef, ()=>{
        const scrollContainer = _preconditions.Preconditions.checkExists(ref.current);
        return {
            getScrollState: ()=>getScrollState(scrollContainer, direction, isRtl),
            scrollBy: (options)=>scrollContainer.scrollBy(options),
            scrollTo: (options)=>{
                const { position, behavior } = options;
                if (!position) {
                    scrollContainer.scrollTo(options);
                    return;
                }
                const state = getScrollState(scrollContainer, direction, isRtl);
                if (state.scrollable) {
                    const { raw } = state;
                    if (direction === 'vertical') {
                        const top = position === 'start' ? 0 : raw.scroll.height - raw.client.height;
                        scrollContainer.scrollTo({
                            top,
                            behavior
                        });
                    } else {
                        const startPosition = isRtl ? raw.scroll.width - raw.client.width : 0;
                        const endPosition = isRtl ? 0 : raw.scroll.width - raw.client.width;
                        const left = position === 'start' ? startPosition : endPosition;
                        scrollContainer.scrollTo({
                            left: (0, _normalized_scroll.getContextualizedScrollLeft)(scrollContainer, left, isRtl),
                            behavior: options === null || options === void 0 ? void 0 : options.behavior
                        });
                    }
                }
            }
        };
    }, [
        direction,
        isRtl
    ]);
    const handleScroll = _react.useCallback(()=>{
        if (onScroll && ref.current) {
            const scrollContainer = ref.current;
            onScroll(()=>getScrollState(scrollContainer, direction, isRtl));
        }
    }, [
        onScroll,
        direction,
        isRtl
    ]);
    return (0, _jsxruntime.jsx)("div", {
        tabIndex: 0,
        "aria-label": ariaLabel,
        role: ariaLabel ? role : undefined,
        className: (0, _classnames.default)(_base_scrollablecss.default.scrollContainer, className, {
            [_base_scrollablecss.default.vertical]: direction === 'vertical',
            [_base_scrollablecss.default.horizontal]: direction === 'horizontal'
        }),
        ref: ref,
        onScroll: handleScroll,
        children: children
    });
});
function getScrollState(scrollContainer, direction, isRtl) {
    const scrollable = direction === 'vertical' ? scrollContainer.scrollHeight > scrollContainer.clientHeight : scrollContainer.scrollWidth > scrollContainer.clientWidth;
    if (!scrollable) return {
        scrollable: false
    };
    const raw = {
        scroll: {
            top: scrollContainer.scrollTop,
            height: scrollContainer.scrollHeight,
            left: (0, _normalized_scroll.getNormalizedScrollLeft)(scrollContainer, isRtl),
            width: scrollContainer.scrollWidth
        },
        client: {
            top: scrollContainer.clientTop,
            height: scrollContainer.clientHeight,
            left: scrollContainer.clientLeft,
            width: scrollContainer.clientWidth
        }
    };
    let atStart, atEnd;
    if (direction === 'vertical') {
        atStart = raw.scroll.top <= 0;
        atEnd = raw.scroll.height - raw.scroll.top - raw.client.height < ROUNDING_ERROR;
    } else {
        const atLeftMost = raw.scroll.left <= 0;
        const atRightMost = raw.scroll.width - raw.scroll.left - raw.client.width < ROUNDING_ERROR;
        atStart = isRtl ? atRightMost : atLeftMost;
        atEnd = isRtl ? atLeftMost : atRightMost;
    }
    return {
        scrollable: true,
        raw,
        atStart,
        atEnd
    };
}
