// @canva/app-ui-kit/dist/cjs/ui/base/scrollable/private/scrollable_with_indicator.js
"use strict"
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "ScrollableWithIndicator", {
    enumerable: true,
    get: function() {
        return ScrollableWithIndicator;
    }
});
const _jsxruntime = require("react/jsx-runtime");
const _composereactrefs = _interop_require_default(require("@seznam/compose-react-refs"));
const _debounce = require('../../../../base/debounce');
const _preconditions = require('../../../../base/preconditions');
const _classnames = _interop_require_default(require("classnames"));
const _react = _interop_require_wildcard(require("react"));
const _reactmeasure = _interop_require_default(require("react-measure"));
const _provider = require('../../provider/provider');
const _base_scrollable = require("./base_scrollable");
const _scrollable_with_indicatorcss = _interop_require_default(require("./scrollable_with_indicator.css"));
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function _getRequireWildcardCache(nodeInterop) {
    if (typeof WeakMap !== "function") return null;
    var cacheBabelInterop = new WeakMap();
    var cacheNodeInterop = new WeakMap();
    return (_getRequireWildcardCache = function(nodeInterop) {
        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
    })(nodeInterop);
}
function _interop_require_wildcard(obj, nodeInterop) {
    if (!nodeInterop && obj && obj.__esModule) {
        return obj;
    }
    if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
        return {
            default: obj
        };
    }
    var cache = _getRequireWildcardCache(nodeInterop);
    if (cache && cache.has(obj)) {
        return cache.get(obj);
    }
    var newObj = {
        __proto__: null
    };
    var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
    for(var key in obj){
        if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
            if (desc && (desc.get || desc.set)) {
                Object.defineProperty(newObj, key, desc);
            } else {
                newObj[key] = obj[key];
            }
        }
    }
    newObj.default = obj;
    if (cache) {
        cache.set(obj, newObj);
    }
    return newObj;
}
const UPDATE_INDICATORS_DEBOUNCE_MS = 16;
const isWebkit = ()=>typeof window !== 'undefined' ? /AppleWebKit\//i.test(window.navigator.userAgent) : undefined;
const runningInWebkit = isWebkit();
const ScrollableWithIndicator = _react.forwardRef((props, forwardedRef)=>{
    const [showStartIndicator, setShowStartIndicator] = _react.useState(false);
    const [showEndIndicator, setShowEndIndicator] = _react.useState(false);
    const [widths, setWidths] = _react.useState(null);
    const scrollableRef = _react.useRef(null);
    const isRtl = (0, _provider.useDirection)() === 'RTL';
    const { children, direction, onScroll, ariaLabel, role, type, background } = props;
    const applyWebkitRtlFix = runningInWebkit && isRtl && direction === 'horizontal';
    const enableAnimations = (0, _provider.useEnableAnimations)();
    const updateIndicators = _react.useMemo(()=>{
        return (0, _debounce.debounce)((getScrollState)=>{
            const scrollState = getScrollState();
            if (!scrollState.scrollable) {
                setShowStartIndicator(false);
                setShowEndIndicator(false);
            } else {
                setShowStartIndicator(!scrollState.atStart);
                setShowEndIndicator(!scrollState.atEnd);
            }
        }, UPDATE_INDICATORS_DEBOUNCE_MS, {
            leading: true
        });
    }, []);
    const updateWidths = _react.useMemo(()=>{
        return (0, _debounce.debounce)((getScrollState)=>{
            const scrollState = getScrollState();
            if (scrollState.scrollable) {
                const { raw } = scrollState;
                setWidths({
                    client: raw.client.width,
                    scroll: raw.scroll.width
                });
            }
        }, UPDATE_INDICATORS_DEBOUNCE_MS, {
            leading: true
        });
    }, []);
    const handleScroll = _react.useCallback((getScrollState)=>{
        updateIndicators(getScrollState);
        onScroll && onScroll(getScrollState);
    }, [
        updateIndicators,
        onScroll
    ]);
    const handleChildrenResize = _react.useCallback(()=>{
        const ref = scrollableRef.current;
        if (!ref) return;
        updateIndicators(ref.getScrollState);
        if (applyWebkitRtlFix)
        updateWidths(ref.getScrollState);
    }, [
        updateIndicators,
        updateWidths,
        applyWebkitRtlFix
    ]);
    const indicatorWrapperSharedStyles = (0, _classnames.default)(_scrollable_with_indicatorcss.default.indicatorWrapper, getDirectionClassName(direction));
    const indicatorSharedStyles = (0, _classnames.default)(_scrollable_with_indicatorcss.default.indicator, getBackgroundClassName(background), getIndicatorTypeClassName(type), getDirectionClassName(direction), {
        [_scrollable_with_indicatorcss.default.animate]: enableAnimations
    });
    return (0, _jsxruntime.jsxs)(_base_scrollable.BaseScrollable, {
        direction: direction,
        onScroll: handleScroll,
        ariaLabel: ariaLabel,
        role: role,
        ref: (0, _composereactrefs.default)(forwardedRef, scrollableRef),
        className: (0, _classnames.default)(_scrollable_with_indicatorcss.default.scrollContainer, getDirectionClassName(direction)),
        children: [
            (0, _jsxruntime.jsx)("div", {
                className: (0, _classnames.default)(indicatorWrapperSharedStyles, _scrollable_with_indicatorcss.default.start),
                style: {
                    right: applyWebkitRtlFix && !!widths ? `${widths.scroll - widths.client}px` : undefined
                },
                children: (0, _jsxruntime.jsx)("div", {
                    className: (0, _classnames.default)(indicatorSharedStyles, _scrollable_with_indicatorcss.default.start, {
                        [_scrollable_with_indicatorcss.default.isVisible]: showStartIndicator
                    })
                })
            }),
            (0, _jsxruntime.jsx)("div", {
                className: (0, _classnames.default)(indicatorWrapperSharedStyles, _scrollable_with_indicatorcss.default.end),
                style: {
                    right: applyWebkitRtlFix && !!widths ? `${widths.scroll}px` : undefined
                },
                children: (0, _jsxruntime.jsx)("div", {
                    className: (0, _classnames.default)(indicatorSharedStyles, _scrollable_with_indicatorcss.default.end, {
                        [_scrollable_with_indicatorcss.default.isVisible]: showEndIndicator
                    })
                })
            }),
            (0, _jsxruntime.jsx)(_reactmeasure.default, {
                onResize: handleChildrenResize,
                children: ({ measureRef })=>(0, _jsxruntime.jsx)("div", {
                        ref: measureRef,
                        className: _scrollable_with_indicatorcss.default.childrenWrapper,
                        children: children
                    })
            })
        ]
    });
});
function getBackgroundClassName(background) {
    switch(background){
        case 'canvas':
            return _scrollable_with_indicatorcss.default.canvas;
        case 'tabdock':
            return _scrollable_with_indicatorcss.default.tabdock;
        case 'surface':
            return _scrollable_with_indicatorcss.default.surface;
        case 'page':
            return _scrollable_with_indicatorcss.default.page;
        default:
            throw new _preconditions.UnreachableError(background);
    }
}
function getIndicatorTypeClassName(type) {
    switch(type){
        case 'shadow':
            return _scrollable_with_indicatorcss.default.shadow;
        case 'fade':
            return _scrollable_with_indicatorcss.default.fade;
        default:
            throw new _preconditions.UnreachableError(type);
    }
}
function getDirectionClassName(direction) {
    switch(direction){
        case 'vertical':
            return _scrollable_with_indicatorcss.default.vertical;
        case 'horizontal':
            return _scrollable_with_indicatorcss.default.horizontal;
        default:
            throw new _preconditions.UnreachableError(direction);
    }
}
