// @canva/app-ui-kit/dist/cjs/ui/base/scrollable/private/scrollable.js
"use strict"
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "Scrollable", {
    enumerable: true,
    get: function() {
        return Scrollable;
    }
});
const _jsxruntime = require("react/jsx-runtime");
const _react = _interop_require_wildcard(require("react"));
const _base_scrollable = require("./base_scrollable");
const _scrollable_with_indicator = require("./scrollable_with_indicator");
function _getRequireWildcardCache(nodeInterop) {
    if (typeof WeakMap !== "function") return null;
    var cacheBabelInterop = new WeakMap();
    var cacheNodeInterop = new WeakMap();
    return (_getRequireWildcardCache = function(nodeInterop) {
        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
    })(nodeInterop);
}
function _interop_require_wildcard(obj, nodeInterop) {
    if (!nodeInterop && obj && obj.__esModule) {
        return obj;
    }
    if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
        return {
            default: obj
        };
    }
    var cache = _getRequireWildcardCache(nodeInterop);
    if (cache && cache.has(obj)) {
        return cache.get(obj);
    }
    var newObj = {
        __proto__: null
    };
    var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
    for(var key in obj){
        if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
            if (desc && (desc.get || desc.set)) {
                Object.defineProperty(newObj, key, desc);
            } else {
                newObj[key] = obj[key];
            }
        }
    }
    newObj.default = obj;
    if (cache) {
        cache.set(obj, newObj);
    }
    return newObj;
}
const Scrollable = _react.forwardRef((props, forwardedRef)=>{
    const { children, direction = 'vertical', onScroll, indicator, ariaLabel, role } = props;
    if (indicator) return (0, _jsxruntime.jsx)(_scrollable_with_indicator.ScrollableWithIndicator, {
        direction: direction,
        onScroll: onScroll,
        ariaLabel: ariaLabel,
        role: role,
        ref: forwardedRef,
        type: indicator.type,
        background: indicator.background ? indicator.background : 'surface',
        children: children
    });
    return (0, _jsxruntime.jsx)(_base_scrollable.BaseScrollable, {
        direction: direction,
        onScroll: onScroll,
        ariaLabel: ariaLabel,
        role: role,
        ref: forwardedRef,
        children: children
    });
});
