// @canva/app-ui-kit/dist/esm/ui/base/form/time_input/private/time_input.js
import { jsx as _jsx } from "react/jsx-runtime";
import classNames from 'classnames';
import * as React from 'react';
import { msToTimeString, timeStringToMs } from '../../../date_time/utils/utils';
import { BaseInput } from '../../base_input/base_input';
import { TimezoneDecorator } from '../../input_decorators/input_decorators';
import styles from './time_input.css';
export const TimeInput = React.forwardRef(
    function TimeInput({ value: valueProp, onChange: onChangeProp, min, max, className, timezone, locale, end = _jsx(TimezoneDecorator, {
        timezone: timezone,
        locale: locale
    }), ...passThroughProps }, ref) {
        const onChange = React.useCallback((text)=>{
            const value = text === '' ? undefined : text;
            onChangeProp === null || onChangeProp === void 0 ? void 0 : onChangeProp(value != null ? timeStringToMs(value) : value);
        }, [
            onChangeProp
        ]);
        return _jsx(BaseInput, {
            ...passThroughProps,
            ref: ref,
            type: "time",
            value: valueProp != null ? msToTimeString(valueProp) : '',
            min: min != null ? msToTimeString(min) : min,
            max: max != null ? msToTimeString(max) : max,
            onChange: onChange,
            end: end,
            className: classNames(styles.timeInput, className),
            placeholder: "HH:mm" 
        });
    }
);
TimeInput.displayName = 'TimeInput';
