// @canva/app-ui-kit/dist/esm/ui/apps/developing/ui_kit/components/app_ui_provider/app_ui_provider.js
var _internalSdkWindow___canva__;
import { jsx as _jsx, jsxs as _jsxs, Fragment as _Fragment } from "react/jsx-runtime";
import * as React from 'react';
import { LATEST_APP_UI_KIT_VERSION } from '../../config/config';
import { EaselProvider } from '../../../../../base/provider/provider';
import { ApplyThemeRoot } from '../../../../../base/theme/theme';
export const toEaselDirection = (direction)=>{
    if (typeof direction === 'string' && direction.toUpperCase() === 'RTL') return 'RTL';
    return 'LTR';
};
const internalSdkWindow = window;
const uiKit = internalSdkWindow === null || internalSdkWindow === void 0 ? void 0 : (_internalSdkWindow___canva__ = internalSdkWindow['__canva__']) === null || _internalSdkWindow___canva__ === void 0 ? void 0 : _internalSdkWindow___canva__['uiKit'];
uiKit === null || uiKit === void 0 ? void 0 : uiKit.registerAppUiKitVersion({
    version: LATEST_APP_UI_KIT_VERSION
}).catch(()=>{});
/** 
 * A wrapper that provides theming and user preferences to App UI Kit components.
 * This component must appear once — and only once — at the root of an app, as the other components depend on it.
 */ export function AppUiProvider({ children }) {
    const [theme, setTheme] = React.useState();
    function setStateFromContext(context) {
        setTheme(context.theme);
    }
    React.useEffect(()=>{
        if (!uiKit) throw new Error("`AppUiProvider` must be used within the context of a Canva app. If you're attempting to use this component in a test environment, please use `TestAppUiProvider` component instead. Use of App UI Kit components outside of a Canva app is forbidden and strictly prohibited as per our License.");
        uiKit.getUiContext().then(setStateFromContext);
        return uiKit.onUiContextChange(setStateFromContext);
    }, []);
    if (!theme)
        return _jsx(_Fragment, {});
    const directionFromDocumentElement = document.documentElement.getAttribute('dir');
    return _jsxs(EaselProvider, {
        enableAnimations: true,
        direction: toEaselDirection(directionFromDocumentElement),
        disableFocusTraps: true,
        enableUserSelection: false,
        disableDialogBlur: false,
        children: [
            _jsx(ApplyThemeRoot, {
                appearance: theme,
                classicLight: theme,
                classicDark: theme,
                light: theme,
                dark: theme
            }),
            children
        ]
    });
}
