// @canva/app-ui-kit/dist/esm/ui/apps/developing/ui_kit/components/app_ui_provider/test_app_ui_provider.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import 'react';
import { EaselProvider } from '../../../../../base/provider/provider';
import { ApplyThemeRoot } from '../../../../../base/theme/theme';
import { toEaselDirection } from './app_ui_provider';
/** 
 * A wrapper that provides theming and user preferences to App UI Kit components in test environments
 * This component must appear once — and only once — at the root of an app, as the other components depend on it.
 *
 * @remarks
 * This component must only be used in test environments. Use {@link AppUiProvider} in production environments.
 */ export function TestAppUiProvider({ children, theme = 'light', direction = 'LTR' }) {
    var _wrappedWindow___canva__;
    const wrappedWindow = window;
    const uiKit = wrappedWindow === null || wrappedWindow === void 0 ? void 0 : (_wrappedWindow___canva__ = wrappedWindow['__canva__']) === null || _wrappedWindow___canva__ === void 0 ? void 0 : _wrappedWindow___canva__['uiKit'];
    if (uiKit != null)
        throw new Error(`'TestAppUiProvider' detected in a non-test environment. Replace 'TestAppUiProvider' usage with 'AppUiProvider' in your app.`);
    return _jsxs(EaselProvider, {
        enableAnimations: true,
        direction: toEaselDirection(direction),
        disableFocusTraps: true,
        enableUserSelection: false,
        disableDialogBlur: false,
        children: [
            _jsx(ApplyThemeRoot, {
                appearance: theme,
                classicLight: theme,
                classicDark: theme,
                light: theme,
                dark: theme
            }),
            children
        ]
    });
}
