// @canva/app-ui-kit/dist/cjs/ui/base/legacy_sheet/internal_sheet.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "InternalSheet", {
    enumerable: true,
    get: function() {
        return InternalSheet;
    }
});
const _jsxruntime = require("react/jsx-runtime");
const _make_observable = require('../../../base/make_observable/make_observable');
const _id_generator = require('../../../base/id_generator/id_generator');
const _dispose_on_unmount = require('../../../base/mobx_react/dispose_on_unmount');
const _observer_for_class = require('../../../base/mobx_react/observer_for_class');
const _classnames = _interop_require_default(require("classnames"));
const _mobx = require("mobx");
const _react = _interop_require_wildcard(require("react"));
const _interaction_manager = require('../a11y/interaction_manager/interaction_manager');
const _device_capabilities = require('../device_capabilities/device_capabilities');
const _layer = require('../layer/layer');
const _mobile_event_handler = require('../mobile_event_handler/mobile_event_handler');
const _theme = require('../theme/theme');
const _layer_view = require("./internal/layer_view");
const _sheetcss = _interop_require_default(require("./internal/sheet.css"));
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function _getRequireWildcardCache(nodeInterop) {
    if (typeof WeakMap !== "function") return null;
    var cacheBabelInterop = new WeakMap();
    var cacheNodeInterop = new WeakMap();
    return (_getRequireWildcardCache = function(nodeInterop) {
        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
    })(nodeInterop);
}
function _interop_require_wildcard(obj, nodeInterop) {
    if (!nodeInterop && obj && obj.__esModule) {
        return obj;
    }
    if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
        return {
            default: obj
        };
    }
    var cache = _getRequireWildcardCache(nodeInterop);
    if (cache && cache.has(obj)) {
        return cache.get(obj);
    }
    var newObj = {
        __proto__: null
    };
    var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
    for(var key in obj){
        if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
            if (desc && (desc.get || desc.set)) {
                Object.defineProperty(newObj, key, desc);
            } else {
                newObj[key] = obj[key];
            }
        }
    }
    newObj.default = obj;
    if (cache) {
        cache.set(obj, newObj);
    }
    return newObj;
}
function applyDecs2203RFactory() {
    function createAddInitializerMethod(initializers, decoratorFinishedRef) {
        return function addInitializer(initializer) {
            assertNotFinished(decoratorFinishedRef, "addInitializer");
            assertCallable(initializer, "An initializer");
            initializers.push(initializer);
        };
    }
    function memberDec(dec, name, desc, initializers, kind, isStatic, isPrivate, metadata, value) {
        var kindStr;
        switch(kind){
            case 1:
                kindStr = "accessor";
                break;
            case 2:
                kindStr = "method";
                break;
            case 3:
                kindStr = "getter";
                break;
            case 4:
                kindStr = "setter";
                break;
            default:
                kindStr = "field";
        }
        var ctx = {
            kind: kindStr,
            name: isPrivate ? "#" + name : name,
            static: isStatic,
            private: isPrivate,
            metadata: metadata
        };
        var decoratorFinishedRef = {
            v: false
        };
        ctx.addInitializer = createAddInitializerMethod(initializers, decoratorFinishedRef);
        var get, set;
        if (kind === 0) {
            if (isPrivate) {
                get = desc.get;
                set = desc.set;
            } else {
                get = function() {
                    return this[name];
                };
                set = function(v) {
                    this[name] = v;
                };
            }
        } else if (kind === 2) {
            get = function() {
                return desc.value;
            };
        } else {
            if (kind === 1 || kind === 3) {
                get = function() {
                    return desc.get.call(this);
                };
            }
            if (kind === 1 || kind === 4) {
                set = function(v) {
                    desc.set.call(this, v);
                };
            }
        }
        ctx.access = get && set ? {
            get: get,
            set: set
        } : get ? {
            get: get
        } : {
            set: set
        };
        try {
            return dec(value, ctx);
        } finally{
            decoratorFinishedRef.v = true;
        }
    }
    function assertNotFinished(decoratorFinishedRef, fnName) {
        if (decoratorFinishedRef.v) {
            throw new Error("attempted to call " + fnName + " after decoration was finished");
        }
    }
    function assertCallable(fn, hint) {
        if (typeof fn !== "function") {
            throw new TypeError(hint + " must be a function");
        }
    }
    function assertValidReturnValue(kind, value) {
        var type = typeof value;
        if (kind === 1) {
            if (type !== "object" || value === null) {
                throw new TypeError("accessor decorators must return an object with get, set, or init properties or void 0");
            }
            if (value.get !== undefined) {
                assertCallable(value.get, "accessor.get");
            }
            if (value.set !== undefined) {
                assertCallable(value.set, "accessor.set");
            }
            if (value.init !== undefined) {
                assertCallable(value.init, "accessor.init");
            }
        } else if (type !== "function") {
            var hint;
            if (kind === 0) {
                hint = "field";
            } else if (kind === 10) {
                hint = "class";
            } else {
                hint = "method";
            }
            throw new TypeError(hint + " decorators must return a function or void 0");
        }
    }
    function applyMemberDec(ret, base, decInfo, name, kind, isStatic, isPrivate, initializers, metadata) {
        var decs = decInfo[0];
        var desc, init, value;
        if (isPrivate) {
            if (kind === 0 || kind === 1) {
                desc = {
                    get: decInfo[3],
                    set: decInfo[4]
                };
            } else if (kind === 3) {
                desc = {
                    get: decInfo[3]
                };
            } else if (kind === 4) {
                desc = {
                    set: decInfo[3]
                };
            } else {
                desc = {
                    value: decInfo[3]
                };
            }
        } else if (kind !== 0) {
            desc = Object.getOwnPropertyDescriptor(base, name);
        }
        if (kind === 1) {
            value = {
                get: desc.get,
                set: desc.set
            };
        } else if (kind === 2) {
            value = desc.value;
        } else if (kind === 3) {
            value = desc.get;
        } else if (kind === 4) {
            value = desc.set;
        }
        var newValue, get, set;
        if (typeof decs === "function") {
            newValue = memberDec(decs, name, desc, initializers, kind, isStatic, isPrivate, metadata, value);
            if (newValue !== void 0) {
                assertValidReturnValue(kind, newValue);
                if (kind === 0) {
                    init = newValue;
                } else if (kind === 1) {
                    init = newValue.init;
                    get = newValue.get || value.get;
                    set = newValue.set || value.set;
                    value = {
                        get: get,
                        set: set
                    };
                } else {
                    value = newValue;
                }
            }
        } else {
            for(var i = decs.length - 1; i >= 0; i--){
                var dec = decs[i];
                newValue = memberDec(dec, name, desc, initializers, kind, isStatic, isPrivate, metadata, value);
                if (newValue !== void 0) {
                    assertValidReturnValue(kind, newValue);
                    var newInit;
                    if (kind === 0) {
                        newInit = newValue;
                    } else if (kind === 1) {
                        newInit = newValue.init;
                        get = newValue.get || value.get;
                        set = newValue.set || value.set;
                        value = {
                            get: get,
                            set: set
                        };
                    } else {
                        value = newValue;
                    }
                    if (newInit !== void 0) {
                        if (init === void 0) {
                            init = newInit;
                        } else if (typeof init === "function") {
                            init = [
                                init,
                                newInit
                            ];
                        } else {
                            init.push(newInit);
                        }
                    }
                }
            }
        }
        if (kind === 0 || kind === 1) {
            if (init === void 0) {
                init = function(instance, init) {
                    return init;
                };
            } else if (typeof init !== "function") {
                var ownInitializers = init;
                init = function(instance, init) {
                    var value = init;
                    for(var i = 0; i < ownInitializers.length; i++){
                        value = ownInitializers[i].call(instance, value);
                    }
                    return value;
                };
            } else {
                var originalInitializer = init;
                init = function(instance, init) {
                    return originalInitializer.call(instance, init);
                };
            }
            ret.push(init);
        }
        if (kind !== 0) {
            if (kind === 1) {
                desc.get = value.get;
                desc.set = value.set;
            } else if (kind === 2) {
                desc.value = value;
            } else if (kind === 3) {
                desc.get = value;
            } else if (kind === 4) {
                desc.set = value;
            }
            if (isPrivate) {
                if (kind === 1) {
                    ret.push(function(instance, args) {
                        return value.get.call(instance, args);
                    });
                    ret.push(function(instance, args) {
                        return value.set.call(instance, args);
                    });
                } else if (kind === 2) {
                    ret.push(value);
                } else {
                    ret.push(function(instance, args) {
                        return value.call(instance, args);
                    });
                }
            } else {
                Object.defineProperty(base, name, desc);
            }
        }
    }
    function applyMemberDecs(Class, decInfos, metadata) {
        var ret = [];
        var protoInitializers;
        var staticInitializers;
        var existingProtoNonFields = new Map();
        var existingStaticNonFields = new Map();
        for(var i = 0; i < decInfos.length; i++){
            var decInfo = decInfos[i];
            if (!Array.isArray(decInfo)) continue;
            var kind = decInfo[1];
            var name = decInfo[2];
            var isPrivate = decInfo.length > 3;
            var isStatic = kind >= 5;
            var base;
            var initializers;
            if (isStatic) {
                base = Class;
                kind = kind - 5;
                staticInitializers = staticInitializers || [];
                initializers = staticInitializers;
            } else {
                base = Class.prototype;
                protoInitializers = protoInitializers || [];
                initializers = protoInitializers;
            }
            if (kind !== 0 && !isPrivate) {
                var existingNonFields = isStatic ? existingStaticNonFields : existingProtoNonFields;
                var existingKind = existingNonFields.get(name) || 0;
                if (existingKind === true || existingKind === 3 && kind !== 4 || existingKind === 4 && kind !== 3) {
                    throw new Error("Attempted to decorate a public method/accessor that has the same name as a previously decorated public method/accessor. This is not currently supported by the decorators plugin. Property name was: " + name);
                } else if (!existingKind && kind > 2) {
                    existingNonFields.set(name, kind);
                } else {
                    existingNonFields.set(name, true);
                }
            }
            applyMemberDec(ret, base, decInfo, name, kind, isStatic, isPrivate, initializers, metadata);
        }
        pushInitializers(ret, protoInitializers);
        pushInitializers(ret, staticInitializers);
        return ret;
    }
    function pushInitializers(ret, initializers) {
        if (initializers) {
            ret.push(function(instance) {
                for(var i = 0; i < initializers.length; i++){
                    initializers[i].call(instance);
                }
                return instance;
            });
        }
    }
    function applyClassDecs(targetClass, classDecs, metadata) {
        if (classDecs.length > 0) {
            var initializers = [];
            var newClass = targetClass;
            var name = targetClass.name;
            for(var i = classDecs.length - 1; i >= 0; i--){
                var decoratorFinishedRef = {
                    v: false
                };
                try {
                    var nextNewClass = classDecs[i](newClass, {
                        kind: "class",
                        name: name,
                        addInitializer: createAddInitializerMethod(initializers, decoratorFinishedRef),
                        metadata
                    });
                } finally{
                    decoratorFinishedRef.v = true;
                }
                if (nextNewClass !== undefined) {
                    assertValidReturnValue(10, nextNewClass);
                    newClass = nextNewClass;
                }
            }
            return [
                defineMetadata(newClass, metadata),
                function() {
                    for(var i = 0; i < initializers.length; i++){
                        initializers[i].call(newClass);
                    }
                }
            ];
        }
    }
    function defineMetadata(Class, metadata) {
        return Object.defineProperty(Class, Symbol.metadata || Symbol.for("Symbol.metadata"), {
            configurable: true,
            enumerable: true,
            value: metadata
        });
    }
    return function applyDecs2203R(targetClass, memberDecs, classDecs, parentClass) {
        if (parentClass !== void 0) {
            var parentMetadata = parentClass[Symbol.metadata || Symbol.for("Symbol.metadata")];
        }
        var metadata = Object.create(parentMetadata === void 0 ? null : parentMetadata);
        var e = applyMemberDecs(targetClass, memberDecs, metadata);
        if (!classDecs.length) defineMetadata(targetClass, metadata);
        return {
            e: e,
            get c () {
                return applyClassDecs(targetClass, classDecs, metadata);
            }
        };
    };
}
function _apply_decs_2203_r(targetClass, memberDecs, classDecs, parentClass) {
    return (_apply_decs_2203_r = applyDecs2203RFactory())(targetClass, memberDecs, classDecs, parentClass);
}
function _identity(x) {
    return x;
}
var _class;
var _initClass, _React_Component;
const defaultPlaceholder = ()=>null;
const InternalSheet = _react.forwardRef((props, ref)=>{
    var _props_overlay;
    const interactionLayerRef = (0, _interaction_manager.useInteractionLayer)({
        markOutsideInert: (_props_overlay = props.overlay) !== null && _props_overlay !== void 0 ? _props_overlay : false
    });
    return (0, _jsxruntime.jsx)(_InternalSheetClass, {
        ...props,
        interactionLayerRef: interactionLayerRef,
        ref: ref
    });
});
let _InternalSheetClass;
new (_class = class extends _identity {
    constructor(){
        super(_InternalSheetClass), _initClass();
    }
}, (()=>{
    class InternalSheetClass extends (_React_Component = _react.Component) {
        static _makeObservable(instance) {
            (0, _make_observable.makeObservable)(instance, {
                contentElement: _mobx.observable.ref,
                outerElement: _mobx.observable.ref,
                parentLayerContainer: _mobx.observable.ref,
                show: _mobx.observable.ref,
                currentContent: _mobx.observable.ref,
                currentOuterContent: _mobx.observable.ref,
                themeData: _mobx.observable.ref,
                themeClass: _mobx.computed,
                window: _mobx.observable.ref,
                supportsSafeAreaInset: _mobx.computed,
                setThemeData: _mobx.action.bound,
                componentDidMount: _mobx.action,
                componentDidUpdate: _mobx.action
            });
        }
        get themeClass() {
            var _this_themeData;
            return (_this_themeData = this.themeData) === null || _this_themeData === void 0 ? void 0 : _this_themeData.className;
        }
        get supportsSafeAreaInset() {
            return this.window ? (0, _device_capabilities.supportsSafeAreaInsetValues)(this.window.navigator) : undefined;
        }
        render() {
            return (0, _jsxruntime.jsxs)(_jsxruntime.Fragment, {
                children: [
                    (0, _jsxruntime.jsx)(_theme.WithThemeData, {
                        children: this.setThemeData
                    }),
                    (0, _jsxruntime.jsx)(_mobile_event_handler.MobileEventHandler, {
                        onBackButton: this.onBackButton,
                        children: (0, _jsxruntime.jsx)(_layer.WithLayerParent, {
                            parentLayer: this.parentLayerContainer,
                            children: this.contentElement && this.outerElement && (0, _jsxruntime.jsxs)(_jsxruntime.Fragment, {
                                children: [
                                    this.renderPortal(this.currentContent, this.contentElement),
                                    this.renderPortal(this.currentOuterContent, this.outerElement)
                                ]
                            })
                        })
                    })
                ]
            });
        }
        setThemeData(data) {
            this.themeData = data;
            return null;
        }
        renderPortal(content, element) {
            if (!content) return null;
            const Content = typeof content === 'function' ? content() : content;
            return this.props.createPortal(Content, element);
        }
        componentDidMount() {
            var _this_props_windowOverride;
            this.window = (_this_props_windowOverride = this.props.windowOverride) !== null && _this_props_windowOverride !== void 0 ? _this_props_windowOverride : window;
            this.outerElement = this.window.document.createElement('div');
            this.contentElement = this.window.document.createElement('div');
            this.outerElement.className = (0, _classnames.default)(_sheetcss.default.sheetElement, _sheetcss.default.scrollable);
            this.contentElement.className = (0, _classnames.default)(_sheetcss.default.sheetElement, {
                [_sheetcss.default.scrollable]: this.props.scrollable
            });
            this.contentElement.style.display = this.props.open ? '' : 'none';
            if (this.currentContent) {
                this.props.layers.addSheet(this);
                this.props.interactionLayerRef(this.contentElement);
            }
            window.clearTimeout(this.closeTimeout);
            (0, _dispose_on_unmount.disposeOnUnmount)(this, [
                (0, _mobx.reaction)(()=>{
                    var _this_themeData;
                    var _this_themeData_classNames;
                    return (_this_themeData_classNames = (_this_themeData = this.themeData) === null || _this_themeData === void 0 ? void 0 : _this_themeData.classNames) !== null && _this_themeData_classNames !== void 0 ? _this_themeData_classNames : [];
                }, (next, previous)=>{
                    if (this.outerElement == null || this.contentElement == null) return;
                    if (previous != null) {
                        this.contentElement.classList.remove(...previous);
                        this.outerElement.classList.remove(...previous);
                    }
                    this.contentElement.classList.add(...next);
                    this.outerElement.classList.add(...next);
                }, {
                    fireImmediately: true
                })
            ]);
        }
        componentWillUnmount() {
            this.props.layers.removeSheet(this);
            this.props.interactionLayerRef(null);
            if (this.props.open && this.props.onCloseAnimationComplete)
            this.closeTimeout = window.setTimeout(this.props.onCloseAnimationComplete, this.enableAnimations ? _layer_view.ANIMATION_TIME : 0);
        }
        componentDidUpdate(prevProps) {
            const { open, content, outerContent, layers, placeholder = defaultPlaceholder, scrollable, onCloseAnimationComplete } = this.props;
            if (this.outerElement == null || this.contentElement == null || this.window == null) return;
            this.contentElement.classList.toggle(_sheetcss.default.scrollable, scrollable);
            if (open && !prevProps.open) {
                this.show = true;
                this.contentElement.style.display = '';
                window.clearTimeout(this.closeTimeout);
                if (!this.currentContent) {
                    this.currentContent = placeholder;
                    this.currentOuterContent = outerContent;
                    window.setTimeout((0, _mobx.action)(()=>{
                        this.currentContent = content;
                    }));
                }
                layers.addSheet(this);
                this.props.interactionLayerRef(this.contentElement);
            } else if (!open && prevProps.open) {
                var _this_sheetDragInternal_sheetBehavior, _this_sheetDragInternal;
                this.show = false;
                (_this_sheetDragInternal = this.sheetDragInternal) === null || _this_sheetDragInternal === void 0 ? void 0 : (_this_sheetDragInternal_sheetBehavior = _this_sheetDragInternal.sheetBehavior) === null || _this_sheetDragInternal_sheetBehavior === void 0 ? void 0 : _this_sheetDragInternal_sheetBehavior.closeSheet();
                this.closeTimeout = window.setTimeout((0, _mobx.action)(()=>{
                    if (this.unmountOnClose) {
                        this.currentContent = undefined;
                        this.currentOuterContent = undefined;
                    }
                    if (this.contentElement != null) this.contentElement.style.display = 'none';
                    onCloseAnimationComplete && onCloseAnimationComplete();
                }), this.enableAnimations ? _layer_view.ANIMATION_TIME : 0);
                if (this.unmountOnClose) {
                    layers.removeSheet(this);
                    this.props.interactionLayerRef(null);
                }
            }
            if (this.contentShouldUpdate()) {
                this.currentContent = content;
                this.currentOuterContent = outerContent;
            }
        }
        contentShouldUpdate() {
            const { placeholder = defaultPlaceholder } = this.props;
            if (placeholder != null && this.currentContent === placeholder)
                return false;
            return this.show;
        }
        get unmountOnClose() {
            return !this.props.mountBehaviour || this.props.mountBehaviour === 'open';
        }
        get applyInsetStyles() {
            const { expandContentIntoUnsafeArea = false, sheetContainsTextInput = false } = this.props;
            return !!this.supportsSafeAreaInset && !expandContentIntoUnsafeArea && !sheetContainsTextInput;
        }
        get from() {
            const { from, easelConfiguration: { direction } } = this.props;
            if (from === 'right' && direction === 'RTL') return 'left';
            else if (from === 'left' && direction === 'RTL') return 'right';
            else return from || 'bottom';
        }
        get onBottom() {
            return this.props.position === 'bottom';
        }
        get roundedCorners() {
            const { roundedCorners = true } = this.props;
            return roundedCorners;
        }
        get autoFocusOnOpen() {
            const { overlay, open, autoFocusOnContent = true } = this.props;
            return !!(overlay && open && autoFocusOnContent);
        }
        get enableAnimations() {
            var _this_props_enableAnimations;
            return ((_this_props_enableAnimations = this.props.enableAnimations) !== null && _this_props_enableAnimations !== void 0 ? _this_props_enableAnimations : true) && this.props.easelConfiguration.enableAnimations;
        }
        get isFullscreen() {
            const { size, minSize } = this.props;
            if (!size) return false;
            const sheetSize = size || minSize;
            return sheetSize === '100%' || sheetSize === '100vh';
        }
        get contentStyles() {
            const { from, size, minSize, maxSize } = this.props;
            const nSize = (minSize != null || maxSize != null) && size == null ? 'inherit' : size;
            const nMaxSize = maxSize || '100%';
            return from === 'left' || from === 'right' ? {
                width: nSize,
                minWidth: minSize,
                maxWidth: nMaxSize
            } : {
                height: nSize,
                minHeight: minSize,
                maxHeight: nMaxSize
            };
        }
        constructor(props){
            super(props), this.id = (_InternalSheetClass._makeObservable(this), _InternalSheetClass.ids.next()), this.closeTimeout = 0, this.onBackButton = (event)=>{
                const { onBackButton, open } = this.props;
                onBackButton && open && onBackButton(event);
            };
            if (props.sheetBehavior) this.sheetDragInternal = {
                sheetBehavior: props.sheetBehavior
            };
            this.easelConfiguration = props.easelConfiguration;
            this.show = !!this.props.open;
            if (this.show) {
                this.currentContent = props.content;
                this.currentOuterContent = props.outerContent;
            }
        }
    }
    ({ c: [_InternalSheetClass, _initClass] } = _apply_decs_2203_r(InternalSheetClass, [], [
        _observer_for_class.observerForClass
    ], _React_Component));
    InternalSheetClass.defaultProps = {
        scrollable: true
    };
    InternalSheetClass.ids = new _id_generator.IdGenerator('sheet__');
})(), _class)();
