// @canva/app-ui-kit/dist/cjs/ui/base/legacy_sheet/legacy_sheet.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: Object.getOwnPropertyDescriptor(all, name).get
    });
}
_export(exports, {
    get LegacyBottomSheet () {
        return LegacyBottomSheet;
    },
    get LegacyDeviceFrameSheet () {
        return LegacyDeviceFrameSheet;
    },
    get LegacySheet () {
        return LegacySheet;
    },
    get createDraggableLegacyBottomSheet () {
        return createDraggableLegacyBottomSheet;
    },
    get getDefaultSheetPeekHeight () {
        return getDefaultSheetPeekHeight;
    }
});
const _jsxruntime = require("react/jsx-runtime");
const _preconditions = require('../../../base/preconditions');
const _react = _interop_require_wildcard(require("react"));
const _reactdom = require("react-dom");
const _gesture_recognizer_view = require('../gestures/gesture_recognizer_view');
const _provider = require('../provider/provider');
const _layers = require("./internal/layers");
const _sheet_behavior = require("./internal/sheet_behavior");
const _internal_sheet = require("./internal_sheet");
function _getRequireWildcardCache(nodeInterop) {
    if (typeof WeakMap !== "function") return null;
    var cacheBabelInterop = new WeakMap();
    var cacheNodeInterop = new WeakMap();
    return (_getRequireWildcardCache = function(nodeInterop) {
        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
    })(nodeInterop);
}
function _interop_require_wildcard(obj, nodeInterop) {
    if (!nodeInterop && obj && obj.__esModule) {
        return obj;
    }
    if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
        return {
            default: obj
        };
    }
    var cache = _getRequireWildcardCache(nodeInterop);
    if (cache && cache.has(obj)) {
        return cache.get(obj);
    }
    var newObj = {
        __proto__: null
    };
    var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
    for(var key in obj){
        if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
            if (desc && (desc.get || desc.set)) {
                Object.defineProperty(newObj, key, desc);
            } else {
                newObj[key] = obj[key];
            }
        }
    }
    newObj.default = obj;
    if (cache) {
        cache.set(obj, newObj);
    }
    return newObj;
}
const getDefaultSheetPeekHeight = (window1)=>Math.round(window1.innerHeight * 0.44);
const DEFAULT_SHEET_PEEK_HEIGHT =
typeof window !== 'undefined' ? getDefaultSheetPeekHeight(window) : undefined;
function useEaselConfiguration() {
    return {
        enableAnimations: (0, _provider.useEnableAnimations)(),
        direction: (0, _provider.useDirection)(),
        disableFocusTraps: (0, _provider.useDisableFocusTraps)(),
        enableUserSelection: (0, _provider.useEnableUserSelection)(),
        theme: (0, _provider.useTheme)(),
        enableHighColorContrast: (0, _provider.useEnableHighColorContrast)(),
        autoplayVideos: (0, _provider.useAutoplayVideos)(),
        disableDialogBlur: (0, _provider.useDisableDialogBlur)()
    };
}
const layers = new _layers.Layers();
const LegacySheet = (props)=>(0, _jsxruntime.jsx)(_internal_sheet.InternalSheet, {
        ...props,
        easelConfiguration: useEaselConfiguration(),
        layers: layers,
        createPortal: _reactdom.createPortal,
        enableLayerDraggable: false
    });
const LegacyBottomSheet = ({ onRequestClose, onCloseAnimationComplete, peekHeightPx, ...restProps })=>{
    const [snapHeights, setSnapHeights] = _react.useState([
        undefined
    ]);
    _react.useEffect(()=>{
        if (peekHeightPx === 'none') {
            setSnapHeights([
                undefined
            ]);
            return;
        }
        setSnapHeights([
            DEFAULT_SHEET_PEEK_HEIGHT,
            undefined
        ]);
    }, [
        peekHeightPx
    ]);
    const [snapIndex, setSnapIndex] = _react.useState(0);
    const handleDragTrigger = _react.useCallback((opts)=>{
        switch(opts.direction){
            case 'opening':
                const isFullyExpanded = snapIndex + 1 === snapHeights.length;
                if (!isFullyExpanded) setSnapIndex(snapIndex + 1);
                break;
            case 'closing':
                const isSmallestHeight = snapIndex === 0;
                if (isSmallestHeight) onRequestClose && onRequestClose({
                    action: 'resize'
                });
                else setSnapIndex(snapIndex - 1);
                break;
            default:
                throw new _preconditions.UnreachableError(opts.direction);
        }
    }, [
        snapIndex,
        snapHeights.length,
        onRequestClose
    ]);
    const handleCloseAnimationComplete = _react.useCallback(()=>{
        setSnapIndex(0);
        onCloseAnimationComplete && onCloseAnimationComplete();
    }, [
        onCloseAnimationComplete
    ]);
    return (0, _jsxruntime.jsx)(_internal_sheet.InternalSheet, {
        ...restProps,
        easelConfiguration: useEaselConfiguration(),
        layers: layers,
        createPortal: _reactdom.createPortal,
        hasShadow: true,
        expandContentIntoUnsafeArea: true,
        from: "bottom",
        maxSize: undefined,
        minSize: undefined,
        scrollable: false,
        enableLayerDraggable: true,
        maxVisibleSizePx: snapHeights[snapIndex],
        onRequestClose: onRequestClose,
        onLayerDraggingThresholdTriggered: handleDragTrigger,
        onCloseAnimationComplete: handleCloseAnimationComplete
    });
};
const LegacyDeviceFrameSheet = (props)=>(0, _jsxruntime.jsx)(_internal_sheet.InternalSheet, {
        ...props,
        windowOverride: props.window,
        easelConfiguration: useEaselConfiguration(),
        layers: layers,
        createPortal: _reactdom.createPortal,
        enableLayerDraggable: false
    });
function createDraggableLegacyBottomSheet({ onStateChanged, onPanning, peekSizePx, transparentOverlayOnPeek, viewportTopOffset, window: window1 }) {
    const sheetBehavior = new _sheet_behavior.BottomSheetDragBehavior(window1, onStateChanged, peekSizePx, transparentOverlayOnPeek, viewportTopOffset, onPanning);
    const recognizers = [
        sheetBehavior.recognizer
    ];
    const Handle = ({ children })=>(0, _jsxruntime.jsx)(_gesture_recognizer_view.GestureRecognizerView, {
            recognizers: recognizers,
            children: ({ gestureRef })=>(0, _jsxruntime.jsx)("div", {
                    ref: gestureRef,
                    children: children
                })
        });
    const Sheet = (props)=>(0, _jsxruntime.jsx)(_internal_sheet.InternalSheet, {
            ...props,
            easelConfiguration: useEaselConfiguration(),
            from: "bottom",
            layers: layers,
            windowOverride: window1,
            createPortal: _reactdom.createPortal,
            sheetBehavior: sheetBehavior,
            enableLayerDraggable: false
        });
    const sheetController = sheetBehavior;
    return {
        Handle,
        Sheet,
        sheetController,
        isSwipingSheet: ()=>sheetBehavior.isSwiping
    };
}
