// @canva/app-ui-kit/dist/esm/ui/base/alert/private/alert.js
import { jsx as _jsx, jsxs as _jsxs, Fragment as _Fragment } from "react/jsx-runtime";
import { UnreachableError } from '../../../../base/preconditions';
import classNames from 'classnames';
import 'react';
import { Box } from '../../box/box';
import { Button } from '../../button/button';
import { useFocusableRef } from '../../focusable/focusable';
import { AlertTriangleIcon } from '../../icons/alert_triangle/icon';
import { CheckIcon } from '../../icons/check/icon';
import { InfoIcon } from '../../icons/info/icon';
import { XIcon } from '../../icons/x/icon';
import { Column, Columns } from '../../layout/layout';
import { customProperties as linkCustomProperties } from '../../link/internal/link.css';
import themeStyles from '../../theme/theme.css';
import { InheritColor, Text } from '../../typography/typography';
import styles from './alert.css';
import { AlertMessages } from './alert.messages';
export const alertTones = [
    'positive',
    'warn',
    'critical',
    'info',
    'neutral'
];
const typographyTones = {
    positive: themeStyles.colorPositiveForeLow,
    warn: themeStyles.colorWarnForeLow,
    critical: themeStyles.colorCriticalForeLow,
    info: themeStyles.colorInfoForeLow,
    neutral: themeStyles.colorNeutralForeLow
};
export const AlertIcon = ({ tone, size = 'medium' })=>{
    switch(tone){
        case 'critical':
        case 'warn':
            return _jsx(AlertTriangleIcon, {
                size: size
            });
        case 'positive':
            return _jsx(CheckIcon, {
                size: size
            });
        case 'info':
            return _jsx(InfoIcon, {
                size: size
            });
        case 'neutral':
            return null;
        default:
            throw new UnreachableError(tone);
    }
};
export const Alert = ({ children, className, title, tone, dismissible, onDismiss, size = 'medium', rounded = true, showIcon = size !== 'small', Icon, focusRef })=>{
    let iconElement;
    if (showIcon) {
        if (Icon) iconElement = _jsx(Icon, {
            size: size
        });
        else iconElement = _jsx(AlertIcon, {
            tone: tone,
            size: size
        });
    }
    const containerRef = useFocusableRef(focusRef);
    return _jsx(Box, {
        background: `${tone}Low`,
        borderRadius: rounded ? 'element' : 'none',
        paddingX: size === 'medium' ? '2u' : '1.5u',
        paddingY: size === 'medium' ? '2u' : '1u',
        width: "full",
        className: classNames(className, styles.alert, {
            [styles.small]: size === 'small'
        }),
        customProperties: {
            [linkCustomProperties.linkColor]: typographyTones[tone]
        },
        role: tone === 'critical' ? 'alert' : undefined,
        ref: containerRef,
        tabIndex: focusRef != null ? -1 : undefined,
        children: _jsx(Text, {
            tagName: "div",
            size: size,
            tone: InheritColor,
            children: _jsxs(Columns, {
                spacing: "1u",
                alignY: "baseline",
                children: [
                    iconElement && _jsx(Column, {
                        width: "content",
                        children: iconElement
                    }),
                    _jsxs(Column, {
                        children: [
                            title && _jsxs(_Fragment, {
                                children: [
                                    _jsx("strong", {
                                        children: title
                                    }),
                                    ' '
                                ]
                            }),
                            children
                        ]
                    }),
                    dismissible && _jsx(Column, {
                        width: "content",
                        children: _jsx(Button, {
                            variant: "tertiary",
                            className: styles.close,
                            onClick: onDismiss,
                            icon: XIcon,
                            ariaLabel: AlertMessages.dismiss(),
                            size: "small",
                            iconSize: size
                        })
                    })
                ]
            })
        })
    });
};
