// @canva/app-ui-kit/dist/cjs/ui/base/tooltip/internal/internal_tooltip.js
"use strict"
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: Object.getOwnPropertyDescriptor(all, name).get
    });
}
_export(exports, {
    get InternalTooltip () {
        return InternalTooltip;
    },
    get StatelessTooltip () {
        return StatelessTooltip;
    },
    get TooltipArrow () {
        return TooltipArrow;
    },
    get TooltipBox () {
        return TooltipBox;
    },
    get TooltipContent () {
        return TooltipContent;
    }
});
const _jsxruntime = require("react/jsx-runtime");
const _preconditions = require('../../../../base/preconditions');
const _classnames = _interop_require_default(require("classnames"));
const _react = _interop_require_wildcard(require("react"));
const _Transition = _interop_require_wildcard(require("react-transition-group/Transition"));
const _access_mode_state = require('../../a11y/access_mode_state/access_mode_state');
const _identifier = require('../../a11y/identifier/identifier');
const _pointer_input_state = require('../../a11y/pointer_input_state/pointer_input_state');
const _screen_reader_content = require('../../a11y/screen_reader_content/screen_reader_content');
const _box = require('../../box/box');
const _layer = require('../../layer/layer');
const _layout = require('../../layout/layout');
const _metrics = require('../../metrics/metrics');
const _responsive_image = require('../../responsive_image/responsive_image');
const _shortcut = require('../../shortcut/shortcut');
const _pin = require('../../surface/pin/pin');
const _theme = require('../../theme/theme');
const _typography = require('../../typography/typography');
const _internal_tooltipcss = _interop_require_default(require("./internal_tooltip.css"));
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function _getRequireWildcardCache(nodeInterop) {
    if (typeof WeakMap !== "function") return null;
    var cacheBabelInterop = new WeakMap();
    var cacheNodeInterop = new WeakMap();
    return (_getRequireWildcardCache = function(nodeInterop) {
        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
    })(nodeInterop);
}
function _interop_require_wildcard(obj, nodeInterop) {
    if (!nodeInterop && obj && obj.__esModule) {
        return obj;
    }
    if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
        return {
            default: obj
        };
    }
    var cache = _getRequireWildcardCache(nodeInterop);
    if (cache && cache.has(obj)) {
        return cache.get(obj);
    }
    var newObj = {
        __proto__: null
    };
    var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
    for(var key in obj){
        if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
            if (desc && (desc.get || desc.set)) {
                Object.defineProperty(newObj, key, desc);
            } else {
                newObj[key] = obj[key];
            }
        }
    }
    newObj.default = obj;
    if (cache) {
        cache.set(obj, newObj);
    }
    return newObj;
}
function isRenderTrigger(children) {
    return typeof children === 'function';
}
const MOTION_DURATION_MS = 150;
const FADE_IN_DELAY = 150;
const FADE_OUT_DELAY = 0;
function InternalTooltip(props) {
    const { label, description, disabled = false, mode } = props;
    const timer = _react.useRef(undefined);
    const [open, setOpen] = _react.useState(false);
    const openRef = _react.useRef(false);
    const withDelay = _react.useCallback((cb, delay)=>{
        window.clearTimeout(timer.current);
        if (!delay) return Promise.resolve(cb());
        return new Promise((resolve)=>{
            timer.current = window.setTimeout(()=>{
                cb();
                resolve();
            }, delay);
        });
    }, []);
    const showTooltip = _react.useCallback((delay = FADE_IN_DELAY)=>{
        return withDelay(()=>setOpen(true), delay);
    }, [
        withDelay
    ]);
    const hideTooltip = _react.useCallback(()=>{
        return withDelay(()=>setOpen(false), FADE_OUT_DELAY);
    }, [
        withDelay
    ]);
    const onDocumentKeydown = _react.useCallback((e)=>{
        if (e.key === 'Escape' && openRef.current) {
            e.stopPropagation();
            hideTooltip();
        }
    }, [
        hideTooltip
    ]);
    _react.useEffect(()=>{
        document.addEventListener('keydown', onDocumentKeydown, {
            capture: true
        });
        return ()=>{
            document.removeEventListener('keydown', onDocumentKeydown, {
                capture: true
            });
        };
    }, [
        onDocumentKeydown
    ]);
    _react.useEffect(()=>{
        openRef.current = open;
    }, [
        open
    ]);
    _react.useEffect(()=>{
        if (disabled) hideTooltip();
    }, [
        disabled,
        hideTooltip
    ]);
    const screenReaderLabel = description != null ? `${label}: ${description}` : label;
    if (mode === 'toggletip') return (0, _jsxruntime.jsx)(BaseToggleTip, {
        ...props,
        hideTooltip: hideTooltip,
        open: open,
        screenReaderLabel: screenReaderLabel,
        showTooltip: showTooltip,
        disabled: disabled
    });
    return (0, _jsxruntime.jsx)(BaseTooltip, {
        ...props,
        hideTooltip: hideTooltip,
        open: open,
        screenReaderLabel: screenReaderLabel,
        showTooltip: showTooltip,
        disabled: disabled
    });
}
function BaseTooltip({ children, closeOnClick = true, disabled, forwardedRef, hideTooltip, open, screenReaderLabel, showTooltip, ...props }) {
    const [tooltipId] = (0, _identifier.useIdentifiers)(1);
    const isMouseInput = (0, _pointer_input_state.useIsMouseInput)();
    const isMouseAccessMode = (0, _access_mode_state.useIsMouseAccessMode)();
    const showTooltipIfMouseInput = _react.useCallback(()=>{
        if (disabled || !isMouseInput) return;
        showTooltip();
    }, [
        disabled,
        showTooltip,
        isMouseInput
    ]);
    const showTooltipIfKeyboardMode = _react.useCallback(()=>{
        if (disabled || isMouseAccessMode) return;
        showTooltip(0);
    }, [
        disabled,
        showTooltip,
        isMouseAccessMode
    ]);
    const renderedChildren = _react.useMemo(()=>{
        if (isRenderTrigger(children)) return children({
            onBlur: hideTooltip,
            onFocus: showTooltipIfKeyboardMode,
            onMouseDown: closeOnClick ? hideTooltip : undefined,
            onMouseEnter: showTooltipIfMouseInput,
            onMouseLeave: hideTooltip,
            tabIndex: 0,
            tooltipId
        });
    }, [
        children,
        closeOnClick,
        hideTooltip,
        showTooltipIfKeyboardMode,
        showTooltipIfMouseInput,
        tooltipId
    ]);
    if (isRenderTrigger(children)) return ((0, _jsxruntime.jsxs)(_jsxruntime.Fragment, {
            children: [
                (0, _jsxruntime.jsx)(StatelessTooltip, {
                    ...props,
                    open: open,
                    children: renderedChildren
                }),
                (0, _jsxruntime.jsx)(_screen_reader_content.ScreenReaderContent, {
                    asLabelTextOnly: true,
                    id: tooltipId,
                    children: screenReaderLabel
                })
            ]
        }));
    return closeOnClick ?
    (0, _jsxruntime.jsx)("div", {
        onMouseEnter: showTooltipIfMouseInput,
        onMouseLeave: hideTooltip,
        onMouseDown: hideTooltip,
        ref: forwardedRef,
        className: props.expandTooltipWrapper ? _internal_tooltipcss.default.expand : undefined,
        children: (0, _jsxruntime.jsx)(StatelessTooltip, {
            ...props,
            children: children,
            open: open
        })
    }) : (0, _jsxruntime.jsx)("div", {
        onMouseEnter: showTooltipIfMouseInput,
        onMouseLeave: hideTooltip,
        ref: forwardedRef,
        className: props.expandTooltipWrapper ? _internal_tooltipcss.default.expand : undefined,
        children: (0, _jsxruntime.jsx)(StatelessTooltip, {
            ...props,
            children: children,
            open: open
        })
    });
}
function BaseToggleTip({ open, hideTooltip, showTooltip, children, screenReaderLabel, ...props }) {
    const [renderScreenReaderLabel, setRenderScreenReaderLabel] = _react.useState(false);
    const showToggletip = _react.useCallback(async (e)=>{
        e.currentTarget.focus({
            preventScroll: true
        });
        setRenderScreenReaderLabel(false);
        await showTooltip(0);
        setTimeout(()=>setRenderScreenReaderLabel(true), 50);
    }, [
        showTooltip,
        setRenderScreenReaderLabel
    ]);
    const renderedChildren = _react.useMemo(()=>{
        return children({
            onClick: showToggletip,
            onBlur: hideTooltip
        });
    }, [
        children,
        hideTooltip,
        showToggletip
    ]);
    return ((0, _jsxruntime.jsxs)(_jsxruntime.Fragment, {
            children: [
                (0, _jsxruntime.jsx)(StatelessTooltip, {
                    ...props,
                    open: open,
                    children: renderedChildren
                }),
                (0, _jsxruntime.jsx)(_screen_reader_content.ScreenReaderContent, {
                    role: "status",
                    ariaRelevant: "all",
                    children: renderScreenReaderLabel ? screenReaderLabel : ''
                })
            ]
        }));
}
function TooltipContent(props) {
    const { id, label, description, align = 'center', shortcut, lineClamp, thumbnail, mode } = props;
    const shortcutElement = shortcut ? (0, _jsxruntime.jsx)(_shortcut.Shortcut, {
        className: _internal_tooltipcss.default.shortcut,
        combination: shortcut,
        size: "small"
    }) : null;
    const lines = lineClamp ? [
        label
    ] : label.trim().split('\n');
    const labelContent = lines.length === 1 ? (0, _jsxruntime.jsx)(_typography.Text, {
        size: "small",
        weight: "bold",
        tagName: "div",
        lineClamp: lineClamp,
        tone: "primary",
        className: _internal_tooltipcss.default.label,
        alignment: align,
        id: id,
        children: label
    }) : (0, _jsxruntime.jsx)("div", {
        className: _internal_tooltipcss.default.label,
        id: id,
        children: lines.map((line, i)=>(0, _jsxruntime.jsx)(_typography.Text, {
                size: "small",
                alignment: align,
                tone: "primary",
                children: line
            }, i))
    });
    const descriptionContent = description != null && (0, _jsxruntime.jsx)(_typography.Text, {
        size: "small",
        tone: "secondary",
        alignment: align,
        className: _internal_tooltipcss.default.description,
        children: description
    });
    const thumbnailContent = thumbnail != null && (0, _jsxruntime.jsx)("div", {
        className: _internal_tooltipcss.default.thumbnail,
        children: typeof thumbnail === 'string' ? (0, _jsxruntime.jsx)("img", {
            src: thumbnail,
            alt: ""
        }) : (0, _jsxruntime.jsx)(_responsive_image.PixelDensityResponsiveImage, {
            sources: thumbnail,
            alt: ""
        })
    });
    const ariaProps = mode === 'toggletip' ? {} : {
        ['role']: 'tooltip'
    };
    return (0, _jsxruntime.jsx)(_theme.ThemeBoundary, {
        light: "dark",
        dark: "dark",
        classicLight: "dark",
        classicDark: "dark",
        children: ({ className })=>(0, _jsxruntime.jsx)(TooltipBox, {
                className: className,
                display: "flex",
                flexDirection: "row",
                ...ariaProps,
                children: (0, _jsxruntime.jsxs)(_layout.Rows, {
                    spacing: "0",
                    align: align,
                    children: [
                        thumbnailContent,
                        labelContent,
                        descriptionContent,
                        shortcutElement
                    ]
                })
            })
    });
}
function TooltipBox({ className, children, ...props }) {
    return (0, _jsxruntime.jsx)(_box.Box, {
        className: (0, _classnames.default)(className, _internal_tooltipcss.default.tooltip),
        background: "surface",
        borderRadius: "element",
        paddingX: "1u",
        paddingY: "0.5u",
        ...props,
        children: children
    });
}
function TooltipArrow({ inset = true }) {
    const paths = (0, _jsxruntime.jsxs)(_jsxruntime.Fragment, {
        children: [
            (0, _jsxruntime.jsx)("path", {
                d: "M15 16L10.3972 11.67335C9.04995 10.40695 6.95006 10.40695 5.60282 11.67335L1 16L15 16Z",
                className: _internal_tooltipcss.default.arrowInner
            }),
            (0, _jsxruntime.jsx)("path", {
                d: "M15 16L10.3972 11.67335C9.04995 10.40695 6.95006 10.40695 5.60282 11.67335L1 16L2.46004 16L6.28773 12.40198C7.25004 11.4974 8.74996 11.4974 9.71228 12.40198L13.54 16L15 16Z",
                className: _internal_tooltipcss.default.arrowOutline
            })
        ]
    });
    return (0, _jsxruntime.jsx)("svg", {
        width: "16",
        height: "16",
        viewBox: "0 0 16 16",
        fill: "none",
        xmlns: "http://www.w3.org/2000/svg",
        className: _internal_tooltipcss.default.arrow,
        children: inset ? (0, _jsxruntime.jsx)("g", {
            transform: "translate(0, -6)",
            children: paths
        }) : paths
    });
}
function StatelessTooltip({ children, placement = 'bottom', open, arrow = true, refObj, direction, ...contentProps }) {
    return (0, _jsxruntime.jsx)(TransitionPin, {
        open: open,
        blockInsidePointerEvents: true,
        reference: refObj !== null && refObj !== void 0 ? refObj : children,
        placement: getPinPlacement(placement, direction),
        offset: {
            main: _metrics.baseUnit,
            cross: 0
        },
        level: _layer.LayerLevel.TOOLTIPS,
        children: ({ style, childRef, setArrow })=>(0, _jsxruntime.jsxs)("div", {
                ref: childRef,
                style: style,
                children: [
                    (0, _jsxruntime.jsx)(TooltipContent, {
                        ...contentProps
                    }),
                    arrow ? (0, _jsxruntime.jsx)("div", {
                        ref: setArrow,
                        className: _internal_tooltipcss.default.arrowWrapper,
                        children: (0, _jsxruntime.jsx)(TooltipArrow, {
                            inset: false
                        })
                    }) : null
                ]
            })
    });
}
const PIN_PLACEMENT_MAP = {
    ['left-start']: {
        ltr: 'start-top',
        rtl: 'end-top'
    },
    ['left']: {
        ltr: 'start-center',
        rtl: 'end-center'
    },
    ['left-end']: {
        ltr: 'start-bottom',
        rtl: 'end-bottom'
    },
    ['right-start']: {
        ltr: 'end-top',
        rtl: 'start-top'
    },
    ['right']: {
        ltr: 'end-center',
        rtl: 'start-center'
    },
    ['right-end']: {
        ltr: 'end-bottom',
        rtl: 'start-bottom'
    },
    ['top']: {
        ltr: 'top-center',
        rtl: 'top-center'
    },
    ['bottom']: {
        ltr: 'bottom-center',
        rtl: 'bottom-center'
    }
};
function getPinPlacement(placement, direction) {
    const mapping = PIN_PLACEMENT_MAP[placement];
    switch(direction){
        case undefined:
            return mapping.ltr;
        case 'LTR':
            return {
                placement: mapping.ltr,
                rtlAware: false
            };
        case 'RTL':
            return {
                placement: mapping.rtl,
                rtlAware: false
            };
        default:
            throw new _preconditions.UnreachableError(direction);
    }
}
function TransitionPin({ open, children, ...props }) {
    const [transitionOpen, setTransitionOpen] = _react.useState(open);
    const childRef = _react.useRef(null);
    return (0, _jsxruntime.jsx)(_pin.Pin, {
        ...props,
        open: open || transitionOpen,
        children: ({ setArrow })=>(0, _jsxruntime.jsx)(_Transition.default, {
                in: open,
                appear: true,
                timeout: MOTION_DURATION_MS,
                onEnter: ()=>setTransitionOpen(true),
                onExited: ()=>setTransitionOpen(false),
                nodeRef: childRef,
                children: (state)=>children({
                        style: {
                            transition: `opacity ${MOTION_DURATION_MS}ms ease-in-out`,
                            opacity: state === _Transition.ENTERING || state === _Transition.ENTERED ? 1 : 0
                        },
                        childRef,
                        setArrow
                    })
            })
    });
}
