// @canva/app-ui-kit/dist/esm/ui/base/color_picker/alpha_slider/private/alpha_slider.js
import { jsx as _jsx } from "react/jsx-runtime";
import classNames from 'classnames';
import * as React from 'react';
import { BaseSlider } from '../../../form/slider/slider';
import styles from './alpha_slider.css';
import { AlphaSliderMessages } from './alpha_slider.messages';
const handle = ({ focused, trackHovering })=>_jsx("div", {
        className: classNames(styles.handle, {
            [styles.handleHoverOrFocus]: focused || trackHovering
        })
    });
export const AlphaSlider = ({ alpha, color, onChange: onChangeProp, onChangeStart, onChangeComplete, ariaLabel = AlphaSliderMessages.colorAlphaLabel() })=>{
    const track = React.useCallback(()=>_jsx("div", {
            className: styles.track,
            children: _jsx("div", {
                className: styles.overlay,
                style: {
                    background: `linear-gradient(to right, transparent, ${color})`
                }
            })
        }), [
        color
    ]);
    const onChange = React.useCallback((value)=>{
        onChangeProp === null || onChangeProp === void 0 ? void 0 : onChangeProp(value);
    }, [
        onChangeProp
    ]);
    return _jsx("div", {
        className: styles.container,
        children: _jsx("div", {
            className: styles.widener,
            children: _jsx(BaseSlider, {
                value: alpha,
                min: 0,
                max: 1,
                step: 0.005,
                onChange: onChange,
                onDragStart: onChangeStart,
                onDragEnd: onChangeComplete,
                ariaLabel: ariaLabel,
                fillColor: "transparent",
                handle: handle,
                track: track
            })
        })
    });
};
