// @canva/app-ui-kit/dist/cjs/ui/base/gestures/pinch_recognizer.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "PinchRecognizer", {
    enumerable: true,
    get: function() {
        return PinchRecognizer;
    }
});
const _pan_recognizer = require("./pan_recognizer");
class PinchRecognizer {
    getBrowserHandledTouchActions() {
        return [
            'pan-x',
            'pan-y'
        ];
    }
    onPointerDown(pointer, allPointers) {
        if (pointer.start.button != null && pointer.start.button !== 0)
            return;
        this.checkPointerState(pointer, allPointers);
    }
    onPointerMove(pointers, allPointers) {
        var _this_opts_onMove, _this_opts;
        if (!this.state) return;
        if (!this.state.recognized) {
            var _this_opts_onStart, _this_opts1;
            this.state.recognized = true;
            (_this_opts_onStart = (_this_opts1 = this.opts).onStart) === null || _this_opts_onStart === void 0 ? void 0 : _this_opts_onStart.call(_this_opts1, getPinchEventProperties(this.state, pointers[0].current, allPointers));
        } else
            (_this_opts_onMove = (_this_opts = this.opts).onMove) === null || _this_opts_onMove === void 0 ? void 0 : _this_opts_onMove.call(_this_opts, getPinchEventProperties(this.state, pointers[0].current, allPointers));
    }
    onPointerUp(pointer, allPointers) {
        this.checkPointerState(pointer, allPointers);
    }
    onUnmount(allPointers) {
        var _this_state, _this_opts_onEnd, _this_opts;
        if ((_this_state = this.state) === null || _this_state === void 0 ? void 0 : _this_state.recognized) (_this_opts_onEnd = (_this_opts = this.opts).onEnd) === null || _this_opts_onEnd === void 0 ? void 0 : _this_opts_onEnd.call(_this_opts, getPinchEventProperties(this.state, this.state.p1Start, allPointers));
    }
    checkPointerState(pointer, allPointers) {
        if (!this.state && allPointers.size === 2) {
            const [p1Start, p2Start] = Array.from(allPointers.values(), (p)=>p.current);
            this.state = {
                startTime: Date.now(),
                recognized: false,
                p1Start,
                p2Start
            };
        } else if (this.state) {
            var _this_opts_onEnd, _this_opts;
            const state = this.state;
            this.state = undefined;
            if (state.recognized) (_this_opts_onEnd = (_this_opts = this.opts).onEnd) === null || _this_opts_onEnd === void 0 ? void 0 : _this_opts_onEnd.call(_this_opts, getPinchEventProperties(state, pointer.current, allPointers));
        }
    }
    constructor(opts){
        this.opts = opts;
        this.shouldDisableTextSelect = true;
    }
}
function getPinchEventProperties(state, currentPointer, allPointers) {
    const { p1Start, p2Start } = state;
    const p1Current = currentPointer.id === p1Start.id ? currentPointer : allPointers.get(p1Start.id).current;
    const p2Current = currentPointer.id === p2Start.id ? currentPointer : allPointers.get(p2Start.id).current;
    const start = midpoint(p1Start, p2Start);
    const current = midpoint(p1Current, p2Current);
    const startDis = dist2(p1Start, p2Start);
    const currentDis = dist2(p1Current, p2Current);
    const scale = Math.sqrt(currentDis / startDis);
    return {
        ...(0, _pan_recognizer.getPanEventProperties)(currentPointer, start, current),
        scale
    };
}
const midpoint = (p1, p2)=>({
        x: (p1.x + p2.x) / 2,
        y: (p1.y + p2.y) / 2,
        timestamp: Math.max(p1.timestamp, p2.timestamp)
    });
const dist2 = (p1, p2)=>(p2.x - p1.x) ** 2 + (p2.y - p1.y) ** 2;
