// @canva/app-ui-kit/dist/cjs/ui/base/gestures/press_recognizer.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: Object.getOwnPropertyDescriptor(all, name).get
    });
}
_export(exports, {
    get DEFAULT_POINTER_DOWN_TIMEOUT () {
        return DEFAULT_POINTER_DOWN_TIMEOUT;
    },
    get PressRecognizer () {
        return PressRecognizer;
    }
});
const _input = require('../dom/input/input');
const _recognizer = require("./recognizer");
const DEFAULT_POINTER_DOWN_TIMEOUT = 251;
class PressRecognizer {
    getBrowserHandledTouchActions() {
        return [
            'auto'
        ];
    }
    onPointerDown(pointer, allPointers) {
        if (pointer.start.button != null && pointer.start.button !== 0)
            return;
        if (allPointers.size === 1) {
            this.setupEventListeners(pointer.start);
            var _this_opts_minTime;
            const timeoutId = window.setTimeout(this.timeout, (_this_opts_minTime = this.opts.minTime) !== null && _this_opts_minTime !== void 0 ? _this_opts_minTime : DEFAULT_POINTER_DOWN_TIMEOUT);
            this.state = {
                pointer,
                timeoutId,
                recognized: false
            };
        } else if (allPointers.size > 1 && this.state) {
            window.clearTimeout(this.state.timeoutId);
            this.state = undefined;
        }
    }
    onPointerMove(pointers, allPointers) {
        if (this.state && !this.isWithinMaxDistance()) {
            var _this_opts_onCancel, _this_opts;
            const state = this.state;
            this.state = undefined;
            window.clearTimeout(state.timeoutId);
            this.clearEventListeners(state.pointer.start);
            state.recognized && ((_this_opts_onCancel = (_this_opts = this.opts).onCancel) === null || _this_opts_onCancel === void 0 ? void 0 : _this_opts_onCancel.call(_this_opts, (0, _recognizer.getCommonEventProperties)(state.pointer.current)));
        }
    }
    onPointerUp(pointer) {
        if (this.state) {
            var _this_opts_onEnd, _this_opts, _this_opts_onCancel, _this_opts1;
            const state = this.state;
            const withinMaxDistance = this.isWithinMaxDistance();
            this.state = undefined;
            window.clearTimeout(state.timeoutId);
            this.clearEventListeners(state.pointer.start);
            if (state.recognized) withinMaxDistance ? (_this_opts_onEnd = (_this_opts = this.opts).onEnd) === null || _this_opts_onEnd === void 0 ? void 0 : _this_opts_onEnd.call(_this_opts, (0, _recognizer.getCommonEventProperties)(pointer.current)) : (_this_opts_onCancel = (_this_opts1 = this.opts).onCancel) === null || _this_opts_onCancel === void 0 ? void 0 : _this_opts_onCancel.call(_this_opts1, (0, _recognizer.getCommonEventProperties)(pointer.current));
        }
    }
    onUnmount() {
        var _this_state, _this_opts_onCancel, _this_opts;
        if ((_this_state = this.state) === null || _this_state === void 0 ? void 0 : _this_state.recognized) (_this_opts_onCancel = (_this_opts = this.opts).onCancel) === null || _this_opts_onCancel === void 0 ? void 0 : _this_opts_onCancel.call(_this_opts, (0, _recognizer.getCommonEventProperties)(this.state.pointer.current));
    }
    isWithinMaxDistance() {
        if (!this.state) return false;
        const { start, current } = this.state.pointer;
        const distanceSqrd = (current.x - start.x) ** 2 + (current.y - start.y) ** 2;
        var _this_opts_maxDistance;
        return distanceSqrd <= ((_this_opts_maxDistance = this.opts.maxDistance) !== null && _this_opts_maxDistance !== void 0 ? _this_opts_maxDistance : 9) ** 2;
    }
    constructor(opts){
        this.opts = opts;
        this.timeout = ()=>{
            if (this.state) {
                var _this_opts_onStart, _this_opts;
                this.state.recognized = true;
                (_this_opts_onStart = (_this_opts = this.opts).onStart) === null || _this_opts_onStart === void 0 ? void 0 : _this_opts_onStart.call(_this_opts, (0, _recognizer.getCommonEventProperties)(this.state.pointer.start));
            }
        };
        this.setupEventListeners = (pointer)=>{
            pointer.currentTarget.addEventListener('contextmenu', this.onContextMenu);
        };
        this.clearEventListeners = (pointer)=>{
            pointer.currentTarget.removeEventListener('contextmenu', this.onContextMenu);
        };
        this.onContextMenu = (e)=>{
            if ((0, _input.isSomeInputActive)(e))
                return;
            e.preventDefault();
        };
    }
}
