// @canva/app-ui-kit/dist/cjs/ui/base/tabs/private/behavior.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: Object.getOwnPropertyDescriptor(all, name).get
    });
}
_export(exports, {
    get TabsContext () {
        return TabsContext;
    },
    get TabsElementsProvider () {
        return TabsElementsProvider;
    },
    get useRegisterTabsElements () {
        return useRegisterTabsElements;
    },
    get useTab () {
        return useTab;
    },
    get useTabList () {
        return useTabList;
    },
    get useTabPanel () {
        return useTabPanel;
    },
    get useTabs () {
        return useTabs;
    },
    get useTabsContext () {
        return useTabsContext;
    },
    get useTabsElements () {
        return useTabsElements;
    }
});
const _jsxruntime = require("react/jsx-runtime");
const _react = require("react");
const _controllable_value = require('../../controllable_value/controllable_value');
const _stable_function = require('../../stable_function/stable_function');
const _cache = require("./cache");
const _focusable_children = require("./focusable_children");
const _generate_id = require("./generate_id");
const useTab = ({ id, active: activeProp, onClick: onClickProp, ariaControls: ariaControlsProp })=>{
    const context = useTabsContext();
    const isControlled = !context || activeProp != null;
    const active = isControlled ? activeProp : context.activeId === id;
    const registerTabId = context === null || context === void 0 ? void 0 : context.registerTabId;
    const unregisterTabId = context === null || context === void 0 ? void 0 : context.unregisterTabId;
    const setActiveId = context === null || context === void 0 ? void 0 : context.setActiveId;
    const role = 'tab';
    const htmlId = context ? (0, _generate_id.generateId)(context.tabsId, id, role) : id;
    const tabIndex = active ? undefined : -1;
    const ariaControls = context && context.tabPanelIds.includes(id) ? (0, _generate_id.generateId)(context.tabsId, id, 'tabpanel') : ariaControlsProp;
    (0, _react.useEffect)(()=>{
        if (isControlled) return;
        registerTabId === null || registerTabId === void 0 ? void 0 : registerTabId(id);
        return ()=>{
            unregisterTabId === null || unregisterTabId === void 0 ? void 0 : unregisterTabId(id);
        };
    }, [
        isControlled,
        registerTabId,
        unregisterTabId,
        id
    ]);
    const onClick = (0, _react.useCallback)(()=>{
        if (active) return;
        onClickProp === null || onClickProp === void 0 ? void 0 : onClickProp(id);
        if (!isControlled) setActiveId === null || setActiveId === void 0 ? void 0 : setActiveId(id);
    }, [
        isControlled,
        active,
        setActiveId,
        id,
        onClickProp
    ]);
    const onFocus = (0, _react.useCallback)(()=>{
        onClick();
    }, [
        onClick
    ]);
    return {
        active,
        id: htmlId,
        role,
        ariaControls,
        tabIndex,
        onClick,
        onFocus
    };
};
const useTabList = ()=>{
    const { ref } = (0, _focusable_children.useKeyboardFocusNavigation)();
    return {
        ref: ref,
        role: 'tablist'
    };
};
const useTabPanel = ({ id, active: activeProp, ariaLabelledBy: ariaLabelledByProp, children: childrenProp, contentRenderStrategy = 'lazy' })=>{
    const context = useTabsContext();
    const { ref, focusableChildren } = (0, _focusable_children.useFocusableChildren)({
        onlyTabbables: true
    });
    const isControlled = !context || activeProp != null;
    const active = isControlled ? activeProp : context.activeId === id;
    const registerTabPanelId = context === null || context === void 0 ? void 0 : context.registerTabPanelId;
    const unregisterTabPanelId = context === null || context === void 0 ? void 0 : context.unregisterTabPanelId;
    const role = 'tabpanel';
    const htmlId = context ? (0, _generate_id.generateId)(context.tabsId, id, role) : id;
    const ariaLabelledBy = context && context.tabIds.includes(id) ? (0, _generate_id.generateId)(context.tabsId, id, 'tab') : ariaLabelledByProp;
    (0, _react.useEffect)(()=>{
        if (isControlled) return;
        registerTabPanelId === null || registerTabPanelId === void 0 ? void 0 : registerTabPanelId(id);
        return ()=>{
            unregisterTabPanelId === null || unregisterTabPanelId === void 0 ? void 0 : unregisterTabPanelId(id);
        };
    }, [
        isControlled,
        registerTabPanelId,
        unregisterTabPanelId,
        id
    ]);
    let children = childrenProp;
    const { setItem, getItem } = (0, _cache.useCache)();
    if (contentRenderStrategy === 'lazy') {
        if (active) setItem(id, children);
        children = getItem(id);
    } else if (contentRenderStrategy === 'always-remount' && !active) children = null;
    return {
        ref: ref,
        active,
        id: htmlId,
        role,
        ariaLabelledBy,
        tabIndex: active ? focusableChildren.length > 0 ? undefined : 0 : -1,
        children
    };
};
const TabsContext = (0, _react.createContext)(null);
function useTabsContext() {
    return (0, _react.useContext)(TabsContext);
}
const useTabs = ({ activeId: activeIdProp, defaultActiveId, onSelect })=>{
    const tabsId = (0, _react.useId)();
    const [activeId, setActiveId] = (0, _controllable_value.useControllableValue)({
        value: activeIdProp,
        defaultValue: defaultActiveId
    });
    const [tabIds, setTabIds] = (0, _react.useState)([]);
    const [tabPanelIds, setTabPanelIds] = (0, _react.useState)([]);
    const registerTabId = (0, _stable_function.useStableFunction)((id)=>{
        setTabIds((prevIds)=>[
                ...prevIds,
                id
            ]);
    });
    const unregisterTabId = (0, _stable_function.useStableFunction)((id)=>{
        setTabIds((prevIds)=>prevIds.filter((i)=>i !== id));
    });
    const registerTabPanelId = (0, _stable_function.useStableFunction)((id)=>{
        setTabPanelIds((prevIds)=>[
                ...prevIds,
                id
            ]);
    });
    const unregisterTabPanelId = (0, _stable_function.useStableFunction)((id)=>{
        setTabPanelIds((prevIds)=>prevIds.filter((i)=>i !== id));
    });
    (0, _react.useEffect)(()=>{
        if (!activeId || tabIds.length > 0 && !tabIds.includes(activeId))
            setActiveId(tabIds[0]);
    }, [
        activeId,
        setActiveId,
        tabIds
    ]);
    const prevActiveId = (0, _react.useRef)(activeId);
    (0, _react.useEffect)(()=>{
        if (activeId == null || activeId === prevActiveId.current) return;
        onSelect === null || onSelect === void 0 ? void 0 : onSelect(activeId, prevActiveId.current);
        prevActiveId.current = activeId;
    }, [
        activeId,
        onSelect
    ]);
    return {
        tabsId,
        activeId,
        setActiveId,
        tabIds,
        tabPanelIds,
        registerTabId,
        unregisterTabId,
        registerTabPanelId,
        unregisterTabPanelId
    };
};
const TabsElementsContext = (0, _react.createContext)(null);
const TabsElementsProvider = ({ children })=>{
    const [activeElementRef, setActiveElementRef] = (0, _react.useState)(null);
    const [elementRefs, setElementRefs] = (0, _react.useState)([]);
    const [isAnimating, setIsAnimating] = (0, _react.useState)(false);
    const registerElementRef = (0, _stable_function.useStableFunction)((element)=>{
        setElementRefs((prevRefs)=>[
                ...prevRefs,
                element
            ].sort(byDocumentPosition));
    });
    const unregisterElementRef = (0, _stable_function.useStableFunction)((element)=>{
        setElementRefs((prevRefs)=>prevRefs.filter((i)=>i !== element));
    });
    return (0, _jsxruntime.jsx)(TabsElementsContext.Provider, {
        value: {
            elementRefs,
            registerElementRef,
            unregisterElementRef,
            activeElementRef,
            setActiveElementRef,
            isAnimating,
            setIsAnimating
        },
        children: children
    });
};
function byDocumentPosition(a, b) {
    var _a_compareDocumentPosition;
    return ((_a_compareDocumentPosition = a.compareDocumentPosition) === null || _a_compareDocumentPosition === void 0 ? void 0 : _a_compareDocumentPosition.call(a, b)) & Node.DOCUMENT_POSITION_PRECEDING ? 1 : -1;
}
function useRegisterTabsElements(active) {
    const ref = (0, _react.useRef)(null);
    const context = (0, _react.useContext)(TabsElementsContext);
    const registerElementRef = context === null || context === void 0 ? void 0 : context.registerElementRef;
    const unregisterElementRef = context === null || context === void 0 ? void 0 : context.unregisterElementRef;
    (0, _react.useLayoutEffect)(()=>{
        const element = ref.current;
        if (!element) return;
        registerElementRef === null || registerElementRef === void 0 ? void 0 : registerElementRef(element);
        return ()=>{
            unregisterElementRef === null || unregisterElementRef === void 0 ? void 0 : unregisterElementRef(element);
        };
    }, [
        registerElementRef,
        unregisterElementRef
    ]);
    (0, _react.useLayoutEffect)(()=>{
        if (context && active && ref.current && context.activeElementRef !== ref.current) context.setActiveElementRef(ref.current);
    }, [
        context,
        active
    ]);
    return {
        ref
    };
}
function useTabsElements() {
    const context = (0, _react.useContext)(TabsElementsContext);
    var _context_activeElementRef, _context_elementRefs;
    return {
        activeElementRef: (_context_activeElementRef = context === null || context === void 0 ? void 0 : context.activeElementRef) !== null && _context_activeElementRef !== void 0 ? _context_activeElementRef : null,
        setActiveElementRef: context === null || context === void 0 ? void 0 : context.setActiveElementRef,
        elementRefs: (_context_elementRefs = context === null || context === void 0 ? void 0 : context.elementRefs) !== null && _context_elementRefs !== void 0 ? _context_elementRefs : [],
        registerElementRef: context === null || context === void 0 ? void 0 : context.registerElementRef,
        unregisterElementRef: context === null || context === void 0 ? void 0 : context.registerElementRef,
        isAnimating: context === null || context === void 0 ? void 0 : context.isAnimating,
        setIsAnimating: context === null || context === void 0 ? void 0 : context.setIsAnimating
    };
}
