// @canva/app-ui-kit/dist/cjs/ui/base/tabs/private/tab_panel.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "TabPanel", {
    enumerable: true,
    get: function() {
        return TabPanel;
    }
});
const _jsxruntime = require("react/jsx-runtime");
const _composereactrefs = _interop_require_default(require("@seznam/compose-react-refs"));
const _classnames = _interop_require_default(require("classnames"));
const _react = require("react");
const _reacttransitiongroup = require("react-transition-group");
const _provider = require('../../provider/provider');
const _behavior = require("./behavior");
const _tab_panelcss = _interop_require_default(require("./tab_panel.css"));
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
const ANIMATION_DURATION = parseInt(_tab_panelcss.default.animationDuration, 10) || 0;
const TabPanel = ({ children: childrenProp, id: idProp, active: activeProp, animate: animateProp = true, ariaLabelledBy: ariaLabelledByProp, contentRenderStrategy = 'lazy' })=>{
    const { active, ref, id, role, ariaLabelledBy, tabIndex, children } = (0, _behavior.useTabPanel)({
        id: idProp,
        active: activeProp,
        ariaLabelledBy: ariaLabelledByProp,
        children: childrenProp,
        contentRenderStrategy
    });
    const { ref: collectionRef } = (0, _behavior.useRegisterTabsElements)(active);
    const { activeElementRef, elementRefs, setIsAnimating } = (0, _behavior.useTabsElements)();
    const animate = (0, _provider.useEnableAnimations)() && activeElementRef && animateProp;
    const elementIndex = ref.current ? elementRefs.indexOf(ref.current) : -1;
    const activeElementIndex = activeElementRef ? elementRefs.indexOf(activeElementRef) : -1;
    const isAfterActive = elementIndex > activeElementIndex;
    const isBeforeActive = elementIndex < activeElementIndex;
    const onAnimationEnter = (0, _react.useCallback)(()=>{
        if (!setIsAnimating) return;
        setIsAnimating(true);
    }, [
        setIsAnimating
    ]);
    const onAnimationEntered = (0, _react.useCallback)(()=>{
        if (!setIsAnimating) return;
        setIsAnimating(false);
    }, [
        setIsAnimating
    ]);
    return (0, _jsxruntime.jsx)(_reacttransitiongroup.CSSTransition, {
        in: active && !!animate,
        timeout: ANIMATION_DURATION,
        classNames: {
            enterActive: _tab_panelcss.default.animating
        },
        nodeRef: ref,
        onEnter: onAnimationEnter,
        onEntered: onAnimationEntered,
        children: (0, _jsxruntime.jsx)("div", {
            ref: (0, _composereactrefs.default)(ref, collectionRef),
            id: id,
            role: role,
            "aria-labelledby": ariaLabelledBy,
            tabIndex: tabIndex,
            className: (0, _classnames.default)(_tab_panelcss.default.panel, active && _tab_panelcss.default.active, animate && _tab_panelcss.default.animate, animate && isAfterActive && _tab_panelcss.default.afterActive, animate && isBeforeActive && _tab_panelcss.default.beforeActive),
            children: children
        })
    });
};
