// @canva/app-ui-kit/dist/esm/ui/base/surface/pin/private/pin.js
import { jsx as _jsx, jsxs as _jsxs, Fragment as _Fragment } from "react/jsx-runtime";
import { Preconditions } from '../../../../../base/preconditions';
import classNames from 'classnames';
import * as React from 'react';
import { Layer, LayerLevel } from '../../../layer/layer';
import { isReferenceObject, useBackend } from './backend';
import styles from './pin.css';
export const Pin = React.forwardRef(
    function Pin({ open, onOutsidePointerDown, blockOutsidePointerEvents = false, blockInsidePointerEvents = false, onCalculateLayout, reference, placement = 'auto', enableFlip = true, offset = {
        main: 0,
        cross: 0
    }, boundary = 'viewport', boundaryPadding = '1u', level = LayerLevel.PINS, parentLayer, children }, handleRef) {
        const resolvedPlacement = useAutoPlacement(placement);
        const { placement: FixedPlacement, rtlAware } = typeof resolvedPlacement === 'string' ? {
            placement: resolvedPlacement,
            rtlAware: true
        } : resolvedPlacement;
        const { setReference: setBackendReference, contentProps: allContentProps, setArrow, update } = useBackend({
            placement: FixedPlacement,
            enableFlip,
            rtlAware,
            offset,
            boundary,
            boundaryPadding,
            onCalculateLayout
        });
        const { className: contentClassName, ...contentProps } = allContentProps;
        React.useImperativeHandle(handleRef, ()=>({
                update
            }));
        const referenceWrapperRef = React.useRef(null);
        const referenceRef = React.useRef(null);
        const updateReferenceRefFromWrapperRef = React.useCallback(()=>{
            if (referenceWrapperRef.current == null)
                return;
            let children = referenceWrapperRef.current.children;
            while (children.length === 1 && window.getComputedStyle(children[0]).display === 'contents')
                children = children[0].children;
            Preconditions.checkState(children.length === 1, `ReactNode references must resolve to 1 DOM element, got ${children.length}. Consider using a ReferenceObject instead.`);
            referenceRef.current = children[0];
            setBackendReference(children[0]);
        }, [
            setBackendReference
        ]);
        React.useLayoutEffect(()=>{
            if (isReferenceObject(reference)) {
                referenceRef.current = reference;
                setBackendReference(reference);
            }
        }, [
            reference,
            setBackendReference
        ]);
        React.useLayoutEffect(()=>{
            if (!isReferenceObject(reference))
                updateReferenceRefFromWrapperRef();
        }, [
            reference,
            updateReferenceRefFromWrapperRef
        ]);
        React.useEffect(()=>{
            return ()=>{
                referenceRef.current = null;
                setBackendReference(null);
            };
        }, [
            setBackendReference
        ]);
        const onOutsideLayerPointerDown = React.useCallback(({ target })=>{
            var _ref_ownerDocument;
            const ref = referenceRef.current;
            var _ref_ownerDocument_defaultView;
            const Node = ((_ref_ownerDocument_defaultView = ref === null || ref === void 0 ? void 0 : (_ref_ownerDocument = ref.ownerDocument) === null || _ref_ownerDocument === void 0 ? void 0 : _ref_ownerDocument.defaultView) !== null && _ref_ownerDocument_defaultView !== void 0 ? _ref_ownerDocument_defaultView : window).Node;
            const isPointerDownOnReference = target instanceof Node && ref instanceof Node && ref.contains(target);
            onOutsidePointerDown === null || onOutsidePointerDown === void 0 ? void 0 : onOutsidePointerDown({
                target: isPointerDownOnReference ? 'reference' : 'other'
            });
        }, [
            onOutsidePointerDown
        ]);
        const maybeOnOutsidePointerDown = open && onOutsidePointerDown != null ? onOutsideLayerPointerDown : undefined;
        const onBackdropMouseDown = React.useCallback(()=>{
            if (!onOutsidePointerDown) return;
            window.setTimeout(()=>onOutsidePointerDown({
                    target: 'other'
                }), 0);
        }, [
            onOutsidePointerDown
        ]);
        const attachReferenceWrapperRef = React.useCallback((ref)=>{
            referenceWrapperRef.current = ref;
            updateReferenceRefFromWrapperRef();
        }, [
            updateReferenceRefFromWrapperRef
        ]);
        return _jsxs(_Fragment, {
            children: [
                reference != null && !isReferenceObject(reference) && _jsx("div", {
                    ref: attachReferenceWrapperRef,
                    className: styles.reference,
                    children: reference
                }),
                open && _jsxs(Layer, {
                    open: true,
                    level: level,
                    parentLayer: parentLayer,
                    onOutsideLayerPointerDown: maybeOnOutsidePointerDown,
                    children: [
                        blockOutsidePointerEvents &&
                        _jsx("div", {
                            className: styles.backdrop,
                            onMouseDown: onBackdropMouseDown
                        }),
                        _jsx("div", {
                            className: classNames(contentClassName, {
                                [styles.blockInsidePointerEvents]: blockInsidePointerEvents
                            }),
                            ...contentProps,
                            children: typeof children === 'function' ? children({
                                setArrow
                            }) : children
                        })
                    ]
                })
            ]
        });
    }
);
function useAutoPlacement(placement) {
    if (placement !== 'auto') return placement;
    return 'bottom-start';
}
