// @canva/app-ui-kit/dist/esm/ui/base/tabs/private/behavior.js
import { jsx as _jsx } from "react/jsx-runtime";
import { createContext, useCallback, useContext, useEffect, useId, useLayoutEffect, useRef, useState } from 'react';
import 'react';
import { useControllableValue } from '../../controllable_value/controllable_value';
import { useStableFunction } from '../../stable_function/stable_function';
import { useCache } from './cache';
import { useFocusableChildren, useKeyboardFocusNavigation } from './focusable_children';
import { generateId } from './generate_id';
export const useTab = ({ id, active: activeProp, onClick: onClickProp, ariaControls: ariaControlsProp })=>{
    const context = useTabsContext();
    const isControlled = !context || activeProp != null;
    const active = isControlled ? activeProp : context.activeId === id;
    const registerTabId = context === null || context === void 0 ? void 0 : context.registerTabId;
    const unregisterTabId = context === null || context === void 0 ? void 0 : context.unregisterTabId;
    const setActiveId = context === null || context === void 0 ? void 0 : context.setActiveId;
    const role = 'tab';
    const htmlId = context ? generateId(context.tabsId, id, role) : id;
    const tabIndex = active ? undefined : -1;
    const ariaControls = context && context.tabPanelIds.includes(id) ? generateId(context.tabsId, id, 'tabpanel') : ariaControlsProp;
    useEffect(()=>{
        if (isControlled) return;
        registerTabId === null || registerTabId === void 0 ? void 0 : registerTabId(id);
        return ()=>{
            unregisterTabId === null || unregisterTabId === void 0 ? void 0 : unregisterTabId(id);
        };
    }, [
        isControlled,
        registerTabId,
        unregisterTabId,
        id
    ]);
    const onClick = useCallback(()=>{
        if (active) return;
        onClickProp === null || onClickProp === void 0 ? void 0 : onClickProp(id);
        if (!isControlled) setActiveId === null || setActiveId === void 0 ? void 0 : setActiveId(id);
    }, [
        isControlled,
        active,
        setActiveId,
        id,
        onClickProp
    ]);
    const onFocus = useCallback(()=>{
        onClick();
    }, [
        onClick
    ]);
    return {
        active,
        id: htmlId,
        role,
        ariaControls,
        tabIndex,
        onClick,
        onFocus
    };
};
export const useTabList = ()=>{
    const { ref } = useKeyboardFocusNavigation();
    return {
        ref: ref,
        role: 'tablist'
    };
};
export const useTabPanel = ({ id, active: activeProp, ariaLabelledBy: ariaLabelledByProp, children: childrenProp, contentRenderStrategy = 'lazy' })=>{
    const context = useTabsContext();
    const { ref, focusableChildren } = useFocusableChildren({
        onlyTabbables: true
    });
    const isControlled = !context || activeProp != null;
    const active = isControlled ? activeProp : context.activeId === id;
    const registerTabPanelId = context === null || context === void 0 ? void 0 : context.registerTabPanelId;
    const unregisterTabPanelId = context === null || context === void 0 ? void 0 : context.unregisterTabPanelId;
    const role = 'tabpanel';
    const htmlId = context ? generateId(context.tabsId, id, role) : id;
    const ariaLabelledBy = context && context.tabIds.includes(id) ? generateId(context.tabsId, id, 'tab') : ariaLabelledByProp;
    useEffect(()=>{
        if (isControlled) return;
        registerTabPanelId === null || registerTabPanelId === void 0 ? void 0 : registerTabPanelId(id);
        return ()=>{
            unregisterTabPanelId === null || unregisterTabPanelId === void 0 ? void 0 : unregisterTabPanelId(id);
        };
    }, [
        isControlled,
        registerTabPanelId,
        unregisterTabPanelId,
        id
    ]);
    let children = childrenProp;
    const { setItem, getItem } = useCache();
    if (contentRenderStrategy === 'lazy') {
        if (active) setItem(id, children);
        children = getItem(id);
    } else if (contentRenderStrategy === 'always-remount' && !active) children = null;
    return {
        ref: ref,
        active,
        id: htmlId,
        role,
        ariaLabelledBy,
        tabIndex: active ? focusableChildren.length > 0 ? undefined : 0 : -1,
        children
    };
};
export const TabsContext = createContext(null);
export function useTabsContext() {
    return useContext(TabsContext);
}
export const useTabs = ({ activeId: activeIdProp, defaultActiveId, onSelect })=>{
    const tabsId = useId();
    const [activeId, setActiveId] = useControllableValue({
        value: activeIdProp,
        defaultValue: defaultActiveId
    });
    const [tabIds, setTabIds] = useState([]);
    const [tabPanelIds, setTabPanelIds] = useState([]);
    const registerTabId = useStableFunction((id)=>{
        setTabIds((prevIds)=>[
                ...prevIds,
                id
            ]);
    });
    const unregisterTabId = useStableFunction((id)=>{
        setTabIds((prevIds)=>prevIds.filter((i)=>i !== id));
    });
    const registerTabPanelId = useStableFunction((id)=>{
        setTabPanelIds((prevIds)=>[
                ...prevIds,
                id
            ]);
    });
    const unregisterTabPanelId = useStableFunction((id)=>{
        setTabPanelIds((prevIds)=>prevIds.filter((i)=>i !== id));
    });
    useEffect(()=>{
        if (!activeId || tabIds.length > 0 && !tabIds.includes(activeId))
            setActiveId(tabIds[0]);
    }, [
        activeId,
        setActiveId,
        tabIds
    ]);
    const prevActiveId = useRef(activeId);
    useEffect(()=>{
        if (activeId == null || activeId === prevActiveId.current) return;
        onSelect === null || onSelect === void 0 ? void 0 : onSelect(activeId, prevActiveId.current);
        prevActiveId.current = activeId;
    }, [
        activeId,
        onSelect
    ]);
    return {
        tabsId,
        activeId,
        setActiveId,
        tabIds,
        tabPanelIds,
        registerTabId,
        unregisterTabId,
        registerTabPanelId,
        unregisterTabPanelId
    };
};
const TabsElementsContext = createContext(null);
export const TabsElementsProvider = ({ children })=>{
    const [activeElementRef, setActiveElementRef] = useState(null);
    const [elementRefs, setElementRefs] = useState([]);
    const [isAnimating, setIsAnimating] = useState(false);
    const registerElementRef = useStableFunction((element)=>{
        setElementRefs((prevRefs)=>[
                ...prevRefs,
                element
            ].sort(byDocumentPosition));
    });
    const unregisterElementRef = useStableFunction((element)=>{
        setElementRefs((prevRefs)=>prevRefs.filter((i)=>i !== element));
    });
    return _jsx(TabsElementsContext.Provider, {
        value: {
            elementRefs,
            registerElementRef,
            unregisterElementRef,
            activeElementRef,
            setActiveElementRef,
            isAnimating,
            setIsAnimating
        },
        children: children
    });
};
function byDocumentPosition(a, b) {
    var _a_compareDocumentPosition;
    return ((_a_compareDocumentPosition = a.compareDocumentPosition) === null || _a_compareDocumentPosition === void 0 ? void 0 : _a_compareDocumentPosition.call(a, b)) & Node.DOCUMENT_POSITION_PRECEDING ? 1 : -1;
}
export function useRegisterTabsElements(active) {
    const ref = useRef(null);
    const context = useContext(TabsElementsContext);
    const registerElementRef = context === null || context === void 0 ? void 0 : context.registerElementRef;
    const unregisterElementRef = context === null || context === void 0 ? void 0 : context.unregisterElementRef;
    useLayoutEffect(()=>{
        const element = ref.current;
        if (!element) return;
        registerElementRef === null || registerElementRef === void 0 ? void 0 : registerElementRef(element);
        return ()=>{
            unregisterElementRef === null || unregisterElementRef === void 0 ? void 0 : unregisterElementRef(element);
        };
    }, [
        registerElementRef,
        unregisterElementRef
    ]);
    useLayoutEffect(()=>{
        if (context && active && ref.current && context.activeElementRef !== ref.current) context.setActiveElementRef(ref.current);
    }, [
        context,
        active
    ]);
    return {
        ref
    };
}
export function useTabsElements() {
    const context = useContext(TabsElementsContext);
    var _context_activeElementRef, _context_elementRefs;
    return {
        activeElementRef: (_context_activeElementRef = context === null || context === void 0 ? void 0 : context.activeElementRef) !== null && _context_activeElementRef !== void 0 ? _context_activeElementRef : null,
        setActiveElementRef: context === null || context === void 0 ? void 0 : context.setActiveElementRef,
        elementRefs: (_context_elementRefs = context === null || context === void 0 ? void 0 : context.elementRefs) !== null && _context_elementRefs !== void 0 ? _context_elementRefs : [],
        registerElementRef: context === null || context === void 0 ? void 0 : context.registerElementRef,
        unregisterElementRef: context === null || context === void 0 ? void 0 : context.registerElementRef,
        isAnimating: context === null || context === void 0 ? void 0 : context.isAnimating,
        setIsAnimating: context === null || context === void 0 ? void 0 : context.setIsAnimating
    };
}
