// @canva/app-ui-kit/dist/esm/ui/base/tabs/private/tab_panel.js
import { jsx as _jsx } from "react/jsx-runtime";
import composeRefs from '@seznam/compose-react-refs';
import classNames from 'classnames';
import 'react';
import { useCallback } from 'react';
import { CSSTransition } from 'react-transition-group';
import { useEnableAnimations } from '../../provider/provider';
import { useRegisterTabsElements, useTabPanel, useTabsElements } from './behavior';
import styles from './tab_panel.css';
const ANIMATION_DURATION = parseInt(styles.animationDuration, 10) || 0;
export const TabPanel = ({ children: childrenProp, id: idProp, active: activeProp, animate: animateProp = true, ariaLabelledBy: ariaLabelledByProp, contentRenderStrategy = 'lazy' })=>{
    const { active, ref, id, role, ariaLabelledBy, tabIndex, children } = useTabPanel({
        id: idProp,
        active: activeProp,
        ariaLabelledBy: ariaLabelledByProp,
        children: childrenProp,
        contentRenderStrategy
    });
    const { ref: collectionRef } = useRegisterTabsElements(active);
    const { activeElementRef, elementRefs, setIsAnimating } = useTabsElements();
    const animate = useEnableAnimations() && activeElementRef && animateProp;
    const elementIndex = ref.current ? elementRefs.indexOf(ref.current) : -1;
    const activeElementIndex = activeElementRef ? elementRefs.indexOf(activeElementRef) : -1;
    const isAfterActive = elementIndex > activeElementIndex;
    const isBeforeActive = elementIndex < activeElementIndex;
    const onAnimationEnter = useCallback(()=>{
        if (!setIsAnimating) return;
        setIsAnimating(true);
    }, [
        setIsAnimating
    ]);
    const onAnimationEntered = useCallback(()=>{
        if (!setIsAnimating) return;
        setIsAnimating(false);
    }, [
        setIsAnimating
    ]);
    return _jsx(CSSTransition, {
        in: active && !!animate,
        timeout: ANIMATION_DURATION,
        classNames: {
            enterActive: styles.animating
        },
        nodeRef: ref,
        onEnter: onAnimationEnter,
        onEntered: onAnimationEntered,
        children: _jsx("div", {
            ref: composeRefs(ref, collectionRef),
            id: id,
            role: role,
            "aria-labelledby": ariaLabelledBy,
            tabIndex: tabIndex,
            className: classNames(styles.panel, active && styles.active, animate && styles.animate, animate && isAfterActive && styles.afterActive, animate && isBeforeActive && styles.beforeActive),
            children: children
        })
    });
};
