// @canva/app-ui-kit/dist/esm/ui/base/surface/popover/private/arrow.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import * as React from 'react';
import styles from './arrow.css';
export const Arrow = React.forwardRef((_props, ref)=>_jsx("div", {
        ref: ref,
        className: styles.wrapper,
        children: _jsxs("svg", {
            width: "44",
            height: "45",
            viewBox: "-6 -27 44 45",
            fill: "none",
            xmlns: "http://www.w3.org/2000/svg",
            className: styles.arrow,
            children: [
                _jsx("g", {
                    filter: "url(#filter0_f_3010_39642)",
                    children: _jsx("rect", {
                        x: "24",
                        y: "12",
                        width: "16",
                        height: "4",
                        transform: "rotate(-180 24 12)",
                        className: styles.shadow
                    })
                }),
                _jsx("path", {
                    d: "M23.8892 15.3409C24.9794 16.4032 26.4582 17 28 17V18H4V17C5.54184 17 7.02055 16.4032 8.11078 15.3409L11.8892 11.6592C14.1595 9.44694 17.8405 9.44694 20.1108 11.6592L23.8892 15.3409Z",
                    className: styles.content
                }),
                _jsx("path", {
                    d: "M28 16C26.7151 16 25.4879 15.5024 24.5871 14.6246L20.8087 10.943C18.1501 8.35235 13.8499 8.35235 11.1913 10.943L7.4129 14.6246C6.51208 15.5024 5.28487 16 4 16H7.29842C7.58721 15.8061 7.85933 15.5859 8.11078 15.3409L11.8892 11.6592C14.1595 9.44694 17.8405 9.44694 20.1108 11.6592L23.8892 15.3409C24.1407 15.5859 24.4128 15.8061 24.7016 16H28Z",
                    className: styles.border
                }),
                _jsx("defs", {
                    children: _jsxs("filter", {
                        id: "filter0_f_3010_39642",
                        x: "0",
                        y: "0",
                        width: "32",
                        height: "20",
                        filterUnits: "userSpaceOnUse",
                        colorInterpolationFilters: "sRGB",
                        children: [
                            _jsx("feFlood", {
                                floodOpacity: "0",
                                result: "BackgroundImageFix"
                            }),
                            _jsx("feBlend", {
                                mode: "normal",
                                in: "SourceGraphic",
                                in2: "BackgroundImageFix",
                                result: "shape"
                            }),
                            _jsx("feGaussianBlur", {
                                stdDeviation: "4",
                                result: "effect1_foregroundBlur_3010_39642"
                            })
                        ]
                    })
                })
            ]
        })
    }));
