// @canva/app-ui-kit/dist/esm/ui/apps/developing/ui_kit/components/flyout_menu/flyout_menu.js
import { jsx as _jsx } from "react/jsx-runtime";
import { FlyoutMenu as EaselFlyoutMenu } from '../../../../../base/flyout_menu/flyout_menu';
/** 
 * `FlyoutMenu` component that complies with WAI-ARIA patterns and recommendations for menus.
 *
 * It'll automatically manage focus between the trigger and the menu,
 * and will also manage focus when the menu is open, to cycle through
 * the menu items. {@link FlyoutMenuItem}
 *
 * Use `FlyoutMenu` when you need a simple dropdown menu that only contains `FlyoutMenuItem` and `FlyoutMenuDivider`.
 * This component manages its own open/close state and sizes itself based on the menu items,
 * making it perfect for standard menus with consistent behavior and minimal setup.
 *
 * Need more flexibility? Try {@link Flyout} instead.
 */ export function FlyoutMenu(props) {
    return _jsx(EaselFlyoutMenu, {
        iconSize: "medium",
        label: props.label,
        title: props.title,
        description: props.description,
        ariaLabel: props.ariaLabel,
        children: props.children,
        icon: props.icon,
        iconPosition: props.iconPosition,
        tone: props.tone,
        onOpen: props.onOpen,
        onClose: props.onClose,
        flyoutPlacement: props.flyoutPlacement,
        tooltipLabel: props.tooltipLabel
    });
}
