// @canva/app-ui-kit/dist/esm/ui/base/form/search_input/private/search_input_menu.js
import { jsx as _jsx, jsxs as _jsxs, Fragment as _Fragment } from "react/jsx-runtime";
import classNames from 'classnames';
import * as React from 'react';
import { ClearableInput } from '../../clearable_input/clearable_input';
import { OutsidePointerDownHandler } from '../../../outside_pointerdown_handler/outside_pointerdown_handler';
import { Popover } from '../../../surface/popover/popover';
import styles from './search_input_menu.css';
const InternalSearchInputMenu = (props, ref)=>{
    const { children, detachContent = false, border, onOutsidePointerDown, end, className, error, type = 'search', usePopover = false, captureFocus, ...searchInputProps } = props;
    const hasContent = !!children;
    const withBordersSetInInput = usePopover || !hasContent || detachContent;
    const withShadow = border === 'faintShadow' || border === 'solid' && hasContent;
    const inputRef = React.useRef(null);
    const outerRef = React.useRef(null);
    React.useImperativeHandle(ref, ()=>({
            focus: ()=>{
                var _inputRef_current;
                return (_inputRef_current = inputRef.current) === null || _inputRef_current === void 0 ? void 0 : _inputRef_current.focus();
            },
            blur: ()=>{
                var _inputRef_current;
                return (_inputRef_current = inputRef.current) === null || _inputRef_current === void 0 ? void 0 : _inputRef_current.blur();
            }
        }));
    const handleOutsideInputInteraction = React.useCallback(()=>{
        inputRef.current && inputRef.current.blur();
        onOutsidePointerDown && onOutsidePointerDown();
    }, [
        onOutsidePointerDown
    ]);
    const onTabOutside = React.useCallback((event)=>{
        if (usePopover)
        return;
        if (event.key === 'Tab') setTimeout(()=>{
            var _outerRef_current;
            if (document.activeElement instanceof HTMLElement && !((_outerRef_current = outerRef.current) === null || _outerRef_current === void 0 ? void 0 : _outerRef_current.contains(document.activeElement))) handleOutsideInputInteraction();
        }, 5);
    }, [
        handleOutsideInputInteraction,
        usePopover
    ]);
    React.useEffect(()=>{
        window.addEventListener('blur', handleOutsideInputInteraction);
        return ()=>window.removeEventListener('blur', handleOutsideInputInteraction);
    }, [
        handleOutsideInputInteraction
    ]);
    const input = _jsx(ClearableInput, {
        ref: inputRef,
        type: type,
        ...searchInputProps,
        borderless: border !== 'solid' || !withBordersSetInInput,
        error: withBordersSetInInput && error,
        inputClassName: styles.input,
        className: classNames({
            [styles.inputContainerWithContent]: !withBordersSetInInput
        }),
        iconClassName: classNames(searchInputProps.iconClassName, styles.icon),
        end: end
    });
    return _jsx(OutsidePointerDownHandler, {
        onOutsidePointerDown: handleOutsideInputInteraction,
        children: _jsx("div", {
            className: classNames(styles.searchInputMenu, className, {
                [styles.faintShadow]: withBordersSetInInput && withShadow
            }),
            ref: outerRef,
            onKeyDown: onTabOutside,
            children: usePopover ? _jsx(Popover, {
                open: hasContent,
                onRequestClose: onOutsidePointerDown,
                reference: input,
                placement: "bottom-start",
                width: "reference",
                offset: "0",
                captureFocus: captureFocus,
                children: _jsx("div", {
                    className: styles.content,
                    children: children
                })
            }) : _jsxs(_Fragment, {
                children: [
                    input,
                    hasContent && _jsx("div", {
                        className: classNames(styles.container, detachContent ? styles.detached : styles.attached, {
                            [styles.faintShadow]: withShadow,
                            [styles.solidBorder]: border === 'solid' && !hasContent,
                            [styles.error]: error
                        }),
                        children: _jsx("div", {
                            className: styles.content,
                            children: children
                        })
                    })
                ]
            })
        })
    });
};
export const SearchInputMenu = React.forwardRef(InternalSearchInputMenu);
SearchInputMenu.displayName = 'SearchInputMenu';
