// @canva/app-ui-kit/dist/cjs/ui/base/typography/internal/typography.js
"use strict"
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: Object.getOwnPropertyDescriptor(all, name).get
    });
}
_export(exports, {
    get InheritColor () {
        return InheritColor;
    },
    get Text () {
        return Text;
    },
    get Title () {
        return Title;
    },
    get getDefaultHeadingTagName () {
        return getDefaultHeadingTagName;
    },
    get typographySizes () {
        return typographySizes;
    },
    get typographyTones () {
        return typographyTones;
    }
});
const _jsxruntime = require("react/jsx-runtime");
const _preconditions = require('../../../../base/preconditions');
const _classnames = _interop_require_default(require("classnames"));
const _react = _interop_require_wildcard(require("react"));
const _access_mode_state = require('../../a11y/access_mode_state/access_mode_state');
const _handle = require('../../handle/handle');
const _typographycss = _interop_require_wildcard(require("./typography.css"));
const _typography_aria_attributes = require("./typography_aria_attributes");
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function _getRequireWildcardCache(nodeInterop) {
    if (typeof WeakMap !== "function") return null;
    var cacheBabelInterop = new WeakMap();
    var cacheNodeInterop = new WeakMap();
    return (_getRequireWildcardCache = function(nodeInterop) {
        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
    })(nodeInterop);
}
function _interop_require_wildcard(obj, nodeInterop) {
    if (!nodeInterop && obj && obj.__esModule) {
        return obj;
    }
    if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
        return {
            default: obj
        };
    }
    var cache = _getRequireWildcardCache(nodeInterop);
    if (cache && cache.has(obj)) {
        return cache.get(obj);
    }
    var newObj = {
        __proto__: null
    };
    var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
    for(var key in obj){
        if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
            if (desc && (desc.get || desc.set)) {
                Object.defineProperty(newObj, key, desc);
            } else {
                newObj[key] = obj[key];
            }
        }
    }
    newObj.default = obj;
    if (cache) {
        cache.set(obj, newObj);
    }
    return newObj;
}
const InheritColor = Symbol('inherit');
const typographyTones = [
    'primary',
    'secondary',
    'tertiary',
    'critical',
    'info',
    'success',
    'warn',
    InheritColor
];
const typographySizes = [
    'xxlarge',
    'xlarge',
    'large',
    'medium',
    'small',
    'xsmall',
    'xxsmall'
];
const Text = _react.memo(_react.forwardRef((props, ref)=>{
    var _props_tagName;
    const tagName = (_props_tagName = props.tagName) !== null && _props_tagName !== void 0 ? _props_tagName : 'p';
    var _props_weight;
    const variant = (_props_weight = props.weight) !== null && _props_weight !== void 0 ? _props_weight : 'regular';
    return (0, _jsxruntime.jsx)(TextComponent, {
        ...props,
        variant: variant,
        tagName: tagName,
        ref: ref
    });
}));
const Title = _react.memo(_react.forwardRef((props, ref)=>{
    var _props_tagName;
    const tagName = (_props_tagName = props.tagName) !== null && _props_tagName !== void 0 ? _props_tagName : getDefaultHeadingTagName(props.size);
    return (0, _jsxruntime.jsx)(TextComponent, {
        ...props,
        variant: "title",
        tagName: tagName,
        ref: ref
    });
}));
const TextComponent = _react.forwardRef((props, ref)=>{
    const { id, alignment = 'start', tone = 'primary', margins = 'none', capitalization = 'default', size = 'medium', variant, tagName: TagName, lineClamp, wrapStyle, children, allowUserSelect, className, elementTiming, ...aria } = props;
    const hideOutline = (0, _access_mode_state.useIsMouseAccessMode)() && props.tabIndex === -1;
    const innerRef = _react.useRef(null);
    _react.useImperativeHandle(ref, ()=>(0, _handle.createBasicHTMLHandle)(innerRef));
    if (children === '' || children == null) return null;
    const textWrapClassName = getTextWrapClassName({
        lineClamp,
        wrapStyle
    });
    const lineClampStyles = lineClamp && lineClamp > 1 ? {
        WebkitLineClamp: lineClamp
    } : {};
    const finalClassName = (0, _classnames.default)(getTextClassName({
        variant,
        size
    }), margins === 'legacy' && _typographycss.default.margin, (0, _typographycss.getStyle)(alignment), tone === InheritColor ? _typographycss.default.inheritColor : tone && (0, _typographycss.getStyle)(tone), textWrapClassName, allowUserSelect && _typographycss.default.allowUserSelect, capitalization === 'uppercase' && _typographycss.default.uppercase, hideOutline && _typographycss.default.hideOutline, className);
    return (0, _jsxruntime.jsx)(TagName, {
        id: id,
        className: finalClassName,
        style: lineClampStyles,
        tabIndex: props.tabIndex,
        elementtiming: elementTiming,
        ref: innerRef,
        ...(0, _typography_aria_attributes.getTypographyAriaAttributes)(aria),
        children: children
    });
});
function getTextWrapClassName({ lineClamp, wrapStyle }) {
    if (lineClamp == null && wrapStyle == null)
        return null;
    if (lineClamp === 1)
    return _typographycss.default.ellipsis;
    return {
        [_typographycss.default.lineClamp]: lineClamp && lineClamp > 1,
        [_typographycss.default.wrapStyleBalance]: wrapStyle === 'balance',
        [_typographycss.default.wrapStylePretty]: wrapStyle === 'pretty'
    };
}
function getVariantClassNamePart(variant) {
    switch(variant){
        case 'regular':
            return '';
        case 'title':
            return 'Title';
        case 'bold':
            return 'Bold';
        default:
            throw new _preconditions.UnreachableError(variant);
    }
}
function getSizeClassNamePart(size) {
    switch(size){
        case 'xxlarge':
            return 'ExtraExtraLarge';
        case 'xlarge':
            return 'ExtraLarge';
        case 'large':
            return 'Large';
        case 'medium':
            return 'Medium';
        case 'small':
            return 'Small';
        case 'xsmall':
            return 'ExtraSmall';
        case 'xxsmall':
            return 'ExtraExtraSmall';
        default:
            throw new _preconditions.UnreachableError(size);
    }
}
function getTextClassName({ variant, size }) {
    return (0, _typographycss.getStyle)(`text${getVariantClassNamePart(variant)}${getSizeClassNamePart(size)}`);
}
function getDefaultHeadingTagName(size = 'medium') {
    switch(size){
        case 'xxlarge':
            return 'h1';
        case 'xlarge':
            return 'h1';
        case 'large':
            return 'h2';
        case 'medium':
            return 'h3';
        case 'small':
            return 'h4';
        case 'xsmall':
            return 'h5';
        case 'xxsmall':
            return 'h6';
        default:
            throw new _preconditions.UnreachableError(size);
    }
}
