// @canva/design/lib/cjs/sdk/design/public.js
// Copyright 2023 Canva Inc. All Rights Reserved.
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    selection: function() {
        return selection;
    },
    overlay: function() {
        return overlay;
    },
    addPage: function() {
        return addPage;
    },
    setCurrentPageBackground: function() {
        return setCurrentPageBackground;
    },
    getDefaultPageDimensions: function() {
        return getDefaultPageDimensions;
    },
    requestExport: function() {
        return requestExport;
    },
    ui: function() {
        return ui;
    },
    addNativeElement: function() {
        return addNativeElement;
    },
    addAudioTrack: function() {
        return addAudioTrack;
    },
    getCurrentPageContext: function() {
        return getCurrentPageContext;
    },
    initAppElement: function() {
        return initAppElement;
    },
    getDesignToken: function() {
        return getDesignToken;
    }
});
const { canva } = window;
const selection = canva.designInteraction.selection;
const overlay = canva.designInteraction.overlay;
function addPage(opts) {
    return canva.designInteraction.addPage(opts);
}
function setCurrentPageBackground(opts) {
    return canva.designInteraction.setCurrentPageBackground(opts);
}
function getDefaultPageDimensions() {
    return canva.designInteraction.getDefaultPageDimensions();
}
function requestExport(request) {
    return canva.export.requestExport(request);
}
const ui = canva.dragAndDrop;
function addNativeElement(element) {
    return canva.designInteraction.addNativeElement(element);
}
function addAudioTrack(audioTrack) {
    return canva.designInteraction.addAudioTrack(audioTrack);
}
function getCurrentPageContext() {
    return canva.designInteraction.getCurrentPageContext();
}
function initAppElement(appElementConfig) {
    return canva.designInteraction.initAppElement(appElementConfig);
}
function getDesignToken() {
    return canva.designInteraction.getDesignToken();
}
