// @canva/app-ui-kit/dist/cjs/ui/apps/developing/ui_kit/components/color_picker/color_selector.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "ColorSelector", {
    enumerable: true,
    get: function() {
        return ColorSelector;
    }
});
const _jsxruntime = require("react/jsx-runtime");
require("react");
const _color_picker = require('../../../../../base/color_picker/color_picker');
const _swatch = require('../../../../../base/swatch/swatch');
function ColorSelector(props) {
    const { color, disabled, onChange, onDeleteColor, onOpen, onClose, triggerMode } = props;
    return (0, _jsxruntime.jsx)(_color_picker.ColorPickerFlyout, {
        onChange: onChange,
        color: color,
        onDeleteColor: onDeleteColor,
        onOpen: onOpen,
        onClose: onClose,
        trigger: (props)=>{
            if (triggerMode === 'addColorButton') return (0, _jsxruntime.jsx)(_swatch.AddColorButton, {
                ...props,
                size: "small",
                disabled: disabled
            });
            return (0, _jsxruntime.jsx)(_swatch.Swatch, {
                ...props,
                fill: [
                    color
                ],
                size: "small",
                disabled: disabled
            });
        }
    });
}
