// @canva/app-ui-kit/dist/esm/ui/base/device_capabilities/keyboard.js
import { useSyncExternalStore } from 'react';
import { getVisualViewport } from './visual_viewport';
const globalNavigator = typeof navigator !== 'undefined' ? navigator : undefined;
export function useIsVirtualKeyboard() {
    return useSyncExternalStore(subscribeIsVirtualKeyboard, getClientIsVirtualKeyboard, getServerIsVirtualKeyboard);
}
export function isVirtualKeyboard(navigator1 = globalNavigator) {
    return navigator1 != null && !!navigator1.userAgent.match(/iphone|ipad|android/gi);
}
export function getVirtualKeyboardOverlayHeight() {
    const visualViewport = getVisualViewport();
    var _visualViewport_height;
    const visualViewportHeight = (_visualViewport_height = visualViewport === null || visualViewport === void 0 ? void 0 : visualViewport.height) !== null && _visualViewport_height !== void 0 ? _visualViewport_height : 0;
    const viewportHeight = Math.max(window.document.documentElement.clientHeight, window.innerHeight);
    return Math.max(viewportHeight - visualViewportHeight, 0);
}
export function guessKeyboardOverlayHeight(navigator1 = globalNavigator) {
    if (!isVirtualKeyboard(navigator1)) return 0;
    if (window.innerHeight < window.screen.height - 240)
     return 0;
    return Math.max(270, Math.min(340, window.screen.height * 0.45))
     + (Internal.isCanvaEditor(navigator1) ? 80 : 0);
}
export const Internal = {
    isCanvaEditor(navigator1 = globalNavigator) {
        return navigator1 != null && navigator1.userAgent.indexOf('canvaeditor') >= 0;
    }
};
export function applyKeyboardViewportOffsetCompensation(height) {
    const visualViewport = getVisualViewport();
    var _visualViewport_offsetTop;
    const visualViewportOffsetTop = (_visualViewport_offsetTop = visualViewport === null || visualViewport === void 0 ? void 0 : visualViewport.offsetTop) !== null && _visualViewport_offsetTop !== void 0 ? _visualViewport_offsetTop : 0;
    return Math.max(0, height - visualViewportOffsetTop);
}
const subscribeIsVirtualKeyboard = (_input)=>()=>{};
const isVirtualKeyboardValue = isVirtualKeyboard();
const getClientIsVirtualKeyboard = ()=>isVirtualKeyboardValue;
const getServerIsVirtualKeyboard = ()=>false;
