// @canva/app-ui-kit/dist/esm/ui/base/device_capabilities/visual_viewport.js
import * as React from 'react';
const globalWindow = typeof window !== 'undefined' ? window : undefined;
let proxied = null;
const proxyVisualViewport = (visualViewport)=>{
    if (!globalWindow) return null;
    if (proxied) return proxied;
    let listeners = [];
    let hasBoundUnload = false;
    const bindUnload = ()=>{
        if (hasBoundUnload) return;
        hasBoundUnload = true;
        globalWindow.addEventListener('unload', ()=>{
            listeners.forEach((args)=>{
                visualViewport.removeEventListener(...args);
            });
            listeners = [];
        });
    };
    proxied = {
        addEventListener (name, fn, options) {
            const useCapture = options && (options === true || options.capture === true);
            listeners.push([
                name,
                fn,
                useCapture
            ]);
            bindUnload();
            return visualViewport.addEventListener(name, fn, options);
        },
        removeEventListener (name, fn, options) {
            const useCapture = options && (options === true || options.capture === true);
            listeners = listeners.filter(([listenerName, listenerFn, listenerUseCapture])=>{
                return !(listenerName === name && listenerFn === fn && listenerUseCapture === useCapture);
            });
            return visualViewport.removeEventListener(name, fn, options);
        },
        get offsetLeft () {
            return visualViewport.offsetLeft;
        },
        get offsetTop () {
            return visualViewport.offsetTop;
        },
        get pageLeft () {
            return visualViewport.pageLeft;
        },
        get pageTop () {
            return visualViewport.pageTop;
        },
        get width () {
            return visualViewport.width;
        },
        get height () {
            return visualViewport.height;
        },
        get scale () {
            return visualViewport.scale;
        }
    };
    return proxied;
};
export const getVisualViewport = (targetWindow = globalWindow)=>{
    if (!targetWindow) return null;
    try {
        if (targetWindow.top && targetWindow.top.visualViewport) return proxyVisualViewport(targetWindow.top.visualViewport);
        return targetWindow.visualViewport;
    } catch (e) {
        return targetWindow.visualViewport;
    }
};
export const getViewportContainerSize = (targetWindow = globalWindow)=>{
    if (!targetWindow) return null;
    const visualViewport = getVisualViewport(targetWindow);
    if (!visualViewport) return null;
    if (!targetWindow.frameElement) return {
        offsetLeft: visualViewport.offsetLeft,
        offsetTop: visualViewport.offsetTop,
        width: visualViewport.width,
        height: visualViewport.height
    };
    const iframeRect = targetWindow.frameElement.getBoundingClientRect();
    const viewportTop = visualViewport.offsetTop - iframeRect.top;
    const viewportLeft = visualViewport.offsetLeft - iframeRect.left;
    const viewportBottom = viewportTop + visualViewport.height;
    const viewportRight = viewportLeft + visualViewport.width;
    const iframeTop = 0;
    const iframeLeft = 0;
    const iframeBottom = iframeTop + iframeRect.height;
    const iframeRight = iframeLeft + iframeRect.width;
    const offsetTop = Math.max(viewportTop, iframeTop);
    const offsetLeft = Math.max(viewportLeft, iframeLeft);
    const offsetBottom = Math.min(viewportBottom, iframeBottom);
    const offsetRight = Math.min(viewportRight, iframeRight);
    return {
        offsetLeft,
        offsetTop,
        width: offsetRight - offsetLeft,
        height: offsetBottom - offsetTop
    };
};
export function useViewportContainerSize() {
    const [containerSize, setContainerSize] = React.useState(null);
    const visualViewport = getVisualViewport();
    React.useLayoutEffect(()=>{
        if (!visualViewport) return;
        const onVisualViewportChange = ()=>{
            setContainerSize(getViewportContainerSize());
        };
        onVisualViewportChange();
        visualViewport.addEventListener('resize', onVisualViewportChange);
        visualViewport.addEventListener('scroll', onVisualViewportChange);
        return ()=>{
            visualViewport === null || visualViewport === void 0 ? void 0 : visualViewport.removeEventListener('resize', onVisualViewportChange);
            visualViewport === null || visualViewport === void 0 ? void 0 : visualViewport.removeEventListener('scroll', onVisualViewportChange);
        };
    }, [
        visualViewport
    ]);
    return containerSize !== null && containerSize !== void 0 ? containerSize : null;
}
