// @canva/app-ui-kit/dist/esm/ui/base/color_picker/eye_dropper_button/private/eye_dropper_button.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import * as React from 'react';
import { Button } from '../../../button/button';
import { EyedropperIcon } from '../../../icons/eyedropper/icon';
import { LegacySheet } from '../../../legacy_sheet/legacy_sheet';
import { MobileEventHandler } from '../../../mobile_event_handler/mobile_event_handler';
import styles from './eye_dropper_button.css';
import { EyeDropperButtonMessages } from './eye_dropper_button.messages';
export function EyeDropperButton(props) {
    const [active, setActive] = React.useState(false);
    const onStart = (event)=>{
        setActive(true);
        props.onStart(event).finally(()=>setActive(false));
    };
    const onBackButton = (e)=>{
        var _props_onBackButton;
        if (!active) return;
        (_props_onBackButton = props.onBackButton) === null || _props_onBackButton === void 0 ? void 0 : _props_onBackButton.call(props);
        e.stopPropagation();
    };
    return _jsxs(MobileEventHandler, {
        onBackButton: onBackButton,
        children: [
            _jsx(Button, {
                variant: "tertiary",
                icon: EyedropperIcon,
                active: active,
                className: styles.eyedropperButton,
                tooltipLabel: EyeDropperButtonMessages.eyedropperTooltip(),
                onClick: onStart
            }),
            props.isSheet && _jsx(LegacySheet, {
                content: null,
                open: active
            })
        ]
    });
}
