// @canva/app-ui-kit/dist/cjs/ui/base/color_picker/private/color_picker.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "ColorPicker", {
    enumerable: true,
    get: function() {
        return ColorPicker;
    }
});
const _jsxruntime = require("react/jsx-runtime");
const _react = _interop_require_wildcard(require("react"));
const _hsv_color = require('../hsv_color/hsv_color');
const _internal_color_picker = require("./internal_color_picker");
function _getRequireWildcardCache(nodeInterop) {
    if (typeof WeakMap !== "function") return null;
    var cacheBabelInterop = new WeakMap();
    var cacheNodeInterop = new WeakMap();
    return (_getRequireWildcardCache = function(nodeInterop) {
        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
    })(nodeInterop);
}
function _interop_require_wildcard(obj, nodeInterop) {
    if (!nodeInterop && obj && obj.__esModule) {
        return obj;
    }
    if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
        return {
            default: obj
        };
    }
    var cache = _getRequireWildcardCache(nodeInterop);
    if (cache && cache.has(obj)) {
        return cache.get(obj);
    }
    var newObj = {
        __proto__: null
    };
    var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
    for(var key in obj){
        if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
            if (desc && (desc.get || desc.set)) {
                Object.defineProperty(newObj, key, desc);
            } else {
                newObj[key] = obj[key];
            }
        }
    }
    newObj.default = obj;
    if (cache) {
        cache.set(obj, newObj);
    }
    return newObj;
}
function ColorPicker(props) {
    const { alpha, sizingMode, isSheet, autoFocusHexInput, showInputSwatch, onChangeStart, onChangeComplete, onStartEyedropper, onEyedropperBackButton, onDeleteColor } = props;
    const prevColorRef = _react.useRef(undefined);
    const [, forceUpdate] = _react.useState(1);
    const setColor = _react.useCallback((newColor)=>{
        prevColorRef.current = newColor;
        forceUpdate((n)=>-n);
    }, []);
    const color = prevColorRef.current ? prevColorRef.current.matchHexString(props.color) : _hsv_color.HsvColor.fromHexString(props.color, alpha);
    const onInputChange = (hex)=>{
        setColor(_hsv_color.HsvColor.fromHexString(hex));
        props.onChange(hex, 'hex');
    };
    const onSatValChange = (color)=>{
        setColor(color);
        props.onChange(color.toRgb().toHexString(), 'satVal', alpha);
    };
    const onHueChange = (hue)=>{
        const newColor = new _hsv_color.HsvColor(hue, color.s, color.v, alpha);
        setColor(newColor);
        if (newColor.v > 0 && newColor.s > 0)
        props.onChange(newColor.toRgb().toHexString(), 'hue', alpha);
    };
    const onAlphaChange = (alpha)=>{
        const newColor = new _hsv_color.HsvColor(color.h, color.s, color.v, alpha);
        setColor(newColor);
        props.onChange(newColor.toRgb().toHexString(), 'alpha', alpha);
    };
    return (0, _jsxruntime.jsx)(_internal_color_picker.InternalColorPicker, {
        color: color,
        sizingMode: sizingMode,
        isSheet: isSheet,
        showInputSwatch: showInputSwatch,
        autoFocusHexInput: autoFocusHexInput,
        onChangeStart: onChangeStart,
        onChangeComplete: onChangeComplete,
        onInputChange: onInputChange,
        onHueChange: onHueChange,
        onSatValChange: onSatValChange,
        onAlphaChange: alpha != null ? onAlphaChange : undefined,
        onStartEyedropper: onStartEyedropper,
        onEyedropperBackButton: onEyedropperBackButton,
        onDeleteColor: onDeleteColor
    });
}
