// @canva/app-ui-kit/dist/cjs/ui/base/color_picker/private/internal_color_picker.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "InternalColorPicker", {
    enumerable: true,
    get: function() {
        return InternalColorPicker;
    }
});
const _jsxruntime = require("react/jsx-runtime");
const _classnames = _interop_require_default(require("classnames"));
const _react = _interop_require_wildcard(require("react"));
const _box = require('../../box/box');
const _button = require('../../button/button');
const _alpha_slider = require('../alpha_slider/alpha_slider');
const _color_code_input = require('../color_code_input/color_code_input');
const _eye_dropper_button = require('../eye_dropper_button/eye_dropper_button');
const _hue_slider = require('../hue_slider/hue_slider');
const _sat_val_picker = require('../sat_val_picker/sat_val_picker');
const _draggable_view = require('../../draggable/draggable_view');
const _icon = require('../../icons/trash/icon');
const _color_pickermessages = require("./color_picker.messages");
const _internal_color_pickercss = _interop_require_default(require("./internal_color_picker.css"));
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function _getRequireWildcardCache(nodeInterop) {
    if (typeof WeakMap !== "function") return null;
    var cacheBabelInterop = new WeakMap();
    var cacheNodeInterop = new WeakMap();
    return (_getRequireWildcardCache = function(nodeInterop) {
        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
    })(nodeInterop);
}
function _interop_require_wildcard(obj, nodeInterop) {
    if (!nodeInterop && obj && obj.__esModule) {
        return obj;
    }
    if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
        return {
            default: obj
        };
    }
    var cache = _getRequireWildcardCache(nodeInterop);
    if (cache && cache.has(obj)) {
        return cache.get(obj);
    }
    var newObj = {
        __proto__: null
    };
    var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
    for(var key in obj){
        if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
            if (desc && (desc.get || desc.set)) {
                Object.defineProperty(newObj, key, desc);
            } else {
                newObj[key] = obj[key];
            }
        }
    }
    newObj.default = obj;
    if (cache) {
        cache.set(obj, newObj);
    }
    return newObj;
}
const InternalColorPicker = _react.memo((props)=>{
    const { color, sizingMode = 'fill', isSheet = sizingMode === 'fill', autoFocusHexInput = 'on-desktop', onStartEyedropper, onEyedropperBackButton, onDeleteColor, onChangeStart, onChangeComplete, onInputChange, onHueChange, onSatValChange, onAlphaChange, showInputSwatch } = props;
    const eyeDropperButton = onStartEyedropper && (0, _jsxruntime.jsx)(_eye_dropper_button.EyeDropperButton, {
        onStart: onStartEyedropper,
        onBackButton: onEyedropperBackButton,
        isSheet: isSheet
    });
    const deleteColorButton = onDeleteColor && (0, _jsxruntime.jsx)(DeleteColorButton, {
        onDeleteColor: onDeleteColor
    });
    if (sizingMode === 'fixed') return (0, _jsxruntime.jsxs)(_jsxruntime.Fragment, {
        children: [
            (0, _jsxruntime.jsx)(ColorPickerBody, {
                color: color,
                onChangeStart: onChangeStart,
                onChangeComplete: onChangeComplete,
                onHueChange: onHueChange,
                onSatValChange: onSatValChange,
                onAlphaChange: onAlphaChange
            }),
            (0, _jsxruntime.jsx)(ColorPickerInputGroup, {
                showInputSwatch: showInputSwatch === 'auto' ? 'never' : showInputSwatch,
                autoFocusHexInput: autoFocusHexInput,
                color: color,
                hexInputSizingMode: "fill",
                onInputChange: onInputChange,
                onAlphaChange: onAlphaChange,
                eyeDropperButton: eyeDropperButton,
                deleteColorButton: deleteColorButton
            })
        ]
    });
    return (0, _jsxruntime.jsxs)(_jsxruntime.Fragment, {
        children: [
            (0, _jsxruntime.jsx)("div", {
                className: _internal_color_pickercss.default.colorPickerHeader,
                children: (0, _jsxruntime.jsx)(ColorPickerInputGroup, {
                    color: color,
                    hexInputSizingMode: "fixed",
                    showInputSwatch: showInputSwatch === 'auto' ? 'always' : showInputSwatch,
                    onInputChange: onInputChange,
                    onAlphaChange: onAlphaChange,
                    eyeDropperButton: eyeDropperButton,
                    deleteColorButton: deleteColorButton
                })
            }),
            (0, _jsxruntime.jsx)(_box.Box, {
                padding: "2u",
                children: (0, _jsxruntime.jsx)(ColorPickerBody, {
                    color: color,
                    showDivider: true,
                    onChangeStart: onChangeStart,
                    onChangeComplete: onChangeComplete,
                    onHueChange: onHueChange,
                    onSatValChange: onSatValChange,
                    onAlphaChange: onAlphaChange
                })
            })
        ]
    });
});
const ColorPickerInputGroup = _react.memo((props)=>(0, _jsxruntime.jsxs)("div", {
        className: _internal_color_pickercss.default.colorInputSection,
        children: [
            props.deleteColorButton,
            (0, _jsxruntime.jsx)("div", {
                className: (0, _classnames.default)(_internal_color_pickercss.default.colorCodeInput, {
                    [_internal_color_pickercss.default.fixedWithoutAlpha]: props.hexInputSizingMode === 'fixed' && props.onAlphaChange == null,
                    [_internal_color_pickercss.default.fixedWithAlpha]: props.hexInputSizingMode === 'fixed' && props.onAlphaChange != null,
                    [_internal_color_pickercss.default.fill]: props.hexInputSizingMode === 'fill'
                }),
                children: (0, _jsxruntime.jsx)(_color_code_input.ColorCodeInput, {
                    autoFocus: props.autoFocusHexInput,
                    color: props.color.toRgb().toHexString(),
                    alpha: props.color.alpha,
                    onAlphaChange: props.onAlphaChange,
                    onChange: props.onInputChange,
                    withSwatch: props.showInputSwatch === 'always',
                    ariaLabel: _color_pickermessages.ColorPickerMessages.colorCodeInputLabel()
                })
            }),
            props.eyeDropperButton
        ]
    }));
const ColorPickerBody = _react.memo((props)=>(0, _jsxruntime.jsxs)(_jsxruntime.Fragment, {
        children: [
            (0, _jsxruntime.jsx)(_sat_val_picker.SatValPicker, {
                color: props.color,
                onChange: props.onSatValChange,
                onChangeStart: props.onChangeStart,
                onChangeComplete: props.onChangeComplete,
                Draggable: _draggable_view.DraggableView
            }),
            props.showDivider && (0, _jsxruntime.jsx)("div", {
                className: _internal_color_pickercss.default.divider
            }),
            (0, _jsxruntime.jsx)(_hue_slider.HueSlider, {
                hue: props.color.h,
                onChangeStart: props.onChangeStart,
                onChangeComplete: props.onChangeComplete,
                onChange: props.onHueChange
            }),
            props.onAlphaChange && (0, _jsxruntime.jsx)(_alpha_slider.AlphaSlider, {
                alpha: props.color.alpha,
                color: props.color.toRgb().toHexString(),
                onChangeStart: props.onChangeStart,
                onChangeComplete: props.onChangeComplete,
                onChange: props.onAlphaChange
            })
        ]
    }));
const DeleteColorButton = (props)=>(0, _jsxruntime.jsx)(_button.Button, {
        variant: "tertiary",
        icon: _icon.TrashIcon,
        className: _internal_color_pickercss.default.deleteColorButton,
        tooltipLabel: _color_pickermessages.ColorPickerMessages.deleteColorTooltip(),
        onClick: props.onDeleteColor
    });
