// @canva/app-ui-kit/dist/cjs/ui/base/badge/private/badge_container.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: Object.getOwnPropertyDescriptor(all, name).get
    });
}
_export(exports, {
    get BadgeContainer () {
        return BadgeContainer;
    },
    get BadgeGroup () {
        return BadgeGroup;
    },
    get ExtendedBadgeVisibilityTrigger () {
        return ExtendedBadgeVisibilityTrigger;
    },
    get SwapOnHover () {
        return SwapOnHover;
    },
    get VisibleOnHover () {
        return VisibleOnHover;
    }
});
const _jsxruntime = require("react/jsx-runtime");
const _classnames = _interop_require_default(require("classnames"));
const _mobxreactlite = require("mobx-react-lite");
const _react = _interop_require_wildcard(require("react"));
const _access_mode_state = require('../../a11y/access_mode_state/access_mode_state');
const _pointer_input_state = require('../../a11y/pointer_input_state/pointer_input_state');
const _badge_containercss = _interop_require_default(require("./badge_container.css"));
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function _getRequireWildcardCache(nodeInterop) {
    if (typeof WeakMap !== "function") return null;
    var cacheBabelInterop = new WeakMap();
    var cacheNodeInterop = new WeakMap();
    return (_getRequireWildcardCache = function(nodeInterop) {
        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
    })(nodeInterop);
}
function _interop_require_wildcard(obj, nodeInterop) {
    if (!nodeInterop && obj && obj.__esModule) {
        return obj;
    }
    if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
        return {
            default: obj
        };
    }
    var cache = _getRequireWildcardCache(nodeInterop);
    if (cache && cache.has(obj)) {
        return cache.get(obj);
    }
    var newObj = {
        __proto__: null
    };
    var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
    for(var key in obj){
        if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
            if (desc && (desc.get || desc.set)) {
                Object.defineProperty(newObj, key, desc);
            } else {
                newObj[key] = obj[key];
            }
        }
    }
    newObj.default = obj;
    if (cache) {
        cache.set(obj, newObj);
    }
    return newObj;
}
const paddingStyleClassMap = {
    ['compact']: _badge_containercss.default.compact,
    ['standard']: undefined
};
const transitionClassMap = {
    ['fade']: _badge_containercss.default.fxFade,
    ['slide']: _badge_containercss.default.fxSlide,
    ['none']: _badge_containercss.default.fxAppear
};
const BadgeContainer = _react.forwardRef(
    ({ children, paddingStyle = 'standard', className, ariaLabel, ariaLabelledBy, ariaDescribedBy, onMouseEnter, onMouseLeave, onMouseUp, onMouseDown, role }, ref)=>(0, _jsxruntime.jsx)(_mobxreactlite.Observer, {
            children: ()=>(0, _jsxruntime.jsx)("div", {
                    ref: ref,
                    className: (0, _classnames.default)(_badge_containercss.default.container, _badge_containercss.default.hoverTrigger, paddingStyleClassMap[paddingStyle], className, {
                        [_badge_containercss.default.focusVisible]: _access_mode_state.accessModeState.isKeyboardMode
                    }),
                    role: role,
                    "aria-label": ariaLabel,
                    "aria-labelledby": ariaLabelledBy,
                    "aria-describedby": ariaDescribedBy,
                    onMouseEnter: onMouseEnter,
                    onMouseLeave: onMouseLeave,
                    onMouseUp: onMouseUp,
                    onMouseDown: onMouseDown,
                    children: children
                })
        })
);
function BadgeGroup({ location, visibility = 'always', transition, children, stretch, className }) {
    return (0, _jsxruntime.jsx)("span", {
        className: (0, _classnames.default)(_badge_containercss.default.badgeGroup, location && getLocationClasses(location), transition ? transitionClassMap[transition] : _badge_containercss.default.defaultFx, {
            [_badge_containercss.default.hoverTarget]: visibility === 'on-hover',
            [_badge_containercss.default.pinning]: location != null,
            [_badge_containercss.default.stretch]: stretch
        }, className),
        children: children
    });
}
const VisibleOnHover = ({ children, transition, showOnTouchDevice = false })=>{
    const isMouseInput = (0, _pointer_input_state.useIsMouseInput)();
    if (!isMouseInput && !showOnTouchDevice) return null;
    return (0, _jsxruntime.jsx)("span", {
        className: (0, _classnames.default)(_badge_containercss.default.badgeGroup, transition ? transitionClassMap[transition] : _badge_containercss.default.defaultFx, {
            [_badge_containercss.default.hoverTarget]: isMouseInput
        }),
        children: children
    });
};
const SwapOnHover = ({ onHover, children })=>(0, _jsxruntime.jsxs)("span", {
        className: _badge_containercss.default.swapOnHoverWrapper,
        children: [
            (0, _jsxruntime.jsx)("span", {
                className: _badge_containercss.default.hideOnHover,
                children: children
            }),
            (0, _jsxruntime.jsx)("span", {
                className: _badge_containercss.default.showOnHover,
                children: onHover
            })
        ]
    });
function getLocationClasses(location) {
    return {
        [_badge_containercss.default.top]: location === 'top-start' || location === 'top-end',
        [_badge_containercss.default.bottom]: location === 'bottom-start' || location === 'bottom-end',
        [_badge_containercss.default.left]: location === 'top-start' || location === 'bottom-start' || location === 'center-start',
        [_badge_containercss.default.right]: location === 'top-end' || location === 'bottom-end' || location === 'center-end',
        [_badge_containercss.default.verticalCenter]: location === 'center-start' || location === 'center-end'
    };
}
const ExtendedBadgeVisibilityTrigger = (0, _mobxreactlite.observer)(_react.forwardRef(
    ({ children, tagName: TagName = 'div', role, className, onMouseEnter, onMouseLeave, onMouseUp, onMouseDown, ariaLabel, ariaLabelledBy, ariaDescribedBy }, ref)=>{
        return (0, _jsxruntime.jsx)(TagName, {
            className: (0, _classnames.default)(className, _badge_containercss.default.hoverTrigger, {
                [_badge_containercss.default.focusVisible]: _access_mode_state.accessModeState.isKeyboardMode
            }),
            onMouseEnter: onMouseEnter,
            onMouseLeave: onMouseLeave,
            onMouseUp: onMouseUp,
            onMouseDown: onMouseDown,
            "aria-label": ariaLabel,
            "aria-labelledby": ariaLabelledBy,
            "aria-describedby": ariaDescribedBy,
            role: role,
            ref: ref,
            children: children
        });
    }
));
