// @canva/app-ui-kit/dist/cjs/ui/base/badge/private/badge.js
"use strict"
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: Object.getOwnPropertyDescriptor(all, name).get
    });
}
_export(exports, {
    get Badge () {
        return Badge;
    },
    get ButtonBadge () {
        return ButtonBadge;
    },
    get PrimaryButtonBadge () {
        return PrimaryButtonBadge;
    },
    get SecondaryButtonBadge () {
        return SecondaryButtonBadge;
    },
    get badgeButtonTones () {
        return badgeButtonTones;
    },
    get tones () {
        return tones;
    }
});
const _jsxruntime = require("react/jsx-runtime");
const _classnames = _interop_require_default(require("classnames"));
const _mobxreactlite = require("mobx-react-lite");
const _react = _interop_require_wildcard(require("react"));
const _access_mode_state = require('../../a11y/access_mode_state/access_mode_state');
const _pointer_input_state = require('../../a11y/pointer_input_state/pointer_input_state');
const _box = require('../../box/box');
const _focusable = require('../../focusable/focusable');
const _badgecss = _interop_require_wildcard(require("./badge.css"));
const _wrapper_badge = require("./wrapper_badge");
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function _getRequireWildcardCache(nodeInterop) {
    if (typeof WeakMap !== "function") return null;
    var cacheBabelInterop = new WeakMap();
    var cacheNodeInterop = new WeakMap();
    return (_getRequireWildcardCache = function(nodeInterop) {
        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
    })(nodeInterop);
}
function _interop_require_wildcard(obj, nodeInterop) {
    if (!nodeInterop && obj && obj.__esModule) {
        return obj;
    }
    if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
        return {
            default: obj
        };
    }
    var cache = _getRequireWildcardCache(nodeInterop);
    if (cache && cache.has(obj)) {
        return cache.get(obj);
    }
    var newObj = {
        __proto__: null
    };
    var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
    for(var key in obj){
        if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
            if (desc && (desc.get || desc.set)) {
                Object.defineProperty(newObj, key, desc);
            } else {
                newObj[key] = obj[key];
            }
        }
    }
    newObj.default = obj;
    if (cache) {
        cache.set(obj, newObj);
    }
    return newObj;
}
const tones = [
    'contrast',
    'neutral',
    'positive',
    'warn',
    'info',
    'critical',
    'neutralLow',
    'assist',
    'assistLow',
    'positiveLow',
    'warnLow',
    'infoLow',
    'criticalLow'
];
const BaseBadge = _react.memo(
    ({ className, icon, text, shape = 'regular', size = 'tiny', tone, ariaLabel, border = 'none', width, ariaHidden = false })=>(0, _jsxruntime.jsxs)(_box.Box, {
            tagName: "span",
            background: tone === 'underlay' ? undefined : tone,
            ariaHidden: ariaHidden,
            className: (0, _classnames.default)(_badgecss.default.badge, {
                [_badgecss.default.tinyBadge]: size === 'tiny',
                [_badgecss.default.smallBadge]: size === 'small',
                [_badgecss.default.circle]: shape === 'circle',
                [_badgecss.default.borderSolid]: border === 'solid',
                [_badgecss.default.underlayBadge]: tone === 'underlay'
            }, className),
            ariaLabel: ariaLabel,
            width: width,
            children: [
                icon && (0, _jsxruntime.jsx)("span", {
                    className: (0, _classnames.default)(_badgecss.default.icon, text && _badgecss.default.withText),
                    children: typeof icon === 'function' ? icon() : icon
                }),
                text
            ]
        })
);
const UnderlayBadge = _react.memo((props)=>{
    return (0, _jsxruntime.jsx)(BaseBadge, {
        ...props,
        tone: "underlay",
        ariaHidden: true
    });
});
const Badge = ({ children, wrapInset = '-0.5u', ...props })=>{
    if (children != null) return (0, _jsxruntime.jsxs)(_wrapper_badge.WrapperBadge, {
        wrapInset: wrapInset,
        badge: (0, _jsxruntime.jsx)(BaseBadge, {
            ...props,
            width: undefined
        }),
        children: [
            children,
            (0, _jsxruntime.jsx)(UnderlayBadge, {
                text: props.text,
                icon: props.icon,
                size: props.size,
                shape: props.shape
            })
        ]
    });
    return (0, _jsxruntime.jsx)(BaseBadge, {
        ...props
    });
};
function preventFocusOutline() {
    _access_mode_state.accessModeState.setAccessMode('mouse');
}
const badgeButtonTones = [
    'primary',
    'secondary'
];
function isMenuButtonProps(props) {
    return props.ariaHasPopup === 'menu';
}
const PrimaryButtonBadge = (props)=>(0, _jsxruntime.jsx)(ButtonBadge, {
        ...props,
        tone: "primary"
    });
const SecondaryButtonBadge = (props)=>(0, _jsxruntime.jsx)(ButtonBadge, {
        ...props,
        tone: "secondary"
    });
const ButtonBadge = (props)=>{
    const { className, disabled, onMouseDown, onClick, children, tone, ariaLabel, focusRef } = props;
    const buttonRef = (0, _focusable.useFocusableRef)(focusRef);
    const onMouseDownHandler = _react.useCallback((e)=>{
        if (onMouseDown) onMouseDown(e);
        preventFocusOutline();
    }, [
        onMouseDown
    ]);
    const hoverSupported = (0, _pointer_input_state.useIsHoveringInput)();
    return (0, _jsxruntime.jsx)(_mobxreactlite.Observer, {
        children: ()=>(0, _jsxruntime.jsx)("button", {
                ref: buttonRef,
                onClick: onClick,
                onMouseDown: onMouseDownHandler,
                disabled: disabled,
                className: (0, _classnames.default)(_badgecss.default.buttonBadge, (0, _badgecss.getStyle)(`${tone}Button`), className, {
                    [_badgecss.default.hoverSupported]: hoverSupported,
                    [_badgecss.default.hideOutline]: _access_mode_state.accessModeState.isMouseMode,
                    [_badgecss.default.disabled]: disabled,
                    [_badgecss.default.expanded]: isMenuButtonProps(props) ? props.expanded : undefined
                }),
                "aria-label": ariaLabel,
                "aria-expanded": isMenuButtonProps(props) ? props.expanded : undefined,
                "aria-controls": isMenuButtonProps(props) ? props.ariaControls : undefined,
                "aria-haspopup": isMenuButtonProps(props) ? props.ariaHasPopup : undefined,
                children: children
            })
    });
};
