// @canva/app-ui-kit/dist/esm/ui/base/accordion/private/accordion.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { UnreachableError } from '../../../../base/preconditions';
import * as React from 'react';
import { Box } from '../../box/box';
import { BaseButton } from '../../button/base_button/base_button';
import { useControllableValue } from '../../controllable_value/controllable_value';
import { Divider } from '../../divider/divider';
import { ChevronDownIcon } from '../../icons/chevron_down/icon';
import { ChevronUpIcon } from '../../icons/chevron_up/icon';
import { Column, Columns, Rows } from '../../layout/layout';
import { InheritColor, Text } from '../../typography/typography';
import styles from './accordion.css';
const AccordionContext = React.createContext(undefined);
export const Accordion = ({ defaultExpanded, display = 'comfortable', noChildPadding, dividers = true, headingSize = 'large', headingLevel, tagName = 'ul', spacing = '3u', children })=>{
    const AccordionDivider = dividers ? _jsx("li", {
        "aria-hidden": "true",
        children: _jsx(Divider, {})
    }) : undefined;
    return _jsx(AccordionContext.Provider, {
        value: {
            defaultExpanded,
            display,
            noChildPadding,
            headingSize,
            headingLevel
        },
        children: _jsxs(Rows, {
            role: "list",
            tagName: tagName,
            spacing: display === 'compact' ? '1u' : spacing,
            children: [
                React.Children.map(children, (child, index)=>_jsxs(React.Fragment, {
                        children: [
                            AccordionDivider,
                            child
                        ]
                    }, index)),
                AccordionDivider
            ]
        })
    });
};
export const AccordionItem = ({ id, title, Icon, defaultExpanded, expanded: expandedProp, onClick, disabled, children })=>{
    const context = React.useContext(AccordionContext);
    var _ref;
    const [expanded, setExpanded] = useControllableValue({
        value: expandedProp,
        defaultValue: (_ref = defaultExpanded !== null && defaultExpanded !== void 0 ? defaultExpanded : context === null || context === void 0 ? void 0 : context.defaultExpanded) !== null && _ref !== void 0 ? _ref : false,
        onChange: onClick
    });
    const ChevronIcon = expanded ? ChevronUpIcon : ChevronDownIcon;
    const iconSize = (context === null || context === void 0 ? void 0 : context.display) === 'compact' ? 'small' : 'medium';
    var _context_headingLevel;
    const headingLevel = (_context_headingLevel = context === null || context === void 0 ? void 0 : context.headingLevel) !== null && _context_headingLevel !== void 0 ? _context_headingLevel : getDefaultHeadingLevel({
        size: (context === null || context === void 0 ? void 0 : context.headingSize) || 'large'
    });
    const buttonId = React.useId();
    const contentId = React.useId();
    return _jsxs(Box, {
        tagName: context ? 'li' : 'div',
        className: styles.accordionCard,
        id: id,
        children: [
            _jsx("div", {
                role: "heading",
                "aria-level": headingLevel,
                children: _jsx(BaseButton, {
                    tone: "tertiary",
                    display: "block",
                    width: "full",
                    paddingX: "1u",
                    paddingY: "0.5u",
                    borderRadius: "elementStandard",
                    id: buttonId,
                    disclosure: true,
                    active: expanded,
                    ariaControls: contentId,
                    className: styles.header,
                    disabled: disabled,
                    onClick: ()=>setExpanded(!expanded),
                    children: _jsxs(Columns, {
                        tagName: "span",
                        spacing: "1u",
                        alignY: "center",
                        children: [
                            Icon && _jsx(Column, {
                                tagName: "span",
                                width: "content",
                                children: _jsx(Icon, {
                                    size: iconSize
                                })
                            }),
                            _jsx(Column, {
                                tagName: "span",
                                children: typeof title === 'string' ? _jsx(Text, {
                                    weight: "bold",
                                    alignment: "start",
                                    size: (context === null || context === void 0 ? void 0 : context.display) === 'compact' ? 'small' : context === null || context === void 0 ? void 0 : context.headingSize,
                                    tone: InheritColor,
                                    children: title
                                }) : title
                            }),
                            _jsx(Column, {
                                tagName: "span",
                                width: "content",
                                children: _jsx(ChevronIcon, {
                                    size: "medium"
                                })
                            })
                        ]
                    })
                })
            }),
            _jsx(Box, {
                id: contentId,
                ariaLabelledBy: buttonId,
                role: "region",
                display: expanded ? 'block' : 'none',
                paddingX: (context === null || context === void 0 ? void 0 : context.noChildPadding) ? undefined : '1u',
                paddingTop: (context === null || context === void 0 ? void 0 : context.noChildPadding) ? undefined : (context === null || context === void 0 ? void 0 : context.display) === 'compact' ? '0.5u' : '2u',
                children: typeof children === 'string' ? _jsx(Text, {
                    children: children
                }) : children
            })
        ]
    });
};
function getDefaultHeadingLevel({ size }) {
    switch(size){
        case 'large':
            return 3;
        case 'medium':
            return 4;
        case 'small':
            return 5;
        default:
            throw new UnreachableError(size);
    }
}
