// @canva/app-ui-kit/dist/cjs/ui/base/button/repeat_click_on_hold/repeat_click_on_hold.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: Object.getOwnPropertyDescriptor(all, name).get
    });
}
_export(exports, {
    get FAST_REPEAT_DELAY () {
        return FAST_REPEAT_DELAY;
    },
    get FAST_REPEAT_INTERVAL () {
        return FAST_REPEAT_INTERVAL;
    },
    get SLOW_REPEAT_DELAY () {
        return SLOW_REPEAT_DELAY;
    },
    get SLOW_REPEAT_INTERVAL () {
        return SLOW_REPEAT_INTERVAL;
    },
    get WithRepeatClickOnHold () {
        return WithRepeatClickOnHold;
    },
    get useRepeatClickOnHold () {
        return useRepeatClickOnHold;
    }
});
const _react = _interop_require_wildcard(require("react"));
function _getRequireWildcardCache(nodeInterop) {
    if (typeof WeakMap !== "function") return null;
    var cacheBabelInterop = new WeakMap();
    var cacheNodeInterop = new WeakMap();
    return (_getRequireWildcardCache = function(nodeInterop) {
        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
    })(nodeInterop);
}
function _interop_require_wildcard(obj, nodeInterop) {
    if (!nodeInterop && obj && obj.__esModule) {
        return obj;
    }
    if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
        return {
            default: obj
        };
    }
    var cache = _getRequireWildcardCache(nodeInterop);
    if (cache && cache.has(obj)) {
        return cache.get(obj);
    }
    var newObj = {
        __proto__: null
    };
    var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
    for(var key in obj){
        if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
            if (desc && (desc.get || desc.set)) {
                Object.defineProperty(newObj, key, desc);
            } else {
                newObj[key] = obj[key];
            }
        }
    }
    newObj.default = obj;
    if (cache) {
        cache.set(obj, newObj);
    }
    return newObj;
}
const SLOW_REPEAT_DELAY = 450;
const FAST_REPEAT_DELAY = 350;
const SLOW_REPEAT_INTERVAL = 100;
const FAST_REPEAT_INTERVAL = 40;
function WithRepeatClickOnHold(props) {
    return props.children(useRepeatClickOnHold(props));
}
function useRepeatClickOnHold({ speed = 'none', disabled, onClick }) {
    const repeatDelayTimeoutIdRef = _react.useRef(undefined);
    const repeatIntervalIdRef = _react.useRef(undefined);
    const isPointerDownRef = _react.useRef(false);
    const onClickRef = _react.useRef(onClick);
    _react.useLayoutEffect(()=>{
        onClickRef.current = onClick;
    });
    const clearTimers = _react.useCallback(()=>{
        if (repeatDelayTimeoutIdRef.current) {
            window.clearTimeout(repeatDelayTimeoutIdRef.current);
            repeatDelayTimeoutIdRef.current = undefined;
        }
        if (repeatIntervalIdRef.current) {
            window.clearInterval(repeatIntervalIdRef.current);
            repeatIntervalIdRef.current = undefined;
        }
    }, []);
    _react.useEffect(()=>{
        return ()=>{
            clearTimers();
        };
    }, [
        clearTimers
    ]);
    const maybeClick = _react.useCallback((e)=>{
        var _onClickRef_current;
        if (!disabled) (_onClickRef_current = onClickRef.current) === null || _onClickRef_current === void 0 ? void 0 : _onClickRef_current.call(onClickRef, e);
    }, [
        disabled
    ]);
    const onPointerDown = _react.useCallback((e)=>{
        isPointerDownRef.current = true;
        const repeatInterval = speed === 'fast' ? FAST_REPEAT_INTERVAL : SLOW_REPEAT_INTERVAL;
        const repeatDelay = speed === 'fast' ? FAST_REPEAT_DELAY : SLOW_REPEAT_DELAY;
        clearTimers();
        repeatDelayTimeoutIdRef.current = window.setTimeout(()=>{
            maybeClick(e);
            repeatIntervalIdRef.current = window.setInterval(()=>{
                maybeClick(e);
            }, repeatInterval);
        }, repeatDelay);
    }, [
        speed,
        maybeClick,
        clearTimers
    ]);
    const onPointerUp = _react.useCallback((e)=>{
        if (isPointerDownRef.current) {
            isPointerDownRef.current = false;
            if (!repeatIntervalIdRef.current) maybeClick(e);
            clearTimers();
        }
        e.preventDefault();
    }, [
        clearTimers,
        maybeClick
    ]);
    const onPointerLeave = _react.useCallback((e)=>{
        isPointerDownRef.current = false;
        clearTimers();
        e.preventDefault();
    }, [
        clearTimers
    ]);
    const onContextMenu = _react.useCallback((e)=>{
        e.preventDefault();
        return false;
    }, []);
    return speed === 'none' ? {
        onClick
    } : {
        onPointerDown,
        onPointerUp,
        onPointerLeave,
        onContextMenu,
        onClick: undefined
    };
}
