// @canva/app-ui-kit/dist/cjs/ui/base/form/select/private/base_multi_select.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: Object.getOwnPropertyDescriptor(all, name).get
    });
}
_export(exports, {
    get BaseMultiSelect () {
        return BaseMultiSelect;
    },
    get BaseMultiSelectStore () {
        return BaseMultiSelectStore;
    }
});
const _jsxruntime = require("react/jsx-runtime");
const _make_observable = require('../../../../../base/make_observable/make_observable');
const _exists = require('../../../../../base/exists');
const _mobx = _interop_require_wildcard(require("mobx"));
const _react = _interop_require_wildcard(require("react"));
const _header = require('../../../surface/header/header');
const _base_select = require("./base_select");
const _base_selectmessages = require("./base_select.messages");
const _base_select_presenter = require("./base_select_presenter");
const _base_select_util = require("./base_select_util");
const _stateless_base_select = require("./stateless_base_select");
function _getRequireWildcardCache(nodeInterop) {
    if (typeof WeakMap !== "function") return null;
    var cacheBabelInterop = new WeakMap();
    var cacheNodeInterop = new WeakMap();
    return (_getRequireWildcardCache = function(nodeInterop) {
        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
    })(nodeInterop);
}
function _interop_require_wildcard(obj, nodeInterop) {
    if (!nodeInterop && obj && obj.__esModule) {
        return obj;
    }
    if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
        return {
            default: obj
        };
    }
    var cache = _getRequireWildcardCache(nodeInterop);
    if (cache && cache.has(obj)) {
        return cache.get(obj);
    }
    var newObj = {
        __proto__: null
    };
    var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
    for(var key in obj){
        if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
            if (desc && (desc.get || desc.set)) {
                Object.defineProperty(newObj, key, desc);
            } else {
                newObj[key] = obj[key];
            }
        }
    }
    newObj.default = obj;
    if (cache) {
        cache.set(obj, newObj);
    }
    return newObj;
}
const selectAllValue = Symbol();
const selectAllOption = {
    value: selectAllValue,
    label: _base_selectmessages.BaseSelectMessages.selectAll()
};
class BaseMultiSelectStore extends _base_select_presenter.BaseSelectStore {
    static _makeObservable(instance) {
        (0, _make_observable.makeObservable)(instance, {
            value: _mobx.observable.ref
        });
    }
    constructor(...args){
        super(...args), this.value = (BaseMultiSelectStore._makeObservable(this), undefined);
    }
}
class BaseMultiSelect extends _react.Component {
    get value() {
        var _this_props_value;
        return (_this_props_value = this.props.value) !== null && _this_props_value !== void 0 ? _this_props_value : this.store.value;
    }
    isAllSelected() {
        if (!this.value)
            return false;
        if (this.props.maxSelectedOptions && this.value.length >= this.props.maxSelectedOptions)
            return true;
        if (this.value.length === (0, _base_select_util.flattenOptions)(this.props.options).length)
            return true;
        let valueWithoutDisabledOptions = this.value;
        this.getDisabledOptions().forEach((option)=>{
            valueWithoutDisabledOptions = valueWithoutDisabledOptions.filter((value)=>value !== option.value);
        });
        return Array.isArray(valueWithoutDisabledOptions) && valueWithoutDisabledOptions.length === this.getAvailableOptions().length;
    }
    getAvailableOptions() {
        return (0, _base_select_util.flattenOptions)(this.props.options).filter((option)=>!option.disabled);
    }
    getDisabledOptions() {
        return (0, _base_select_util.flattenOptions)(this.props.options).filter((option)=>option.disabled);
    }
    render() {
        const { showSelectAll, options, maxSelectedOptions } = this.props;
        const finalShowSelectAll = showSelectAll && (!maxSelectedOptions || maxSelectedOptions >= options.length);
        return (0, _jsxruntime.jsx)(_base_select.BaseSelect, {
            store: this.store,
            presenter: this.presenter,
            Trigger: this.Trigger,
            Item: this.Item,
            onSelection: this.onSelection,
            closeOnSelection: false,
            headerEnd: ({ mode })=>mode === 'sheet' && (0, _jsxruntime.jsx)(_header.SurfaceHeaderTextButton, {
                    onClick: this.closeMenu,
                    children: _base_selectmessages.BaseSelectMessages.done()
                }),
            ...this.props,
            isSelected: this.isSelected,
            options: finalShowSelectAll ? [
                selectAllOption,
                ...options
            ] : options
        });
    }
    constructor(...args){
        var _this_props_store, _this_props_presenter;
        super(...args), this.store = (_this_props_store = this.props.store) !== null && _this_props_store !== void 0 ? _this_props_store : new BaseMultiSelectStore(), this.presenter = (_this_props_presenter = this.props.presenter) !== null && _this_props_presenter !== void 0 ? _this_props_presenter : new _base_select_presenter.BaseSelectPresenter(), this.isSelected = (option, value)=>{
            if (option.value === selectAllValue) return this.isAllSelected();
            const { isSelected = _base_select_util.defaultIsSelected } = this.props;
            return isSelected(option, value);
        }, this.onSelection = (option)=>{
            var _this_value, _this_props_onChange, _this_props;
            if (option.value === selectAllValue) return this.onSelectAll();
            if (option.value == null) return;
            const { isSelected = _base_select_util.defaultIsSelected, areEqual = _base_select_util.defaultAreEqual } = this.props;
            var _this_value_slice;
            const value = (_this_value_slice = (_this_value = this.value) === null || _this_value === void 0 ? void 0 : _this_value.slice()) !== null && _this_value_slice !== void 0 ? _this_value_slice : [];
            const newValue = isSelected(option, value) ? value.filter((v)=>!areEqual(v, option.value)) : value.concat(option.value);
            if (this.props.maxSelectedOptions && this.isAllSelected() && newValue.length > value.length)
                return;
            this.presenter.setValue(this.store, newValue);
            (_this_props_onChange = (_this_props = this.props).onChange) === null || _this_props_onChange === void 0 ? void 0 : _this_props_onChange.call(_this_props, newValue);
        }, this.onSelectAll = ()=>{
            var _this_props_onChange, _this_props;
            let newValue = this.isAllSelected() ? [] : this.getAvailableOptions().map((option)=>option.value).filter(_exists.exists);
            const disabledSelectedOptions = this.getDisabledOptions().map((option)=>this.isSelected(option, this.value) ? option.value : undefined).filter(_exists.exists);
            if (this.props.maxSelectedOptions)
                newValue = newValue.slice(0, this.props.maxSelectedOptions - disabledSelectedOptions.length);
            newValue = newValue.concat(disabledSelectedOptions);
            this.presenter.setValue(this.store, newValue);
            (_this_props_onChange = (_this_props = this.props).onChange) === null || _this_props_onChange === void 0 ? void 0 : _this_props_onChange.call(_this_props, newValue);
        }, this.closeMenu = ()=>{
            var _this_props_onClose, _this_props;
            this.presenter.toggle(this.store, false);
            (_this_props_onClose = (_this_props = this.props).onClose) === null || _this_props_onClose === void 0 ? void 0 : _this_props_onClose.call(_this_props);
        }, this.Trigger = _react.memo((props)=>{
            var _props_selectedOptions;
            const { Trigger = _stateless_base_select.BaseSelectTrigger } = this.props;
            const selectedOptions = (_props_selectedOptions = props.selectedOptions) === null || _props_selectedOptions === void 0 ? void 0 : _props_selectedOptions.filter((option)=>option.value !== selectAllValue);
            return (0, _jsxruntime.jsx)(Trigger, {
                ...props,
                selectedOptions: selectedOptions
            });
        }), this.Item = _react.memo((props)=>{
            return (0, _jsxruntime.jsx)(_stateless_base_select.BaseSelectItem, {
                ...props,
                multiSelectable: true
            });
        });
    }
}
