// @canva/app-ui-kit/dist/esm/ui/base/badge/private/badge.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import classNames from 'classnames';
import { Observer } from 'mobx-react-lite';
import * as React from 'react';
import { accessModeState } from '../../a11y/access_mode_state/access_mode_state';
import { useIsHoveringInput } from '../../a11y/pointer_input_state/pointer_input_state';
import { Box } from '../../box/box';
import { useFocusableRef } from '../../focusable/focusable';
import styles, { getStyle } from './badge.css';
import { WrapperBadge } from './wrapper_badge';
export const tones = [
    'contrast',
    'neutral',
    'positive',
    'warn',
    'info',
    'critical',
    'neutralLow',
    'assist',
    'assistLow',
    'positiveLow',
    'warnLow',
    'infoLow',
    'criticalLow'
];
const BaseBadge = React.memo(
    ({ className, icon, text, shape = 'regular', size = 'tiny', tone, ariaLabel, border = 'none', width, ariaHidden = false })=>_jsxs(Box, {
            tagName: "span",
            background: tone === 'underlay' ? undefined : tone,
            ariaHidden: ariaHidden,
            className: classNames(styles.badge, {
                [styles.tinyBadge]: size === 'tiny',
                [styles.smallBadge]: size === 'small',
                [styles.circle]: shape === 'circle',
                [styles.borderSolid]: border === 'solid',
                [styles.underlayBadge]: tone === 'underlay'
            }, className),
            ariaLabel: ariaLabel,
            width: width,
            children: [
                icon && _jsx("span", {
                    className: classNames(styles.icon, text && styles.withText),
                    children: typeof icon === 'function' ? icon() : icon
                }),
                text
            ]
        })
);
const UnderlayBadge = React.memo((props)=>{
    return _jsx(BaseBadge, {
        ...props,
        tone: "underlay",
        ariaHidden: true
    });
});
export const Badge = ({ children, wrapInset = '-0.5u', ...props })=>{
    if (children != null) return _jsxs(WrapperBadge, {
        wrapInset: wrapInset,
        badge: _jsx(BaseBadge, {
            ...props,
            width: undefined
        }),
        children: [
            children,
            _jsx(UnderlayBadge, {
                text: props.text,
                icon: props.icon,
                size: props.size,
                shape: props.shape
            })
        ]
    });
    return _jsx(BaseBadge, {
        ...props
    });
};
function preventFocusOutline() {
    accessModeState.setAccessMode('mouse');
}
export const badgeButtonTones = [
    'primary',
    'secondary'
];
function isMenuButtonProps(props) {
    return props.ariaHasPopup === 'menu';
}
export const PrimaryButtonBadge = (props)=>_jsx(ButtonBadge, {
        ...props,
        tone: "primary"
    });
export const SecondaryButtonBadge = (props)=>_jsx(ButtonBadge, {
        ...props,
        tone: "secondary"
    });
export const ButtonBadge = (props)=>{
    const { className, disabled, onMouseDown, onClick, children, tone, ariaLabel, focusRef } = props;
    const buttonRef = useFocusableRef(focusRef);
    const onMouseDownHandler = React.useCallback((e)=>{
        if (onMouseDown) onMouseDown(e);
        preventFocusOutline();
    }, [
        onMouseDown
    ]);
    const hoverSupported = useIsHoveringInput();
    return _jsx(Observer, {
        children: ()=>_jsx("button", {
                ref: buttonRef,
                onClick: onClick,
                onMouseDown: onMouseDownHandler,
                disabled: disabled,
                className: classNames(styles.buttonBadge, getStyle(`${tone}Button`), className, {
                    [styles.hoverSupported]: hoverSupported,
                    [styles.hideOutline]: accessModeState.isMouseMode,
                    [styles.disabled]: disabled,
                    [styles.expanded]: isMenuButtonProps(props) ? props.expanded : undefined
                }),
                "aria-label": ariaLabel,
                "aria-expanded": isMenuButtonProps(props) ? props.expanded : undefined,
                "aria-controls": isMenuButtonProps(props) ? props.ariaControls : undefined,
                "aria-haspopup": isMenuButtonProps(props) ? props.ariaHasPopup : undefined,
                children: children
            })
    });
};
