// @canva/app-ui-kit/dist/cjs/ui/base/accordion/private/accordion.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: Object.getOwnPropertyDescriptor(all, name).get
    });
}
_export(exports, {
    get Accordion () {
        return Accordion;
    },
    get AccordionItem () {
        return AccordionItem;
    }
});
const _jsxruntime = require("react/jsx-runtime");
const _preconditions = require('../../../../base/preconditions');
const _react = _interop_require_wildcard(require("react"));
const _box = require('../../box/box');
const _base_button = require('../../button/base_button/base_button');
const _controllable_value = require('../../controllable_value/controllable_value');
const _divider = require('../../divider/divider');
const _icon = require('../../icons/chevron_down/icon');
const _icon1 = require('../../icons/chevron_up/icon');
const _layout = require('../../layout/layout');
const _typography = require('../../typography/typography');
const _accordioncss = _interop_require_default(require("./accordion.css"));
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function _getRequireWildcardCache(nodeInterop) {
    if (typeof WeakMap !== "function") return null;
    var cacheBabelInterop = new WeakMap();
    var cacheNodeInterop = new WeakMap();
    return (_getRequireWildcardCache = function(nodeInterop) {
        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
    })(nodeInterop);
}
function _interop_require_wildcard(obj, nodeInterop) {
    if (!nodeInterop && obj && obj.__esModule) {
        return obj;
    }
    if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
        return {
            default: obj
        };
    }
    var cache = _getRequireWildcardCache(nodeInterop);
    if (cache && cache.has(obj)) {
        return cache.get(obj);
    }
    var newObj = {
        __proto__: null
    };
    var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
    for(var key in obj){
        if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
            if (desc && (desc.get || desc.set)) {
                Object.defineProperty(newObj, key, desc);
            } else {
                newObj[key] = obj[key];
            }
        }
    }
    newObj.default = obj;
    if (cache) {
        cache.set(obj, newObj);
    }
    return newObj;
}
const AccordionContext = _react.createContext(undefined);
const Accordion = ({ defaultExpanded, display = 'comfortable', noChildPadding, dividers = true, headingSize = 'large', headingLevel, tagName = 'ul', spacing = '3u', children })=>{
    const AccordionDivider = dividers ? (0, _jsxruntime.jsx)("li", {
        "aria-hidden": "true",
        children: (0, _jsxruntime.jsx)(_divider.Divider, {})
    }) : undefined;
    return (0, _jsxruntime.jsx)(AccordionContext.Provider, {
        value: {
            defaultExpanded,
            display,
            noChildPadding,
            headingSize,
            headingLevel
        },
        children: (0, _jsxruntime.jsxs)(_layout.Rows, {
            role: "list",
            tagName: tagName,
            spacing: display === 'compact' ? '1u' : spacing,
            children: [
                _react.Children.map(children, (child, index)=>(0, _jsxruntime.jsxs)(_react.Fragment, {
                        children: [
                            AccordionDivider,
                            child
                        ]
                    }, index)),
                AccordionDivider
            ]
        })
    });
};
const AccordionItem = ({ id, title, Icon, defaultExpanded, expanded: expandedProp, onClick, disabled, children })=>{
    const context = _react.useContext(AccordionContext);
    var _ref;
    const [expanded, setExpanded] = (0, _controllable_value.useControllableValue)({
        value: expandedProp,
        defaultValue: (_ref = defaultExpanded !== null && defaultExpanded !== void 0 ? defaultExpanded : context === null || context === void 0 ? void 0 : context.defaultExpanded) !== null && _ref !== void 0 ? _ref : false,
        onChange: onClick
    });
    const ChevronIcon = expanded ? _icon1.ChevronUpIcon : _icon.ChevronDownIcon;
    const iconSize = (context === null || context === void 0 ? void 0 : context.display) === 'compact' ? 'small' : 'medium';
    var _context_headingLevel;
    const headingLevel = (_context_headingLevel = context === null || context === void 0 ? void 0 : context.headingLevel) !== null && _context_headingLevel !== void 0 ? _context_headingLevel : getDefaultHeadingLevel({
        size: (context === null || context === void 0 ? void 0 : context.headingSize) || 'large'
    });
    const buttonId = _react.useId();
    const contentId = _react.useId();
    return (0, _jsxruntime.jsxs)(_box.Box, {
        tagName: context ? 'li' : 'div',
        className: _accordioncss.default.accordionCard,
        id: id,
        children: [
            (0, _jsxruntime.jsx)("div", {
                role: "heading",
                "aria-level": headingLevel,
                children: (0, _jsxruntime.jsx)(_base_button.BaseButton, {
                    tone: "tertiary",
                    display: "block",
                    width: "full",
                    paddingX: "1u",
                    paddingY: "0.5u",
                    borderRadius: "elementStandard",
                    id: buttonId,
                    disclosure: true,
                    active: expanded,
                    ariaControls: contentId,
                    className: _accordioncss.default.header,
                    disabled: disabled,
                    onClick: ()=>setExpanded(!expanded),
                    children: (0, _jsxruntime.jsxs)(_layout.Columns, {
                        tagName: "span",
                        spacing: "1u",
                        alignY: "center",
                        children: [
                            Icon && (0, _jsxruntime.jsx)(_layout.Column, {
                                tagName: "span",
                                width: "content",
                                children: (0, _jsxruntime.jsx)(Icon, {
                                    size: iconSize
                                })
                            }),
                            (0, _jsxruntime.jsx)(_layout.Column, {
                                tagName: "span",
                                children: typeof title === 'string' ? (0, _jsxruntime.jsx)(_typography.Text, {
                                    weight: "bold",
                                    alignment: "start",
                                    size: (context === null || context === void 0 ? void 0 : context.display) === 'compact' ? 'small' : context === null || context === void 0 ? void 0 : context.headingSize,
                                    tone: _typography.InheritColor,
                                    children: title
                                }) : title
                            }),
                            (0, _jsxruntime.jsx)(_layout.Column, {
                                tagName: "span",
                                width: "content",
                                children: (0, _jsxruntime.jsx)(ChevronIcon, {
                                    size: "medium"
                                })
                            })
                        ]
                    })
                })
            }),
            (0, _jsxruntime.jsx)(_box.Box, {
                id: contentId,
                ariaLabelledBy: buttonId,
                role: "region",
                display: expanded ? 'block' : 'none',
                paddingX: (context === null || context === void 0 ? void 0 : context.noChildPadding) ? undefined : '1u',
                paddingTop: (context === null || context === void 0 ? void 0 : context.noChildPadding) ? undefined : (context === null || context === void 0 ? void 0 : context.display) === 'compact' ? '0.5u' : '2u',
                children: typeof children === 'string' ? (0, _jsxruntime.jsx)(_typography.Text, {
                    children: children
                }) : children
            })
        ]
    });
};
function getDefaultHeadingLevel({ size }) {
    switch(size){
        case 'large':
            return 3;
        case 'medium':
            return 4;
        case 'small':
            return 5;
        default:
            throw new _preconditions.UnreachableError(size);
    }
}
