// @canva/app-ui-kit/dist/cjs/ui/apps/developing/ui_kit/components/box/box.js
"use strict"
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: Object.getOwnPropertyDescriptor(all, name).get
    });
}
_export(exports, {
    get Box () {
        return Box;
    },
    get alignItems () {
        return alignItems;
    },
    get backgroundTones () {
        return backgroundTones;
    },
    get borderRadiuses () {
        return borderRadiuses;
    },
    get borders () {
        return borders;
    },
    get displays () {
        return displays;
    },
    get flexDirections () {
        return flexDirections;
    },
    get flexWraps () {
        return flexWraps;
    },
    get justifyContents () {
        return justifyContents;
    },
    get sizes () {
        return sizes;
    }
});
const _jsxruntime = require("react/jsx-runtime");
const _preconditions = require('../../../../../../base/preconditions');
const _box = require('../../../../../base/box/box');
const backgroundTones = [
    'contrast',
    'neutral',
    'neutralLow'
];
const borders = [
    'none',
    'standard',
    'critical',
    'low'
];
const borderRadiuses = [
    'none',
    'standard',
    'large'
];
const sizes = [
    'unset',
    'full'
];
const displays = [
    'block',
    'none',
    'flex',
    'inline-flex'
];
const flexDirections = [
    'row',
    'column'
];
const flexWraps = [
    'noWrap',
    'wrap'
];
const justifyContents = [
    'normal',
    'center',
    'start',
    'end',
    'spaceBetween'
];
const alignItems = [
    'stretch',
    'center',
    'start',
    'end'
];
function Box(props) {
    return (0, _jsxruntime.jsx)(_box.Box, {
        display: props.display,
        flexDirection: props.flexDirection,
        flexWrap: props.flexWrap,
        alignItems: props.alignItems,
        justifyContent: props.justifyContent,
        background: props.background,
        border: props.border,
        borderRadius: mapBorderRadius(props.borderRadius),
        padding: props.padding,
        paddingX: props.paddingX,
        paddingY: props.paddingY,
        paddingTop: props.paddingTop,
        paddingBottom: props.paddingBottom,
        paddingStart: props.paddingStart,
        paddingEnd: props.paddingEnd,
        width: props.width,
        height: props.height,
        children: props.children,
        className: props.className,
        id: props.id
    });
}
function mapBorderRadius(borderRadius) {
    switch(borderRadius){
        case undefined:
            return undefined;
        case 'none':
            return 'none';
        case 'standard':
            return 'legacyStandard';
        case 'large':
            return 'legacyLarge';
        default:
            throw new _preconditions.UnreachableError(borderRadius);
    }
}
