// @canva/app-ui-kit/dist/esm/ui/base/color_picker/hue_slider/private/hue_slider.js
import { jsx as _jsx } from "react/jsx-runtime";
import classNames from 'classnames';
import * as React from 'react';
import { BaseSlider } from '../../../form/slider/slider';
import styles from './hue_slider.css';
import { HueSliderMessages } from './hue_slider.messages';
const track = ()=>_jsx("div", {
        className: styles.track
    });
const handle = ({ focused, trackHovering })=>_jsx("div", {
        className: classNames(styles.handle, {
            [styles.handleHoverOrFocus]: focused || trackHovering
        })
    });
export function HueSlider(
    { hue, onChange: onChangeProp, onChangeStart, onChangeComplete, ariaLabel = HueSliderMessages.colorHueLabel() }
) {
    const handleAtEndRef = React.useRef(false);
    const onChange = React.useCallback((value)=>{
        handleAtEndRef.current = value === 1;
        onChangeProp === null || onChangeProp === void 0 ? void 0 : onChangeProp(value);
    }, [
        onChangeProp
    ]);
    const handlePosition = handleAtEndRef.current && hue === 0 ? 1 : hue;
    return _jsx("div", {
        className: styles.container,
        children: _jsx("div", {
            className: styles.widener,
            children: _jsx(BaseSlider, {
                value: handlePosition,
                min: 0,
                max: 1,
                step: 0.005,
                onChange: onChange,
                onDragStart: onChangeStart,
                onDragEnd: onChangeComplete,
                ariaLabel: ariaLabel,
                fillColor: "transparent",
                handle: handle,
                track: track
            })
        })
    });
}
