// @canva/app-ui-kit/dist/esm/ui/base/date_time/date_time_picker/private/date_time_picker_presenter.js
import { makeObservable } from '../../../../../base/make_observable/make_observable';
import * as mobx from 'mobx';
import { fromTimezoneToUTC, fromUTCToTimezone, getMonth, getTimeOfDay, getTimezone, getTimezoneOffset, getYear, setTimeToDate, toISO, toUTCDateTimeObject } from '../../utils/utils';
export class DateTimePickerStore {
    static _makeObservable(instance) {
        makeObservable(instance, {
            _date: mobx.observable.ref,
            viewDate: mobx.observable.ref,
            focusedDate: mobx.observable.ref,
            timezone: mobx.computed,
            dateInTimezone: mobx.computed,
            focusedDateInTimezone: mobx.computed,
            todayInTimezone: mobx.computed,
            date: mobx.computed
        });
    }
    get timezone() {
        return this.defaultTimezone || getTimezone(this.viewDate, this.locale);
    }
    get todayTimezoneOffset() {
        return this.defaultTimezone ? this.defaultTimezone.offset : getTimezoneOffset(this.today);
    }
    get dateInTimezone() {
        return toUTCDateTimeObject(fromUTCToTimezone(this.viewDate, this.timezone.offset));
    }
    get focusedDateInTimezone() {
        if (this.focusedDate == null) return;
        return toUTCDateTimeObject(fromUTCToTimezone(this.focusedDate, this.timezone.offset));
    }
    get todayInTimezone() {
        return toUTCDateTimeObject(fromUTCToTimezone(this.today, this.todayTimezoneOffset));
    }
    get date() {
        return this._date;
    }
    set date(date) {
        this._date = date;
        if (date != null) this.viewDate = date;
    }
    constructor(utcDate, locale, utcToday, defaultTimezone){
        this.locale = locale;
        this.defaultTimezone = defaultTimezone;
        this._date = (DateTimePickerStore._makeObservable(this), undefined);
        this.today = utcToday || toISO(new Date());
        this._date = utcDate;
        this.viewDate = utcDate !== null && utcDate !== void 0 ? utcDate : this.today;
    }
}
export class DateTimePickerPresenter {
    static _makeObservable(instance) {
        makeObservable(instance, {
            unsetDate: mobx.action,
            updateFocusedDate: mobx.action,
            unsetFocusedDate: mobx.action,
            updateDateTime: mobx.action
        });
    }
    createStore(date, locale, today, timeZone) {
        return new DateTimePickerStore(date, locale, today, timeZone);
    }
    getDate(store) {
        return store.dateInTimezone;
    }
    getMonth(store) {
        return getMonth(store.dateInTimezone);
    }
    getYear(store) {
        return getYear(store.dateInTimezone);
    }
    getTime(store) {
        return getTimeOfDay(store.dateInTimezone);
    }
    getUTCDate(store) {
        return toISO(store.viewDate);
    }
    updateDate(store, dateInTimezone, options) {
        this.updateDateTime(store, dateInTimezone, getTimeOfDay(store.dateInTimezone), options === null || options === void 0 ? void 0 : options.updateFocus);
    }
    updateTime(store, timeMs) {
        this.updateDateTime(store, store.dateInTimezone, timeMs);
    }
    unsetDate(store) {
        store.date = undefined;
    }
    updateFocusedDate(store, dateInTimezone) {
        const utcDate = this.getUtcDate(store, dateInTimezone, getTimeOfDay(store.dateInTimezone));
        store.focusedDate = toISO(utcDate);
        store.viewDate = store.focusedDate;
    }
    unsetFocusedDate(store) {
        store.focusedDate = undefined;
    }
    getUtcDate(store, dateInTimezone, dayTime) {
        const localDate = setTimeToDate(dateInTimezone, dayTime);
        const timezoneOffset = store.defaultTimezone ? store.defaultTimezone.offset : getTimezoneOffset(localDate);
        let utcDate = fromTimezoneToUTC(localDate, timezoneOffset);
        const utcDateTimezoneOffset = store.defaultTimezone ? store.defaultTimezone.offset : getTimezoneOffset(utcDate);
        if (timezoneOffset !== utcDateTimezoneOffset)
            utcDate = fromTimezoneToUTC(localDate, utcDateTimezoneOffset);
        return utcDate;
    }
    updateDateTime(store, dateInTimezone, dayTime, updateFocus) {
        const utcDate = this.getUtcDate(store, dateInTimezone, dayTime);
        store.date = toISO(utcDate);
        store.viewDate = store.date;
        if (updateFocus) store.focusedDate = store.date;
    }
    constructor(){
        DateTimePickerPresenter._makeObservable(this);
    }
}
