// @canva/app-ui-kit/dist/cjs/ui/base/form/upload/file_input/private/dropzone.js
"use strict"
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "Dropzone", {
    enumerable: true,
    get: function() {
        return Dropzone;
    }
});
const _jsxruntime = require("react/jsx-runtime");
const _react = _interop_require_wildcard(require("react"));
const _reactdnd = require("react-dnd");
const _reactdndhtml5backend = require("react-dnd-html5-backend");
const _dropzonecss = _interop_require_default(require("./dropzone.css"));
const _dropzone_context = require("./dropzone_context");
const _dropzone_overlay = require("./dropzone_overlay");
const _file_validation = require("./file_validation");
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function _getRequireWildcardCache(nodeInterop) {
    if (typeof WeakMap !== "function") return null;
    var cacheBabelInterop = new WeakMap();
    var cacheNodeInterop = new WeakMap();
    return (_getRequireWildcardCache = function(nodeInterop) {
        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
    })(nodeInterop);
}
function _interop_require_wildcard(obj, nodeInterop) {
    if (!nodeInterop && obj && obj.__esModule) {
        return obj;
    }
    if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
        return {
            default: obj
        };
    }
    var cache = _getRequireWildcardCache(nodeInterop);
    if (cache && cache.has(obj)) {
        return cache.get(obj);
    }
    var newObj = {
        __proto__: null
    };
    var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
    for(var key in obj){
        if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
            if (desc && (desc.get || desc.set)) {
                Object.defineProperty(newObj, key, desc);
            } else {
                newObj[key] = obj[key];
            }
        }
    }
    newObj.default = obj;
    if (cache) {
        cache.set(obj, newObj);
    }
    return newObj;
}
const Dropzone = ({ disabled, multiple, accept, onDropAcceptedFiles, onDropRejectedFiles, children, showOverlay = true })=>{
    const onDrop = _react.useCallback((files)=>{
        if (disabled) return;
        const { acceptedFiles, rejectedFiles } = (0, _file_validation.getValidatedFiles)(files, multiple, accept);
        acceptedFiles.length > 0 && (onDropAcceptedFiles === null || onDropAcceptedFiles === void 0 ? void 0 : onDropAcceptedFiles(acceptedFiles));
        rejectedFiles.length > 0 && (onDropRejectedFiles === null || onDropRejectedFiles === void 0 ? void 0 : onDropRejectedFiles(rejectedFiles));
    }, [
        accept,
        disabled,
        multiple,
        onDropAcceptedFiles,
        onDropRejectedFiles
    ]);
    const [{ isOver }, dropRef] = (0, _reactdnd.useDrop)({
        accept: [
            _reactdndhtml5backend.NativeTypes.FILE
        ],
        collect: (monitor)=>({
                isOver: monitor.isOver()
            }),
        drop (e) {
            const { files } = e;
            files && onDrop(files);
        }
    });
    const dropzoneContext = _react.useMemo(()=>({
            isOver,
            disabled,
            accept,
            onDropAcceptedFiles,
            onDropRejectedFiles
        }), [
        accept,
        disabled,
        isOver,
        onDropAcceptedFiles,
        onDropRejectedFiles
    ]);
    return (0, _jsxruntime.jsx)("div", {
        className: _dropzonecss.default.dropzone,
        ref: _react.useCallback((node)=>{
            dropRef(node);
        }, [
            dropRef
        ]),
        children: (0, _jsxruntime.jsxs)(_dropzone_context.DropzoneContext.Provider, {
            value: dropzoneContext,
            children: [
                children,
                showOverlay && (0, _jsxruntime.jsx)(_dropzone_overlay.DropzoneOverlay, {})
            ]
        })
    });
};
