// @canva/app-ui-kit/dist/esm/ui/apps/developing/ui_kit/components/card/typography/typography.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import * as React from 'react';
import { Box } from '../../../../../../base/box/box';
import { BasicButton } from '../../../../../../base/button/button';
import { AnimatedSpinnerIcon } from '../../../../../../base/icons/animated_spinner/icon';
import styles from './typography.css';
export const TypographyCard = ({ ariaLabel, onClick, onDragStart, children, loading })=>{
    const [isDragging, setIsDragging] = React.useState(false);
    const [isHovering, setIsHovering] = React.useState(false);
    const boxBackground = isHovering && !loading ? 'neutral' : 'surface';
    return _jsx(BasicButton, {
        ariaLabel: ariaLabel,
        onClick: loading ? undefined : onClick,
        draggable: loading !== true && onDragStart !== undefined,
        onDragStart: (event)=>{
            if (onDragStart === undefined || loading === true) return;
            setIsDragging(true);
            onDragStart(event);
        },
        onDragEnd: ()=>setIsDragging(false),
        onMouseEnter: ()=>setIsHovering(true),
        onMouseLeave: ()=>setIsHovering(false),
        noChildWrapper: true,
        children: _jsxs(Box, {
            padding: "2u",
            borderRadius: "legacyStandard",
            background: boxBackground,
            border: "standard",
            className: loading ? styles.wrapperLoading : styles.wrapper,
            children: [
                _jsx("span", {
                    style: {
                        opacity: isDragging ? 0 : 1
                    },
                    children: children
                }),
                loading && _jsx(Box, {
                    className: styles.loading,
                    display: "flex",
                    justifyContent: "center",
                    alignItems: "center",
                    width: "full",
                    height: "full",
                    borderRadius: "legacyStandard",
                    children: _jsx(AnimatedSpinnerIcon, {
                        size: "medium",
                        className: styles.spinner
                    })
                })
            ]
        })
    });
};
