// @canva/app-ui-kit/dist/esm/ui/base/button/private/button.js
import { jsx as _jsx } from "react/jsx-runtime";
import * as React from 'react';
import { CircleButtonWithLabel, CircleButtonWithoutLabel, InternalAnchor, InternalButton, renderButtonOrAnchor } from './internal_button';
export const Button = React.forwardRef((props, ref)=>{
    return _jsx(InternalButton, {
        ...props,
        ref: ref
    });
});
export const ButtonLink = React.forwardRef((props, ref)=>{
    return _jsx(InternalAnchor, {
        ...props,
        ref: ref
    });
});
export const BasicButton = React.forwardRef((props, ref)=>{
    return renderButtonOrAnchor({
        ...props,
        variant: 'basic'
    }, ref);
});
export const BasicButtonLink = React.forwardRef((props, ref)=>{
    return _jsx(InternalAnchor, {
        ...props,
        variant: "basic",
        ref: ref
    });
});
export const PrimaryButton = (props)=>renderButtonOrAnchor({
        ...props,
        variant: 'primary'
    });
export const SecondaryButton = (props)=>renderButtonOrAnchor({
        ...props,
        variant: 'secondary'
    });
export const TertiaryButton = (props)=>renderButtonOrAnchor({
        ...props,
        variant: 'tertiary'
    });
export const LinkButton = (props)=>renderButtonOrAnchor({
        ...props,
        variant: 'linkButton'
    });
function isCircleButtonWithChildren(props) {
    return typeof props.children !== 'undefined';
}
export const CircleButton = React.forwardRef((props, ref)=>{
    if (isCircleButtonWithChildren(props)) return _jsx(CircleButtonWithLabel, {
        ref: ref,
        ...props
    });
    return _jsx(CircleButtonWithoutLabel, {
        ref: ref,
        ...props
    });
});
