// @canva/app-ui-kit/dist/cjs/ui/base/surface/flyout/private/flyout.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "Flyout", {
    enumerable: true,
    get: function() {
        return Flyout;
    }
});
const _jsxruntime = require("react/jsx-runtime");
const _preconditions = require('../../../../../base/preconditions');
const _react = _interop_require_wildcard(require("react"));
const _device_capabilities = require('../../../device_capabilities/device_capabilities');
const _skip_render_on_client = require('../../../hydration/skip_render_on_client');
const _popover = require('../../popover/popover');
const _sheet = require('../../sheet/sheet');
function _getRequireWildcardCache(nodeInterop) {
    if (typeof WeakMap !== "function") return null;
    var cacheBabelInterop = new WeakMap();
    var cacheNodeInterop = new WeakMap();
    return (_getRequireWildcardCache = function(nodeInterop) {
        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
    })(nodeInterop);
}
function _interop_require_wildcard(obj, nodeInterop) {
    if (!nodeInterop && obj && obj.__esModule) {
        return obj;
    }
    if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
        return {
            default: obj
        };
    }
    var cache = _getRequireWildcardCache(nodeInterop);
    if (cache && cache.has(obj)) {
        return cache.get(obj);
    }
    var newObj = {
        __proto__: null
    };
    var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
    for(var key in obj){
        if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
            if (desc && (desc.get || desc.set)) {
                Object.defineProperty(newObj, key, desc);
            } else {
                newObj[key] = obj[key];
            }
        }
    }
    newObj.default = obj;
    if (cache) {
        cache.set(obj, newObj);
    }
    return newObj;
}
const modes = [
    'sheet',
    'popover'
];
function Flyout({ open, onRequestClose, onCloseComplete, blockOutsidePointerEvents, trigger, width = '32u', dangerouslySetWidthMode = 'fixed', header, footer, children, title, description, headerStart, headerEnd, headerDivider = 'always', headerAlignment = 'auto', ...contentContainerProps }) {
    const mode = useMode();
    const resolveContent = (content)=>typeof content === 'function' ? modes.map((m, i)=>(0, _jsxruntime.jsx)(_skip_render_on_client.SkipRenderOnClient, {
                shouldRenderOnClient: ()=>mode === m,
                initiallyVisible: i === 0,
                children: content({
                    mode: m
                })
            }, m)) : content;
    const resolvedHeader = resolveContent(header);
    const resolvedFooter = resolveContent(footer);
    const endDecorator = resolveContent(headerEnd);
    const startDecorator = resolveContent(headerStart);
    const headerAndFooterProps = {
        footer: hasContent(resolvedFooter) ? resolvedFooter : null,
        header: hasContent(resolvedHeader) ? resolvedHeader : null,
        title,
        description,
        headerStart: startDecorator,
        headerEnd: endDecorator,
        headerDivider,
        headerAlignment: headerAlignment === 'auto' ? mode === 'sheet' ? 'center' : 'start' : headerAlignment
    };
    return (0, _jsxruntime.jsx)("div", {
        children: [
            ...modes.map((m, i)=>{
                switch(m){
                    case 'popover':
                        const popoverWidth = width === 'trigger' ? 'reference' : width;
                        return (0, _jsxruntime.jsx)(_skip_render_on_client.SkipRenderOnClient, {
                            shouldRenderOnClient: ()=>mode === 'popover',
                            initiallyVisible: i === 0,
                            children: (0, _jsxruntime.jsx)(_popover.Popover, {
                                open: open,
                                onRequestClose: onRequestClose,
                                onCloseComplete: onCloseComplete,
                                blockOutsidePointerEvents: blockOutsidePointerEvents,
                                reference: trigger,
                                width: popoverWidth,
                                widthMode: dangerouslySetWidthMode,
                                ...headerAndFooterProps,
                                ...contentContainerProps,
                                children: resolveContent(children)
                            })
                        }, m);
                    case 'sheet':
                        return (0, _jsxruntime.jsx)(_skip_render_on_client.SkipRenderOnClient, {
                            shouldRenderOnClient: ()=>mode === 'sheet',
                            initiallyVisible: i === 0,
                            children: (0, _jsxruntime.jsx)(FlyoutSheet, {
                                open: open,
                                onRequestClose: onRequestClose,
                                onCloseComplete: onCloseComplete,
                                trigger: trigger,
                                ...headerAndFooterProps,
                                ...contentContainerProps,
                                children: resolveContent(children)
                            })
                        }, m);
                    default:
                        throw new _preconditions.UnreachableError(m);
                }
            })
        ]
    });
}
function useMode() {
    return (0, _device_capabilities.isPortraitMobileScreen)() ? 'sheet' : 'popover';
}
const hasContent = (node)=>_react.Children.count(node) > 0;
function FlyoutSheet({ id: idProp, open, onRequestClose, onCloseComplete, onScroll, trigger, handleSafeAreaInsetBottom = true, children, ...contentContainerProps }) {
    const defaultId = _react.useId();
    const id = idProp || defaultId;
    return (0, _jsxruntime.jsxs)(_jsxruntime.Fragment, {
        children: [
            typeof trigger === 'function' ? trigger({
                ariaControls: open ? id : undefined,
                active: open,
                disclosure: true,
                ariaHasPopup: contentContainerProps.role
            }) : !(0, _popover.isReferenceObject)(trigger) && trigger,
            (0, _jsxruntime.jsx)(_sheet.Sheet, {
                open: open,
                onRequestClose: onRequestClose,
                onCloseComplete: onCloseComplete,
                onScroll: onScroll,
                id: id,
                handleSafeAreaInsetBottom: handleSafeAreaInsetBottom,
                ...contentContainerProps,
                children: children
            })
        ]
    });
}
