// @canva/app-ui-kit/dist/esm/ui/base/card/private/decorators/card_button.js
import { jsx as _jsx } from "react/jsx-runtime";
import { action } from 'mobx';
import { observer, useLocalObservable } from 'mobx-react-lite';
import * as React from 'react';
import { Button, ConjoinedButtons } from '../../../button/button';
import { MoreHorizontalIcon } from '../../../icons/more_horizontal/icon';
import { Flyout as BaseFlyout } from '../../../surface/flyout/flyout';
import { useCardConfiguration } from '../card_context';
import { CardMessages } from './card_button.messages';
import { CardDecorator } from './card_decorator';
export const CardButtonGroup = (props)=>{
    return _jsx(CardButtonGroupContextProvider, {
        children: _jsx(InternalCardButtonGroup, {
            ...props
        })
    });
};
const InternalCardButtonGroup = observer(function InternalCardButtonGroup(props) {
    const configuration = useCardConfiguration();
    const context = React.useContext(CardButtonGroupContext);
    var _props_disabled;
    const disabled = (_props_disabled = props.disabled) !== null && _props_disabled !== void 0 ? _props_disabled : configuration.disabled;
    const buttons = React.Children.map(props.children, (child)=>{
        if (React.isValidElement(child))
        return React.cloneElement(child, {
            size: props.size,
            disabled
        });
    });
    var _props_visibility;
    return _jsx(CardDecorator, {
        location: props.location,
        visibility: (context === null || context === void 0 ? void 0 : context.isFlyoutOpen) ? 'always' : (_props_visibility = props.visibility) !== null && _props_visibility !== void 0 ? _props_visibility : 'on-hover',
        transition: props.transition,
        isInteractive: true,
        children: _jsx(ConjoinedButtons, {
            children: buttons
        })
    });
});
export const CardButton = React.memo(function CardButton(props) {
    const { className, icon, ariaLabel, onClick, disabled, size, variant = 'contrast' } = props;
    const buttonSize = size === 'small' ? 'tiny' : 'small';
    return _jsx(Button, {
        variant: variant,
        icon: icon,
        iconSize: size,
        size: buttonSize,
        ariaLabel: ariaLabel,
        onClick: onClick,
        className: className,
        disabled: disabled
    });
});
export const CardMenuButton = observer(function CardMenuButton(props) {
    const context = React.useContext(CardButtonGroupContext);
    const flyoutStore = useLocalObservable(()=>({
            isOpen: false,
            toggle: action(()=>{
                context === null || context === void 0 ? void 0 : context.setIsFlyoutOpen(!flyoutStore.isOpen);
                flyoutStore.isOpen = !flyoutStore.isOpen;
            })
        }));
    return _jsx(StatelessCardMenuButton, {
        ...props,
        isOpen: flyoutStore.isOpen,
        toggleFlyout: flyoutStore.toggle
    });
});
export const StatelessCardMenuButton = (props)=>{
    const { Flyout = BaseFlyout, flyoutContent, isOpen, toggleFlyout, ariaLabel = CardMessages.more(), size, disabled, icon = MoreHorizontalIcon, className, variant = 'contrast', onContextMenu } = props;
    const buttonSize = size === 'small' ? 'tiny' : 'small';
    return _jsx(Flyout, {
        open: isOpen,
        onRequestClose: isOpen ? toggleFlyout : undefined,
        placement: "bottom-start",
        trigger: (triggerProps)=>_jsx(Button, {
                ...triggerProps,
                variant: variant,
                icon: icon,
                iconSize: size,
                size: buttonSize,
                disabled: disabled,
                className: className,
                ariaLabel: ariaLabel,
                onClick: toggleFlyout,
                onContextMenu: onContextMenu
            }),
        children: typeof flyoutContent === 'function' ? flyoutContent({
            toggleFlyout
        }) : flyoutContent
    });
};
const CardButtonGroupContext = React.createContext(undefined);
const CardButtonGroupContextProvider = ({ children })=>{
    const [isOpen, setIsOpen] = React.useState(false);
    const contextValue = React.useMemo(()=>{
        const setIsFlyoutOpen = (isOpen)=>{
            setIsOpen(isOpen);
        };
        return {
            isFlyoutOpen: isOpen,
            setIsFlyoutOpen
        };
    }, [
        isOpen,
        setIsOpen
    ]);
    return _jsx(CardButtonGroupContext.Provider, {
        value: contextValue,
        children: children
    });
};
