// @canva/app-ui-kit/dist/esm/ui/base/menu/private/placeholder.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import * as React from 'react';
import { Box } from '../../box/box';
import { Column, Columns } from '../../layout/layout';
import { Placeholder } from '../../placeholder/placeholder';
import { TextPlaceholder } from '../../typography/placeholder/placeholder';
import styles from './placeholder.css';
export const MenuItemPlaceholder = React.memo(
    function MenuItemPlaceholder({ index, startIcon = false, width: _width }) {
        const width = _width || getPlaceholderWidth(index || 0);
        return _jsx(Box, {
            tagName: "li",
            role: "none",
            paddingX: "2u",
            className: styles.menuItemPlaceholder,
            children: _jsxs(Columns, {
                spacing: "1u",
                align: "start",
                alignY: "center",
                children: [
                    startIcon && _jsx(Column, {
                        width: "content",
                        children: _jsx(Box, {
                            className: styles.startIcon,
                            children: _jsx(Placeholder, {
                                shape: "square",
                                index: index
                            })
                        })
                    }),
                    _jsx(Column, {
                        width: width,
                        children: _jsx(TextPlaceholder, {
                            size: "medium",
                            index: index
                        })
                    })
                ]
            })
        });
    }
);
MenuItemPlaceholder.displayName = 'MenuItemPlaceholder';
const placeholderWidthDefaultValues = [
    '2/3',
    '1/2'
];
const getPlaceholderWidth = (index)=>{
    return index != null ? placeholderWidthDefaultValues[index % 2] : '2/3';
};
