// @canva/app-ui-kit/dist/cjs/base/graphemes/split.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "splitByGraphemeClusters", {
    enumerable: true,
    get: function() {
        return splitByGraphemeClusters;
    }
});
const _codepoint_data = require("./codepoint_data");
function splitByGraphemeClusters(str) {
    const result = [];
    let brk;
    let index = 0;
    while(index < str.length && (brk = nextBreak(str, index)) < str.length){
        result.push(str.slice(index, brk));
        index = brk;
    }
    if (index < str.length) result.push(str.slice(index));
    return result;
}
function nextBreak(str, index) {
    let prev = codePointType(str.codePointAt(index));
    const mids = [];
    for(let i = index + 1; i < str.length; i++){
        if (isSurrogate(str, i - 1))
            continue;
        const next = codePointType(str.codePointAt(i));
        if (shouldBreak(prev, mids, next)) return i;
        mids.push(next);
        prev = next;
    }
    return str.length;
}
function shouldBreak(previous, mids, current) {
    if (previous === _codepoint_data.CR && current === _codepoint_data.LF)
        return false;
    if ([
        _codepoint_data.Control,
        _codepoint_data.CR,
        _codepoint_data.LF
    ].includes(previous))
        return true;
    if ([
        _codepoint_data.Control,
        _codepoint_data.CR,
        _codepoint_data.LF
    ].includes(current))
        return true;
    if (current === _codepoint_data.Extend || current === _codepoint_data.ZWJ)
        return false;
    if (current === _codepoint_data.SpacingMark)
        return false;
    if (previous === _codepoint_data.Prepend)
        return false;
    if (previous === _codepoint_data.ZWJ)
        return false;
    if (mids.includes(_codepoint_data.Regional_Indicator))
        return true;
    if (previous === _codepoint_data.Regional_Indicator && current === _codepoint_data.Regional_Indicator)
        return false;
    return true;
}
function codePointType(value) {
    if (value == null) return _codepoint_data.Other;
    else if (value === _codepoint_data.CR_value) return _codepoint_data.CR;
    else if (value === _codepoint_data.LF_value) return _codepoint_data.LF;
    else if (value === _codepoint_data.ZWJ_value) return _codepoint_data.ZWJ;
    if (checkRange(value, _codepoint_data.Prepend_values, _codepoint_data.Prepend_ranges)) return _codepoint_data.Prepend;
    else if (checkRange(value, _codepoint_data.Control_values, _codepoint_data.Control_ranges)) return _codepoint_data.Control;
    else if (checkRange(value, _codepoint_data.SpacingMark_values, _codepoint_data.SpacingMark_ranges)) return _codepoint_data.SpacingMark;
    else if (checkRange(value, _codepoint_data.Extend_values, _codepoint_data.Extend_ranges)) return _codepoint_data.Extend;
    else if (checkRange(value, new Set([]), _codepoint_data.Regional_Indicator_ranges)) return _codepoint_data.Regional_Indicator;
    return _codepoint_data.Other;
}
function checkRange(value, values, ranges) {
    if (values.has(value)) return true;
    const firstRange = ranges[0];
    const lastRange = ranges.length > 1 ? ranges[ranges.length - 1] : firstRange;
    if (firstRange[0] <= value && value <= lastRange[1]) return ranges.some(([a, b])=>a <= value && value <= b);
    return false;
}
function isSurrogate(str, pos) {
    return 0xd800 <= str.charCodeAt(pos) && str.charCodeAt(pos) <= 0xdbff && 0xdc00 <= str.charCodeAt(pos + 1) && str.charCodeAt(pos + 1) <= 0xdfff;
}
