// @canva/app-ui-kit/dist/cjs/ui/apps/developing/ui_kit/components/button/button.js
"use strict"
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "Button", {
    enumerable: true,
    get: function() {
        return Button;
    }
});
const _jsxruntime = require("react/jsx-runtime");
const _button = require('../../../../../base/button/button');
function Button(props) {
    const commonProps = {
        ariaLabel: props.ariaLabel,
        children: props.children,
        disabled: props.disabled,
        selected: props.selected,
        icon: props.icon,
        iconPosition: props.iconPosition,
        alignment: props.alignment,
        loading: props.loading,
        onClick: props.onClick,
        tooltipLabel: props.tooltipLabel,
        stretch: props.stretch,
        type: props.type || 'button',
        variant: props.variant || 'tertiary'
    };
    const shouldIncludeSize = ()=>{
        return !!(props.icon && (!props.variant || props.variant === 'tertiary' || props.variant === 'contrast') && !props.children);
    };
    const shouldIncludePressed = ()=>{
        return props.variant !== 'primary';
    };
    return (0, _jsxruntime.jsx)(_button.Button, {
        ...commonProps,
        size: shouldIncludeSize() ? props.size || 'medium' : undefined,
        pressed: shouldIncludePressed() ? props.pressed : undefined
    });
}
