// @canva/app-ui-kit/dist/esm/ui/apps/developing/ui_kit/components/flyout/flyout.js
import { jsx as _jsx } from "react/jsx-runtime";
import { Flyout as EaselFlyout } from '../../../../../base/surface/flyout/flyout';
/** 
 * `Flyout` provides a mechanism for rendering content in a "floating" surface that
 * is triggered by content in the primary document flow.
 *
 * Use `Flyout` when you need a flexible container that can hold any content.
 * This component requires you to manage the open/close state and specify the size,
 * making it ideal for custom implementations where you need more control over the behavior and appearance.
 *
 * If you only require a simple dropdown menu, try {@link FlyoutMenu} instead.
 */ export function Flyout(props) {
    return _jsx(EaselFlyout, {
        title: props.title,
        description: props.description,
        headerStart: props.headerStart,
        headerEnd: props.headerEnd,
        open: props.open,
        onRequestClose: props.onRequestClose,
        onCloseComplete: props.onCloseComplete,
        placement: props.placement,
        width: props.width,
        trigger: props.trigger,
        footer: props.footer,
        children: props.children
    });
}
