// @canva/app-ui-kit/dist/esm/ui/base/handle/private/extended_functionality_handle/dom_element_handle.js
export function createDOMElementHandle(ref) {
    return {
        dangerouslyGetElement: ()=>getElementRecursive(ref.current)
    };
}
function getElementRecursive(elementOrHandle) {
    if (isDOMElementHandleRecursive(elementOrHandle)) return getElementRecursive(elementOrHandle.dangerouslyGetElement());
    return elementOrHandle;
}
export function dangerouslyGetElement(elementOrHandle) {
    return isDOMElementHandle(elementOrHandle) ? elementOrHandle.dangerouslyGetElement() : elementOrHandle;
}
export function dangerouslyGetHTMLElement(elementOrHandle) {
    if (!isDOMElementHandle(elementOrHandle)) return elementOrHandle;
    const el = elementOrHandle.dangerouslyGetElement();
    return el instanceof HTMLElement ? el : null;
}
function isDOMElementHandle(elementOrHandle) {
    if (elementOrHandle == null) return false;
    return elementOrHandle.dangerouslyGetElement != null;
}
function isDOMElementHandleRecursive(elementOrHandle) {
    if (elementOrHandle == null) return false;
    return elementOrHandle.dangerouslyGetElement != null;
}
