// @canva/app-ui-kit/dist/esm/ui/apps/developing/ui_kit/styles/tokens.js
import styles from './tokens.css';
export const tokens = {
    space0: styles.uiKitSpace0,
    space050: styles.uiKitSpace050,
    space1: styles.uiKitSpace1,
    space150: styles.uiKitSpace150,
    space2: styles.uiKitSpace2,
    space3: styles.uiKitSpace3,
    space4: styles.uiKitSpace4,
    space6: styles.uiKitSpace6,
    space8: styles.uiKitSpace8,
    space12: styles.uiKitSpace12,
    baseUnit: styles.uiKitBaseUnit,
    borderRadius: styles.uiKitBorderRadius,
    minTouchableArea: styles.uiKitMinTouchableArea,
    shadowSurface: styles.uiKitShadowSurface,
    shadowSurfaceHigh: styles.uiKitShadowSurfaceHigh,
    /**  @deprecated use shadowSurface instead. */ boxShadowFaint: styles.uiKitBoxShadowFaint,
    /**  @deprecated use shadowSurface instead. */ boxShadowMedium: styles.uiKitBoxShadowMedium,
    /**  @deprecated use shadowSurfaceHigh instead. */ boxShadowHeavy: styles.uiKitBoxShadowHeavy,
    hoverTransition: styles.uiKitHoverTransition,
    fadeTransition: styles.uiKitFadeTransition,
    colorOverlay: styles.uiKitColorOverlay,
    colorPrimary: styles.uiKitColorPrimary,
    colorPrimaryHover: styles.uiKitColorPrimaryHover,
    colorPrimaryActive: styles.uiKitColorPrimaryActive,
    colorPrimaryDisabled: styles.uiKitColorPrimaryDisabled,
    colorPrimaryFore: styles.uiKitColorPrimaryFore,
    colorPrimaryForeDisabled: styles.uiKitColorPrimaryForeDisabled,
    colorSecondary: styles.uiKitColorSecondary,
    colorSecondaryHover: styles.uiKitColorSecondaryHover,
    colorSecondaryActive: styles.uiKitColorSecondaryActive,
    colorSecondaryDisabled: styles.uiKitColorSecondaryDisabled,
    colorSecondaryFore: styles.uiKitColorSecondaryFore,
    colorSecondaryForeDisabled: styles.uiKitColorSecondaryForeDisabled,
    colorTertiary: styles.uiKitColorTertiary,
    colorTertiaryHover: styles.uiKitColorTertiaryHover,
    colorTertiaryActive: styles.uiKitColorTertiaryActive,
    colorTertiaryDisabled: styles.uiKitColorTertiaryDisabled,
    colorTertiaryFore: styles.uiKitColorTertiaryFore,
    colorTertiaryForeDisabled: styles.uiKitColorTertiaryForeDisabled,
    colorContrast: styles.uiKitColorContrast,
    colorContrastHover: styles.uiKitColorContrastHover,
    colorContrastActive: styles.uiKitColorContrastActive,
    colorContrastDisabled: styles.uiKitColorContrastDisabled,
    colorContrastFore: styles.uiKitColorContrastFore,
    colorContrastForeDisabled: styles.uiKitColorContrastForeDisabled,
    colorNeutral: styles.uiKitColorNeutral,
    colorNeutralHover: styles.uiKitColorNeutralHover,
    colorNeutralActive: styles.uiKitColorNeutralActive,
    colorNeutralDisabled: styles.uiKitColorNeutralDisabled,
    colorNeutralLow: styles.uiKitColorNeutralLow,
    colorNeutralFore: styles.uiKitColorNeutralFore,
    colorNeutralForeLow: styles.uiKitColorNeutralForeLow,
    colorNeutralForeDisabled: styles.uiKitColorNeutralForeDisabled,
    colorPositive: styles.uiKitColorPositive,
    colorPositiveHover: styles.uiKitColorPositiveHover,
    colorPositiveActive: styles.uiKitColorPositiveActive,
    colorPositiveDisabled: styles.uiKitColorPositiveDisabled,
    colorPositiveLow: styles.uiKitColorPositiveLow,
    colorPositiveFore: styles.uiKitColorPositiveFore,
    colorPositiveForeLow: styles.uiKitColorPositiveForeLow,
    colorPositiveForeDisabled: styles.uiKitColorPositiveForeDisabled,
    colorInfo: styles.uiKitColorInfo,
    colorInfoHover: styles.uiKitColorInfoHover,
    colorInfoActive: styles.uiKitColorInfoActive,
    colorInfoDisabled: styles.uiKitColorInfoDisabled,
    colorInfoLow: styles.uiKitColorInfoLow,
    colorInfoFore: styles.uiKitColorInfoFore,
    colorInfoForeLow: styles.uiKitColorInfoForeLow,
    colorInfoForeDisabled: styles.uiKitColorInfoForeDisabled,
    colorWarn: styles.uiKitColorWarn,
    colorWarnHover: styles.uiKitColorWarnHover,
    colorWarnActive: styles.uiKitColorWarnActive,
    colorWarnDisabled: styles.uiKitColorWarnDisabled,
    colorWarnLow: styles.uiKitColorWarnLow,
    colorWarnFore: styles.uiKitColorWarnFore,
    colorWarnForeLow: styles.uiKitColorWarnForeLow,
    colorWarnForeDisabled: styles.uiKitColorWarnForeDisabled,
    colorCritical: styles.uiKitColorCritical,
    colorCriticalHover: styles.uiKitColorCriticalHover,
    colorCriticalActive: styles.uiKitColorCriticalActive,
    colorCriticalDisabled: styles.uiKitColorCriticalDisabled,
    colorCriticalLow: styles.uiKitColorCriticalLow,
    colorCriticalFore: styles.uiKitColorCriticalFore,
    colorCriticalForeLow: styles.uiKitColorCriticalForeLow,
    colorCriticalForeDisabled: styles.uiKitColorCriticalForeDisabled,
    colorCanvas: styles.uiKitColorCanvas,
    colorTabdock: styles.uiKitColorTabdock,
    colorPage: styles.uiKitColorPage,
    colorSurface: styles.uiKitColorSurface,
    colorTypographyPrimary: styles.uiKitColorTypographyPrimary,
    colorTypographySecondary: styles.uiKitColorTypographySecondary,
    colorTypographyTertiary: styles.uiKitColorTypographyTertiary,
    /**  @deprecated - use `colorTypographyPlaceholder` instead */ colorTypographyQuaternary: styles.uiKitColorTypographyQuaternary,
    colorTypographyPlaceholder: styles.uiKitColorTypographyPlaceholder,
    colorTypographyPositive: styles.uiKitColorTypographyPositive,
    colorTypographyCritical: styles.uiKitColorTypographyCritical,
    colorTypographyCriticalHover: styles.uiKitColorTypographyCriticalHover,
    colorTypographyCriticalActive: styles.uiKitColorTypographyCriticalActive,
    colorTypographyWarn: styles.uiKitColorTypographyWarn,
    colorTypographyInfo: styles.uiKitColorTypographyInfo,
    colorTypographyLink: styles.uiKitColorTypographyLink,
    colorTypographyLinkHover: styles.uiKitColorTypographyLinkHover,
    colorTypographyLinkActive: styles.uiKitColorTypographyLinkActive,
    colorBorder: styles.uiKitColorBorder,
    colorBorderHover: styles.uiKitColorBorderHover,
    colorBorderActive: styles.uiKitColorBorderActive,
    colorBorderDisabled: styles.uiKitColorBorderDisabled,
    colorBorderCritical: styles.uiKitColorBorderCritical,
    colorBorderLow: styles.uiKitColorBorderLow,
    colorBorderStrong: styles.uiKitColorBorderStrong
};
