// @canva/app-ui-kit/dist/cjs/ui/base/menu/menu.js
"use strict"
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: Object.getOwnPropertyDescriptor(all, name).get
    });
}
_export(exports, {
    get DIVIDER_HEIGHT_MULTIPIER () {
        return _menu.DIVIDER_HEIGHT_MULTIPIER;
    },
    get ITEM_HEIGHT_MULTIPIER () {
        return _menu.ITEM_HEIGHT_MULTIPIER;
    },
    get Menu () {
        return _menu.Menu;
    },
    get MenuDivider () {
        return _menu.MenuDivider;
    },
    get MenuItem () {
        return _menu.MenuItem;
    },
    get MenuItemButton () {
        return _menu.MenuItemButton;
    },
    get MenuItemPlaceholder () {
        return _placeholder.MenuItemPlaceholder;
    },
    get TEXT_DIVIDER_HEIGHT_MULTIPIER () {
        return _menu.TEXT_DIVIDER_HEIGHT_MULTIPIER;
    },
    get ToggleMenu () {
        return _menu.ToggleMenu;
    }
});
const _menu = require("./private/menu");
const _placeholder = require("./private/placeholder");
