// @canva/app-ui-kit/dist/esm/ui/base/form/radio_group/private/radio_item.js
import { jsx as _jsx, jsxs as _jsxs, Fragment as _Fragment } from "react/jsx-runtime";
import composeRefs from '@seznam/compose-react-refs';
import classNames from 'classnames';
import { Observer } from 'mobx-react-lite';
import * as React from 'react';
import { accessModeState } from '../../../a11y/access_mode_state/access_mode_state';
import { createBasicHTMLHandle } from '../../../handle/handle';
import { WithKeyCombinationHandler } from '../../../key_combinations/key_combinations';
import { Text } from '../../../typography/typography';
import styles from './radio_item.css';
const setToKeyboardMode = ()=>{
    accessModeState.setAccessMode('keyboard');
};
const inputKeyMap = [
    [
        'ArrowUp',
        setToKeyboardMode
    ],
    [
        'ArrowDown',
        setToKeyboardMode
    ],
    [
        'ArrowLeft',
        setToKeyboardMode
    ],
    [
        'ArrowRight',
        setToKeyboardMode
    ]
];
export const RadioItem = React.forwardRef(RadioItemImpl);
function RadioItemImpl({ label, description, className, disabled, checked, ariaLabel, name, value, onChange: _onChange, onFocus, onBlur }, ref) {
    const formatValue = ()=>{
        if (value === undefined) return undefined;
        if (typeof value === 'string' || typeof value === 'number') return value;
        return JSON.stringify(value);
    };
    const onChange = (event)=>{
        if (_onChange != null && !checked) _onChange(value, event);
    };
    const inputId = React.useId();
    const labelId = React.useId();
    const descriptionId = React.useId();
    const inputRef = React.useRef(null);
    React.useImperativeHandle(ref, ()=>createBasicHTMLHandle(inputRef));
    return _jsx(WithKeyCombinationHandler, {
        map: inputKeyMap,
        handleInputs: true,
        children: ({ ref: keyCombinationRef })=>_jsx(Observer, {
                children: ()=>_jsxs(_Fragment, {
                        children: [
                            _jsx("input", {
                                id: inputId,
                                "aria-label": ariaLabel,
                                "aria-labelledby": !ariaLabel && label ? labelId : undefined,
                                "aria-describedby": description ? descriptionId : undefined,
                                type: "radio",
                                className: styles.radio,
                                onChange: onChange,
                                ref: composeRefs(inputRef, keyCombinationRef),
                                value: formatValue(),
                                disabled: disabled,
                                checked: checked,
                                name: name,
                                onFocus: onFocus,
                                onBlur: onBlur
                            }),
                            _jsxs("label", {
                                htmlFor: inputId,
                                "aria-hidden": true,
                                onClick: (event)=>event.stopPropagation(),
                                className: classNames(styles.label, className, {
                                    [styles.hideOutline]: accessModeState.isMouseMode,
                                    [styles.disabled]: disabled
                                }),
                                children: [
                                    _jsx("span", {
                                        "aria-hidden": true,
                                        className: classNames(styles.uiRadio, {
                                            [styles.hasLabelText]: !!label
                                        })
                                    }),
                                    _jsx(RadioItemLabel, {
                                        label: label,
                                        description: description,
                                        labelId: labelId,
                                        descriptionId: descriptionId,
                                        disabled: disabled
                                    })
                                ]
                            })
                        ]
                    })
            })
    });
}
function RadioItemLabel({ label, description, disabled, labelId, descriptionId }) {
    const labelElement = typeof label === 'string' ? _jsx(Text, {
        id: labelId,
        tagName: "span",
        margins: "none",
        tone: disabled ? 'tertiary' : 'primary',
        children: label
    }) : _jsx("div", {
        id: labelId,
        children: label
    });
    if (!description) return labelElement;
    return _jsxs("div", {
        children: [
            labelElement,
            _jsx(Text, {
                id: descriptionId,
                tone: "tertiary",
                size: "small",
                children: description
            })
        ]
    });
}
