// @canva/app-ui-kit/dist/esm/ui/base/form/radio_group/private/radio_group.js
import { jsx as _jsx, jsxs as _jsxs, Fragment as _Fragment } from "react/jsx-runtime";
import { createElement as _createElement } from "react";
import composeRefs from '@seznam/compose-react-refs';
import classNames from 'classnames';
import * as React from 'react';
import { BaseButton } from '../../../button/base_button/base_button';
import { useControllableValue } from '../../../controllable_value/controllable_value';
import { createDOMElementHandle, createFocusHandle, createLayoutHandle, createNodeHandle, createScrollHandle, createTraverseHandle } from '../../../handle/handle';
import { useKeyCombinationHandler } from '../../../key_combinations/key_combinations';
import { Grid } from '../../../layout/layout';
import themes from '../../../theme/theme.css';
import { Text } from '../../../typography/typography';
import styles from './radio_group.css';
import { RadioItem } from './radio_item';
export const RadioGroup = React.forwardRef(RadioGroupImpl);
function RadioGroupImpl(props, ref) {
    const { value: valueProp, defaultValue: defaultValueProp, options, onChange: onChangeProp, disabled, id, ariaLabel, ariaLabelledBy, ariaDescribedBy, itemsPerRow, className, name, variant = 'default', onFocus, onBlur } = props;
    const radioGroupRef = React.useRef(null);
    const currentFocusRef = React.useRef(null);
    const radioGroupOrCurrentFocusRef = variant === 'button' ? radioGroupRef : currentFocusRef;
    React.useImperativeHandle(ref, ()=>({
            ...createLayoutHandle(radioGroupRef),
            ...createDOMElementHandle(radioGroupRef),
            ...createNodeHandle(radioGroupRef),
            ...createScrollHandle(radioGroupRef),
            ...createTraverseHandle(radioGroupRef),
            ...createFocusHandle(radioGroupOrCurrentFocusRef)
        }));
    const [value, setValue] = useControllableValue({
        value: valueProp,
        defaultValue: defaultValueProp
    });
    const onChange = React.useCallback((value, e)=>{
        setValue(value);
        onChangeProp === null || onChangeProp === void 0 ? void 0 : onChangeProp(value, e);
    }, [
        onChangeProp,
        setValue
    ]);
    const onItemFocus = React.useCallback((e)=>{
        var _radioGroupRef_current;
        const previouslyFocussedElement = e.relatedTarget;
        if (previouslyFocussedElement && (radioGroupRef === null || radioGroupRef === void 0 ? void 0 : (_radioGroupRef_current = radioGroupRef.current) === null || _radioGroupRef_current === void 0 ? void 0 : _radioGroupRef_current.contains(previouslyFocussedElement))) return;
        onFocus === null || onFocus === void 0 ? void 0 : onFocus(e);
    }, [
        onFocus
    ]);
    const onItemBlur = React.useCallback((e)=>{
        var _radioGroupRef_current;
        const nextElementToBeFocussed = e.relatedTarget;
        if (nextElementToBeFocussed && (radioGroupRef === null || radioGroupRef === void 0 ? void 0 : (_radioGroupRef_current = radioGroupRef.current) === null || _radioGroupRef_current === void 0 ? void 0 : _radioGroupRef_current.contains(nextElementToBeFocussed))) return;
        onBlur === null || onBlur === void 0 ? void 0 : onBlur(e);
    }, [
        onBlur
    ]);
    if (variant === 'button') return _jsx(ButtonRadioGroup, {
        options: options,
        value: value,
        onChange: onChange,
        className: className,
        disabled: disabled,
        id: id,
        ariaLabel: ariaLabel,
        ariaLabelledBy: ariaLabelledBy,
        ariaDescribedBy: ariaDescribedBy,
        itemsPerRow: itemsPerRow,
        ref: radioGroupRef,
        onFocus: onFocus && onItemFocus,
        onBlur: onBlur && onItemBlur
    });
    const radioGroup = _jsx("div", {
        role: "radiogroup",
        id: id,
        "aria-label": ariaLabel,
        "aria-labelledby": ariaLabelledBy,
        "aria-describedby": ariaDescribedBy,
        ref: radioGroupRef,
        className: styles.radioGroup,
        children: options.map((option, index)=>{
            const hasFocusRef = value != null ? option.value === value : index === 0;
            var _option_disabled;
            return _createElement(RadioItem, {
                ...option,
                disabled: (_option_disabled = option.disabled) !== null && _option_disabled !== void 0 ? _option_disabled : disabled,
                checked: option.value === value,
                key: index,
                onChange: onChange,
                name: name,
                ref: hasFocusRef ? currentFocusRef : undefined,
                onFocus: onFocus && onItemFocus,
                onBlur: onBlur && onItemBlur
            });
        })
    });
    if (className) return _jsx("div", {
        className: className,
        children: radioGroup
    });
    return radioGroup;
}
const ButtonItem = React.memo(React.forwardRef(
    ({ selected, disabled, label, description, onClick, tabbable, className, ariaLabel, onFocus, onBlur }, ref)=>{
        const descriptionId = React.useId();
        return _jsx(BaseButton, {
            ref: ref,
            className: classNames(styles.buttonItem, {
                [styles.buttonItemActive]: selected
            }, className),
            ariaLabel: ariaLabel,
            disabled: disabled,
            onClick: onClick,
            type: "button",
            role: "radio",
            selected: selected,
            tabIndex: selected || tabbable ? undefined : -1,
            onFocus: onFocus,
            onBlur: onBlur,
            ariaDescribedBy: description ? descriptionId : undefined,
            display: "block",
            borderRadius: "element",
            padding: "1u",
            backgroundColor: themes.colorSurface,
            children: _jsx(ButtonItemLabel, {
                label: label,
                description: description,
                disabled: disabled,
                selected: selected,
                descriptionId: descriptionId
            })
        });
    }
));
function ButtonItemLabel({ label, description, disabled, selected, descriptionId }) {
    const labelElement = typeof label === 'string' ? _jsx(Text, {
        tagName: "span",
        margins: "none",
        alignment: "center",
        tone: disabled ? 'tertiary' : 'primary',
        weight: selected ? 'bold' : 'regular',
        children: label
    }) : label;
    if (!description) return labelElement;
    return _jsxs(_Fragment, {
        children: [
            labelElement,
            _jsx("div", {
                "aria-hidden": true,
                children: _jsx(Text, {
                    id: descriptionId,
                    tagName: "span",
                    tone: "tertiary",
                    size: "small",
                    children: description
                })
            })
        ]
    });
}
const ButtonItemContainer = React.forwardRef(function ButtonItemContainer(props, ref) {
    const { selected, option, onChange, disabled, tabbable, className, onFocus, onBlur } = props;
    const onClick = React.useMemo(()=>{
        return (event)=>!selected && (onChange === null || onChange === void 0 ? void 0 : onChange(option.value, event));
    }, [
        selected,
        option,
        onChange
    ]);
    var _option_disabled;
    return _jsx(ButtonItem, {
        ref: ref,
        ariaLabel: option.ariaLabel,
        selected: selected,
        disabled: (_option_disabled = option.disabled) !== null && _option_disabled !== void 0 ? _option_disabled : disabled,
        tabbable: tabbable,
        className: className,
        onClick: onClick,
        onFocus: onFocus,
        onBlur: onBlur,
        label: option.label,
        description: option.description
    });
});
const ButtonRadioGroup = React.forwardRef(function(props, ref) {
    const { itemsPerRow, disabled, options, value, id, ariaLabel, ariaLabelledBy, ariaDescribedBy, className, onChange, onFocus, onBlur } = props;
    const shouldFocus = React.useRef(false);
    const activeRef = React.useRef(null);
    const divRef = React.useRef(null);
    React.useImperativeHandle(ref, ()=>({
            ...createDOMElementHandle(divRef),
            ...createLayoutHandle(divRef),
            ...createNodeHandle(divRef),
            ...createScrollHandle(divRef),
            ...createTraverseHandle(divRef),
            ...createFocusHandle(activeRef)
        }));
    React.useEffect(()=>{
        var _activeRef_current;
        if (shouldFocus.current) (_activeRef_current = activeRef.current) === null || _activeRef_current === void 0 ? void 0 : _activeRef_current.focus();
    }, [
        shouldFocus,
        value
    ]);
    const availableOptions = options.filter((option)=>{
        return !option.disabled;
    });
    const firstAvailableOption = availableOptions[0];
    const moveSelection = (delta)=>{
        const currentIndex = availableOptions.findIndex((option)=>option.value === value);
        const newIndex = ((currentIndex > -1 ? currentIndex : 0) + availableOptions.length + delta) % availableOptions.length;
        shouldFocus.current = true;
        onChange === null || onChange === void 0 ? void 0 : onChange(availableOptions[newIndex].value);
    };
    const { ref: keyRef } = useKeyCombinationHandler([
        [
            'ArrowUp',
            ()=>moveSelection(-1)
        ],
        [
            'ArrowDown',
            ()=>moveSelection(1)
        ],
        [
            'previous',
            ()=>moveSelection(-1)
        ],
        [
            'next',
            ()=>moveSelection(1)
        ]
    ], {
        handleInputs: true
    });
    return _jsx("div", {
        ref: composeRefs(keyRef, divRef),
        className: className,
        role: "radiogroup",
        id: id,
        "aria-label": ariaLabel,
        "aria-labelledby": ariaLabelledBy,
        "aria-describedby": ariaDescribedBy,
        children: _jsx(Grid, {
            spacing: "1u",
            columns: itemsPerRow || options.length,
            children: options.map((option, idx)=>{
                const tabbable = option === firstAvailableOption && value === undefined;
                const isCurrentItemSelected = option.value === value;
                const hasFocusRef = value != null ? isCurrentItemSelected : option === firstAvailableOption;
                const buttonItemContainerRef = hasFocusRef || isCurrentItemSelected ? activeRef : undefined;
                return _jsx(ButtonItemContainer, {
                    disabled: disabled,
                    selected: isCurrentItemSelected,
                    option: option,
                    tabbable: tabbable,
                    className: option.className,
                    ref: buttonItemContainerRef,
                    onChange: onChange,
                    onFocus: onFocus,
                    onBlur: onBlur
                }, idx);
            })
        })
    });
});
