// @canva/app-ui-kit/dist/cjs/ui/base/layer/private/layer.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
function _export(target, all) {
  for (var name in all) Object.defineProperty(target, name, {
    enumerable: true,
    get: Object.getOwnPropertyDescriptor(all, name).get
  });
}
_export(exports, {
  get Layer() {
    return Layer;
  },
  get LayerError() {
    return LayerError;
  },
  get LayerLevel() {
    return LayerLevel;
  },
  get WithLayerParent() {
    return WithLayerParent;
  }
});
const _jsxruntime = require("react/jsx-runtime");
const _make_observable = require('../../../../base/make_observable/make_observable');
const _dispose_on_unmount = require('../../../../base/mobx_react/dispose_on_unmount');
const _observer_for_class = require('../../../../base/mobx_react/observer_for_class');
const _mobx = _interop_require_wildcard(require("mobx"));
const _react = _interop_require_wildcard(require("react"));
const _reactdom = _interop_require_wildcard(require("react-dom"));
const _interaction_manager = require('../../a11y/interaction_manager/interaction_manager');
const _theme = require('../../theme/theme');
const _internal = require("./internal");
const _layercss = _interop_require_default(require("./layer.css"));
function _interop_require_default(obj) {
  return obj && obj.__esModule ? obj : {
    default: obj
  };
}
function _getRequireWildcardCache(nodeInterop) {
  if (typeof WeakMap !== "function") return null;
  var cacheBabelInterop = new WeakMap();
  var cacheNodeInterop = new WeakMap();
  return (_getRequireWildcardCache = function (nodeInterop) {
    return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
  })(nodeInterop);
}
function _interop_require_wildcard(obj, nodeInterop) {
  if (!nodeInterop && obj && obj.__esModule) {
    return obj;
  }
  if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
    return {
      default: obj
    };
  }
  var cache = _getRequireWildcardCache(nodeInterop);
  if (cache && cache.has(obj)) {
    return cache.get(obj);
  }
  var newObj = {
    __proto__: null
  };
  var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
  for (var key in obj) {
    if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
      var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
      if (desc && (desc.get || desc.set)) {
        Object.defineProperty(newObj, key, desc);
      } else {
        newObj[key] = obj[key];
      }
    }
  }
  newObj.default = obj;
  if (cache) {
    cache.set(obj, newObj);
  }
  return newObj;
}
function applyDecs2203RFactory() {
  function createAddInitializerMethod(initializers, decoratorFinishedRef) {
    return function addInitializer(initializer) {
      assertNotFinished(decoratorFinishedRef, "addInitializer");
      assertCallable(initializer, "An initializer");
      initializers.push(initializer);
    };
  }
  function memberDec(dec, name, desc, initializers, kind, isStatic, isPrivate, metadata, value) {
    var kindStr;
    switch (kind) {
      case 1:
        kindStr = "accessor";
        break;
      case 2:
        kindStr = "method";
        break;
      case 3:
        kindStr = "getter";
        break;
      case 4:
        kindStr = "setter";
        break;
      default:
        kindStr = "field";
    }
    var ctx = {
      kind: kindStr,
      name: isPrivate ? "#" + name : name,
      static: isStatic,
      private: isPrivate,
      metadata: metadata
    };
    var decoratorFinishedRef = {
      v: false
    };
    ctx.addInitializer = createAddInitializerMethod(initializers, decoratorFinishedRef);
    var get, set;
    if (kind === 0) {
      if (isPrivate) {
        get = desc.get;
        set = desc.set;
      } else {
        get = function () {
          return this[name];
        };
        set = function (v) {
          this[name] = v;
        };
      }
    } else if (kind === 2) {
      get = function () {
        return desc.value;
      };
    } else {
      if (kind === 1 || kind === 3) {
        get = function () {
          return desc.get.call(this);
        };
      }
      if (kind === 1 || kind === 4) {
        set = function (v) {
          desc.set.call(this, v);
        };
      }
    }
    ctx.access = get && set ? {
      get: get,
      set: set
    } : get ? {
      get: get
    } : {
      set: set
    };
    try {
      return dec(value, ctx);
    } finally {
      decoratorFinishedRef.v = true;
    }
  }
  function assertNotFinished(decoratorFinishedRef, fnName) {
    if (decoratorFinishedRef.v) {
      throw new Error("attempted to call " + fnName + " after decoration was finished");
    }
  }
  function assertCallable(fn, hint) {
    if (typeof fn !== "function") {
      throw new TypeError(hint + " must be a function");
    }
  }
  function assertValidReturnValue(kind, value) {
    var type = typeof value;
    if (kind === 1) {
      if (type !== "object" || value === null) {
        throw new TypeError("accessor decorators must return an object with get, set, or init properties or void 0");
      }
      if (value.get !== undefined) {
        assertCallable(value.get, "accessor.get");
      }
      if (value.set !== undefined) {
        assertCallable(value.set, "accessor.set");
      }
      if (value.init !== undefined) {
        assertCallable(value.init, "accessor.init");
      }
    } else if (type !== "function") {
      var hint;
      if (kind === 0) {
        hint = "field";
      } else if (kind === 10) {
        hint = "class";
      } else {
        hint = "method";
      }
      throw new TypeError(hint + " decorators must return a function or void 0");
    }
  }
  function applyMemberDec(ret, base, decInfo, name, kind, isStatic, isPrivate, initializers, metadata) {
    var decs = decInfo[0];
    var desc, init, value;
    if (isPrivate) {
      if (kind === 0 || kind === 1) {
        desc = {
          get: decInfo[3],
          set: decInfo[4]
        };
      } else if (kind === 3) {
        desc = {
          get: decInfo[3]
        };
      } else if (kind === 4) {
        desc = {
          set: decInfo[3]
        };
      } else {
        desc = {
          value: decInfo[3]
        };
      }
    } else if (kind !== 0) {
      desc = Object.getOwnPropertyDescriptor(base, name);
    }
    if (kind === 1) {
      value = {
        get: desc.get,
        set: desc.set
      };
    } else if (kind === 2) {
      value = desc.value;
    } else if (kind === 3) {
      value = desc.get;
    } else if (kind === 4) {
      value = desc.set;
    }
    var newValue, get, set;
    if (typeof decs === "function") {
      newValue = memberDec(decs, name, desc, initializers, kind, isStatic, isPrivate, metadata, value);
      if (newValue !== void 0) {
        assertValidReturnValue(kind, newValue);
        if (kind === 0) {
          init = newValue;
        } else if (kind === 1) {
          init = newValue.init;
          get = newValue.get || value.get;
          set = newValue.set || value.set;
          value = {
            get: get,
            set: set
          };
        } else {
          value = newValue;
        }
      }
    } else {
      for (var i = decs.length - 1; i >= 0; i--) {
        var dec = decs[i];
        newValue = memberDec(dec, name, desc, initializers, kind, isStatic, isPrivate, metadata, value);
        if (newValue !== void 0) {
          assertValidReturnValue(kind, newValue);
          var newInit;
          if (kind === 0) {
            newInit = newValue;
          } else if (kind === 1) {
            newInit = newValue.init;
            get = newValue.get || value.get;
            set = newValue.set || value.set;
            value = {
              get: get,
              set: set
            };
          } else {
            value = newValue;
          }
          if (newInit !== void 0) {
            if (init === void 0) {
              init = newInit;
            } else if (typeof init === "function") {
              init = [init, newInit];
            } else {
              init.push(newInit);
            }
          }
        }
      }
    }
    if (kind === 0 || kind === 1) {
      if (init === void 0) {
        init = function (instance, init) {
          return init;
        };
      } else if (typeof init !== "function") {
        var ownInitializers = init;
        init = function (instance, init) {
          var value = init;
          for (var i = 0; i < ownInitializers.length; i++) {
            value = ownInitializers[i].call(instance, value);
          }
          return value;
        };
      } else {
        var originalInitializer = init;
        init = function (instance, init) {
          return originalInitializer.call(instance, init);
        };
      }
      ret.push(init);
    }
    if (kind !== 0) {
      if (kind === 1) {
        desc.get = value.get;
        desc.set = value.set;
      } else if (kind === 2) {
        desc.value = value;
      } else if (kind === 3) {
        desc.get = value;
      } else if (kind === 4) {
        desc.set = value;
      }
      if (isPrivate) {
        if (kind === 1) {
          ret.push(function (instance, args) {
            return value.get.call(instance, args);
          });
          ret.push(function (instance, args) {
            return value.set.call(instance, args);
          });
        } else if (kind === 2) {
          ret.push(value);
        } else {
          ret.push(function (instance, args) {
            return value.call(instance, args);
          });
        }
      } else {
        Object.defineProperty(base, name, desc);
      }
    }
  }
  function applyMemberDecs(Class, decInfos, metadata) {
    var ret = [];
    var protoInitializers;
    var staticInitializers;
    var existingProtoNonFields = new Map();
    var existingStaticNonFields = new Map();
    for (var i = 0; i < decInfos.length; i++) {
      var decInfo = decInfos[i];
      if (!Array.isArray(decInfo)) continue;
      var kind = decInfo[1];
      var name = decInfo[2];
      var isPrivate = decInfo.length > 3;
      var isStatic = kind >= 5;
      var base;
      var initializers;
      if (isStatic) {
        base = Class;
        kind = kind - 5;
        staticInitializers = staticInitializers || [];
        initializers = staticInitializers;
      } else {
        base = Class.prototype;
        protoInitializers = protoInitializers || [];
        initializers = protoInitializers;
      }
      if (kind !== 0 && !isPrivate) {
        var existingNonFields = isStatic ? existingStaticNonFields : existingProtoNonFields;
        var existingKind = existingNonFields.get(name) || 0;
        if (existingKind === true || existingKind === 3 && kind !== 4 || existingKind === 4 && kind !== 3) {
          throw new Error("Attempted to decorate a public method/accessor that has the same name as a previously decorated public method/accessor. This is not currently supported by the decorators plugin. Property name was: " + name);
        } else if (!existingKind && kind > 2) {
          existingNonFields.set(name, kind);
        } else {
          existingNonFields.set(name, true);
        }
      }
      applyMemberDec(ret, base, decInfo, name, kind, isStatic, isPrivate, initializers, metadata);
    }
    pushInitializers(ret, protoInitializers);
    pushInitializers(ret, staticInitializers);
    return ret;
  }
  function pushInitializers(ret, initializers) {
    if (initializers) {
      ret.push(function (instance) {
        for (var i = 0; i < initializers.length; i++) {
          initializers[i].call(instance);
        }
        return instance;
      });
    }
  }
  function applyClassDecs(targetClass, classDecs, metadata) {
    if (classDecs.length > 0) {
      var initializers = [];
      var newClass = targetClass;
      var name = targetClass.name;
      for (var i = classDecs.length - 1; i >= 0; i--) {
        var decoratorFinishedRef = {
          v: false
        };
        try {
          var nextNewClass = classDecs[i](newClass, {
            kind: "class",
            name: name,
            addInitializer: createAddInitializerMethod(initializers, decoratorFinishedRef),
            metadata
          });
        } finally {
          decoratorFinishedRef.v = true;
        }
        if (nextNewClass !== undefined) {
          assertValidReturnValue(10, nextNewClass);
          newClass = nextNewClass;
        }
      }
      return [defineMetadata(newClass, metadata), function () {
        for (var i = 0; i < initializers.length; i++) {
          initializers[i].call(newClass);
        }
      }];
    }
  }
  function defineMetadata(Class, metadata) {
    return Object.defineProperty(Class, Symbol.metadata || Symbol.for("Symbol.metadata"), {
      configurable: true,
      enumerable: true,
      value: metadata
    });
  }
  return function applyDecs2203R(targetClass, memberDecs, classDecs, parentClass) {
    if (parentClass !== void 0) {
      var parentMetadata = parentClass[Symbol.metadata || Symbol.for("Symbol.metadata")];
    }
    var metadata = Object.create(parentMetadata === void 0 ? null : parentMetadata);
    var e = applyMemberDecs(targetClass, memberDecs, metadata);
    if (!classDecs.length) defineMetadata(targetClass, metadata);
    return {
      e: e,
      get c() {
        return applyClassDecs(targetClass, classDecs, metadata);
      }
    };
  };
}
function _apply_decs_2203_r(targetClass, memberDecs, classDecs, parentClass) {
  return (_apply_decs_2203_r = applyDecs2203RFactory())(targetClass, memberDecs, classDecs, parentClass);
}
function _identity(x) {
  return x;
}
var _class;
var _initClass, _React_Component;
const LayerLevel = {
  CONTENT: 0,
  PINS: 1,
  LIGHTBOXES: 2,
  MODALS: 3,
  COMPLIANCE_NOTICES: 4,
  TOASTS: 5,
  TOOLTIPS: 6
};
class LayerError extends Error {}
const LayerContext = _react.createContext({});
function WithLayerParent({
  parentLayer,
  children
}) {
  return (0, _jsxruntime.jsx)(LayerContext.Provider, {
    value: {
      parentLayer
    },
    children: children
  });
}
function Layer({
  open,
  children,
  level,
  onOutsideLayerPointerDown,
  parentLayer,
  markOutsideInert
}) {
  const interactionLayerRef = (0, _interaction_manager.useInteractionLayer)({
    markOutsideInert: markOutsideInert !== null && markOutsideInert !== void 0 ? markOutsideInert : false
  });
  return (0, _jsxruntime.jsx)(_LayerInternal, {
    open: open,
    children: children,
    level: level,
    onOutsideLayerPointerDown: onOutsideLayerPointerDown,
    parentLayer: parentLayer,
    interactionLayerRef: interactionLayerRef,
    markOutsideInert: markOutsideInert
  });
}
let _LayerInternal;
new (_class = class extends _identity {
  constructor() {
    super(_LayerInternal), _initClass();
  }
}, (() => {
  class LayerInternal extends (_React_Component = _react.Component) {
    static _makeObservable(instance) {
      (0, _make_observable.makeObservable)(instance, {
        layerContent: _mobx.observable.ref,
        close: _mobx.action,
        show: _mobx.action
      });
    }
    componentDidMount() {
      this.parentLayer = this.props.parentLayer || this.context.parentLayer || document.body;
      this.parentDocument = this.parentLayer && this.parentLayer.ownerDocument || document;
      (0, _dispose_on_unmount.disposeOnUnmount)(this, [_mobx.reaction(() => this.props.open, open => open ? this.show(this.props.level) : this.close(), {
        fireImmediately: true
      }), _mobx.reaction(() => this.props.onOutsideLayerPointerDown, outsideLayerPointerDown => {
        outsideLayerPointerDown != null ? this.maybeAddPointerEventListeners() : this.maybeRemovePointerEventListeners();
      }, {
        fireImmediately: true
      })]);
    }
    componentWillUnmount() {
      this.maybeRemovePointerEventListeners();
      this.close();
    }
    maybeAddPointerEventListeners() {
      if (!this.hasEventListeners && this.layerRoot != null && this.parentDocument) {
        this.parentDocument.addEventListener('mousedown', this.onDocumentMouseDown);
        this.parentDocument.addEventListener('touchstart', this.onDocumentTouchStart);
        this.hasEventListeners = true;
      }
    }
    maybeRemovePointerEventListeners() {
      if (this.hasEventListeners && this.parentDocument) {
        this.parentDocument.removeEventListener('mousedown', this.onDocumentMouseDown);
        this.parentDocument.removeEventListener('touchstart', this.onDocumentTouchStart);
        this.hasEventListeners = false;
      }
    }
    close() {
      var
      _this_parentDocument;
      if (!this.layerRoot || !this.layerContent) return;
      if (this.parentLayer == null || this.layerRoot.parentNode !== this.parentLayer) throw new LayerError('Closed before mount.');
      this.props.interactionLayerRef(null);
      this.parentLayer.removeChild(this.layerRoot);
      this.layerRoot = undefined;
      this.layerContent = undefined;
      (_this_parentDocument = this.parentDocument) === null || _this_parentDocument === void 0 ? void 0 : _this_parentDocument.dispatchEvent(new this.parentGlobal.CustomEvent('layerclose'));
    }
    get parentGlobal() {
      var _this_parentDocument;
      return ((_this_parentDocument = this.parentDocument) === null || _this_parentDocument === void 0 ? void 0 : _this_parentDocument.defaultView) || window;
    }
    show(level) {
      if (!this.parentDocument) return;
      if (!this.layerRoot || !this.layerContent) {
        if (!this.parentLayer) throw new LayerError('Shown before mount.');
        this.layerContent = this.parentDocument.createElement('div');
        this.layerContent.classList.add(_layercss.default.layerContent);
        this.layerRoot = this.parentDocument.createElement('div');
        if (this.parentLayer === this.parentDocument.body)
          this.layerRoot.addEventListener('click', () => 0);
        this.layerRoot.classList.add(_internal.LAYER_MARKER_CLASS_NAME);
        this.layerRoot.classList.add(_layercss.default.layerRoot);
        this.layerRoot.appendChild(this.layerContent);
        (0, _internal.insertLayerIntoParent)(this.layerRoot, this.parentLayer);
        this.props.interactionLayerRef(this.layerContent);
      }
      this.layerRoot.style.zIndex = String(level != null ? level : 1);
    }
    render() {
      if (this.layerContent) return _reactdom.createPortal((0, _jsxruntime.jsx)(LayerContext.Provider, {
        value: {
          parentLayer: this.layerRoot
        },
        children: (0, _jsxruntime.jsx)(_theme.WithThemeData, {
          children: data => (0, _jsxruntime.jsx)("div", {
            className: data.className,
            children: this.props.children
          })
        })
      }), this.layerContent);
      return null;
    }
    constructor(...args) {
      super(...args), this.layerRoot = (_LayerInternal._makeObservable(this), undefined), this.layerContent = undefined, this.parentDocument = typeof document === 'undefined' ? undefined : document, this.hasEventListeners = false, this.onDocumentMouseDown = event => {
        if (this.props.onOutsideLayerPointerDown == null) return;
        if (this.layerRoot == null)
          return;
        if (!(event.target instanceof this.parentGlobal.Node))
          return;
        if (!this.layerRoot.contains(event.target)) this.props.onOutsideLayerPointerDown(event);
      }, this.onDocumentTouchStart = event => {
        this.onDocumentMouseDown(event);
      };
    }
  }
  ({
    c: [_LayerInternal, _initClass]
  } = _apply_decs_2203_r(LayerInternal, [], [_observer_for_class.observerForClass], _React_Component));
  LayerInternal.contextType = LayerContext;
})(), _class)();
