// @canva/app-ui-kit/dist/esm/ui/base/theme/internal.js
import { makeObservable } from '../../../base/make_observable/make_observable';
import { Preconditions, UnreachableError } from '../../../base/preconditions';
import classNames from 'classnames';
import { action, computed, observable } from 'mobx';
import * as React from 'react';
import preloadSelectors from './preload.css';
import styles from './theme.css';
export const stripGlobal = (className)=>{
    var _className_replace;
    return (_className_replace = className === null || className === void 0 ? void 0 : className.replace(/^\s*:global\(\.(.+?)\)\s*$/, '$1')) !== null && _className_replace !== void 0 ? _className_replace : '';
};
let themeClasses;
let themedClass;
function getThemeClassMap() {
    if (themeClasses == null) themeClasses = new Map([
        [
            'light',
            stripGlobal(styles.themeLight)
        ],
        [
            'dark',
            stripGlobal(styles.themeDark)
        ]
    ]);
    return themeClasses;
}
function getThemedClass() {
    if (themedClass == null) themedClass = stripGlobal(styles.themed);
    return themedClass;
}
export function getThemeClasses(theme) {
    const themeClasses = getThemeClassMap();
    return [
        getThemedClass(),
        Preconditions.checkExists(themeClasses.get(theme))
    ];
}
export function getAllPreloadClasses() {
    return [
        getThemedClass(),
        ...getThemeClassMap().values(),
        stripGlobal(preloadSelectors.themePreloadModeClassic)
    ];
}
export class ThemeStore {
    static _makeObservable(instance) {
        makeObservable(instance, {
            parent: observable.ref,
            _appearance: observable.ref,
            themeMapping: observable.deep,
            setParent: action,
            setAppearance: action,
            appearance: computed,
            setThemeMapping: action,
            currentTheme: computed
        });
    }
    setParent(store) {
        this.parent = store;
    }
    setAppearance(appearance) {
        this._appearance = appearance;
    }
    get appearance() {
        var _this_parent;
        var _this__appearance;
        return (_this__appearance = this._appearance) !== null && _this__appearance !== void 0 ? _this__appearance : (_this_parent = this.parent) === null || _this_parent === void 0 ? void 0 : _this_parent.appearance;
    }
    setThemeMapping(themeMapping) {
        this.themeMapping = themeMapping;
    }
    get currentTheme() {
        var _this_parent;
        const appearance = this.appearance;
        if (appearance == null)
            return undefined;
        var _this_themeMapping_appearance;
        return (_this_themeMapping_appearance = this.themeMapping[appearance]) !== null && _this_themeMapping_appearance !== void 0 ? _this_themeMapping_appearance : (_this_parent = this.parent) === null || _this_parent === void 0 ? void 0 : _this_parent.currentTheme;
    }
    constructor(){
        this.parent = (ThemeStore._makeObservable(this), undefined);
        this.themeMapping = {};
    }
}
export class ThemeData {
    static _makeObservable(instance) {
        makeObservable(instance, {
            classNames: computed,
            className: computed,
            mode: computed
        });
    }
    get classNames() {
        const { currentTheme } = this.store;
        if (currentTheme == null) return [];
        return getThemeClasses(currentTheme);
    }
    get className() {
        const classes = this.classNames;
        if (classes.length === 0) return undefined;
        return classNames(classes);
    }
    get mode() {
        switch(this.store.appearance){
            case 'dark':
            case 'light':
                return 'modern';
            case 'classicDark':
            case 'classicLight':
                return 'classic';
            case undefined:
                return undefined;
            default:
                throw new UnreachableError(this.store.appearance);
        }
    }
    constructor(store){
        this.store = (ThemeData._makeObservable(this), undefined);
        this.store = store;
    }
}
export const rootThemeStore = new ThemeStore();
const rootThemeData = new ThemeData(rootThemeStore);
export const ThemeContext = React.createContext({
    store: rootThemeStore,
    data: rootThemeData
});
