// @canva/app-ui-kit/dist/cjs/ui/base/typography/truncated/private/truncated.js
"use strict"
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: Object.getOwnPropertyDescriptor(all, name).get
    });
}
_export(exports, {
    get TruncatedText () {
        return TruncatedText;
    },
    get TruncatedTitle () {
        return TruncatedTitle;
    }
});
const _jsxruntime = require("react/jsx-runtime");
const _react = _interop_require_wildcard(require("react"));
const _tooltip = require('../../../tooltip/tooltip');
const _typography = require('../../internal/typography');
const _typographycss = _interop_require_default(require('../../internal/typography.css'));
const _typography1 = require('../../typography');
const _is_truncated = require("./is_truncated");
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function _getRequireWildcardCache(nodeInterop) {
    if (typeof WeakMap !== "function") return null;
    var cacheBabelInterop = new WeakMap();
    var cacheNodeInterop = new WeakMap();
    return (_getRequireWildcardCache = function(nodeInterop) {
        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
    })(nodeInterop);
}
function _interop_require_wildcard(obj, nodeInterop) {
    if (!nodeInterop && obj && obj.__esModule) {
        return obj;
    }
    if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
        return {
            default: obj
        };
    }
    var cache = _getRequireWildcardCache(nodeInterop);
    if (cache && cache.has(obj)) {
        return cache.get(obj);
    }
    var newObj = {
        __proto__: null
    };
    var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
    for(var key in obj){
        if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
            if (desc && (desc.get || desc.set)) {
                Object.defineProperty(newObj, key, desc);
            } else {
                newObj[key] = obj[key];
            }
        }
    }
    newObj.default = obj;
    if (cache) {
        cache.set(obj, newObj);
    }
    return newObj;
}
const TruncatedText = _react.memo((props)=>{
    const { lineClamp = 1, tooltipPlacement, tooltipAlignment, children, tagName: TagName = 'p', measure = 'onResize', ...textProps } = props;
    const { ref, isTruncated } = (0, _is_truncated.useIsTruncated)(children, measure);
    return ((0, _jsxruntime.jsx)(_tooltip.Tooltip, {
                label: children,
                placement: tooltipPlacement,
                disabled: !isTruncated,
                align: tooltipAlignment,
                children: ({ tooltipId, ...triggerProps })=>{
                    return ((0, _jsxruntime.jsx)(TagName, {
                                ...(isTruncated ? triggerProps : {}),
                                className: _typographycss.default.base,
                                children: (0, _jsxruntime.jsx)(_typography1.Text, {
                                    ...textProps,
                                    lineClamp: lineClamp,
                                    tagName: "span",
                                    ref: ref,
                                    children: children
                                })
                            }));
                }
            }));
});
const TruncatedTitle = _react.memo((props)=>{
    const { lineClamp = 1, tooltipPlacement, children, tagName: TagName = (0, _typography.getDefaultHeadingTagName)(props.size), measure = 'onResize', ...titleProps } = props;
    const { ref, isTruncated } = (0, _is_truncated.useIsTruncated)(children, measure);
    return ((0, _jsxruntime.jsx)(_tooltip.Tooltip, {
                label: children,
                placement: tooltipPlacement,
                disabled: !isTruncated,
                children: ({ tooltipId, ...triggerProps })=>{
                    return ((0, _jsxruntime.jsx)(TagName, {
                                ...(isTruncated ? triggerProps : {}),
                                className: _typographycss.default.base,
                                children: (0, _jsxruntime.jsx)(_typography1.Title, {
                                    ...titleProps,
                                    lineClamp: lineClamp,
                                    tagName: "span",
                                    ref: ref,
                                    children: children
                                })
                            }));
                }
            }));
});
