// @canva/app-ui-kit/dist/cjs/ui/base/color_picker/alpha_slider/private/alpha_slider.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "AlphaSlider", {
    enumerable: true,
    get: function() {
        return AlphaSlider;
    }
});
const _jsxruntime = require("react/jsx-runtime");
const _classnames = _interop_require_default(require("classnames"));
const _react = _interop_require_wildcard(require("react"));
const _slider = require('../../../form/slider/slider');
const _alpha_slidercss = _interop_require_default(require("./alpha_slider.css"));
const _alpha_slidermessages = require("./alpha_slider.messages");
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function _getRequireWildcardCache(nodeInterop) {
    if (typeof WeakMap !== "function") return null;
    var cacheBabelInterop = new WeakMap();
    var cacheNodeInterop = new WeakMap();
    return (_getRequireWildcardCache = function(nodeInterop) {
        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
    })(nodeInterop);
}
function _interop_require_wildcard(obj, nodeInterop) {
    if (!nodeInterop && obj && obj.__esModule) {
        return obj;
    }
    if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
        return {
            default: obj
        };
    }
    var cache = _getRequireWildcardCache(nodeInterop);
    if (cache && cache.has(obj)) {
        return cache.get(obj);
    }
    var newObj = {
        __proto__: null
    };
    var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
    for(var key in obj){
        if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
            if (desc && (desc.get || desc.set)) {
                Object.defineProperty(newObj, key, desc);
            } else {
                newObj[key] = obj[key];
            }
        }
    }
    newObj.default = obj;
    if (cache) {
        cache.set(obj, newObj);
    }
    return newObj;
}
const handle = ({ focused, trackHovering })=>(0, _jsxruntime.jsx)("div", {
        className: (0, _classnames.default)(_alpha_slidercss.default.handle, {
            [_alpha_slidercss.default.handleHoverOrFocus]: focused || trackHovering
        })
    });
const AlphaSlider = ({ alpha, color, onChange: onChangeProp, onChangeStart, onChangeComplete, ariaLabel = _alpha_slidermessages.AlphaSliderMessages.colorAlphaLabel() })=>{
    const track = _react.useCallback(()=>(0, _jsxruntime.jsx)("div", {
            className: _alpha_slidercss.default.track,
            children: (0, _jsxruntime.jsx)("div", {
                className: _alpha_slidercss.default.overlay,
                style: {
                    background: `linear-gradient(to right, transparent, ${color})`
                }
            })
        }), [
        color
    ]);
    const onChange = _react.useCallback((value)=>{
        onChangeProp === null || onChangeProp === void 0 ? void 0 : onChangeProp(value);
    }, [
        onChangeProp
    ]);
    return (0, _jsxruntime.jsx)("div", {
        className: _alpha_slidercss.default.container,
        children: (0, _jsxruntime.jsx)("div", {
            className: _alpha_slidercss.default.widener,
            children: (0, _jsxruntime.jsx)(_slider.BaseSlider, {
                value: alpha,
                min: 0,
                max: 1,
                step: 0.005,
                onChange: onChange,
                onDragStart: onChangeStart,
                onDragEnd: onChangeComplete,
                ariaLabel: ariaLabel,
                fillColor: "transparent",
                handle: handle,
                track: track
            })
        })
    });
};
