// @canva/app-ui-kit/dist/esm/ui/base/a11y/announcer/private/accessibility_announcer_region.js
import { jsx as _jsx, jsxs as _jsxs, Fragment as _Fragment } from "react/jsx-runtime";
import { useEffect, useRef } from 'react';
import styles from './accessibility_announcer_region.css';
export const AccessibilityAnnouncerRegion = ({ onMount, onUnmount })=>{
    const assertiveRef = useRef(null);
    const politeRef = useRef(null);
    useEffect(()=>{
        if (assertiveRef.current && politeRef.current) onMount({
            assertiveRegion: assertiveRef,
            politeRegion: politeRef
        });
        return ()=>{
            onUnmount();
        };
    }, [
        onMount,
        onUnmount
    ]);
    return _jsxs(_Fragment, {
        children: [
            _jsx("div", {
                ref: assertiveRef,
                className: styles.visuallyHidden,
                "aria-atomic": "true",
                "aria-live": "assertive"
            }),
            _jsx("div", {
                ref: politeRef,
                className: styles.visuallyHidden,
                "aria-atomic": "true",
                "aria-live": "polite"
            })
        ]
    });
};
