// @canva/app-ui-kit/dist/esm/ui/base/a11y/announcer/private/announcer_fallback.js
import { useCallback, useEffect, useMemo, useRef } from 'react';
import { useWindow } from '../../screen_reader_content/screen_reader_content';
import styles from './accessibility_announcer_region.css';
import { updateLiveRegionContent } from './announcer_utils';
const FALLBACK_ASSERTIVE_ID = 'canva-a11y-fallback-assertive';
const FALLBACK_POLITE_ID = 'canva-a11y-fallback-polite';
function createLiveRegion(ariaLive, id, document) {
    const element = document.createElement('div');
    element.id = id;
    element.setAttribute('aria-live', ariaLive);
    element.setAttribute('aria-atomic', 'true');
    element.className = styles.visuallyHidden;
    return element;
}
function ensureFallbackRegions(document) {
    let assertiveRegion = null;
    let politeRegion = null;
    assertiveRegion = document.getElementById(FALLBACK_ASSERTIVE_ID);
    if (!assertiveRegion) {
        assertiveRegion = createLiveRegion('assertive', FALLBACK_ASSERTIVE_ID, document);
        document.body.prepend(assertiveRegion);
    }
    politeRegion = document.getElementById(FALLBACK_POLITE_ID);
    if (!politeRegion) {
        politeRegion = createLiveRegion('polite', FALLBACK_POLITE_ID, document);
        document.body.prepend(politeRegion);
    }
    return {
        assertiveRegion,
        politeRegion
    };
}
async function announceFallback(message, document, regions, priority = 'medium') {
    if (!message) return;
    let { assertiveRegion, politeRegion } = regions;
    if (!assertiveRegion || !politeRegion) {
        const newRegions = ensureFallbackRegions(document);
        assertiveRegion = newRegions.assertiveRegion;
        politeRegion = newRegions.politeRegion;
    }
    if (assertiveRegion && (priority === 'critical' || priority === 'high')) await updateLiveRegionContent(assertiveRegion, message);
    else if (politeRegion) await updateLiveRegionContent(politeRegion, message);
}
function deferFallbackAnnouncement(message, document, regions, priority) {
    let announced = false;
    return ()=>{
        if (announced) return;
        announced = true;
        announceFallback(message, document, regions, priority);
    };
}
const NO_OP_CONTROLLER = {
    announce: ()=>{},
    deferAnnouncement: ()=>()=>{}
};
export function useFallbackAnnouncer(shouldUseFallback = true) {
    const window = useWindow();
    const assertiveRegionRef = useRef(null);
    const politeRegionRef = useRef(null);
    useEffect(()=>{
        if (shouldUseFallback && (window === null || window === void 0 ? void 0 : window.document)) {
            const { assertiveRegion, politeRegion } = ensureFallbackRegions(window.document);
            assertiveRegionRef.current = assertiveRegion;
            politeRegionRef.current = politeRegion;
        }
    }, [
        shouldUseFallback,
        window
    ]);
    const announce = useCallback((message, priority = 'medium')=>{
        if (shouldUseFallback && (window === null || window === void 0 ? void 0 : window.document)) {
            const regions = {
                assertiveRegion: assertiveRegionRef.current,
                politeRegion: politeRegionRef.current
            };
            announceFallback(message, window.document, regions, priority);
        }
    }, [
        shouldUseFallback,
        window
    ]);
    const deferAnnouncement = useCallback((message, priority = 'medium')=>{
        if (shouldUseFallback && (window === null || window === void 0 ? void 0 : window.document)) {
            const regions = {
                assertiveRegion: assertiveRegionRef.current,
                politeRegion: politeRegionRef.current
            };
            return deferFallbackAnnouncement(message, window.document, regions, priority);
        } else
            return ()=>{};
    }, [
        shouldUseFallback,
        window
    ]);
    return useMemo(()=>{
        return shouldUseFallback ? {
            announce,
            deferAnnouncement
        } : NO_OP_CONTROLLER;
    }, [
        shouldUseFallback,
        announce,
        deferAnnouncement
    ]);
}
