// @canva/app-ui-kit/dist/cjs/ui/apps/developing/ui_kit/components/card/horizontal/horizontal.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "HorizontalCard", {
    enumerable: true,
    get: function() {
        return HorizontalCard;
    }
});
const _jsxruntime = require("react/jsx-runtime");
require("react");
const _box = require('../../../../../../base/box/box');
const _card = require('../../../../../../base/card/card');
const _icon = require('../../../../../../base/icons/animated_spinner/icon');
const _icon1 = require('../../../../../../base/icons/template/icon');
const _internal_card = require("../internal_card");
const _horizontalcss = _interop_require_default(require("./horizontal.css"));
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
const HorizontalCard = ({ title, ariaLabel, description, thumbnail, onClick, bottomEnd, bottomEndVisibility, disabled, loading })=>{
    return (0, _jsxruntime.jsx)(_card.Card, {
        layout: "horizontal",
        hoverBoundary: "card",
        padding: "1u",
        hoverEffect: "bgColorTransition",
        title: title,
        description: description,
        ariaLabel: ariaLabel,
        onClick: disabled !== true && loading !== true ? onClick : undefined,
        disabled: disabled,
        decorators: (0, _jsxruntime.jsxs)(_jsxruntime.Fragment, {
            children: [
                bottomEnd && (0, _jsxruntime.jsx)(_card.CardDecorator, {
                    location: "bottom-end",
                    visibility: bottomEndVisibility,
                    isInteractive: true,
                    children: bottomEnd
                }),
                loading && (0, _jsxruntime.jsx)(_box.Box, {
                    className: _horizontalcss.default.loading,
                    display: "flex",
                    justifyContent: "center",
                    alignItems: "center",
                    width: "full",
                    height: "full",
                    children: (0, _jsxruntime.jsx)(_icon.AnimatedSpinnerIcon, {
                        size: "medium",
                        className: _horizontalcss.default.spinner
                    })
                })
            ]
        }),
        thumbnail: (0, _jsxruntime.jsxs)(_jsxruntime.Fragment, {
            children: [
                thumbnail && ('url' in thumbnail ? (0, _jsxruntime.jsx)(_card.CardImageThumbnail, {
                    src: thumbnail.url,
                    alt: thumbnail.alt,
                    width: _internal_card.CARD_THUMBNAIL_SIDE_LENGTH_PX,
                    height: _internal_card.CARD_THUMBNAIL_SIDE_LENGTH_PX,
                    borderRadius: "elementSmall"
                }) : (0, _jsxruntime.jsx)(_card.CardIconThumbnail, {
                    Icon: thumbnail.icon,
                    width: _internal_card.CARD_THUMBNAIL_SIDE_LENGTH_PX,
                    height: _internal_card.CARD_THUMBNAIL_SIDE_LENGTH_PX,
                    borderRadius: "elementSmall",
                    border: "none"
                })),
                !thumbnail && (0, _jsxruntime.jsx)(_card.CardIconThumbnail, {
                    Icon: _icon1.TemplateIcon,
                    width: _internal_card.CARD_THUMBNAIL_SIDE_LENGTH_PX,
                    height: _internal_card.CARD_THUMBNAIL_SIDE_LENGTH_PX,
                    borderRadius: "elementSmall",
                    border: "none"
                })
            ]
        })
    });
};
