// @canva/app-ui-kit/dist/esm/ui/base/form/input_decorators/private/chars_remaining.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import classNames from 'classnames';
import 'react';
import { useAnnounceOnUpdate } from '../../../a11y/announcer/announcer';
import { Text } from '../../../typography/typography';
import styles from './chars_remaining.css';
import { InputDecoratorsMessages } from './input_decorators.messages';
export const CharsRemaining = ({ count, max, margins = 'default', textSize, id })=>{
    const priority = count > max ? 'high' : 'medium';
    useAnnounceOnUpdate(InputDecoratorsMessages.charsUsed(count, max), count, {
        priority,
        debounceMs: 200
    });
    return _jsx("div", {
        className: classNames(styles.charsRemaining, {
            [styles.noMargins]: margins === 'none',
            [styles.verticalMargins]: margins === 'vertical',
            [styles.horizontalMargins]: margins === 'horizontal'
        }),
        children: _jsx(Text, {
            size: textSize,
            margins: "legacy",
            alignment: "end",
            tone: count > max ? 'critical' : 'tertiary',
            children: _jsxs("span", {
                id: id,
                "aria-label": InputDecoratorsMessages.charsUsed(count, max),
                children: [
                    count,
                    "/",
                    max
                ]
            })
        })
    });
};
