// @canva/app-ui-kit/dist/cjs/ui/base/carousel/private/scroll_button.js
"use strict"
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: Object.getOwnPropertyDescriptor(all, name).get
    });
}
_export(exports, {
    get ChevronScrollButton () {
        return ChevronScrollButton;
    },
    get CircularScrollButton () {
        return CircularScrollButton;
    }
});
const _jsxruntime = require("react/jsx-runtime");
const _classnames = _interop_require_default(require("classnames"));
const _mobxreactlite = require("mobx-react-lite");
const _react = _interop_require_wildcard(require("react"));
const _access_mode_state = require('../../a11y/access_mode_state/access_mode_state');
const _pointer_input_state = require('../../a11y/pointer_input_state/pointer_input_state');
const _icon = require('../../icons/chevron_left/icon');
const _icon1 = require('../../icons/chevron_right/icon');
const _scroll_buttoncss = _interop_require_wildcard(require("./scroll_button.css"));
const _scroll_buttonmessages = require("./scroll_button.messages");
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function _getRequireWildcardCache(nodeInterop) {
    if (typeof WeakMap !== "function") return null;
    var cacheBabelInterop = new WeakMap();
    var cacheNodeInterop = new WeakMap();
    return (_getRequireWildcardCache = function(nodeInterop) {
        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
    })(nodeInterop);
}
function _interop_require_wildcard(obj, nodeInterop) {
    if (!nodeInterop && obj && obj.__esModule) {
        return obj;
    }
    if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
        return {
            default: obj
        };
    }
    var cache = _getRequireWildcardCache(nodeInterop);
    if (cache && cache.has(obj)) {
        return cache.get(obj);
    }
    var newObj = {
        __proto__: null
    };
    var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
    for(var key in obj){
        if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
            if (desc && (desc.get || desc.set)) {
                Object.defineProperty(newObj, key, desc);
            } else {
                newObj[key] = obj[key];
            }
        }
    }
    newObj.default = obj;
    if (cache) {
        cache.set(obj, newObj);
    }
    return newObj;
}
const getClassName = ({ className, active, direction, variant, hoverSupported })=>(0, _classnames.default)(_scroll_buttoncss.default.base, (0, _scroll_buttoncss.getStyle)(variant), {
        [_scroll_buttoncss.default.active]: active,
        [_scroll_buttoncss.default.hoverSupported]: hoverSupported,
        [_scroll_buttoncss.default.forward]: direction === 'forward',
        [_scroll_buttoncss.default.backward]: direction === 'backward'
    }, className);
function getButtonContent(props) {
    const { direction } = props;
    if (direction === 'forward') return (0, _jsxruntime.jsx)(_icon1.ChevronRightIcon, {
        size: "medium",
        className: _scroll_buttoncss.default.icon
    });
    else return (0, _jsxruntime.jsx)(_icon.ChevronLeftIcon, {
        size: "medium",
        className: _scroll_buttoncss.default.icon
    });
}
const focusElement = (ref)=>{
    var _ref_current;
    ref === null || ref === void 0 ? void 0 : (_ref_current = ref.current) === null || _ref_current === void 0 ? void 0 : _ref_current.focus();
};
const Button = (0, _mobxreactlite.observer)((props)=>{
    const shouldFocusAfterClick = _react.useRef(false);
    _react.useLayoutEffect(()=>{
        if (shouldFocusAfterClick.current) {
            if (props.scrollState.atStart) {
                if (_access_mode_state.accessModeState.isMouseMode)
                    focusElement(props.oppositeButtonRef);
                else
                    focusElement(props.startLinkRef);
            }
            if (props.scrollState.atEnd) {
                if (_access_mode_state.accessModeState.isMouseMode)
                    focusElement(props.oppositeButtonRef);
                else
                    focusElement(props.endLinkRef);
            }
        }
    }, [
        props.scrollState.atStart,
        props.scrollState.atEnd,
        props.oppositeButtonRef,
        props.direction,
        props.buttonRef,
        props.startLinkRef,
        props.endLinkRef
    ]);
    const handleClick = ()=>{
        shouldFocusAfterClick.current = true;
        props.onClick();
    };
    const hoverSupported = (0, _pointer_input_state.useIsHoveringInput)();
    if (props.scrollState.atStart === false && props.direction === 'backward' || props.scrollState.atEnd === false && props.direction === 'forward') return (0, _jsxruntime.jsx)("button", {
        onClick: handleClick,
        className: getClassName({
            active: props.active,
            variant: props.variant,
            direction: props.direction,
            hoverSupported
        }),
        "aria-controls": props.scrollableId,
        "aria-label": props.direction === 'forward' ? _scroll_buttonmessages.ButtonMessages.screenReaderNext() : _scroll_buttonmessages.ButtonMessages.screenReaderPrevious(),
        style: props.style,
        type: "button",
        ref: props.buttonRef,
        children: getButtonContent(props)
    });
    return null;
});
function getButtonStyle(verticalOffsetPx, horizontalOpts) {
    const horizontalDirection = horizontalOpts && (horizontalOpts.direction === 'backward' ? 1 : -1);
    const horizontalOffsetPx = horizontalDirection && horizontalOpts.horizontalOffsetPx * horizontalDirection;
    return {
        [_scroll_buttoncss.customProperties.horizontalOffset]: horizontalOffsetPx ? `${horizontalOffsetPx}px` : undefined,
        [_scroll_buttoncss.customProperties.verticalOffset]: verticalOffsetPx ? `${verticalOffsetPx}px` : undefined
    };
}
const CircularScrollButton = (props)=>{
    return (0, _jsxruntime.jsx)(Button, {
        ...props,
        variant: "circular",
        style: getButtonStyle(props.verticalOffsetPx, props.horizontalOffsetPx ? {
            horizontalOffsetPx: props.horizontalOffsetPx,
            direction: props.direction
        } : undefined)
    });
};
const ChevronScrollButton = (props)=>(0, _jsxruntime.jsx)(Button, {
        ...props,
        variant: "chevron",
        style: getButtonStyle(props.verticalOffsetPx)
    });
