// @canva/app-ui-kit/dist/cjs/ui/apps/developing/ui_kit/components/pill/pill.js
"use strict"
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "Pill", {
    enumerable: true,
    get: function() {
        return Pill;
    }
});
const _jsxruntime = require("react/jsx-runtime");
const _preconditions = require('../../../../../../base/preconditions');
const _pill = require('../../../../../base/pill/pill');
function Pill({ truncateText, ...props }) {
    var _props_showTooltip, _props_showTooltip1, _props_showTooltip2;
    var _props_showTooltip_label;
    return (0, _jsxruntime.jsx)(_pill.Pill, {
        ...props,
        maxWidth: mapMaxWidth(truncateText),
        size: props.size ? props.size : 'medium',
        shape: "rectangle",
        tooltipLabel: typeof props.showTooltip === 'object' ? (_props_showTooltip_label = (_props_showTooltip = props.showTooltip) === null || _props_showTooltip === void 0 ? void 0 : _props_showTooltip.label) !== null && _props_showTooltip_label !== void 0 ? _props_showTooltip_label : props.text : props.showTooltip === true ? props.text : undefined,
        tooltipPlacement: typeof props.showTooltip === 'object' ? (_props_showTooltip1 = props.showTooltip) === null || _props_showTooltip1 === void 0 ? void 0 : _props_showTooltip1.placement : undefined,
        tooltipDisabled: typeof props.showTooltip === 'object' ? (_props_showTooltip2 = props.showTooltip) === null || _props_showTooltip2 === void 0 ? void 0 : _props_showTooltip2.disabled : undefined
    });
}
function mapMaxWidth(truncateText) {
    switch(truncateText){
        case undefined:
            return undefined;
        case true:
            return '25u';
        case false:
            return '100%';
        default:
            throw new _preconditions.UnreachableError(truncateText);
    }
}
