// @canva/app-ui-kit/dist/esm/ui/base/typography/truncated/private/truncated.js
import { jsx as _jsx } from "react/jsx-runtime";
import * as React from 'react';
import { Tooltip } from '../../../tooltip/tooltip';
import { getDefaultHeadingTagName } from '../../internal/typography';
import styles from '../../internal/typography.css';
import { Text, Title } from '../../typography';
import { useIsTruncated } from './is_truncated';
export const TruncatedText = React.memo((props)=>{
    const { lineClamp = 1, tooltipPlacement, tooltipAlignment, children, tagName: TagName = 'p', measure = 'onResize', ...textProps } = props;
    const { ref, isTruncated } = useIsTruncated(children, measure);
    return (_jsx(Tooltip, {
                label: children,
                placement: tooltipPlacement,
                disabled: !isTruncated,
                align: tooltipAlignment,
                children: ({ tooltipId, ...triggerProps })=>{
                    return (_jsx(TagName, {
                                ...(isTruncated ? triggerProps : {}),
                                className: styles.base,
                                children: _jsx(Text, {
                                    ...textProps,
                                    lineClamp: lineClamp,
                                    tagName: "span",
                                    ref: ref,
                                    children: children
                                })
                            }));
                }
            }));
});
export const TruncatedTitle = React.memo((props)=>{
    const { lineClamp = 1, tooltipPlacement, children, tagName: TagName = getDefaultHeadingTagName(props.size), measure = 'onResize', ...titleProps } = props;
    const { ref, isTruncated } = useIsTruncated(children, measure);
    return (_jsx(Tooltip, {
                label: children,
                placement: tooltipPlacement,
                disabled: !isTruncated,
                children: ({ tooltipId, ...triggerProps })=>{
                    return (_jsx(TagName, {
                                ...(isTruncated ? triggerProps : {}),
                                className: styles.base,
                                children: _jsx(Title, {
                                    ...titleProps,
                                    lineClamp: lineClamp,
                                    tagName: "span",
                                    ref: ref,
                                    children: children
                                })
                            }));
                }
            }));
});
