// @canva/app-ui-kit/dist/esm/ui/base/legacy_sheet/internal/layers.js
import { jsx as _jsx } from "react/jsx-runtime";
import { makeObservable } from '../../../../base/make_observable/make_observable';
import * as mobx from 'mobx';
import 'react';
import * as ReactDOMClient from 'react-dom/client';
import { LayerLevel } from '../../layer/layer';
import { insertLayerIntoParent, LAYER_MARKER_CLASS_NAME } from '../../layer/temporary_exports';
import { EaselProvider } from '../../provider/provider';
import { LayerView } from './layer_view';
export class Layers {
    addSheet(sheet) {
        if (this.allSheets.has(sheet) || sheet.window == null)
            return;
        const { from, onBottom } = sheet;
        const openLayers = this.layers.filter((l)=>l.isOpen);
        let layer = onBottom ? openLayers[openLayers.length - 1] : openLayers[0];
        if (!layer || layer.forced || layer.from !== from || sheet.props.forceLayer) {
            var _sheet_props_layerLevel;
            const layerLevel = this.calculateNewLayerLevel(onBottom, (_sheet_props_layerLevel = sheet.props.layerLevel) !== null && _sheet_props_layerLevel !== void 0 ? _sheet_props_layerLevel : LayerLevel.PINS);
            layer = new Layer(from, !!sheet.props.forceLayer, sheet.window, layerLevel);
            layer.root.render(_jsx(EaselProvider, {
                enableAnimations: sheet.easelConfiguration.enableAnimations,
                direction: sheet.easelConfiguration.direction,
                disableFocusTraps: sheet.easelConfiguration.disableFocusTraps,
                enableUserSelection: sheet.easelConfiguration.enableUserSelection,
                theme: sheet.easelConfiguration.theme,
                enableHighColorContrast: sheet.easelConfiguration.enableHighColorContrast,
                autoplayVideos: sheet.easelConfiguration.autoplayVideos,
                disableDialogBlur: sheet.easelConfiguration.disableDialogBlur,
                children: _jsx(LayerView, {
                    layer: layer,
                    onLayerClose: this.onLayerClose,
                    window: sheet.window
                })
            }));
            if (onBottom) {
                const bottomLayer = this.layers[this.layers.length - 1];
                sheet.window.document.body.insertBefore(layer.element, bottomLayer && bottomLayer.element);
                this.layers.push(layer);
            } else {
                insertLayerIntoParent(layer.element, sheet.window.document.body);
                this.layers.unshift(layer);
            }
        }
        layer.add(sheet);
        this.allSheets.add(sheet);
    }
    removeSheet(sheet) {
        const layer = this.layers.find((l)=>l.sheets.includes(sheet));
        if (layer) {
            layer.remove(sheet);
            this.allSheets.delete(sheet);
            !sheet.enableAnimations && this.onLayerClose(layer);
        }
    }
    constructor(){
        this.layers = [];
        this.allSheets = new Set();
        this.calculateNewLayerLevel = (onBottom, layerLevel)=>{
            if (this.layers.length === 0) return layerLevel;
            if (onBottom) {
                const bottomLayerLevel = this.layers[this.layers.length - 1].layerLevel;
                return layerLevel > bottomLayerLevel ? bottomLayerLevel : layerLevel;
            }
            const topLayerLevel = this.layers[0].layerLevel;
            return layerLevel < topLayerLevel ? topLayerLevel : layerLevel;
        };
        this.onLayerClose = (layer)=>{
            if (layer.sheets.length === 0) {
                layer.root.unmount();
                layer.element.remove();
                this.layers.splice(this.layers.indexOf(layer), 1);
            }
        };
    }
}
export class Layer {
    static _makeObservable(instance) {
        makeObservable(instance, {
            current: mobx.observable.ref,
            previous: mobx.observable.ref,
            update: mobx.action
        });
    }
    remove(sheet) {
        this.sheets.splice(this.sheets.indexOf(sheet), 1);
        this.update();
    }
    add(sheet) {
        if (sheet.onBottom) this.sheets.push(sheet);
        else this.sheets.unshift(sheet);
        this.update();
    }
    get isOpen() {
        return this.sheets.some((s)=>s.show);
    }
    update() {
        const newCurrent = this.sheets[0];
        const oldCurrent = this.current;
        if (newCurrent !== oldCurrent) {
            this.previous = this.previous || !this.currentMounted ? undefined : oldCurrent && !oldCurrent.enableAnimations ? undefined : oldCurrent;
            this.current = newCurrent;
        }
    }
    get enableAnimations() {
        const { current, previous } = this;
        return current ? current.enableAnimations : previous ? previous.enableAnimations : false;
    }
    get hasOuterContent() {
        const { current, previous } = this;
        return !!(current && current.outerElement && current.outerElement.hasChildNodes()) || !!(previous && previous.outerElement && previous.outerElement.hasChildNodes());
    }
    constructor(from, forced, argWindow, layerLevel = LayerLevel.PINS){
        this.from = from;
        this.forced = forced;
        this.argWindow = argWindow;
        this.layerLevel = layerLevel;
        this.sheets = (Layer._makeObservable(this), []);
        this.currentMounted = false;
        this.isHorizontal = from === 'left' || from === 'right';
        this.element = argWindow.document.createElement('div');
        this.element.classList.add(LAYER_MARKER_CLASS_NAME);
        this.element.addEventListener('click', ()=>0);
        this.element.style.position = 'relative';
        this.element.style.zIndex = String(layerLevel);
        this.root = ReactDOMClient.createRoot(this.element);
    }
}
