// @canva/app-ui-kit/dist/cjs/ui/base/thumbnail/private/thumbnail_container.js
"use strict"
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: Object.getOwnPropertyDescriptor(all, name).get
    });
}
_export(exports, {
    get SizingContainer () {
        return SizingContainer;
    },
    get ThumbnailContainer () {
        return ThumbnailContainer;
    },
    get ThumbnailContainerContext () {
        return ThumbnailContainerContext;
    },
    get resolveAspectRatio () {
        return resolveAspectRatio;
    },
    get useInThumbnailContainerContext () {
        return useInThumbnailContainerContext;
    }
});
const _jsxruntime = require("react/jsx-runtime");
const _classnames = _interop_require_default(require("classnames"));
const _react = _interop_require_wildcard(require("react"));
const _aspect_ratio = require('../../aspect_ratio/aspect_ratio');
const _metrics = require('../../metrics/metrics');
const _thumbnailcss = _interop_require_default(require("./thumbnail.css"));
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function _getRequireWildcardCache(nodeInterop) {
    if (typeof WeakMap !== "function") return null;
    var cacheBabelInterop = new WeakMap();
    var cacheNodeInterop = new WeakMap();
    return (_getRequireWildcardCache = function(nodeInterop) {
        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
    })(nodeInterop);
}
function _interop_require_wildcard(obj, nodeInterop) {
    if (!nodeInterop && obj && obj.__esModule) {
        return obj;
    }
    if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
        return {
            default: obj
        };
    }
    var cache = _getRequireWildcardCache(nodeInterop);
    if (cache && cache.has(obj)) {
        return cache.get(obj);
    }
    var newObj = {
        __proto__: null
    };
    var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
    for(var key in obj){
        if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
            if (desc && (desc.get || desc.set)) {
                Object.defineProperty(newObj, key, desc);
            } else {
                newObj[key] = obj[key];
            }
        }
    }
    newObj.default = obj;
    if (cache) {
        cache.set(obj, newObj);
    }
    return newObj;
}
const ThumbnailContainerContext = _react.createContext(undefined);
const useInThumbnailContainerContext = ()=>{
    const containerContext = _react.useContext(ThumbnailContainerContext);
    return containerContext != null;
};
const ThumbnailContainer = ({ children, className, background = 'secondary', border = 'none', borderRadius = 'elementSmall', fit = 'cover', padding = '2u', height, width, aspectRatio: aspectRatioProp })=>{
    const aspectRatio = resolveAspectRatio({
        width,
        height,
        aspectRatio: aspectRatioProp
    });
    return (0, _jsxruntime.jsx)(SizingContainer, {
        height: height,
        width: width,
        aspectRatio: aspectRatio,
        children: (0, _jsxruntime.jsx)("div", {
            className: (0, _classnames.default)(_thumbnailcss.default.thumbnailContainer, className, getThumbnailContainerClassNames({
                background,
                border,
                borderRadius,
                padding
            })),
            children: (0, _jsxruntime.jsx)(ThumbnailContainerContext.Provider, {
                value: {
                    aspectRatio,
                    padding,
                    fit
                },
                children: children
            })
        })
    });
};
const getThumbnailContainerClassNames = ({ background, border, borderRadius, padding })=>{
    const backgroundClasses = {
        [_thumbnailcss.default.secondaryBackground]: background === 'secondary',
        [_thumbnailcss.default.contrastOnLight]: background === 'contrastOnLight',
        [_thumbnailcss.default.contrastOnDark]: background === 'contrastOnDark',
        [_thumbnailcss.default.secondaryBackgroundHover]: background === 'secondaryHover',
        [_thumbnailcss.default.contrastOnLightHover]: background === 'contrastOnLightHover',
        [_thumbnailcss.default.contrastOnDarkHover]: background === 'contrastOnDarkHover'
    };
    const borderClasses = {
        [_thumbnailcss.default.border]: border !== 'none',
        [_thumbnailcss.default.lowBorder]: border === 'low',
        [_thumbnailcss.default.elementSmallBorderRadius]: borderRadius === 'elementSmall',
        [_thumbnailcss.default.elementBorderRadius]: borderRadius === 'element',
        [_thumbnailcss.default.elementRelaxedBorderRadius]: borderRadius === 'elementRelaxed'
    };
    const paddingClasses = {
        [_thumbnailcss.default.paddingOneUnit]: padding === '1u',
        [_thumbnailcss.default.paddingTwoUnit]: padding === '2u'
    };
    return (0, _classnames.default)(backgroundClasses, borderClasses, paddingClasses);
};
const SizingContainer = ({ aspectRatio, width, height, children })=>{
    const widthInPixel = (0, _metrics.resolveUnitOrNumberSize)(width);
    const heightInPixel = (0, _metrics.resolveUnitOrNumberSize)(height);
    if (aspectRatio != null) return (0, _jsxruntime.jsx)(_aspect_ratio.AspectRatio, {
        ratio: aspectRatio,
        width: width,
        height: height,
        children: children
    });
    if (width != null || height != null) return (0, _jsxruntime.jsx)("div", {
        style: {
            width: widthInPixel !== null && widthInPixel !== void 0 ? widthInPixel : '100%',
            height: heightInPixel !== null && heightInPixel !== void 0 ? heightInPixel : '100%',
            flexShrink: 0
        },
        children: children
    });
    return children;
};
const resolveAspectRatio = ({ aspectRatio, width, height })=>{
    const widthInPixel = (0, _metrics.resolveUnitOrNumberSize)(width);
    const heightInPixel = (0, _metrics.resolveUnitOrNumberSize)(height);
    return aspectRatio !== null && aspectRatio !== void 0 ? aspectRatio : widthInPixel != null && heightInPixel != null ? widthInPixel / heightInPixel : undefined;
};
