// @canva/app-ui-kit/dist/cjs/ui/base/card/private/decorators/card_button.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: Object.getOwnPropertyDescriptor(all, name).get
    });
}
_export(exports, {
    get CardButton () {
        return CardButton;
    },
    get CardButtonGroup () {
        return CardButtonGroup;
    },
    get CardMenuButton () {
        return CardMenuButton;
    },
    get StatelessCardMenuButton () {
        return StatelessCardMenuButton;
    }
});
const _jsxruntime = require("react/jsx-runtime");
const _mobx = require("mobx");
const _mobxreactlite = require("mobx-react-lite");
const _react = _interop_require_wildcard(require("react"));
const _button = require('../../../button/button');
const _icon = require('../../../icons/more_horizontal/icon');
const _flyout = require('../../../surface/flyout/flyout');
const _card_context = require("../card_context");
const _card_buttonmessages = require("./card_button.messages");
const _card_decorator = require("./card_decorator");
function _getRequireWildcardCache(nodeInterop) {
    if (typeof WeakMap !== "function") return null;
    var cacheBabelInterop = new WeakMap();
    var cacheNodeInterop = new WeakMap();
    return (_getRequireWildcardCache = function(nodeInterop) {
        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
    })(nodeInterop);
}
function _interop_require_wildcard(obj, nodeInterop) {
    if (!nodeInterop && obj && obj.__esModule) {
        return obj;
    }
    if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
        return {
            default: obj
        };
    }
    var cache = _getRequireWildcardCache(nodeInterop);
    if (cache && cache.has(obj)) {
        return cache.get(obj);
    }
    var newObj = {
        __proto__: null
    };
    var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
    for(var key in obj){
        if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
            if (desc && (desc.get || desc.set)) {
                Object.defineProperty(newObj, key, desc);
            } else {
                newObj[key] = obj[key];
            }
        }
    }
    newObj.default = obj;
    if (cache) {
        cache.set(obj, newObj);
    }
    return newObj;
}
const CardButtonGroup = (props)=>{
    return (0, _jsxruntime.jsx)(CardButtonGroupContextProvider, {
        children: (0, _jsxruntime.jsx)(InternalCardButtonGroup, {
            ...props
        })
    });
};
const InternalCardButtonGroup = (0, _mobxreactlite.observer)(function InternalCardButtonGroup(props) {
    const configuration = (0, _card_context.useCardConfiguration)();
    const context = _react.useContext(CardButtonGroupContext);
    var _props_disabled;
    const disabled = (_props_disabled = props.disabled) !== null && _props_disabled !== void 0 ? _props_disabled : configuration.disabled;
    const buttons = _react.Children.map(props.children, (child)=>{
        if (_react.isValidElement(child))
        return _react.cloneElement(child, {
            size: props.size,
            disabled
        });
    });
    var _props_visibility;
    return (0, _jsxruntime.jsx)(_card_decorator.CardDecorator, {
        location: props.location,
        visibility: (context === null || context === void 0 ? void 0 : context.isFlyoutOpen) ? 'always' : (_props_visibility = props.visibility) !== null && _props_visibility !== void 0 ? _props_visibility : 'on-hover',
        transition: props.transition,
        isInteractive: true,
        children: (0, _jsxruntime.jsx)(_button.ConjoinedButtons, {
            children: buttons
        })
    });
});
const CardButton = _react.memo(function CardButton(props) {
    const { className, icon, ariaLabel, onClick, disabled, size, variant = 'contrast' } = props;
    const buttonSize = size === 'small' ? 'tiny' : 'small';
    return (0, _jsxruntime.jsx)(_button.Button, {
        variant: variant,
        icon: icon,
        iconSize: size,
        size: buttonSize,
        ariaLabel: ariaLabel,
        onClick: onClick,
        className: className,
        disabled: disabled
    });
});
const CardMenuButton = (0, _mobxreactlite.observer)(function CardMenuButton(props) {
    const context = _react.useContext(CardButtonGroupContext);
    const flyoutStore = (0, _mobxreactlite.useLocalObservable)(()=>({
            isOpen: false,
            toggle: (0, _mobx.action)(()=>{
                context === null || context === void 0 ? void 0 : context.setIsFlyoutOpen(!flyoutStore.isOpen);
                flyoutStore.isOpen = !flyoutStore.isOpen;
            })
        }));
    return (0, _jsxruntime.jsx)(StatelessCardMenuButton, {
        ...props,
        isOpen: flyoutStore.isOpen,
        toggleFlyout: flyoutStore.toggle
    });
});
const StatelessCardMenuButton = (props)=>{
    const { Flyout = _flyout.Flyout, flyoutContent, isOpen, toggleFlyout, ariaLabel = _card_buttonmessages.CardMessages.more(), size, disabled, icon = _icon.MoreHorizontalIcon, className, variant = 'contrast', onContextMenu } = props;
    const buttonSize = size === 'small' ? 'tiny' : 'small';
    return (0, _jsxruntime.jsx)(Flyout, {
        open: isOpen,
        onRequestClose: isOpen ? toggleFlyout : undefined,
        placement: "bottom-start",
        trigger: (triggerProps)=>(0, _jsxruntime.jsx)(_button.Button, {
                ...triggerProps,
                variant: variant,
                icon: icon,
                iconSize: size,
                size: buttonSize,
                disabled: disabled,
                className: className,
                ariaLabel: ariaLabel,
                onClick: toggleFlyout,
                onContextMenu: onContextMenu
            }),
        children: typeof flyoutContent === 'function' ? flyoutContent({
            toggleFlyout
        }) : flyoutContent
    });
};
const CardButtonGroupContext = _react.createContext(undefined);
const CardButtonGroupContextProvider = ({ children })=>{
    const [isOpen, setIsOpen] = _react.useState(false);
    const contextValue = _react.useMemo(()=>{
        const setIsFlyoutOpen = (isOpen)=>{
            setIsOpen(isOpen);
        };
        return {
            isFlyoutOpen: isOpen,
            setIsFlyoutOpen
        };
    }, [
        isOpen,
        setIsOpen
    ]);
    return (0, _jsxruntime.jsx)(CardButtonGroupContext.Provider, {
        value: contextValue,
        children: children
    });
};
