// @canva/app-ui-kit/dist/esm/ui/base/layout/private/columns/columns.js
import { jsx as _jsx } from "react/jsx-runtime";
import { UnreachableError } from '../../../../../base/preconditions';
import classNames from 'classnames';
import * as React from 'react';
import { commonPropsToDOMProps } from '../../../box/common_props/common_props';
import { getSpaceValue } from '../../../metrics/metrics';
import { generateStyle } from '../../../responsive/responsive';
import styles, { customProperties } from './columns.css';
export const alignXs = [
    'start',
    'center',
    'end',
    'spaceBetween',
    'spaceAround'
];
export const alignYs = [
    'start',
    'center',
    'end',
    'stretch',
    'baseline'
];
export const heights = [
    'unset',
    'full'
];
export const breakpoints = [
    'small',
    'medium',
    'large',
    'xLarge'
];
export const Column = React.forwardRef(({ width = 'fluid', children, tagName: TagName = 'div', ...rest }, ref)=>{
    return _jsx(TagName, {
        className: classNames(styles.column, getWidthClassName(width)),
        style: getWidthStyle(width),
        ref: ref,
        ...commonPropsToDOMProps(rest),
        children: children
    });
});
const spacingPropertyMap = {
    default: customProperties.spacing,
    smallUp: customProperties.smallUpSpacing,
    mediumUp: customProperties.mediumUpSpacing,
    largeUp: customProperties.largeUpSpacing,
    xLargeUp: customProperties.xLargeUpSpacing
};
const alignPropertyMap = {
    default: customProperties.align,
    smallUp: customProperties.smallUpAlign,
    mediumUp: customProperties.mediumUpAlign,
    largeUp: customProperties.largeUpAlign,
    xLargeUp: customProperties.xLargeUpAlign
};
const alignYPropertyMap = {
    default: customProperties.alignY,
    smallUp: customProperties.smallUpAlignY,
    mediumUp: customProperties.mediumUpAlignY,
    largeUp: customProperties.largeUpAlignY,
    xLargeUp: customProperties.xLargeUpAlignY
};
const heightPropertyMap = {
    default: customProperties.heightValue,
    smallUp: customProperties.smallUpHeight,
    mediumUp: customProperties.mediumUpHeight,
    largeUp: customProperties.largeUpHeight,
    xLargeUp: customProperties.xLargeUpHeight
};
const getAlignValue = (align)=>{
    switch(align){
        case 'start':
            return 'flex-start';
        case 'center':
            return 'center';
        case 'end':
            return 'flex-end';
        case 'spaceBetween':
            return 'space-between';
        case 'spaceAround':
            return 'space-around';
        default:
            throw new UnreachableError(align);
    }
};
const getAlignYValue = (alignY)=>alignY === 'start' || alignY === 'end' ? `flex-${alignY}` : alignY;
const getHeightValue = (height)=>{
    switch(height){
        case 'full':
            return '100%';
        case 'unset':
            return 'unset';
        default:
            throw new UnreachableError(height);
    }
};
export const Columns = React.forwardRef(
    ({ children, spacing, align, alignY, height, tagName: TagName = 'div', reversed, collapseBelow, collapsed, ...rest }, ref)=>{
        return (_jsx(TagName, {
                    className: classNames(styles.columns, reversed && (collapsed || collapseBelow) && styles.rowReverse, collapseBelow && collapseBelowClassNames[collapseBelow], collapsed && styles.collapsed),
                    style: {
                        ...(spacing && generateStyle(spacingPropertyMap, spacing, getSpaceValue)),
                        ...(align && generateStyle(alignPropertyMap, align, getAlignValue)),
                        ...(alignY && generateStyle(alignYPropertyMap, alignY, getAlignYValue)),
                        ...(height && generateStyle(heightPropertyMap, height, getHeightValue))
                    },
                    ref: ref,
                    ...commonPropsToDOMProps(rest),
                    children: reversed && !collapsed && !collapseBelow ? React.Children.toArray(children).reverse() : children
                }));
    }
);
const collapseBelowClassNames = {
    ['small']: styles.collapseBelowSmall,
    ['medium']: styles.collapseBelowMedium,
    ['large']: styles.collapseBelowLarge,
    ['xLarge']: styles.collapseBelowXLarge
};
function getWidthClassName(width) {
    if (isFixedWidth(width)) return styles.widthFixed;
    switch(width){
        case 'fluid':
            return null;
        case 'content':
            return styles.widthContent;
        case 'containedContent':
            return styles.widthContainedContent;
        default:
            throw new UnreachableError(width);
    }
}
function getWidthStyle(width) {
    if (isFixedWidth(width)) return {
        [customProperties.columnWidth]: width
    };
    return null;
}
function isFixedWidth(width) {
    if (width !== 'fluid' && width !== 'content' && width !== 'containedContent') return true;
    return false;
}
