// @canva/app-ui-kit/dist/esm/ui/apps/developing/ui_kit/components/scrollable/scrollable.js
import { jsx as _jsx } from "react/jsx-runtime";
import { Scrollable as EaselScrollable } from '../../../../../base/scrollable/scrollable';
/** 
 * A component that wraps content to make it scrollable.
 *
 * @remarks
 * Use when you need to make an area within the page independently scrollable,
 * such as paragraphs of long text or a list of items.
 *
 * For cycling through related items like images or slides, consider using `<Carousel/>` instead.
 *
 * Note that Scrollable always takes up 100% of its parent's width and height.
 * To constrain the scroll container's dimensions, wrap it in a `<Box/>` component.
 *
 * @example
 * ```tsx
 * <Box height={200}>
 *   <Scrollable>
 *     <Text>Long scrollable content here...</Text>
 *   </Scrollable>
 * </Box>
 * ```
 *
 * @param props - The component props
 * @returns A JSX element
 */ export function Scrollable(props) {
    return _jsx(EaselScrollable, {
        children: props.children,
        direction: props.direction,
        onScroll: props.onScroll,
        indicator: props.indicator ? {
            background: props.indicator.background,
            type: 'fade'
        } : undefined
    });
}
