// @canva/app-ui-kit/dist/cjs/ui/base/device_capabilities/device_pixel_ratio/device_pixel_ratio.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: Object.getOwnPropertyDescriptor(all, name).get
    });
}
_export(exports, {
    get DevicePixelRatioMonitor () {
        return DevicePixelRatioMonitor;
    },
    get getDevicePixelRatio () {
        return getDevicePixelRatio;
    },
    get useDevicePixelRatio () {
        return useDevicePixelRatio;
    }
});
const _mobx = require("mobx");
const _react = require("react");
let prop;
class DevicePixelRatioMonitor {
    startTracking() {
        var _this_window;
        if (((_this_window = this.window) === null || _this_window === void 0 ? void 0 : _this_window.matchMedia) && !this.match) {
            this.match = this.window.matchMedia(`(resolution: ${this.window.devicePixelRatio}dppx)`);
            this.match.addListener(this.onChanged);
        }
    }
    stopTracking() {
        if (this.match) {
            this.match.removeListener(this.onChanged);
            this.match = undefined;
        }
    }
    get value() {
        var _this_window;
        this[_mobx.$mobx].reportObserved();
        var _this_window_devicePixelRatio;
        return (_this_window_devicePixelRatio = (_this_window = this.window) === null || _this_window === void 0 ? void 0 : _this_window.devicePixelRatio) !== null && _this_window_devicePixelRatio !== void 0 ? _this_window_devicePixelRatio : 1;
    }
    constructor(window1){
        this.window = window1;
        this[_$mobx] = (0, _mobx.createAtom)('DevicePixelRatio', ()=>this.startTracking(), ()=>this.stopTracking());
        this.onChanged = ()=>{
            this[_mobx.$mobx].reportChanged();
            this.stopTracking();
            this.startTracking();
        };
    }
}
prop = _$mobx = _mobx.$mobx;
const globalWindow = typeof window !== 'undefined' ? window : undefined;
const monitor = new DevicePixelRatioMonitor(globalWindow);
function useDevicePixelRatio() {
    return (0, _react.useSyncExternalStore)(subscribeDevicePixelRatio, getClientDevicePixelRatioValue, getServerDevicePixelRatioValue);
}
function getDevicePixelRatio() {
    return monitor.value;
}
const computedDevicePixelRatio = (0, _mobx.computed)(getDevicePixelRatio);
const subscribeDevicePixelRatio = (callback)=>(0, _mobx.observe)(computedDevicePixelRatio, callback);
const getClientDevicePixelRatioValue = ()=>computedDevicePixelRatio.get();
const getServerDevicePixelRatioValue = ()=>1;
var _$mobx;
