// @canva/app-ui-kit/dist/cjs/ui/base/date_time/date_time_picker/private/date_time_picker.js
"use strict"
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "DateTimePicker", {
    enumerable: true,
    get: function() {
        return DateTimePicker;
    }
});
const _jsxruntime = require("react/jsx-runtime");
const _observer_for_class = require('../../../../../base/mobx_react/observer_for_class');
const _preconditions = require('../../../../../base/preconditions');
const _mobxreactlite = require("mobx-react-lite");
const _react = _interop_require_wildcard(require("react"));
const _screen_reader_content = require('../../../a11y/screen_reader_content/screen_reader_content');
const _button = require('../../../button/button');
const _calendar = require('../../calendar/calendar');
const _months_navigation = require('../months_navigation/months_navigation');
const _utils = require('../../utils/utils');
const _time_input = require('../../../form/time_input/time_input');
const _key_combinations = require('../../../key_combinations/key_combinations');
const _layout = require('../../../layout/layout');
const _date_time_pickercss = _interop_require_default(require("./date_time_picker.css"));
const _date_time_pickermessages = require("./date_time_picker.messages");
const _date_time_picker_presenter = require("./date_time_picker_presenter");
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function _getRequireWildcardCache(nodeInterop) {
    if (typeof WeakMap !== "function") return null;
    var cacheBabelInterop = new WeakMap();
    var cacheNodeInterop = new WeakMap();
    return (_getRequireWildcardCache = function(nodeInterop) {
        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
    })(nodeInterop);
}
function _interop_require_wildcard(obj, nodeInterop) {
    if (!nodeInterop && obj && obj.__esModule) {
        return obj;
    }
    if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
        return {
            default: obj
        };
    }
    var cache = _getRequireWildcardCache(nodeInterop);
    if (cache && cache.has(obj)) {
        return cache.get(obj);
    }
    var newObj = {
        __proto__: null
    };
    var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
    for(var key in obj){
        if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
            if (desc && (desc.get || desc.set)) {
                Object.defineProperty(newObj, key, desc);
            } else {
                newObj[key] = obj[key];
            }
        }
    }
    newObj.default = obj;
    if (cache) {
        cache.set(obj, newObj);
    }
    return newObj;
}
function applyDecs2203RFactory() {
    function createAddInitializerMethod(initializers, decoratorFinishedRef) {
        return function addInitializer(initializer) {
            assertNotFinished(decoratorFinishedRef, "addInitializer");
            assertCallable(initializer, "An initializer");
            initializers.push(initializer);
        };
    }
    function memberDec(dec, name, desc, initializers, kind, isStatic, isPrivate, metadata, value) {
        var kindStr;
        switch(kind){
            case 1:
                kindStr = "accessor";
                break;
            case 2:
                kindStr = "method";
                break;
            case 3:
                kindStr = "getter";
                break;
            case 4:
                kindStr = "setter";
                break;
            default:
                kindStr = "field";
        }
        var ctx = {
            kind: kindStr,
            name: isPrivate ? "#" + name : name,
            static: isStatic,
            private: isPrivate,
            metadata: metadata
        };
        var decoratorFinishedRef = {
            v: false
        };
        ctx.addInitializer = createAddInitializerMethod(initializers, decoratorFinishedRef);
        var get, set;
        if (kind === 0) {
            if (isPrivate) {
                get = desc.get;
                set = desc.set;
            } else {
                get = function() {
                    return this[name];
                };
                set = function(v) {
                    this[name] = v;
                };
            }
        } else if (kind === 2) {
            get = function() {
                return desc.value;
            };
        } else {
            if (kind === 1 || kind === 3) {
                get = function() {
                    return desc.get.call(this);
                };
            }
            if (kind === 1 || kind === 4) {
                set = function(v) {
                    desc.set.call(this, v);
                };
            }
        }
        ctx.access = get && set ? {
            get: get,
            set: set
        } : get ? {
            get: get
        } : {
            set: set
        };
        try {
            return dec(value, ctx);
        } finally{
            decoratorFinishedRef.v = true;
        }
    }
    function assertNotFinished(decoratorFinishedRef, fnName) {
        if (decoratorFinishedRef.v) {
            throw new Error("attempted to call " + fnName + " after decoration was finished");
        }
    }
    function assertCallable(fn, hint) {
        if (typeof fn !== "function") {
            throw new TypeError(hint + " must be a function");
        }
    }
    function assertValidReturnValue(kind, value) {
        var type = typeof value;
        if (kind === 1) {
            if (type !== "object" || value === null) {
                throw new TypeError("accessor decorators must return an object with get, set, or init properties or void 0");
            }
            if (value.get !== undefined) {
                assertCallable(value.get, "accessor.get");
            }
            if (value.set !== undefined) {
                assertCallable(value.set, "accessor.set");
            }
            if (value.init !== undefined) {
                assertCallable(value.init, "accessor.init");
            }
        } else if (type !== "function") {
            var hint;
            if (kind === 0) {
                hint = "field";
            } else if (kind === 10) {
                hint = "class";
            } else {
                hint = "method";
            }
            throw new TypeError(hint + " decorators must return a function or void 0");
        }
    }
    function applyMemberDec(ret, base, decInfo, name, kind, isStatic, isPrivate, initializers, metadata) {
        var decs = decInfo[0];
        var desc, init, value;
        if (isPrivate) {
            if (kind === 0 || kind === 1) {
                desc = {
                    get: decInfo[3],
                    set: decInfo[4]
                };
            } else if (kind === 3) {
                desc = {
                    get: decInfo[3]
                };
            } else if (kind === 4) {
                desc = {
                    set: decInfo[3]
                };
            } else {
                desc = {
                    value: decInfo[3]
                };
            }
        } else if (kind !== 0) {
            desc = Object.getOwnPropertyDescriptor(base, name);
        }
        if (kind === 1) {
            value = {
                get: desc.get,
                set: desc.set
            };
        } else if (kind === 2) {
            value = desc.value;
        } else if (kind === 3) {
            value = desc.get;
        } else if (kind === 4) {
            value = desc.set;
        }
        var newValue, get, set;
        if (typeof decs === "function") {
            newValue = memberDec(decs, name, desc, initializers, kind, isStatic, isPrivate, metadata, value);
            if (newValue !== void 0) {
                assertValidReturnValue(kind, newValue);
                if (kind === 0) {
                    init = newValue;
                } else if (kind === 1) {
                    init = newValue.init;
                    get = newValue.get || value.get;
                    set = newValue.set || value.set;
                    value = {
                        get: get,
                        set: set
                    };
                } else {
                    value = newValue;
                }
            }
        } else {
            for(var i = decs.length - 1; i >= 0; i--){
                var dec = decs[i];
                newValue = memberDec(dec, name, desc, initializers, kind, isStatic, isPrivate, metadata, value);
                if (newValue !== void 0) {
                    assertValidReturnValue(kind, newValue);
                    var newInit;
                    if (kind === 0) {
                        newInit = newValue;
                    } else if (kind === 1) {
                        newInit = newValue.init;
                        get = newValue.get || value.get;
                        set = newValue.set || value.set;
                        value = {
                            get: get,
                            set: set
                        };
                    } else {
                        value = newValue;
                    }
                    if (newInit !== void 0) {
                        if (init === void 0) {
                            init = newInit;
                        } else if (typeof init === "function") {
                            init = [
                                init,
                                newInit
                            ];
                        } else {
                            init.push(newInit);
                        }
                    }
                }
            }
        }
        if (kind === 0 || kind === 1) {
            if (init === void 0) {
                init = function(instance, init) {
                    return init;
                };
            } else if (typeof init !== "function") {
                var ownInitializers = init;
                init = function(instance, init) {
                    var value = init;
                    for(var i = 0; i < ownInitializers.length; i++){
                        value = ownInitializers[i].call(instance, value);
                    }
                    return value;
                };
            } else {
                var originalInitializer = init;
                init = function(instance, init) {
                    return originalInitializer.call(instance, init);
                };
            }
            ret.push(init);
        }
        if (kind !== 0) {
            if (kind === 1) {
                desc.get = value.get;
                desc.set = value.set;
            } else if (kind === 2) {
                desc.value = value;
            } else if (kind === 3) {
                desc.get = value;
            } else if (kind === 4) {
                desc.set = value;
            }
            if (isPrivate) {
                if (kind === 1) {
                    ret.push(function(instance, args) {
                        return value.get.call(instance, args);
                    });
                    ret.push(function(instance, args) {
                        return value.set.call(instance, args);
                    });
                } else if (kind === 2) {
                    ret.push(value);
                } else {
                    ret.push(function(instance, args) {
                        return value.call(instance, args);
                    });
                }
            } else {
                Object.defineProperty(base, name, desc);
            }
        }
    }
    function applyMemberDecs(Class, decInfos, metadata) {
        var ret = [];
        var protoInitializers;
        var staticInitializers;
        var existingProtoNonFields = new Map();
        var existingStaticNonFields = new Map();
        for(var i = 0; i < decInfos.length; i++){
            var decInfo = decInfos[i];
            if (!Array.isArray(decInfo)) continue;
            var kind = decInfo[1];
            var name = decInfo[2];
            var isPrivate = decInfo.length > 3;
            var isStatic = kind >= 5;
            var base;
            var initializers;
            if (isStatic) {
                base = Class;
                kind = kind - 5;
                staticInitializers = staticInitializers || [];
                initializers = staticInitializers;
            } else {
                base = Class.prototype;
                protoInitializers = protoInitializers || [];
                initializers = protoInitializers;
            }
            if (kind !== 0 && !isPrivate) {
                var existingNonFields = isStatic ? existingStaticNonFields : existingProtoNonFields;
                var existingKind = existingNonFields.get(name) || 0;
                if (existingKind === true || existingKind === 3 && kind !== 4 || existingKind === 4 && kind !== 3) {
                    throw new Error("Attempted to decorate a public method/accessor that has the same name as a previously decorated public method/accessor. This is not currently supported by the decorators plugin. Property name was: " + name);
                } else if (!existingKind && kind > 2) {
                    existingNonFields.set(name, kind);
                } else {
                    existingNonFields.set(name, true);
                }
            }
            applyMemberDec(ret, base, decInfo, name, kind, isStatic, isPrivate, initializers, metadata);
        }
        pushInitializers(ret, protoInitializers);
        pushInitializers(ret, staticInitializers);
        return ret;
    }
    function pushInitializers(ret, initializers) {
        if (initializers) {
            ret.push(function(instance) {
                for(var i = 0; i < initializers.length; i++){
                    initializers[i].call(instance);
                }
                return instance;
            });
        }
    }
    function applyClassDecs(targetClass, classDecs, metadata) {
        if (classDecs.length > 0) {
            var initializers = [];
            var newClass = targetClass;
            var name = targetClass.name;
            for(var i = classDecs.length - 1; i >= 0; i--){
                var decoratorFinishedRef = {
                    v: false
                };
                try {
                    var nextNewClass = classDecs[i](newClass, {
                        kind: "class",
                        name: name,
                        addInitializer: createAddInitializerMethod(initializers, decoratorFinishedRef),
                        metadata
                    });
                } finally{
                    decoratorFinishedRef.v = true;
                }
                if (nextNewClass !== undefined) {
                    assertValidReturnValue(10, nextNewClass);
                    newClass = nextNewClass;
                }
            }
            return [
                defineMetadata(newClass, metadata),
                function() {
                    for(var i = 0; i < initializers.length; i++){
                        initializers[i].call(newClass);
                    }
                }
            ];
        }
    }
    function defineMetadata(Class, metadata) {
        return Object.defineProperty(Class, Symbol.metadata || Symbol.for("Symbol.metadata"), {
            configurable: true,
            enumerable: true,
            value: metadata
        });
    }
    return function applyDecs2203R(targetClass, memberDecs, classDecs, parentClass) {
        if (parentClass !== void 0) {
            var parentMetadata = parentClass[Symbol.metadata || Symbol.for("Symbol.metadata")];
        }
        var metadata = Object.create(parentMetadata === void 0 ? null : parentMetadata);
        var e = applyMemberDecs(targetClass, memberDecs, metadata);
        if (!classDecs.length) defineMetadata(targetClass, metadata);
        return {
            e: e,
            get c () {
                return applyClassDecs(targetClass, classDecs, metadata);
            }
        };
    };
}
function _apply_decs_2203_r(targetClass, memberDecs, classDecs, parentClass) {
    return (_apply_decs_2203_r = applyDecs2203RFactory())(targetClass, memberDecs, classDecs, parentClass);
}
function _identity(x) {
    return x;
}
var _class;
var _initClass, _React_Component;
let _DateTimePickerComponent;
new (_class = class extends _identity {
    constructor(){
        super(_DateTimePickerComponent), _initClass();
    }
}, (()=>{
    class DateTimePickerComponent extends (_React_Component = _react.Component) {
        get date() {
            var _this__date, _this__date1;
            if (((_this__date = this._date) === null || _this__date === void 0 ? void 0 : _this__date.utc) !== this.props.date) {
                if (this.props.date == null)
                    this._date = undefined;
                else this._date = {
                    utc: this.props.date,
                    value: (0, _utils.toUTCDateTimeObject)((0, _utils.fromUTCToTimezone)(this.props.date, this.props.timezone ? this.props.timezone.offset : (0, _utils.getTimezoneOffset)(this.props.date)))
                };
            }
            return (_this__date1 = this._date) === null || _this__date1 === void 0 ? void 0 : _this__date1.value;
        }
        get time() {
            return this.presenter.getTime(this.store);
        }
        get displayedDate() {
            return this.presenter.getDate(this.store);
        }
        get month() {
            return this.presenter.getMonth(this.store);
        }
        get year() {
            return this.presenter.getYear(this.store);
        }
        get navigationDisabled() {
            const { canSelectPastDates, minSelectableDate, maxSelectableDate } = this.props;
            let disabledNav;
            if ((!canSelectPastDates || minSelectableDate === 'today') && (0, _utils.compareMonthAndYear)(this.displayedDate, this.store.todayInTimezone) <= 0) disabledNav = 'prev';
            if (minSelectableDate && minSelectableDate !== 'today' && (0, _utils.compareMonthAndYear)(this.displayedDate, minSelectableDate) <= 0) disabledNav = 'prev';
            if (maxSelectableDate && (0, _utils.compareMonthAndYear)(this.displayedDate, maxSelectableDate) >= 0) disabledNav = disabledNav === 'prev' ? 'both' : 'next';
            return disabledNav;
        }
        moveFocusBy(delta, deltaType) {
            var _this_store_focusedDateInTimezone;
            const currentDate = (_this_store_focusedDateInTimezone = this.store.focusedDateInTimezone) !== null && _this_store_focusedDateInTimezone !== void 0 ? _this_store_focusedDateInTimezone : this.presenter.getDate(this.store);
            let newDate = currentDate;
            switch(deltaType){
                case 'day':
                    newDate = (0, _utils.addDays)(currentDate, delta);
                    break;
                case 'month':
                    newDate = (0, _utils.addMonths)(currentDate, delta);
                    break;
                case 'year':
                    newDate = (0, _utils.addYears)(currentDate, delta);
                    break;
                default:
                    throw new _preconditions.UnreachableError(deltaType);
            }
            const { canSelectPastDates, minSelectableDate, maxSelectableDate } = this.props;
            if ((!canSelectPastDates || minSelectableDate === 'today') && (0, _utils.compareDates)(newDate, this.store.todayInTimezone) < 0) newDate = this.store.todayInTimezone;
            if (minSelectableDate && minSelectableDate !== 'today' && (0, _utils.compareDates)(newDate, minSelectableDate) < 0) newDate = minSelectableDate;
            if (maxSelectableDate && (0, _utils.compareDates)(newDate, maxSelectableDate) > 0) newDate = maxSelectableDate;
            this.presenter.updateFocusedDate(this.store, (0, _utils.toUTCDateTimeObject)(newDate));
        }
        moveFocusToStartOfWeek() {
            var _this_store_focusedDateInTimezone;
            const currentDate = (_this_store_focusedDateInTimezone = this.store.focusedDateInTimezone) !== null && _this_store_focusedDateInTimezone !== void 0 ? _this_store_focusedDateInTimezone : this.presenter.getDate(this.store);
            const currentDayOfWeek = (0, _utils.dayOfWeek)(currentDate);
            if (currentDayOfWeek > 0) this.moveFocusBy(-currentDayOfWeek, 'day');
        }
        moveFocusToEndOfWeek() {
            var _this_store_focusedDateInTimezone;
            const currentDate = (_this_store_focusedDateInTimezone = this.store.focusedDateInTimezone) !== null && _this_store_focusedDateInTimezone !== void 0 ? _this_store_focusedDateInTimezone : this.presenter.getDate(this.store);
            const currentDayOfWeek = (0, _utils.dayOfWeek)(currentDate);
            if (currentDayOfWeek < 6) this.moveFocusBy(6 - currentDayOfWeek, 'day');
        }
        getKeyMap() {
            const keyMap = [
                [
                    'ArrowRight',
                    (e)=>{
                        e.preventDefault();
                        this.moveFocusBy(1, 'day');
                    }
                ],
                [
                    'ArrowLeft',
                    (e)=>{
                        e.preventDefault();
                        this.moveFocusBy(-1, 'day');
                    }
                ],
                [
                    'ArrowDown',
                    (e)=>{
                        e.preventDefault();
                        this.moveFocusBy(7, 'day');
                    }
                ],
                [
                    'ArrowUp',
                    (e)=>{
                        e.preventDefault();
                        this.moveFocusBy(-7, 'day');
                    }
                ],
                [
                    'Home',
                    (e)=>{
                        e.preventDefault();
                        this.moveFocusToStartOfWeek();
                    }
                ],
                [
                    'End',
                    (e)=>{
                        e.preventDefault();
                        this.moveFocusToEndOfWeek();
                    }
                ],
                [
                    'PageUp',
                    (e)=>{
                        e.preventDefault();
                        this.moveFocusBy(-1, 'month');
                    }
                ],
                [
                    'PageDown',
                    (e)=>{
                        e.preventDefault();
                        this.moveFocusBy(1, 'month');
                    }
                ],
                [
                    (0, _key_combinations.shift)('PageUp'),
                    (e)=>{
                        e.preventDefault();
                        this.moveFocusBy(-1, 'year');
                    }
                ],
                [
                    (0, _key_combinations.shift)('PageDown'),
                    (e)=>{
                        e.preventDefault();
                        this.moveFocusBy(1, 'year');
                    }
                ]
            ];
            return keyMap;
        }
        render() {
            return (0, _jsxruntime.jsx)(_key_combinations.WithKeyCombinationHandler, {
                map: this.keyMap,
                handleInputs: false,
                children: ({ ref })=>(0, _jsxruntime.jsx)(_mobxreactlite.Observer, {
                        children: ()=>{
                            const calendarLabelId = _react.useId();
                            const calendarDescriptionId = _react.useId();
                            var _this_store_focusedDateInTimezone, _ref;
                            return (0, _jsxruntime.jsxs)("div", {
                                className: _date_time_pickercss.default.dateTimePicker,
                                children: [
                                    (0, _jsxruntime.jsx)(_months_navigation.MonthsNavigation, {
                                        locale: this.props.locale,
                                        month: this.month,
                                        year: this.year,
                                        today: this.store.todayInTimezone,
                                        disabled: this.navigationDisabled,
                                        animate: this.props.animate,
                                        onChange: this.onChangeMonth,
                                        titleId: calendarLabelId
                                    }),
                                    (0, _jsxruntime.jsxs)("div", {
                                        ref: ref,
                                        children: [
                                            (0, _jsxruntime.jsx)(_screen_reader_content.ScreenReaderContent, {
                                                id: calendarDescriptionId,
                                                children: _date_time_pickermessages.DateTimePickerMessages.calendarDescription()
                                            }),
                                            (0, _jsxruntime.jsx)(_calendar.Calendar, {
                                                locale: this.props.locale,
                                                month: this.month,
                                                year: this.year,
                                                animate: this.props.animate,
                                                today: this.store.todayInTimezone,
                                                selected: this.date,
                                                focused: this.store.focusedDateInTimezone,
                                                includedInTabOrder: (_ref = (_this_store_focusedDateInTimezone = this.store.focusedDateInTimezone) !== null && _this_store_focusedDateInTimezone !== void 0 ? _this_store_focusedDateInTimezone : this.store.dateInTimezone) !== null && _ref !== void 0 ? _ref : this.store.todayInTimezone,
                                                selectedInvalid: this.props.selectedDayInvalid,
                                                canSelectPastDates: this.props.canSelectPastDates,
                                                minSelectableDate: this.props.minSelectableDate,
                                                maxSelectableDate: this.props.maxSelectableDate,
                                                onSelect: this.onSelectDay,
                                                ariaLabelledBy: calendarLabelId,
                                                ariaDescribedBy: calendarDescriptionId
                                            })
                                        ]
                                    }),
                                    this.props.mode === 'datetime' && (0, _jsxruntime.jsxs)(_jsxruntime.Fragment, {
                                        children: [
                                            (0, _jsxruntime.jsx)(_layout.Spacer, {
                                                direction: "vertical",
                                                size: "2u"
                                            }),
                                            (0, _jsxruntime.jsx)(_time_input.TimeInput, {
                                                value: this.time,
                                                error: this.props.timeInvalid,
                                                onChange: this.onChangeTime,
                                                ariaLabel: _date_time_pickermessages.DateTimePickerMessages.timePickerLabel(),
                                                timezone: this.store.timezone,
                                                locale: this.props.locale
                                            })
                                        ]
                                    }),
                                    (0, _jsxruntime.jsx)(this.Footer, {})
                                ]
                            });
                        }
                    })
            });
        }
        constructor(props){
            super(props), this.onChangeMonth = (date)=>{
                this.presenter.unsetFocusedDate(this.store);
                this.presenter.updateDate(this.store, date);
                const navDisabled = this.navigationDisabled;
                if (navDisabled != null) {
                    this.presenter.updateFocusedDate(this.store, date);
                    this.moveFocusBy(0, 'day');
                }
            }, this.onSelectDay = (date)=>{
                this.presenter.updateDate(this.store, date, {
                    updateFocus: true
                });
                this.props.onChange(this.presenter.getUTCDate(this.store));
            }, this.selectToday = ()=>{
                this.presenter.updateDate(this.store, this.store.todayInTimezone, {
                    updateFocus: true
                });
                this.props.onChange(this.presenter.getUTCDate(this.store));
            }, this.clearSelectedDay = ()=>{
                this.presenter.unsetDate(this.store);
                this.presenter.unsetFocusedDate(this.store);
                this.props.onChange(undefined);
            }, this.onChangeTime = (time)=>{
                const timeMs = time !== null && time !== void 0 ? time : 0;
                this.presenter.unsetFocusedDate(this.store);
                this.presenter.updateTime(this.store, timeMs);
                this.props.onChange(this.presenter.getUTCDate(this.store));
            }, this.Footer = ()=>{
                const { clearable, showTodayButton } = this.props;
                if (!clearable && !showTodayButton) return null;
                const columnAlignment = clearable && showTodayButton ? 'spaceBetween' : clearable ? 'start' : 'end';
                return (0, _jsxruntime.jsxs)(_jsxruntime.Fragment, {
                    children: [
                        (0, _jsxruntime.jsx)(_layout.Spacer, {
                            direction: "vertical",
                            size: "1.5u"
                        }),
                        (0, _jsxruntime.jsxs)(_layout.Columns, {
                            spacing: "0",
                            align: columnAlignment,
                            children: [
                                clearable && (0, _jsxruntime.jsx)(_layout.Column, {
                                    width: "content",
                                    children: (0, _jsxruntime.jsx)(_button.Button, {
                                        variant: "tertiary",
                                        onClick: this.clearSelectedDay,
                                        children: _date_time_pickermessages.DateTimePickerMessages.clearLabel()
                                    })
                                }),
                                showTodayButton && (0, _jsxruntime.jsx)(_layout.Column, {
                                    width: "content",
                                    children: (0, _jsxruntime.jsx)(_button.Button, {
                                        variant: "tertiary",
                                        onClick: this.selectToday,
                                        children: _date_time_pickermessages.DateTimePickerMessages.todayLabel()
                                    })
                                })
                            ]
                        })
                    ]
                });
            };
            this.presenter = props.presenter || new _date_time_picker_presenter.DateTimePickerPresenter();
            this.store = props.store || this.presenter.createStore(props.date, props.locale, props.today, props.timezone);
            this.props.focusToday && this.presenter.updateFocusedDate(this.store, this.store.todayInTimezone);
            this.keyMap = this.getKeyMap();
        }
    }
    ({ c: [_DateTimePickerComponent, _initClass] } = _apply_decs_2203_r(DateTimePickerComponent, [], [
        _observer_for_class.observerForClass
    ], _React_Component));
    DateTimePickerComponent.defaultProps = {
        animate: false,
        showTodayButton: true,
        clearable: true,
        mode: 'datetime'
    };
})(), _class)();
class DateTimePicker extends _DateTimePickerComponent {
}
