// @canva/app-ui-kit/dist/esm/ui/base/surface/header/private/header.js
import { jsx as _jsx, jsxs as _jsxs, Fragment as _Fragment } from "react/jsx-runtime";
import classNames from 'classnames';
import * as React from 'react';
import { Box } from '../../../box/box';
import { Button } from '../../../button/button';
import { ArrowLeftIcon } from '../../../icons/arrow_left/icon';
import { XIcon } from '../../../icons/x/icon';
import { Column, Columns } from '../../../layout/layout';
import { Text, Title } from '../../../typography/typography';
import styles from './header.css';
import { HeaderMessages } from './header.messages';
export const SurfaceHeader = ({ title, titleId, description, descriptionId, alignment = 'start', start, end, divider = true, padding = '2u', background = 'surface' })=>{
    const hasStructuredPart = title != null || description != null || start != null || end != null;
    return _jsx(Box, {
        padding: padding,
        width: "full",
        background: background === 'none' ? undefined : background,
        borderBottom: divider ? 'low' : undefined,
        children: hasStructuredPart && _jsx(Box, {
            className: styles.bleed,
            display: "flex",
            children: _jsx(StructuredHeaderContent, {
                title: title,
                titleId: titleId,
                description: description,
                descriptionId: descriptionId,
                alignment: alignment,
                start: start,
                end: end
            })
        })
    });
};
const TitleAndDescription = ({ title, titleId, description, descriptionId, alignment })=>{
    return _jsxs(_Fragment, {
        children: [
            typeof title === 'string' ? _jsx(SurfaceHeaderTitle, {
                id: titleId,
                alignment: alignment,
                children: title
            }) : titleId ? _jsx("div", {
                id: titleId,
                children: title
            }) : title,
            typeof description === 'string' ? _jsx(SurfaceHeaderDescription, {
                id: descriptionId,
                alignment: alignment,
                children: description
            }) : descriptionId ? _jsx("div", {
                id: descriptionId,
                children: description
            }) : description
        ]
    });
};
const renderDecorator = (decorator)=>
    React.isValidElement(decorator) && decorator.type === React.Fragment ? _jsx(_Fragment, {
        children: React.Children.map(decorator.props.children, (child)=>child && renderDecorator(child))
    }) : _jsx(Column, {
        width: "content",
        children: decorator
    });
const renderDecorators = (decorator)=>{
    return _jsx(Box, {
        className: styles.minHeight,
        display: "flex",
        alignItems: "center",
        children: _jsx(Columns, {
            spacing: "0.5u",
            alignY: "center",
            children: renderDecorator(decorator)
        })
    });
};
const StructuredHeaderContent = ({ title, titleId, description, descriptionId, alignment, start, end })=>{
    const titleAndDescriptionClassName = classNames(styles.minHeight, {
        [styles.widthFluid]: alignment === 'start'
    });
    return _jsxs(_Fragment, {
        children: [
            _jsx(Box, {
                justifyContent: "start",
                className: alignment === 'center' && styles.widthFluid,
                children: start != null ? _jsx(Box, {
                    paddingEnd: "0.5u",
                    children: renderDecorators(start)
                }) : null
            }),
            title || description ? _jsx(Box, {
                className: titleAndDescriptionClassName,
                minWidth: "0",
                display: "flex",
                flexDirection: "column",
                justifyContent: "center",
                paddingX: "0.5u",
                children: _jsx(TitleAndDescription, {
                    title: title,
                    titleId: titleId,
                    description: description,
                    descriptionId: descriptionId,
                    alignment: alignment
                })
            }) : _jsx(Box, {
                className: titleAndDescriptionClassName
            }),
            _jsx(Box, {
                display: "flex",
                justifyContent: "end",
                alignItems: "start",
                className: alignment === 'center' && styles.widthFluid,
                children: end != null ? _jsx(Box, {
                    paddingStart: "0.5u",
                    children: renderDecorators(end)
                }) : null
            })
        ]
    });
};
export const SurfaceHeaderTitle = ({ children, id, alignment })=>{
    return _jsx(Title, {
        size: "xsmall",
        lineClamp: 1,
        alignment: alignment,
        id: id,
        children: children
    });
};
export const SurfaceHeaderDescription = ({ children, alignment = 'start', id })=>{
    return _jsx(Text, {
        size: "small",
        tone: "tertiary",
        alignment: alignment,
        id: id,
        children: children
    });
};
export const SurfaceHeaderIconButton = (props)=>{
    return _jsx(Button, {
        ...props,
        variant: "tertiary",
        size: "small"
    });
};
export const SurfaceHeaderBackButton = ({ ariaLabel, ...props })=>{
    return _jsx(SurfaceHeaderIconButton, {
        icon: ArrowLeftIcon,
        ariaLabel: ariaLabel || HeaderMessages.defaultGoBack(),
        ...props
    });
};
export const SurfaceHeaderCloseButton = ({ ariaLabel, ...props })=>{
    return _jsx(SurfaceHeaderIconButton, {
        icon: XIcon,
        ariaLabel: ariaLabel || HeaderMessages.defaultClose(),
        ...props
    });
};
export const SurfaceHeaderTextButton = (props)=>{
    return _jsx(Button, {
        variant: "linkButton",
        size: "small",
        ...props
    });
};
