// @canva/app-ui-kit/dist/cjs/ui/base/tooltip/tooltip.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: Object.getOwnPropertyDescriptor(all, name).get
    });
}
_export(exports, {
    get FloatingTooltip () {
        return _tooltip.FloatingTooltip;
    },
    get Tooltip () {
        return _tooltip.Tooltip;
    },
    get TooltipArrow () {
        return _internal_tooltip.TooltipArrow;
    },
    get TooltipBox () {
        return _internal_tooltip.TooltipBox;
    },
    get TooltipContent () {
        return _internal_tooltip.TooltipContent;
    },
    get TooltipPresenter () {
        return _tooltip_presenter.TooltipPresenter;
    },
    get TooltipState () {
        return _tooltip_presenter.TooltipState;
    },
    get TriggeredTooltip () {
        return _tooltip.TriggeredTooltip;
    }
});
const _internal_tooltip = require("./internal/internal_tooltip");
const _tooltip_presenter = require("./private/tooltip_presenter");
const _tooltip = require("./private/tooltip");
