// @canva/app-ui-kit/dist/cjs/ui/base/form/multiline_input/private/multiline_input_with_footer.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "MultilineInputWithFooter", {
    enumerable: true,
    get: function() {
        return MultilineInputWithFooter;
    }
});
const _jsxruntime = require("react/jsx-runtime");
const _classnames = _interop_require_default(require("classnames"));
const _react = _interop_require_wildcard(require("react"));
const _multiline_input = require("./multiline_input");
const _multiline_inputcss = _interop_require_default(require("./multiline_input.css"));
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function _getRequireWildcardCache(nodeInterop) {
    if (typeof WeakMap !== "function") return null;
    var cacheBabelInterop = new WeakMap();
    var cacheNodeInterop = new WeakMap();
    return (_getRequireWildcardCache = function(nodeInterop) {
        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
    })(nodeInterop);
}
function _interop_require_wildcard(obj, nodeInterop) {
    if (!nodeInterop && obj && obj.__esModule) {
        return obj;
    }
    if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
        return {
            default: obj
        };
    }
    var cache = _getRequireWildcardCache(nodeInterop);
    if (cache && cache.has(obj)) {
        return cache.get(obj);
    }
    var newObj = {
        __proto__: null
    };
    var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
    for(var key in obj){
        if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
            if (desc && (desc.get || desc.set)) {
                Object.defineProperty(newObj, key, desc);
            } else {
                newObj[key] = obj[key];
            }
        }
    }
    newObj.default = obj;
    if (cache) {
        cache.set(obj, newObj);
    }
    return newObj;
}
const MultilineInputWithFooter = _react.forwardRef(function MultilineInputWithFooter(props, ref) {
    const { EndContent, StartContent, ...multilineProps } = props;
    return (0, _jsxruntime.jsx)(_multiline_input.MultilineInput, {
        ...multilineProps,
        ref: ref,
        footer: (0, _jsxruntime.jsxs)("div", {
            className: (0, _classnames.default)(_multiline_inputcss.default.footer, {
                [_multiline_inputcss.default.footerDisabled]: props.disabled,
                [_multiline_inputcss.default.footerReadOnly]: props.readOnly,
                [_multiline_inputcss.default.footerFadein]: !props.borderless
            }),
            role: "none",
            children: [
                (0, _jsxruntime.jsx)("span", {
                    children: StartContent
                }),
                (0, _jsxruntime.jsx)("span", {
                    children: EndContent
                })
            ]
        })
    });
});
