// @canva/app-ui-kit/dist/cjs/ui/base/form/multiline_input/private/multiline_input.js
"use strict"
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "MultilineInput", {
    enumerable: true,
    get: function() {
        return MultilineInput;
    }
});
const _jsxruntime = require("react/jsx-runtime");
const _composereactrefs = _interop_require_default(require("@seznam/compose-react-refs"));
const _classnames = _interop_require_default(require("classnames"));
const _mobxreactlite = require("mobx-react-lite");
const _react = _interop_require_wildcard(require("react"));
const _resizeobserverpolyfill = _interop_require_default(require("resize-observer-polyfill"));
const _pointer_input_state = require('../../../a11y/pointer_input_state/pointer_input_state');
const _controllable_value = require('../../../controllable_value/controllable_value');
const _base_input = require('../../base_input/base_input');
const _temporary_exportscss = _interop_require_default(require('../../base_input/temporary_exports.css'));
const _input = require('../../internal/input');
const _behavior = require('../../../surface/sheet/behavior/behavior');
const _multiline_inputcss = _interop_require_default(require("./multiline_input.css"));
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function _getRequireWildcardCache(nodeInterop) {
    if (typeof WeakMap !== "function") return null;
    var cacheBabelInterop = new WeakMap();
    var cacheNodeInterop = new WeakMap();
    return (_getRequireWildcardCache = function(nodeInterop) {
        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
    })(nodeInterop);
}
function _interop_require_wildcard(obj, nodeInterop) {
    if (!nodeInterop && obj && obj.__esModule) {
        return obj;
    }
    if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
        return {
            default: obj
        };
    }
    var cache = _getRequireWildcardCache(nodeInterop);
    if (cache && cache.has(obj)) {
        return cache.get(obj);
    }
    var newObj = {
        __proto__: null
    };
    var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
    for(var key in obj){
        if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
            if (desc && (desc.get || desc.set)) {
                Object.defineProperty(newObj, key, desc);
            } else {
                newObj[key] = obj[key];
            }
        }
    }
    newObj.default = obj;
    if (cache) {
        cache.set(obj, newObj);
    }
    return newObj;
}
const DEFAULT_BUFFER_ROWS = 1;
const DEFAULT_MAX_ROWS = Number.POSITIVE_INFINITY;
function useAutoGrow(
    { bufferRows = DEFAULT_BUFFER_ROWS, maxRows = DEFAULT_MAX_ROWS, minRows = 1, value }
) {
    const innerRef = _react.useRef(null);
    const [rows, setRows] = _react.useState();
    const remeasure = _react.useCallback(()=>{
        const refElement = innerRef.current;
        if (!refElement) return;
        const rowHeight = refElement.clientHeight;
        const contentRows = rowHeight === 0 ? 0 : Math.round(refElement.scrollHeight / rowHeight);
        const rows = Math.min(maxRows, Math.max(minRows + bufferRows, contentRows + bufferRows));
        setRows(rows);
    }, [
        bufferRows,
        maxRows,
        minRows
    ]);
    _react.useLayoutEffect(remeasure, [
        remeasure,
        value
    ]);
    const stableRemeasure = _react.useRef(remeasure);
    const observerRef = _react.useRef(null);
    const setRef = _react.useCallback((element)=>{
        if (observerRef.current == null)
            observerRef.current = new _resizeobserverpolyfill.default(()=>stableRemeasure.current());
        if (innerRef.current != null)
            observerRef.current.unobserve(innerRef.current);
        innerRef.current = element;
        if (innerRef.current != null) observerRef.current.observe(innerRef.current);
    }, []);
    _react.useEffect(()=>{
        return ()=>{
            var _observerRef_current;
            return (_observerRef_current = observerRef.current) === null || _observerRef_current === void 0 ? void 0 : _observerRef_current.disconnect();
        };
    }, []);
    return {
        ref: setRef,
        rows
    };
}
const MultilineInput = (0, _mobxreactlite.observer)(_react.forwardRef(function MultilineInput(props_, forwardedRef) {
    const { props, setPropOverrides } = (0, _input.usePropOverrides)(props_);
    const { onChange: onChangeProp, onFocus: onFocusProp, onBlur: onBlurProp, onChangeComplete: onChangeCompleteProp, onKeyDown: onKeyDownProp, dir = 'auto' } = props;
    const { ref: inputRef, onMouseDown: onWrapperMouseDown } = (0, _base_input.useOnWrapperMouseDown)();
    const { isFocused, isKeyboardMode, setFocused } = (0, _base_input.useFocusState)();
    const [value, setValue] = (0, _controllable_value.useControllableValue)({
        value: props.value
    });
    const { ref: autoGrowRef, rows } = useAutoGrow({
        ...props,
        value
    });
    const controls = (0, _input.useCreateInputControls)({
        focus: ()=>{
            var _inputRef_current;
            return (_inputRef_current = inputRef.current) === null || _inputRef_current === void 0 ? void 0 : _inputRef_current.focus();
        },
        setDisabled: (disabled)=>setPropOverrides({
                disabled
            }),
        setValue: (value)=>{
            setValue(value);
            onChangeProp === null || onChangeProp === void 0 ? void 0 : onChangeProp(value);
        }
    });
    const onChange = _react.useCallback((e)=>{
        setValue(e.target.value);
        onChangeProp === null || onChangeProp === void 0 ? void 0 : onChangeProp(e.target.value, e);
    }, [
        onChangeProp,
        setValue
    ]);
    const onFocus = _react.useCallback((e)=>{
        onFocusProp === null || onFocusProp === void 0 ? void 0 : onFocusProp(e);
        setFocused(true);
    }, [
        onFocusProp,
        setFocused
    ]);
    const onBlur = _react.useCallback((e)=>{
        onBlurProp === null || onBlurProp === void 0 ? void 0 : onBlurProp(e);
        onChangeCompleteProp === null || onChangeCompleteProp === void 0 ? void 0 : onChangeCompleteProp(e.target.value);
        setFocused(false);
    }, [
        onBlurProp,
        onChangeCompleteProp,
        setFocused
    ]);
    const onKeyDown = _react.useCallback((e)=>{
        if (e.keyCode === 229) {
            e.stopPropagation();
            return;
        }
        onKeyDownProp === null || onKeyDownProp === void 0 ? void 0 : onKeyDownProp(e);
    }, [
        onKeyDownProp
    ]);
    const hoverSupported = (0, _pointer_input_state.useIsHoveringInput)();
    const wrapperClassName = (0, _classnames.default)(_temporary_exportscss.default.wrapper, _multiline_inputcss.default.multilineWrapper, {
        [_temporary_exportscss.default.hoverSupported]: hoverSupported,
        [_temporary_exportscss.default.focusOutline]: isFocused && isKeyboardMode,
        [_temporary_exportscss.default.borderless]: props.borderless
    }, props.disabled && _temporary_exportscss.default.disabled || isFocused && _temporary_exportscss.default.active || props.error && _temporary_exportscss.default.error, props.className);
    const inputClassName = (0, _classnames.default)(_temporary_exportscss.default.textField, _multiline_inputcss.default.multilineTextField, !props.autoGrow && props.resize && _multiline_inputcss.default.resize, props.inputClassName);
    const sheetIsolationRef = (0, _behavior.useSheetScrollIsolation)();
    return ((0, _jsxruntime.jsx)("div", {
            ref: sheetIsolationRef,
            className: wrapperClassName,
            onMouseDown: onWrapperMouseDown,
            children: (0, _jsxruntime.jsxs)(_input.BaseInputContextProvider, {
                controls: controls,
                value: value !== null && value !== void 0 ? value : '',
                children: [
                    (0, _jsxruntime.jsx)("textarea", {
                        className: inputClassName,
                        value: value,
                        onChange: onChange,
                        onFocus: onFocus,
                        onBlur: onBlur,
                        onKeyPress: props.onKeyPress,
                        onKeyDown: onKeyDown,
                        onKeyUp: props.onKeyUp,
                        onPaste: props.onPaste,
                        onClick: props.onClick,
                        ref: (0, _composereactrefs.default)(inputRef, forwardedRef),
                        dir: dir,
                        autoComplete: props.autoComplete,
                        disabled: props.disabled,
                        placeholder: props.placeholder,
                        maxLength: props.maxLength,
                        lang: props.lang,
                        rows: rows !== null && rows !== void 0 ? rows : props.minRows,
                        id: props.id,
                        "aria-label": props.ariaLabel,
                        "aria-labelledby": props.ariaLabelledBy,
                        "aria-describedby": props.ariaDescribedBy,
                        required: props.required,
                        readOnly: props.readOnly
                    }),
                    props.autoGrow && (0, _jsxruntime.jsx)("textarea", {
                        "aria-hidden": "true",
                        className: (0, _classnames.default)(inputClassName, _multiline_inputcss.default.hiddenInput),
                        dir: dir,
                        ref: autoGrowRef,
                        readOnly: true,
                        rows: 1,
                        value: value
                    }),
                    props.footer
                ]
            })
        }));
}));
MultilineInput.displayName = 'MultilineInput';
