// @canva/app-ui-kit/dist/cjs/ui/base/surface/sheet/behavior/private/resize.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "useResize", {
    enumerable: true,
    get: function() {
        return useResize;
    }
});
const _react = _interop_require_wildcard(require("react"));
const _stable_function = require('../../../../stable_function/stable_function');
const _snap_points = require("./snap_points");
function _getRequireWildcardCache(nodeInterop) {
    if (typeof WeakMap !== "function") return null;
    var cacheBabelInterop = new WeakMap();
    var cacheNodeInterop = new WeakMap();
    return (_getRequireWildcardCache = function(nodeInterop) {
        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
    })(nodeInterop);
}
function _interop_require_wildcard(obj, nodeInterop) {
    if (!nodeInterop && obj && obj.__esModule) {
        return obj;
    }
    if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
        return {
            default: obj
        };
    }
    var cache = _getRequireWildcardCache(nodeInterop);
    if (cache && cache.has(obj)) {
        return cache.get(obj);
    }
    var newObj = {
        __proto__: null
    };
    var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
    for(var key in obj){
        if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
            if (desc && (desc.get || desc.set)) {
                Object.defineProperty(newObj, key, desc);
            } else {
                newObj[key] = obj[key];
            }
        }
    }
    newObj.default = obj;
    if (cache) {
        cache.set(obj, newObj);
    }
    return newObj;
}
const FLICK_VELOCITY_BREAKPOINT = 1;
function useResize({ snapPointControls, snapPointIndex, setSnapPointIndex, setOffset }) {
    const [resizing, setResizing] = _react.useState(false);
    const resizeState = _react.useRef(undefined);
    const start = (0, _stable_function.useStableFunction)(()=>{
        const points = snapPointControls.get();
        const heights = points.map((point)=>point.height);
        const current = (0, _snap_points.pointForIndex)(points, snapPointIndex);
        resizeState.current = {
            points,
            maxOffset: Math.max(...heights) - current.height,
            minOffset: Math.min(...heights) - current.height,
            height: current.height,
            offset: 0,
            startTime: Date.now()
        };
    });
    const update = (0, _stable_function.useStableFunction)((offset)=>{
        const state = resizeState.current;
        if (state == null) return;
        const newOffset = Math.min(state.maxOffset, Math.max(state.minOffset, offset));
        let updateResizing = resizing;
        if (!updateResizing && newOffset !== 0) {
            updateResizing = true;
            setResizing(true);
        }
        if (updateResizing) {
            setOffset(newOffset, state.height);
            state.offset = newOffset;
        }
    });
    const end = (0, _stable_function.useStableFunction)(()=>{
        const state = resizeState.current;
        if (state == null) return;
        const timeDelta = Date.now() - state.startTime;
        const velocity = state.offset / timeDelta;
        const flicked = Math.abs(velocity) >= FLICK_VELOCITY_BREAKPOINT;
        const current = (0, _snap_points.pointForIndex)(state.points, snapPointIndex);
        const newSize = current.height + state.offset;
        const [index] = state.points.reduce((acc, { height, indices })=>{
            const delta = height - newSize;
            if (flicked && delta < 0 !== velocity < 0)
                return acc;
            const absoluteDelta = Math.abs(delta);
            return absoluteDelta < acc[1] ? [
                indices[0],
                absoluteDelta
            ] : acc;
        }, [
            0,
            Infinity
        ]);
        setOffset(undefined, undefined);
        resizeState.current = undefined;
        snapPointControls.clear();
        setSnapPointIndex(index);
        setResizing(false);
    });
    return {
        resizing,
        resizeControls: {
            start,
            update,
            end
        }
    };
}
