// @canva/app-ui-kit/dist/cjs/ui/base/form/input_decorators/private/button_decorator.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "ButtonDecorator", {
    enumerable: true,
    get: function() {
        return ButtonDecorator;
    }
});
const _jsxruntime = require("react/jsx-runtime");
const _classnames = _interop_require_default(require("classnames"));
const _react = _interop_require_wildcard(require("react"));
const _badge = require('../../../badge/badge');
const _button = require('../../../button/button');
const _button_decoratorcss = _interop_require_default(require("./button_decorator.css"));
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function _getRequireWildcardCache(nodeInterop) {
    if (typeof WeakMap !== "function") return null;
    var cacheBabelInterop = new WeakMap();
    var cacheNodeInterop = new WeakMap();
    return (_getRequireWildcardCache = function(nodeInterop) {
        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
    })(nodeInterop);
}
function _interop_require_wildcard(obj, nodeInterop) {
    if (!nodeInterop && obj && obj.__esModule) {
        return obj;
    }
    if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
        return {
            default: obj
        };
    }
    var cache = _getRequireWildcardCache(nodeInterop);
    if (cache && cache.has(obj)) {
        return cache.get(obj);
    }
    var newObj = {
        __proto__: null
    };
    var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
    for(var key in obj){
        if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
            if (desc && (desc.get || desc.set)) {
                Object.defineProperty(newObj, key, desc);
            } else {
                newObj[key] = obj[key];
            }
        }
    }
    newObj.default = obj;
    if (cache) {
        cache.set(obj, newObj);
    }
    return newObj;
}
const ButtonDecorator = _react.memo(
    function ButtonDecorator({ label, icon, badge, onClick, active, disclosure, disabled, className }) {
        const button = (0, _jsxruntime.jsx)(_button.Button, {
            size: "small",
            variant: "tertiary",
            iconSize: "medium",
            icon: icon,
            onClick: onClick,
            tooltipLabel: label,
            tooltipPlacement: "top",
            ariaLabel: label,
            active: active,
            disclosure: disclosure,
            disabled: disabled,
            className: (0, _classnames.default)(badge != null && _button_decoratorcss.default.enablePointerEvents, className)
        });
        if (badge != null) return (0, _jsxruntime.jsxs)("div", {
            className: (0, _classnames.default)(_button_decoratorcss.default.badgeContainer, _button_decoratorcss.default.disablePointerEvents),
            children: [
                button,
                (0, _jsxruntime.jsx)("div", {
                    className: _button_decoratorcss.default.badge,
                    children: typeof badge === 'string' ? (0, _jsxruntime.jsx)(_badge.Badge, {
                        tone: "assist",
                        shape: "circle",
                        text: badge,
                        ariaLabel: badge
                    }) : badge
                })
            ]
        });
        return button;
    }
);
