// @canva/app-ui-kit/dist/cjs/ui/base/form/input_decorators/private/chars_remaining.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "CharsRemaining", {
    enumerable: true,
    get: function() {
        return CharsRemaining;
    }
});
const _jsxruntime = require("react/jsx-runtime");
const _classnames = _interop_require_default(require("classnames"));
require("react");
const _announcer = require('../../../a11y/announcer/announcer');
const _typography = require('../../../typography/typography');
const _chars_remainingcss = _interop_require_default(require("./chars_remaining.css"));
const _input_decoratorsmessages = require("./input_decorators.messages");
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
const CharsRemaining = ({ count, max, margins = 'default', textSize, id })=>{
    const priority = count > max ? 'high' : 'medium';
    (0, _announcer.useAnnounceOnUpdate)(_input_decoratorsmessages.InputDecoratorsMessages.charsUsed(count, max), count, {
        priority,
        debounceMs: 200
    });
    return (0, _jsxruntime.jsx)("div", {
        className: (0, _classnames.default)(_chars_remainingcss.default.charsRemaining, {
            [_chars_remainingcss.default.noMargins]: margins === 'none',
            [_chars_remainingcss.default.verticalMargins]: margins === 'vertical',
            [_chars_remainingcss.default.horizontalMargins]: margins === 'horizontal'
        }),
        children: (0, _jsxruntime.jsx)(_typography.Text, {
            size: textSize,
            margins: "legacy",
            alignment: "end",
            tone: count > max ? 'critical' : 'tertiary',
            children: (0, _jsxruntime.jsxs)("span", {
                id: id,
                "aria-label": _input_decoratorsmessages.InputDecoratorsMessages.charsUsed(count, max),
                children: [
                    count,
                    "/",
                    max
                ]
            })
        })
    });
};
