// @canva/app-ui-kit/dist/cjs/ui/base/form/input_decorators/private/words_remaining.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "WordsRemaining", {
    enumerable: true,
    get: function() {
        return WordsRemaining;
    }
});
const _jsxruntime = require("react/jsx-runtime");
require("react");
const _box = require('../../../box/box');
const _typography = require('../../../typography/typography');
const _input_decoratorsmessages = require("./input_decorators.messages");
const WordsRemaining = ({ count, max, margins = 'default', textSize, id })=>{
    return (0, _jsxruntime.jsx)(_box.Box, {
        padding: margins === 'default' ? '1u' : undefined,
        paddingX: margins === 'horizontal' ? '1u' : undefined,
        paddingY: margins === 'vertical' ? '1u' : undefined,
        children: (0, _jsxruntime.jsx)(_typography.Text, {
            size: textSize,
            alignment: "end",
            tone: count > max ? 'critical' : 'tertiary',
            children: (0, _jsxruntime.jsxs)("span", {
                id: id,
                "aria-label": _input_decoratorsmessages.InputDecoratorsMessages.wordsUsed(count, max),
                "aria-live": "polite",
                children: [
                    count,
                    "/",
                    max
                ]
            })
        })
    });
};
