// @canva/app-ui-kit/dist/esm/ui/base/swatch/private/internal_swatch.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { UnreachableError } from '../../../../base/preconditions';
import classnames from 'classnames';
import { observer } from 'mobx-react-lite';
import * as React from 'react';
import { pointerInputState, useIsHoveringInput } from '../../a11y/pointer_input_state/pointer_input_state';
import { BadgeContainer, BadgeGroup } from '../../badge/badge';
import { BaseButton } from '../../button/base_button/base_button';
import { CircleButton } from '../../button/button';
import { createFocusHandle, useHandleRef } from '../../handle/handle';
import { XIcon } from '../../icons/x/icon';
import styles from './swatch.css';
import { SwatchMessages } from './swatch.messages';
export function getBorderRadiusStyles(shape, size) {
    if (shape === 'circle') return undefined;
    switch(size){
        case 'xxsmall':
        case 'xsmall':
            return styles.borderRadiusElementSmall;
        case 'small':
        case 'medium':
        case 'large':
            return styles.borderRadiusElement;
        default:
            throw new UnreachableError(size);
    }
}
export const StaticSwatch = React.forwardRef(
    ({ className, children, ariaLabel, ariaDescribedBy, ariaLabelledBy, id, isRoot }, ref)=>{
        const spanRef = useHandleRef(ref, createFocusHandle);
        return _jsx("span", {
            ref: spanRef,
            "aria-label": ariaLabelledBy ? undefined : ariaLabel,
            "aria-labelledby": ariaLabelledBy,
            "aria-describedby": ariaDescribedBy,
            id: id,
            className: classnames(className, isRoot && styles.swatchRoot),
            role: "img",
            children: children
        });
    }
);
export const ClickableSwatch = observer(React.forwardRef(
    function _ClickableSwatch({ size = 'medium', shape = 'circle', onClick, active, selected, pressed, disabled, ariaLabel, disclosure, role, tabIndex, stretch, aspectRatio, children, ariaHasPopup, ariaControls, tooltipLabel, tooltipDescription, disableTooltip, isRoot }, ref) {
        const buttonRef = useHandleRef(ref, createFocusHandle);
        const borderRadiusStyles = getBorderRadiusStyles(shape, size);
        const hoverSupported = useIsHoveringInput();
        return _jsx(BaseButton, {
            onClick: onClick,
            active: active,
            selected: selected,
            pressed: pressed,
            disclosure: disclosure,
            disabled: disabled,
            role: role,
            ref: buttonRef,
            tabIndex: tabIndex,
            ariaLabel: ariaLabel,
            ariaHasPopup: ariaHasPopup,
            ariaControls: ariaControls,
            tooltipLabel: tooltipLabel,
            tooltipDescription: tooltipDescription,
            tooltipDisabled: disableTooltip,
            position: "relative",
            width: stretch ? 'full' : undefined,
            height: stretch && aspectRatio ? 'full' : undefined,
            className: classnames(styles.swatchButton, borderRadiusStyles, {
                [styles.swatchRoot]: isRoot,
                [styles.active]: selected || pressed || active,
                [styles.disabled]: disabled,
                [styles.fixed]: !stretch,
                [styles.hoverSupported]: hoverSupported
            }),
            children: children
        });
    }
));
export const DeletableSwatch = observer(React.forwardRef(
    function deletableSwatch({ onDelete, deleteButtonVisibility, stretch, aspectRatio, disableTooltip, children, isRoot }, ref) {
        const buttonRef = useHandleRef(ref, createFocusHandle);
        return _jsxs(BadgeContainer, {
            className: classnames(styles.swatchWrapper, {
                [styles.swatchRoot]: isRoot,
                [styles.stretch]: stretch,
                [styles.fixed]: !stretch,
                [styles.aspectRatio]: aspectRatio
            }),
            children: [
                children,
                _jsx(BadgeGroup, {
                    location: "top-end",
                    visibility: !pointerInputState.isMouseInput ? 'always' : deleteButtonVisibility,
                    children: _jsx(CircleButton, {
                        ref: buttonRef,
                        icon: XIcon,
                        variant: "contrast",
                        onClick: onDelete,
                        tooltipLabel: SwatchMessages.deleteTooltip(),
                        tooltipDisabled: disableTooltip,
                        size: "tiny",
                        iconSize: "tiny",
                        className: styles.deleteButton
                    })
                })
            ]
        });
    }
));
