// @canva/app-ui-kit/dist/cjs/ui/base/card/private/card_content.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: Object.getOwnPropertyDescriptor(all, name).get
    });
}
_export(exports, {
    get CardContent () {
        return CardContent;
    },
    get CardDescription () {
        return CardDescription;
    },
    get CardPreTitle () {
        return CardPreTitle;
    },
    get CardTitle () {
        return CardTitle;
    }
});
const _jsxruntime = require("react/jsx-runtime");
const _classnames = _interop_require_default(require("classnames"));
require("react");
const _layout = require('../../layout/layout');
const _typography = require('../../typography/typography');
const _cardcss = _interop_require_default(require("./card.css"));
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
const CardContent = function CardContent({ titleId, descriptionId, title, description, preTitle, content }) {
    const hasStandardContent = !!(title || description || preTitle);
    const customContent = typeof content === 'function' ? content({
        titleId,
        descriptionId
    }) : content;
    if (hasStandardContent) return (0, _jsxruntime.jsxs)(_layout.Rows, {
        spacing: "0",
        children: [
            typeof preTitle !== 'string' ? preTitle : (0, _jsxruntime.jsx)(CardPreTitle, {
                children: preTitle
            }),
            title != null && typeof title !== 'string' ? (0, _jsxruntime.jsx)("div", {
                className: _cardcss.default.title,
                id: titleId,
                children: title
            }) : (0, _jsxruntime.jsx)(CardTitle, {
                id: titleId,
                lineClamp: description ? 1 : 2,
                children: title
            }),
            description != null && typeof description !== 'string' ? (0, _jsxruntime.jsx)("div", {
                className: _cardcss.default.description,
                id: descriptionId,
                children: description
            }) : (0, _jsxruntime.jsx)(CardDescription, {
                id: descriptionId,
                children: description
            }),
            customContent
        ]
    });
    return customContent;
};
const CardTitle = function CardTitle({ className, weight = 'bold', lineClamp = 2, ...rest }) {
    return (0, _jsxruntime.jsx)(_typography.Text, {
        className: (0, _classnames.default)(_cardcss.default.title, className),
        weight: weight,
        lineClamp: lineClamp,
        ...rest
    });
};
const CardDescription = function CardDescription({ size = 'small', tone = 'secondary', lineClamp = 1, ...rest }) {
    return (0, _jsxruntime.jsx)(_typography.Text, {
        size: size,
        tone: tone,
        lineClamp: lineClamp,
        className: _cardcss.default.description,
        ...rest
    });
};
const CardPreTitle = function CardPreTitle({ size = 'xsmall', tone = 'secondary', capitalization = 'uppercase', lineClamp = 1, ...rest }) {
    return (0, _jsxruntime.jsx)(_typography.Text, {
        size: size,
        tone: tone,
        capitalization: capitalization,
        lineClamp: lineClamp,
        ...rest
    });
};
