// @canva/app-ui-kit/dist/cjs/ui/base/card/private/card_thumbnail.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: Object.getOwnPropertyDescriptor(all, name).get
    });
}
_export(exports, {
    get CardIconThumbnail () {
        return CardIconThumbnail;
    },
    get CardImageThumbnail () {
        return CardImageThumbnail;
    },
    get CardPlaceholder () {
        return CardPlaceholder;
    },
    get CardThumbnail () {
        return CardThumbnail;
    },
    get CardThumbnailContainer () {
        return CardThumbnailContainer;
    },
    get CardVideoThumbnail () {
        return CardVideoThumbnail;
    }
});
const _jsxruntime = require("react/jsx-runtime");
const _classnames = _interop_require_default(require("classnames"));
const _mobx = require("mobx");
const _mobxreactlite = require("mobx-react-lite");
const _react = _interop_require_wildcard(require("react"));
const _supports_animation = require('../../animation/supports_animation');
const _provider = require('../../provider/provider');
const _thumbnail = require('../../thumbnail/thumbnail');
const _card = require("./card");
const _cardcss = _interop_require_default(require("./card.css"));
const _card_context = require("./card_context");
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function _getRequireWildcardCache(nodeInterop) {
    if (typeof WeakMap !== "function") return null;
    var cacheBabelInterop = new WeakMap();
    var cacheNodeInterop = new WeakMap();
    return (_getRequireWildcardCache = function(nodeInterop) {
        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
    })(nodeInterop);
}
function _interop_require_wildcard(obj, nodeInterop) {
    if (!nodeInterop && obj && obj.__esModule) {
        return obj;
    }
    if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
        return {
            default: obj
        };
    }
    var cache = _getRequireWildcardCache(nodeInterop);
    if (cache && cache.has(obj)) {
        return cache.get(obj);
    }
    var newObj = {
        __proto__: null
    };
    var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
    for(var key in obj){
        if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
            if (desc && (desc.get || desc.set)) {
                Object.defineProperty(newObj, key, desc);
            } else {
                newObj[key] = obj[key];
            }
        }
    }
    newObj.default = obj;
    if (cache) {
        cache.set(obj, newObj);
    }
    return newObj;
}
const CardThumbnailContainer = (0, _mobxreactlite.observer)(({ border, borderRadius: borderRadiusProp, background: backgroundProp = 'secondary', aspectRatio, padding, children, className, width, height, fit })=>{
    const shouldBgColorChangeOnHover = useShouldBgColorChangeOnHover();
    const background = useCardThumbnailBackground(backgroundProp, !!shouldBgColorChangeOnHover);
    const borderRadius = useBorderRadius(borderRadiusProp);
    return (0, _jsxruntime.jsx)(_card.SelectableWrapper, {
        boundary: "thumbnail",
        borderRadius: borderRadius,
        children: (0, _jsxruntime.jsx)(_thumbnail.ThumbnailContainer, {
            border: border,
            borderRadius: borderRadius,
            padding: padding,
            background: background,
            width: width,
            height: height,
            aspectRatio: aspectRatio,
            fit: fit,
            className: (0, _classnames.default)(className, {
                [_cardcss.default.bgColorTransition]: shouldBgColorChangeOnHover
            }),
            children: children
        })
    });
});
const CardThumbnail = (0, _mobxreactlite.observer)(({ border = 'low', borderRadius: borderRadiusProp, children, ...props })=>{
    const borderRadius = useBorderRadius(borderRadiusProp);
    const thumbnail = (0, _jsxruntime.jsx)(_thumbnail.Thumbnail, {
        ...props,
        border: border,
        borderRadius: borderRadius,
        children: children
    });
    return (0, _jsxruntime.jsx)(_card.SelectableWrapper, {
        boundary: "thumbnail",
        borderRadius: borderRadius,
        children: thumbnail
    });
});
const CardPlaceholder = (0, _mobxreactlite.observer)(({ border = 'low', borderRadius: borderRadiusProp, ...cardPlaceholderProps })=>{
    const borderRadius = useBorderRadius(borderRadiusProp);
    const placeholderThumbnail = (0, _jsxruntime.jsx)(_thumbnail.ImageThumbnailPlaceholder, {
        ...cardPlaceholderProps,
        border: border,
        borderRadius: borderRadius
    });
    return (0, _jsxruntime.jsx)(_card.SelectableWrapper, {
        boundary: "thumbnail",
        borderRadius: borderRadius,
        children: placeholderThumbnail
    });
});
const CardIconThumbnail = (0, _mobxreactlite.observer)(({ border = 'low', borderRadius: borderRadiusProp, ...iconThumbnailProps })=>{
    const borderRadius = useBorderRadius(borderRadiusProp);
    const iconThumbnail = (0, _jsxruntime.jsx)(_thumbnail.IconThumbnail, {
        ...iconThumbnailProps,
        border: border,
        borderRadius: borderRadius
    });
    return (0, _jsxruntime.jsx)(_card.SelectableWrapper, {
        boundary: "thumbnail",
        borderRadius: borderRadius,
        children: iconThumbnail
    });
});
const CardImageThumbnail = (0, _mobxreactlite.observer)(({ renderImage = _thumbnail.StaticImage, border = 'low', borderRadius: borderRadiusProp, ...props })=>{
    const borderRadius = useBorderRadius(borderRadiusProp);
    const imageThumbnail = (0, _jsxruntime.jsx)(_thumbnail.ImageThumbnail, {
        ...props,
        border: border,
        borderRadius: borderRadius,
        src: props.src,
        renderImage: ({ ...imageProps })=>renderImage({
                ...imageProps,
                className: (0, _classnames.default)(props.className, _cardcss.default.image),
                crossOrigin: props.crossOrigin,
                src: props.src,
                alt: props.alt || '',
                elementTiming: props.elementTiming
            })
    });
    return (0, _jsxruntime.jsx)(_card.SelectableWrapper, {
        boundary: "thumbnail",
        borderRadius: borderRadius,
        children: imageThumbnail
    });
});
const CardVideoThumbnail = (0, _mobxreactlite.observer)(_react.forwardRef((props, ref)=>{
    const { border = 'low', borderRadius: borderRadiusProp, className, videoSrc, imageSrc, imageLoadingState, imageFallback, imagePlaceholder, imageAlt, videoLoadingState, VideoComponent, renderImage, loadVideo, onVideoLoad, onImageLoad, autoplay } = props;
    const isFocused = (0, _card_context.useCardFocus)();
    const borderRadius = useBorderRadius(borderRadiusProp);
    const autoplayVideos = (0, _provider.useAutoplayVideos)() || 'ADAPTIVE';
    const prefersReducedMotion = (0, _supports_animation.usePrefersReducedMotion)();
    const enableAutoplay = !prefersReducedMotion && autoplayVideos === 'ADAPTIVE' || autoplayVideos === 'AUTOPLAY';
    const whenToPlay = imageSrc == null ? 'always' : props.whenToPlay === 'always' ? enableAutoplay ? 'always' : 'on-hover' : props.whenToPlay;
    const isHovered = (0, _card_context.useCardHover)({
        track: whenToPlay === 'on-hover'
    });
    const store = (0, _mobxreactlite.useLocalObservable)(()=>({
            playState: 'pause',
            setPlayState: (0, _mobx.action)((value)=>{
                store.playState = value;
            })
        }));
    const { setPlayState } = store;
    _react.useEffect(()=>(0, _mobx.autorun)(()=>{
            if ((whenToPlay === 'always' || whenToPlay === 'now') && store.playState !== 'play') setPlayState('play');
            if (whenToPlay === 'never' && store.playState === 'play') setPlayState('pause');
        }), [
        whenToPlay,
        store.playState,
        setPlayState
    ]);
    _react.useEffect(()=>(0, _mobx.autorun)(()=>{
            if (whenToPlay === 'on-hover') {
                if ((isHovered || isFocused) && store.playState !== 'play') setPlayState('play');
                else if (!isHovered && !isFocused && store.playState === 'play') setPlayState('pause');
            }
        }), [
        whenToPlay,
        isHovered,
        isFocused,
        store.playState,
        setPlayState
    ]);
    const videoThumbnail = (0, _jsxruntime.jsx)(_thumbnail.VideoThumbnail, {
        ...props,
        ref: ref,
        border: border,
        borderRadius: borderRadius,
        videoSrc: videoSrc,
        imageSrc: imageSrc,
        playState: store.playState,
        autoplay: autoplay,
        delayVideoLoad: whenToPlay === 'on-hover' ? 150 : 0,
        VideoComponent: VideoComponent,
        videoClassName: (0, _classnames.default)(_cardcss.default.video, className),
        videoLoadingState: videoLoadingState,
        loadVideo: loadVideo,
        onVideoLoad: onVideoLoad,
        imageAlt: imageAlt,
        renderImage: renderImage,
        imageClassName: (0, _classnames.default)(className, _cardcss.default.image),
        imagePlaceholder: imagePlaceholder,
        imageLoadingState: imageLoadingState,
        imageFallback: imageFallback,
        onImageLoad: onImageLoad
    });
    return (0, _jsxruntime.jsx)(_card.SelectableWrapper, {
        boundary: "thumbnail",
        borderRadius: borderRadius,
        children: videoThumbnail
    });
}));
const backgroundMapping = {
    ['contrastOnLight']: {
        hover: 'contrastOnLightHover',
        default: 'contrastOnLight'
    },
    ['contrastOnDark']: {
        hover: 'contrastOnDarkHover',
        default: 'contrastOnDark'
    },
    ['secondary']: {
        hover: 'secondaryHover',
        default: 'secondary'
    },
    ['none']: {
        hover: 'secondaryHover',
        default: 'none'
    }
};
function getBackground(background, isHover) {
    return backgroundMapping[background][isHover ? 'hover' : 'default'];
}
const useShouldBgColorChangeOnHover = ()=>{
    const { hoverEffect, hoverBoundary } = (0, _card_context.useCardConfiguration)();
    return hoverBoundary === 'thumbnail' && hoverEffect === 'bgColorTransition';
};
function useBorderRadius(borderRadius) {
    const { borderRadius: cardBorderRadius } = (0, _card_context.useCardConfiguration)();
    return borderRadius !== null && borderRadius !== void 0 ? borderRadius : cardBorderRadius;
}
function useCardThumbnailBackground(background, shouldBgColorChangeOnHover) {
    const isHovered = (0, _card_context.useCardHover)({
        track: shouldBgColorChangeOnHover
    });
    return shouldBgColorChangeOnHover ? getBackground(background, isHovered) : background;
}
