// @canva/app-ui-kit/dist/cjs/ui/base/card/private/card_context.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: Object.getOwnPropertyDescriptor(all, name).get
    });
}
_export(exports, {
    get useCardConfiguration () {
        return useCardConfiguration;
    },
    get useCardConfigurationSetter () {
        return useCardConfigurationSetter;
    },
    get useCardEventHandlers () {
        return useCardEventHandlers;
    },
    get useCardFocus () {
        return useCardFocus;
    },
    get useCardHover () {
        return useCardHover;
    },
    get useCardMouseDown () {
        return useCardMouseDown;
    },
    get withCardContextProvider () {
        return withCardContextProvider;
    }
});
const _jsxruntime = require("react/jsx-runtime");
const _make_observable = require('../../../../base/make_observable/make_observable');
const _mobx = _interop_require_wildcard(require("mobx"));
const _react = _interop_require_wildcard(require("react"));
const _pointer_input_state = require('../../a11y/pointer_input_state/pointer_input_state');
function _getRequireWildcardCache(nodeInterop) {
    if (typeof WeakMap !== "function") return null;
    var cacheBabelInterop = new WeakMap();
    var cacheNodeInterop = new WeakMap();
    return (_getRequireWildcardCache = function(nodeInterop) {
        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
    })(nodeInterop);
}
function _interop_require_wildcard(obj, nodeInterop) {
    if (!nodeInterop && obj && obj.__esModule) {
        return obj;
    }
    if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
        return {
            default: obj
        };
    }
    var cache = _getRequireWildcardCache(nodeInterop);
    if (cache && cache.has(obj)) {
        return cache.get(obj);
    }
    var newObj = {
        __proto__: null
    };
    var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
    for(var key in obj){
        if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
            if (desc && (desc.get || desc.set)) {
                Object.defineProperty(newObj, key, desc);
            } else {
                newObj[key] = obj[key];
            }
        }
    }
    newObj.default = obj;
    if (cache) {
        cache.set(obj, newObj);
    }
    return newObj;
}
const useCardConfiguration = ()=>useCardConfigurationStore().configuration;
function useCardConfigurationSetter(configuration) {
    const { setConfiguration } = useCardConfigurationStore();
    const willMount = _react.useRef(true);
    if (willMount.current) setConfiguration(configuration);
    willMount.current = false;
    _react.useEffect(()=>{
        setConfiguration(configuration);
    }, [
        setConfiguration,
        configuration
    ]);
}
function useCardHover(args) {
    var _args_track;
    const track = (_args_track = args === null || args === void 0 ? void 0 : args.track) !== null && _args_track !== void 0 ? _args_track : true;
    const interactionStore = useCardInteractionStore();
    _react.useEffect(()=>{
        if (track) _mobx.runInAction(()=>interactionStore.trackHover = true);
    }, [
        interactionStore,
        track
    ]);
    return interactionStore.isHovered;
}
function useCardFocus() {
    const interactionStore = useCardInteractionStore();
    _react.useEffect(()=>{
        _mobx.runInAction(()=>interactionStore.trackFocus = true);
    }, [
        interactionStore
    ]);
    return interactionStore.isFocused;
}
function useCardMouseDown() {
    const interactionStore = useCardInteractionStore();
    _react.useEffect(()=>{
        _mobx.runInAction(()=>interactionStore.trackMouseDown = true);
    }, [
        interactionStore
    ]);
    return interactionStore.isMouseDown;
}
const useCardEventHandlers = ()=>useCardInteractionStore().eventHandlers;
function withCardContextProvider(Component) {
    const WithProvider = (props)=>{
        const [context] = _react.useState(createContext);
        return (0, _jsxruntime.jsx)(CardContext.Provider, {
            value: context,
            children: (0, _jsxruntime.jsx)(Component, {
                ...props
            })
        });
    };
    var _Component_displayName;
    WithProvider.displayName = `withCardContext(${(_Component_displayName = Component.displayName) !== null && _Component_displayName !== void 0 ? _Component_displayName : Component.name})`;
    return WithProvider;
}
class CardInteractionStore {
    static _makeObservable(instance) {
        (0, _make_observable.makeObservable)(instance, {
            isHovered: _mobx.observable.ref,
            isFocused: _mobx.observable.ref,
            isMouseDown: _mobx.observable.ref,
            trackHover: _mobx.observable.ref,
            trackFocus: _mobx.observable.ref,
            trackMouseDown: _mobx.observable.ref,
            eventHandlers: _mobx.computed,
            setHovered: _mobx.action.bound,
            setFocused: _mobx.action.bound,
            setMouseDown: _mobx.action.bound
        });
    }
    get eventHandlers() {
        const trackHover = _pointer_input_state.pointerInputState.isHoveringInput && this.trackHover;
        return {
            onMouseEnter: trackHover ? ()=>this.setHovered(true) : undefined,
            onMouseLeave: trackHover ? ()=>this.setHovered(false) : undefined,
            onFocus: this.trackFocus ? ()=>this.setFocused(true) : undefined,
            onBlur: this.trackFocus ? ()=>this.setFocused(false) : undefined,
            onMouseDown: this.trackMouseDown ? ()=>this.setMouseDown(true) : undefined,
            onMouseUp: this.trackMouseDown ? ()=>this.setMouseDown(false) : undefined
        };
    }
    setHovered(isHovered) {
        this.isHovered = isHovered;
    }
    setFocused(isFocused) {
        this.isFocused = isFocused;
    }
    setMouseDown(isMouseDown) {
        this.isMouseDown = isMouseDown;
    }
    constructor(){
        this.isHovered = (CardInteractionStore._makeObservable(this), false);
        this.isFocused = false;
        this.isMouseDown = false;
        this.trackHover = false;
        this.trackFocus = false;
        this.trackMouseDown = false;
    }
}
class CardConfigurationStore {
    static _makeObservable(instance) {
        (0, _make_observable.makeObservable)(instance, {
            configuration: _mobx.observable.struct,
            setConfiguration: _mobx.action.bound
        });
    }
    setConfiguration(configuration) {
        this.configuration = configuration;
    }
    constructor(){
        this.configuration = (CardConfigurationStore._makeObservable(this), {
            disabled: false,
            border: 'none',
            borderRadius: 'element',
            layout: 'vertical',
            selected: false,
            selectableMode: 'none',
            selectableBoundary: 'thumbnail',
            hoverEffect: 'none',
            hoverBoundary: 'thumbnail'
        });
    }
}
function createContext() {
    return {
        interactionStore: new CardInteractionStore(),
        configurationStore: new CardConfigurationStore()
    };
}
const CardContext = _react.createContext(undefined);
const useCardContext = ()=>{
    var _React_useContext;
    return (_React_useContext = _react.useContext(CardContext)) !== null && _React_useContext !== void 0 ? _React_useContext : createContext();
};
const useCardConfigurationStore = ()=>useCardContext().configurationStore;
const useCardInteractionStore = ()=>useCardContext().interactionStore;
