// @canva/app-ui-kit/dist/cjs/ui/base/legacy_sheet/internal/layers.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: Object.getOwnPropertyDescriptor(all, name).get
    });
}
_export(exports, {
    get Layer () {
        return Layer;
    },
    get Layers () {
        return Layers;
    }
});
const _jsxruntime = require("react/jsx-runtime");
const _make_observable = require('../../../../base/make_observable/make_observable');
const _mobx = _interop_require_wildcard(require("mobx"));
require("react");
const _client = _interop_require_wildcard(require("react-dom/client"));
const _layer = require('../../layer/layer');
const _temporary_exports = require('../../layer/temporary_exports');
const _provider = require('../../provider/provider');
const _layer_view = require("./layer_view");
function _getRequireWildcardCache(nodeInterop) {
    if (typeof WeakMap !== "function") return null;
    var cacheBabelInterop = new WeakMap();
    var cacheNodeInterop = new WeakMap();
    return (_getRequireWildcardCache = function(nodeInterop) {
        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
    })(nodeInterop);
}
function _interop_require_wildcard(obj, nodeInterop) {
    if (!nodeInterop && obj && obj.__esModule) {
        return obj;
    }
    if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
        return {
            default: obj
        };
    }
    var cache = _getRequireWildcardCache(nodeInterop);
    if (cache && cache.has(obj)) {
        return cache.get(obj);
    }
    var newObj = {
        __proto__: null
    };
    var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
    for(var key in obj){
        if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
            if (desc && (desc.get || desc.set)) {
                Object.defineProperty(newObj, key, desc);
            } else {
                newObj[key] = obj[key];
            }
        }
    }
    newObj.default = obj;
    if (cache) {
        cache.set(obj, newObj);
    }
    return newObj;
}
class Layers {
    addSheet(sheet) {
        if (this.allSheets.has(sheet) || sheet.window == null)
            return;
        const { from, onBottom } = sheet;
        const openLayers = this.layers.filter((l)=>l.isOpen);
        let layer = onBottom ? openLayers[openLayers.length - 1] : openLayers[0];
        if (!layer || layer.forced || layer.from !== from || sheet.props.forceLayer) {
            var _sheet_props_layerLevel;
            const layerLevel = this.calculateNewLayerLevel(onBottom, (_sheet_props_layerLevel = sheet.props.layerLevel) !== null && _sheet_props_layerLevel !== void 0 ? _sheet_props_layerLevel : _layer.LayerLevel.PINS);
            layer = new Layer(from, !!sheet.props.forceLayer, sheet.window, layerLevel);
            layer.root.render((0, _jsxruntime.jsx)(_provider.EaselProvider, {
                enableAnimations: sheet.easelConfiguration.enableAnimations,
                direction: sheet.easelConfiguration.direction,
                disableFocusTraps: sheet.easelConfiguration.disableFocusTraps,
                enableUserSelection: sheet.easelConfiguration.enableUserSelection,
                theme: sheet.easelConfiguration.theme,
                enableHighColorContrast: sheet.easelConfiguration.enableHighColorContrast,
                autoplayVideos: sheet.easelConfiguration.autoplayVideos,
                disableDialogBlur: sheet.easelConfiguration.disableDialogBlur,
                children: (0, _jsxruntime.jsx)(_layer_view.LayerView, {
                    layer: layer,
                    onLayerClose: this.onLayerClose,
                    window: sheet.window
                })
            }));
            if (onBottom) {
                const bottomLayer = this.layers[this.layers.length - 1];
                sheet.window.document.body.insertBefore(layer.element, bottomLayer && bottomLayer.element);
                this.layers.push(layer);
            } else {
                (0, _temporary_exports.insertLayerIntoParent)(layer.element, sheet.window.document.body);
                this.layers.unshift(layer);
            }
        }
        layer.add(sheet);
        this.allSheets.add(sheet);
    }
    removeSheet(sheet) {
        const layer = this.layers.find((l)=>l.sheets.includes(sheet));
        if (layer) {
            layer.remove(sheet);
            this.allSheets.delete(sheet);
            !sheet.enableAnimations && this.onLayerClose(layer);
        }
    }
    constructor(){
        this.layers = [];
        this.allSheets = new Set();
        this.calculateNewLayerLevel = (onBottom, layerLevel)=>{
            if (this.layers.length === 0) return layerLevel;
            if (onBottom) {
                const bottomLayerLevel = this.layers[this.layers.length - 1].layerLevel;
                return layerLevel > bottomLayerLevel ? bottomLayerLevel : layerLevel;
            }
            const topLayerLevel = this.layers[0].layerLevel;
            return layerLevel < topLayerLevel ? topLayerLevel : layerLevel;
        };
        this.onLayerClose = (layer)=>{
            if (layer.sheets.length === 0) {
                layer.root.unmount();
                layer.element.remove();
                this.layers.splice(this.layers.indexOf(layer), 1);
            }
        };
    }
}
class Layer {
    static _makeObservable(instance) {
        (0, _make_observable.makeObservable)(instance, {
            current: _mobx.observable.ref,
            previous: _mobx.observable.ref,
            update: _mobx.action
        });
    }
    remove(sheet) {
        this.sheets.splice(this.sheets.indexOf(sheet), 1);
        this.update();
    }
    add(sheet) {
        if (sheet.onBottom) this.sheets.push(sheet);
        else this.sheets.unshift(sheet);
        this.update();
    }
    get isOpen() {
        return this.sheets.some((s)=>s.show);
    }
    update() {
        const newCurrent = this.sheets[0];
        const oldCurrent = this.current;
        if (newCurrent !== oldCurrent) {
            this.previous = this.previous || !this.currentMounted ? undefined : oldCurrent && !oldCurrent.enableAnimations ? undefined : oldCurrent;
            this.current = newCurrent;
        }
    }
    get enableAnimations() {
        const { current, previous } = this;
        return current ? current.enableAnimations : previous ? previous.enableAnimations : false;
    }
    get hasOuterContent() {
        const { current, previous } = this;
        return !!(current && current.outerElement && current.outerElement.hasChildNodes()) || !!(previous && previous.outerElement && previous.outerElement.hasChildNodes());
    }
    constructor(from, forced, argWindow, layerLevel = _layer.LayerLevel.PINS){
        this.from = from;
        this.forced = forced;
        this.argWindow = argWindow;
        this.layerLevel = layerLevel;
        this.sheets = (Layer._makeObservable(this), []);
        this.currentMounted = false;
        this.isHorizontal = from === 'left' || from === 'right';
        this.element = argWindow.document.createElement('div');
        this.element.classList.add(_temporary_exports.LAYER_MARKER_CLASS_NAME);
        this.element.addEventListener('click', ()=>0);
        this.element.style.position = 'relative';
        this.element.style.zIndex = String(layerLevel);
        this.root = _client.createRoot(this.element);
    }
}
