// @canva/app-ui-kit/dist/cjs/ui/base/legacy_sheet/internal/overlay.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "Overlay", {
    enumerable: true,
    get: function() {
        return Overlay;
    }
});
const _jsxruntime = require("react/jsx-runtime");
const _classnames = _interop_require_default(require("classnames"));
const _react = _interop_require_wildcard(require("react"));
const _reacttransitiongroup = require("react-transition-group");
const _overlaycss = _interop_require_default(require("./overlay.css"));
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function _getRequireWildcardCache(nodeInterop) {
    if (typeof WeakMap !== "function") return null;
    var cacheBabelInterop = new WeakMap();
    var cacheNodeInterop = new WeakMap();
    return (_getRequireWildcardCache = function(nodeInterop) {
        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
    })(nodeInterop);
}
function _interop_require_wildcard(obj, nodeInterop) {
    if (!nodeInterop && obj && obj.__esModule) {
        return obj;
    }
    if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
        return {
            default: obj
        };
    }
    var cache = _getRequireWildcardCache(nodeInterop);
    if (cache && cache.has(obj)) {
        return cache.get(obj);
    }
    var newObj = {
        __proto__: null
    };
    var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
    for(var key in obj){
        if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
            if (desc && (desc.get || desc.set)) {
                Object.defineProperty(newObj, key, desc);
            } else {
                newObj[key] = obj[key];
            }
        }
    }
    newObj.default = obj;
    if (cache) {
        cache.set(obj, newObj);
    }
    return newObj;
}
const cssTransitionClassnames = {
    enter: _overlaycss.default.overlayEnter,
    enterActive: _overlaycss.default.overlayEnterActive,
    exit: _overlaycss.default.overlayExit
};
const Overlay = _react.memo(
    function Overlay({ ariaLabel, show, opacity = 'default', transitionTimeout = 0, onClick }) {
        const nodeRef = (0, _react.useRef)(null);
        const captureRef = (0, _react.useCallback)((newNode)=>{
            var _nodeRef_current;
            newNode === null || newNode === void 0 ? void 0 : newNode.addEventListener('touchmove', preventDefault, {
                passive: false
            });
            (_nodeRef_current = nodeRef.current) === null || _nodeRef_current === void 0 ? void 0 : _nodeRef_current.removeEventListener('touchmove', preventDefault);
            nodeRef.current = newNode;
        }, []);
        const maybeOnClickAccessibleProp = onClick ? {
            tabIndex: 0,
            ['aria-label']: ariaLabel,
            role: 'button',
            onClick
        } : {};
        return (0, _jsxruntime.jsx)(_reacttransitiongroup.CSSTransition, {
            in: show,
            timeout: transitionTimeout,
            unmountOnExit: true,
            mountOnEnter: true,
            classNames: cssTransitionClassnames,
            nodeRef: nodeRef,
            children: (0, _jsxruntime.jsx)("div", {
                ref: captureRef,
                className: (0, _classnames.default)(_overlaycss.default.overlay, {
                    [_overlaycss.default.enableAnimation]: transitionTimeout !== 0,
                    [_overlaycss.default.overlayTransparent]: opacity === 'transparent'
                }),
                ...maybeOnClickAccessibleProp
            })
        });
    }
);
function preventDefault(event) {
    event.preventDefault();
}
