// @canva/app-ui-kit/dist/cjs/ui/apps/developing/ui_kit/entry.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: Object.getOwnPropertyDescriptor(all, name).get
    });
}
_export(exports, {
    get Accordion () {
        return _accordion.Accordion;
    },
    get AccordionItem () {
        return _accordion_item.AccordionItem;
    },
    get Alert () {
        return _alert.Alert;
    },
    get AppUiProvider () {
        return _app_ui_provider.AppUiProvider;
    },
    get AudioCard () {
        return _audio.AudioCard;
    },
    get AudioContextProvider () {
        return _audio_context.AudioContextProvider;
    },
    get Avatar () {
        return _avatar.Avatar;
    },
    get AvatarGroup () {
        return _avatar_group.AvatarGroup;
    },
    get AvatarPlaceholder () {
        return _avatar_placeholder.AvatarPlaceholder;
    },
    get Badge () {
        return _badge.Badge;
    },
    get Box () {
        return _box.Box;
    },
    get Button () {
        return _button.Button;
    },
    get Carousel () {
        return _carousel.Carousel;
    },
    get CharacterCountDecorator () {
        return _character_count.CharacterCountDecorator;
    },
    get Checkbox () {
        return _checkbox.Checkbox;
    },
    get CheckboxGroup () {
        return _checkbox_group.CheckboxGroup;
    },
    get ClearDecorator () {
        return _clear_decorator.ClearDecorator;
    },
    get ColorSelector () {
        return _color_selector.ColorSelector;
    },
    get Column () {
        return _columns.Column;
    },
    get Columns () {
        return _columns.Columns;
    },
    get CustomizableSelect () {
        return _select.CustomizableSelect;
    },
    get DateInput () {
        return _date_input.DateInput;
    },
    get EmbedCard () {
        return _embed.EmbedCard;
    },
    get FileInput () {
        return _file_input.FileInput;
    },
    get FileInputItem () {
        return _file_input_item.FileInputItem;
    },
    get Flyout () {
        return _flyout.Flyout;
    },
    get FlyoutMenu () {
        return _flyout_menu.FlyoutMenu;
    },
    get FlyoutMenuDivider () {
        return _flyout_menu_divider.FlyoutMenuDivider;
    },
    get FlyoutMenuItem () {
        return _flyout_menu_item.FlyoutMenuItem;
    },
    get FormField () {
        return _form_field.FormField;
    },
    get Grid () {
        return _grid.Grid;
    },
    get HorizontalCard () {
        return _horizontal.HorizontalCard;
    },
    get ImageCard () {
        return _image.ImageCard;
    },
    get InputPill () {
        return _input_pill.InputPill;
    },
    get Link () {
        return _link.Link;
    },
    get LoadingIndicator () {
        return _loading_indicator.LoadingIndicator;
    },
    get Masonry () {
        return _masonry.Masonry;
    },
    get MasonryItem () {
        return _masonry.MasonryItem;
    },
    get Menu () {
        return _menu.Menu;
    },
    get MenuDivider () {
        return _menu_divider.MenuDivider;
    },
    get MenuItem () {
        return _menu_item.MenuItem;
    },
    get MultilineInput () {
        return _multiline_input.MultilineInput;
    },
    get NumberInput () {
        return _number_input.NumberInput;
    },
    get Pill () {
        return _pill.Pill;
    },
    get PillsInput () {
        return _pills_input.PillsInput;
    },
    get Placeholder () {
        return _placeholder.Placeholder;
    },
    get ProgressBar () {
        return _progress_bar.ProgressBar;
    },
    get RadioGroup () {
        return _radio_group.RadioGroup;
    },
    get Rows () {
        return _rows.Rows;
    },
    get Scrollable () {
        return _scrollable.Scrollable;
    },
    get SearchInputMenu () {
        return _search_input_menu.SearchInputMenu;
    },
    get SegmentedControl () {
        return _segmented_control.SegmentedControl;
    },
    get Select () {
        return _select.Select;
    },
    get Slider () {
        return _slider.Slider;
    },
    get Swatch () {
        return _swatch.Swatch;
    },
    get Switch () {
        return _switch.Switch;
    },
    get Tab () {
        return _tab.Tab;
    },
    get TabList () {
        return _tab_list.TabList;
    },
    get TabPanel () {
        return _tab_panel.TabPanel;
    },
    get TabPanels () {
        return _tab_panels.TabPanels;
    },
    get Tabs () {
        return _tabs.Tabs;
    },
    get TestAppUiProvider () {
        return _test_app_ui_provider.TestAppUiProvider;
    },
    get Text () {
        return _typography.Text;
    },
    get TextInput () {
        return _text_input.TextInput;
    },
    get TextPlaceholder () {
        return _placeholder1.TextPlaceholder;
    },
    get TimeInput () {
        return _time_input.TimeInput;
    },
    get Title () {
        return _typography.Title;
    },
    get TitlePlaceholder () {
        return _placeholder1.TitlePlaceholder;
    },
    get TruncatedText () {
        return _truncated.TruncatedText;
    },
    get TypographyCard () {
        return _typography1.TypographyCard;
    },
    get VideoCard () {
        return _video.VideoCard;
    },
    get WordCountDecorator () {
        return _word_count.WordCountDecorator;
    },
    get tokens () {
        return _tokens.tokens;
    }
});
require("./initialize");
require("./styles/defaults.css");
require("./styles/tokens.css");
const _accordion = require("./components/accordion/accordion");
const _accordion_item = require("./components/accordion/accordion_item");
const _alert = require("./components/alert/alert");
const _app_ui_provider = require("./components/app_ui_provider/app_ui_provider");
const _test_app_ui_provider = require("./components/app_ui_provider/test_app_ui_provider");
const _audio = require("./components/card/audio/audio");
const _audio_context = require("./components/card/audio/audio_context");
const _avatar = require("./components/avatar/avatar");
const _avatar_group = require("./components/avatar/avatar_group");
const _avatar_placeholder = require("./components/avatar/avatar_placeholder");
const _badge = require("./components/badge/badge");
const _box = require("./components/box/box");
const _button = require("./components/button/button");
const _carousel = require("./components/carousel/carousel");
const _character_count = require("./components/form/input/decorators/character_count");
const _checkbox = require("./components/form/checkbox/checkbox");
const _checkbox_group = require("./components/form/checkbox/checkbox_group");
const _clear_decorator = require("./components/form/input/decorators/clear_decorator");
const _color_selector = require("./components/color_picker/color_selector");
const _columns = require("./components/layout/columns/columns");
const _date_input = require("./components/form/input/date_input");
const _embed = require("./components/card/embed/embed");
const _file_input = require("./components/form/upload/file_input");
const _file_input_item = require("./components/form/upload/file_input_item");
const _flyout = require("./components/flyout/flyout");
const _flyout_menu = require("./components/flyout_menu/flyout_menu");
const _flyout_menu_divider = require("./components/flyout_menu/flyout_menu_divider");
const _flyout_menu_item = require("./components/flyout_menu/flyout_menu_item");
const _form_field = require("./components/form/form_field/form_field");
const _grid = require("./components/layout/grid/grid");
const _horizontal = require("./components/card/horizontal/horizontal");
const _image = require("./components/card/image/image");
const _input_pill = require("./components/form/pills_input/input_pill/input_pill");
const _link = require("./components/link/link");
const _loading_indicator = require("./components/loading_indicator/loading_indicator");
const _menu = require("./components/menu/menu");
const _menu_divider = require("./components/menu/menu_divider");
const _menu_item = require("./components/menu/menu_item");
const _masonry = require("./components/masonry/masonry");
const _multiline_input = require("./components/form/input/multiline_input");
const _number_input = require("./components/form/input/number_input");
const _pill = require("./components/pill/pill");
const _pills_input = require("./components/form/pills_input/pills_input");
const _placeholder = require("./components/placeholder/placeholder");
const _progress_bar = require("./components/progress_bar/progress_bar");
const _radio_group = require("./components/form/radio_group/radio_group");
const _rows = require("./components/layout/rows/rows");
const _scrollable = require("./components/scrollable/scrollable");
const _search_input_menu = require("./components/form/input/search_input_menu");
const _segmented_control = require("./components/form/segmented_control/segmented_control");
const _select = require("./components/form/select/select");
const _slider = require("./components/form/slider/slider");
const _swatch = require("./components/swatch/swatch");
const _switch = require("./components/form/switch/switch");
const _tab = require("./components/tabs/tab");
const _tab_list = require("./components/tabs/tab_list");
const _tab_panel = require("./components/tabs/tab_panel");
const _tab_panels = require("./components/tabs/tab_panels");
const _tabs = require("./components/tabs/tabs");
const _typography = require("./components/typography/typography");
const _text_input = require("./components/form/input/text_input");
const _placeholder1 = require("./components/typography/placeholder/placeholder");
const _time_input = require("./components/form/input/time_input");
const _truncated = require("./components/typography/truncated/truncated");
const _typography1 = require("./components/card/typography/typography");
const _video = require("./components/card/video/video");
const _word_count = require("./components/form/input/decorators/word_count");
_export_star(require("./icons/icons"), exports);
const _tokens = require("./styles/tokens");
function _export_star(from, to) {
    Object.keys(from).forEach(function(k) {
        if (k !== "default" && !Object.prototype.hasOwnProperty.call(to, k)) {
            Object.defineProperty(to, k, {
                enumerable: true,
                get: function() {
                    return from[k];
                }
            });
        }
    });
    return from;
}
