// @canva/app-ui-kit/dist/cjs/ui/base/button/button.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: Object.getOwnPropertyDescriptor(all, name).get
    });
}
_export(exports, {
    get BasicButton () {
        return _button.BasicButton;
    },
    get BasicButtonLink () {
        return _button.BasicButtonLink;
    },
    get Button () {
        return _button.Button;
    },
    get ButtonLink () {
        return _button.ButtonLink;
    },
    get CircleButton () {
        return _button.CircleButton;
    },
    get ConjoinedButtons () {
        return _conjoined_buttons.ConjoinedButtons;
    },
    get NeutralButton () {
        return _neutral_button.NeutralButton;
    },
    get NeutralButtonLink () {
        return _neutral_button.NeutralButtonLink;
    }
});
const _button = require("./private/button");
const _neutral_button = require("./private/neutral_button");
const _conjoined_buttons = require("./private/conjoined_buttons");
