// @canva/app-ui-kit/dist/esm/ui/base/device_capabilities/device_pixel_ratio/device_pixel_ratio.js
import { $mobx, computed, createAtom, observe } from 'mobx';
import { useSyncExternalStore } from 'react';
let prop;
export class DevicePixelRatioMonitor {
    startTracking() {
        var _this_window;
        if (((_this_window = this.window) === null || _this_window === void 0 ? void 0 : _this_window.matchMedia) && !this.match) {
            this.match = this.window.matchMedia(`(resolution: ${this.window.devicePixelRatio}dppx)`);
            this.match.addListener(this.onChanged);
        }
    }
    stopTracking() {
        if (this.match) {
            this.match.removeListener(this.onChanged);
            this.match = undefined;
        }
    }
    get value() {
        var _this_window;
        this[$mobx].reportObserved();
        var _this_window_devicePixelRatio;
        return (_this_window_devicePixelRatio = (_this_window = this.window) === null || _this_window === void 0 ? void 0 : _this_window.devicePixelRatio) !== null && _this_window_devicePixelRatio !== void 0 ? _this_window_devicePixelRatio : 1;
    }
    constructor(window1){
        this.window = window1;
        this[_$mobx] = createAtom('DevicePixelRatio', ()=>this.startTracking(), ()=>this.stopTracking());
        this.onChanged = ()=>{
            this[$mobx].reportChanged();
            this.stopTracking();
            this.startTracking();
        };
    }
}
prop = _$mobx = $mobx;
const globalWindow = typeof window !== 'undefined' ? window : undefined;
const monitor = new DevicePixelRatioMonitor(globalWindow);
export function useDevicePixelRatio() {
    return useSyncExternalStore(subscribeDevicePixelRatio, getClientDevicePixelRatioValue, getServerDevicePixelRatioValue);
}
export function getDevicePixelRatio() {
    return monitor.value;
}
const computedDevicePixelRatio = computed(getDevicePixelRatio);
const subscribeDevicePixelRatio = (callback)=>observe(computedDevicePixelRatio, callback);
const getClientDevicePixelRatioValue = ()=>computedDevicePixelRatio.get();
const getServerDevicePixelRatioValue = ()=>1;
var _$mobx;
