// @canva/app-ui-kit/dist/cjs/ui/base/date_time/calendar/private/day.js
"use strict"
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: Object.getOwnPropertyDescriptor(all, name).get
    });
}
_export(exports, {
    get CalendarDay () {
        return CalendarDay;
    },
    get CalendarDayWithWrapper () {
        return CalendarDayWithWrapper;
    }
});
const _jsxruntime = require("react/jsx-runtime");
const _classnames = _interop_require_default(require("classnames"));
const _react = _interop_require_wildcard(require("react"));
const _utils = require('../../utils/utils');
const _device_capabilities = require('../../../device_capabilities/device_capabilities');
const _daycss = _interop_require_default(require("./day.css"));
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function _getRequireWildcardCache(nodeInterop) {
    if (typeof WeakMap !== "function") return null;
    var cacheBabelInterop = new WeakMap();
    var cacheNodeInterop = new WeakMap();
    return (_getRequireWildcardCache = function(nodeInterop) {
        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
    })(nodeInterop);
}
function _interop_require_wildcard(obj, nodeInterop) {
    if (!nodeInterop && obj && obj.__esModule) {
        return obj;
    }
    if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
        return {
            default: obj
        };
    }
    var cache = _getRequireWildcardCache(nodeInterop);
    if (cache && cache.has(obj)) {
        return cache.get(obj);
    }
    var newObj = {
        __proto__: null
    };
    var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
    for(var key in obj){
        if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
            if (desc && (desc.get || desc.set)) {
                Object.defineProperty(newObj, key, desc);
            } else {
                newObj[key] = obj[key];
            }
        }
    }
    newObj.default = obj;
    if (cache) {
        cache.set(obj, newObj);
    }
    return newObj;
}
const CalendarDay = _react.memo(
    ({ locale, date, current, today, past, selected, selectedInvalid, disabled, className: classNameProp, dataDate: shouldIncludeDataDate, onSelect, onDayRef, includeInTabOrder, label: labelProp })=>{
        const label = labelProp || (0, _utils.format)(date, locale, 'DD MMMM');
        const dayNumber = (0, _utils.toDate)(date).getUTCDate();
        const dataDate = shouldIncludeDataDate ? date.split('T')[0] : undefined;
        const className = (0, _classnames.default)(_daycss.default.day, classNameProp, {
            [_daycss.default.hoverSupported]: (0, _device_capabilities.canHover)(),
            [_daycss.default.current]: !disabled && current,
            [_daycss.default.disabled]: disabled,
            [_daycss.default.today]: today,
            [_daycss.default.selected]: selected,
            [_daycss.default.invalid]: selected && selectedInvalid
        });
        const getText = ()=>{
            return (0, _jsxruntime.jsx)("span", {
                className: className,
                "aria-label": label,
                ref: onDayRef,
                "data-date": dataDate,
                children: dayNumber
            });
        };
        const onClick = _react.useCallback((event)=>{
            event.preventDefault();
            onSelect === null || onSelect === void 0 ? void 0 : onSelect(date);
        }, [
            date,
            onSelect
        ]);
        const getButton = ()=>{
            return (0, _jsxruntime.jsx)("button", {
                className: className,
                onClick: onClick,
                ref: onDayRef,
                "data-date": dataDate,
                tabIndex: includeInTabOrder ? 0 : -1,
                "aria-label": label,
                "aria-current": today ? 'date' : undefined,
                "aria-pressed": selected ? 'true' : undefined,
                children: dayNumber
            });
        };
        return !disabled && onSelect ? getButton() : getText();
    }
);
const CalendarDayWithWrapper = ({ locale, date, current, today, past, selected, selectedInvalid, disabled, className, onSelect, onDayRef, includeInTabOrder })=>(0, _jsxruntime.jsx)("div", {
        className: _daycss.default.dayWrapper,
        role: "gridcell",
        "aria-selected": selected ? 'true' : undefined,
        "aria-disabled": disabled ? 'true' : undefined,
        children: (0, _jsxruntime.jsx)(CalendarDay, {
            locale: locale,
            date: date,
            current: current,
            today: today,
            past: past,
            selected: selected,
            selectedInvalid: selectedInvalid,
            disabled: disabled,
            className: className,
            onSelect: onSelect,
            onDayRef: onDayRef,
            includeInTabOrder: includeInTabOrder
        })
    });
