// @canva/app-ui-kit/dist/cjs/ui/base/date_time/calendar/private/days_week_labels.js
"use strict"
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: Object.getOwnPropertyDescriptor(all, name).get
    });
}
_export(exports, {
    get CalendarDayOfWeekLabel () {
        return CalendarDayOfWeekLabel;
    },
    get CalendarDaysOfWeekLabels () {
        return CalendarDaysOfWeekLabels;
    }
});
const _jsxruntime = require("react/jsx-runtime");
const _classnames = _interop_require_default(require("classnames"));
const _react = _interop_require_wildcard(require("react"));
const _utils = require('../../utils/utils');
const _days_week_labelscss = _interop_require_default(require("./days_week_labels.css"));
const _week = require("./week");
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function _getRequireWildcardCache(nodeInterop) {
    if (typeof WeakMap !== "function") return null;
    var cacheBabelInterop = new WeakMap();
    var cacheNodeInterop = new WeakMap();
    return (_getRequireWildcardCache = function(nodeInterop) {
        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
    })(nodeInterop);
}
function _interop_require_wildcard(obj, nodeInterop) {
    if (!nodeInterop && obj && obj.__esModule) {
        return obj;
    }
    if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
        return {
            default: obj
        };
    }
    var cache = _getRequireWildcardCache(nodeInterop);
    if (cache && cache.has(obj)) {
        return cache.get(obj);
    }
    var newObj = {
        __proto__: null
    };
    var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
    for(var key in obj){
        if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
            if (desc && (desc.get || desc.set)) {
                Object.defineProperty(newObj, key, desc);
            } else {
                newObj[key] = obj[key];
            }
        }
    }
    newObj.default = obj;
    if (cache) {
        cache.set(obj, newObj);
    }
    return newObj;
}
const CalendarDaysOfWeekLabels = _react.memo(
    ({ locale, className, DayOfWeekLabel = CalendarDayOfWeekLabel, startingDate = new Date(_utils.MONDAY_DATE) })=>{
        let date = startingDate;
        const days = [];
        for(let i = 0; i < _week.DAYS_IN_WEEK; i++){
            days.push({
                weekDayDate: (0, _utils.toISO)(date),
                name: (0, _utils.format)(date, locale, 'weekdayShort'),
                label: (0, _utils.format)(date, locale, 'weekday')
            });
            date = (0, _utils.nextDay)(date);
        }
        return (0, _jsxruntime.jsx)("div", {
            className: (0, _classnames.default)(_days_week_labelscss.default.daysOfWeek, className),
            children: (0, _jsxruntime.jsx)(_week.CalendarWeek, {
                children: days.map((day, i)=>(0, _jsxruntime.jsx)(DayOfWeekLabel, {
                        weekDayDate: day.weekDayDate,
                        name: day.name,
                        label: day.label
                    }, `dayWeek${i}`))
            })
        });
    }
);
const CalendarDayOfWeekLabel = _react.memo(({ name, label, className })=>(0, _jsxruntime.jsx)("div", {
        className: (0, _classnames.default)(_days_week_labelscss.default.dayWeek, className),
        "aria-label": label,
        "aria-readonly": "true",
        role: "columnheader",
        children: name
    }));
