// @canva/app-ui-kit/dist/cjs/ui/base/theme/theme.js
"use strict"

Object.defineProperty(exports, "__esModule", {
  value: true
});
function _export(target, all) {
  for (var name in all) Object.defineProperty(target, name, {
    enumerable: true,
    get: Object.getOwnPropertyDescriptor(all, name).get
  });
}
_export(exports, {
  get ApplyThemeRoot() {
    return ApplyThemeRoot;
  },
  get ThemeBoundary() {
    return ThemeBoundary;
  },
  get WithThemeData() {
    return WithThemeData;
  },
  get cleanThemePreload() {
    return cleanThemePreload;
  },
  get useThemeData() {
    return useThemeData;
  }
});
const _jsxruntime = require("react/jsx-runtime");
const _preconditions = require('../../../base/preconditions');
const _mobx = require("mobx");
const _mobxreactlite = require("mobx-react-lite");
const _react = _interop_require_wildcard(require("react"));
const _provider = require('../provider/provider');
const _adaptive = require("./adaptive");
const _internal = require("./internal");
function _getRequireWildcardCache(nodeInterop) {
  if (typeof WeakMap !== "function") return null;
  var cacheBabelInterop = new WeakMap();
  var cacheNodeInterop = new WeakMap();
  return (_getRequireWildcardCache = function (nodeInterop) {
    return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
  })(nodeInterop);
}
function _interop_require_wildcard(obj, nodeInterop) {
  if (!nodeInterop && obj && obj.__esModule) {
    return obj;
  }
  if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
    return {
      default: obj
    };
  }
  var cache = _getRequireWildcardCache(nodeInterop);
  if (cache && cache.has(obj)) {
    return cache.get(obj);
  }
  var newObj = {
    __proto__: null
  };
  var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
  for (var key in obj) {
    if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
      var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
      if (desc && (desc.get || desc.set)) {
        Object.defineProperty(newObj, key, desc);
      } else {
        newObj[key] = obj[key];
      }
    }
  }
  newObj.default = obj;
  if (cache) {
    cache.set(obj, newObj);
  }
  return newObj;
}
const ADAPTIVE_APPEARANCES = ['classicAdaptive', 'adaptive'];
function cleanThemePreload() {
  const preloadClasses = (0, _internal.getAllPreloadClasses)();
  document.documentElement.classList.remove(...preloadClasses);
  document.body.classList.remove(...preloadClasses);
}
function ApplyThemeRoot({
  appearance: givenAppearance,
  ...ccTheme
}) {
  const appearance = useAppearance({
    theme: (0, _provider.useTheme)(),
    enableHighColorContrast: (0, _provider.useEnableHighColorContrast)()
  }, givenAppearance);
  const theme = useThemeMapping(ccTheme);
  const adaptiveColorScheme = (0, _adaptive.useAdaptiveColorScheme)({
    enabled: ADAPTIVE_APPEARANCES.includes(appearance)
  });
  const resolvedAppearance = resolveAdaptiveAppearance({
    colorScheme: adaptiveColorScheme,
    appearance
  });
  const mountedRef = _react.useRef(false);
  if (!mountedRef.current) {
    mountedRef.current = true;
    (0, _mobx.runInAction)(() => {
      _internal.rootThemeStore.setAppearance(resolvedAppearance);
      _internal.rootThemeStore.setThemeMapping(theme);
    });
  }
  useIsomorphicLayoutEffect((0, _mobx.action)(() => {
    _internal.rootThemeStore.setAppearance(resolvedAppearance);
    _internal.rootThemeStore.setThemeMapping(theme);
  }), [resolvedAppearance, theme]);
  useIsomorphicLayoutEffect(() => (0, _mobx.action)(() => {
    _internal.rootThemeStore.setAppearance(undefined);
    _internal.rootThemeStore.setThemeMapping({});
  }), []);
  useIsomorphicLayoutEffect(cleanThemePreload, []);
  useIsomorphicLayoutEffect(() => (0, _mobx.reaction)(() => _internal.rootThemeStore.currentTheme, (newTheme, prevTheme) => {
    const {
      classList
    } = document.documentElement;
    if (prevTheme != null) classList.remove(...(0, _internal.getThemeClasses)(prevTheme));
    if (newTheme != null) classList.add(...(0, _internal.getThemeClasses)(newTheme));
  }, {
    fireImmediately: true
  }), []);
  return null;
}
function resolveAdaptiveAppearance(
  {
    colorScheme,
    appearance
  }
) {
  switch (appearance) {
    case 'classicAdaptive':
      return colorScheme === 'dark' ? 'classicDark' : 'classicLight';
    case 'adaptive':
      return colorScheme !== null && colorScheme !== void 0 ? colorScheme : 'light';
    default:
      return appearance;
  }
}
const canUseDOM = () => typeof window !== 'undefined' && typeof window.document !== 'undefined' && typeof window.document.createElement !== 'undefined';
const useIsomorphicLayoutEffect = canUseDOM() ? _react.useLayoutEffect : _react.useEffect;
const ThemeBoundary = (0, _mobxreactlite.observer)(function ThemeBoundary({
  children,
  ...ccTheme
}) {
  const theme = useThemeMapping(ccTheme);
  const [themeContext] = _react.useState(() => {
    const store = new _internal.ThemeStore();
    return {
      store,
      data: new _internal.ThemeData(store)
    };
  });
  const {
    store: parentStore
  } = _react.useContext(_internal.ThemeContext);
  themeContext.store.setParent(parentStore);
  themeContext.store.setThemeMapping(theme);
  return (0, _jsxruntime.jsx)(_internal.ThemeContext.Provider, {
    value: themeContext,
    children: children(themeContext.data)
  });
});
function useThemeMapping(theme) {
  return _react.useMemo(() => ({
    ['classicLight']: theme.classicLight,
    ['classicDark']: theme.classicDark,
    ['light']: theme.light,
    ['dark']: theme.dark
  }), [theme.classicLight, theme.classicDark, theme.light, theme.dark]);
}
const useThemeData = () => _react.useContext(_internal.ThemeContext).data;
const WithThemeData = (0, _mobxreactlite.observer)(function WithThemeData({
  children
}) {
  const data = useThemeData();
  return children(data);
});
function useAppearance(
  {
    theme,
    enableHighColorContrast
  },
  givenAppearance
) {
  switch (theme) {
    case undefined:
      return givenAppearance;
    case 'ADAPTIVE':
      return enableHighColorContrast ? 'adaptive' : 'classicAdaptive';
    case 'LIGHT':
      return enableHighColorContrast ? 'light' : 'classicLight';
    case 'DARK':
      return enableHighColorContrast ? 'dark' : 'classicDark';
    default:
      throw new _preconditions.UnreachableError(theme);
  }
}
