// @canva/app-ui-kit/dist/esm/ui/base/date_time/calendar/private/day.js
import { jsx as _jsx } from "react/jsx-runtime";
import classNames from 'classnames';
import * as React from 'react';
import { format, toDate } from '../../utils/utils';
import { canHover } from '../../../device_capabilities/device_capabilities';
import styles from './day.css';
export const CalendarDay = React.memo(
    ({ locale, date, current, today, past, selected, selectedInvalid, disabled, className: classNameProp, dataDate: shouldIncludeDataDate, onSelect, onDayRef, includeInTabOrder, label: labelProp })=>{
        const label = labelProp || format(date, locale, 'DD MMMM');
        const dayNumber = toDate(date).getUTCDate();
        const dataDate = shouldIncludeDataDate ? date.split('T')[0] : undefined;
        const className = classNames(styles.day, classNameProp, {
            [styles.hoverSupported]: canHover(),
            [styles.current]: !disabled && current,
            [styles.disabled]: disabled,
            [styles.today]: today,
            [styles.selected]: selected,
            [styles.invalid]: selected && selectedInvalid
        });
        const getText = ()=>{
            return _jsx("span", {
                className: className,
                "aria-label": label,
                ref: onDayRef,
                "data-date": dataDate,
                children: dayNumber
            });
        };
        const onClick = React.useCallback((event)=>{
            event.preventDefault();
            onSelect === null || onSelect === void 0 ? void 0 : onSelect(date);
        }, [
            date,
            onSelect
        ]);
        const getButton = ()=>{
            return _jsx("button", {
                className: className,
                onClick: onClick,
                ref: onDayRef,
                "data-date": dataDate,
                tabIndex: includeInTabOrder ? 0 : -1,
                "aria-label": label,
                "aria-current": today ? 'date' : undefined,
                "aria-pressed": selected ? 'true' : undefined,
                children: dayNumber
            });
        };
        return !disabled && onSelect ? getButton() : getText();
    }
);
export const CalendarDayWithWrapper = ({ locale, date, current, today, past, selected, selectedInvalid, disabled, className, onSelect, onDayRef, includeInTabOrder })=>_jsx("div", {
        className: styles.dayWrapper,
        role: "gridcell",
        "aria-selected": selected ? 'true' : undefined,
        "aria-disabled": disabled ? 'true' : undefined,
        children: _jsx(CalendarDay, {
            locale: locale,
            date: date,
            current: current,
            today: today,
            past: past,
            selected: selected,
            selectedInvalid: selectedInvalid,
            disabled: disabled,
            className: className,
            onSelect: onSelect,
            onDayRef: onDayRef,
            includeInTabOrder: includeInTabOrder
        })
    });
