// @canva/app-ui-kit/dist/esm/ui/base/date_time/calendar/private/days_week_labels.js
import { jsx as _jsx } from "react/jsx-runtime";
import classNames from 'classnames';
import * as React from 'react';
import { format, MONDAY_DATE, nextDay, toISO } from '../../utils/utils';
import styles from './days_week_labels.css';
import { CalendarWeek, DAYS_IN_WEEK } from './week';
export const CalendarDaysOfWeekLabels = React.memo(
    ({ locale, className, DayOfWeekLabel = CalendarDayOfWeekLabel, startingDate = new Date(MONDAY_DATE) })=>{
        let date = startingDate;
        const days = [];
        for(let i = 0; i < DAYS_IN_WEEK; i++){
            days.push({
                weekDayDate: toISO(date),
                name: format(date, locale, 'weekdayShort'),
                label: format(date, locale, 'weekday')
            });
            date = nextDay(date);
        }
        return _jsx("div", {
            className: classNames(styles.daysOfWeek, className),
            children: _jsx(CalendarWeek, {
                children: days.map((day, i)=>_jsx(DayOfWeekLabel, {
                        weekDayDate: day.weekDayDate,
                        name: day.name,
                        label: day.label
                    }, `dayWeek${i}`))
            })
        });
    }
);
export const CalendarDayOfWeekLabel = React.memo(({ name, label, className })=>_jsx("div", {
        className: classNames(styles.dayWeek, className),
        "aria-label": label,
        "aria-readonly": "true",
        role: "columnheader",
        children: name
    }));
