// @canva/app-ui-kit/dist/cjs/ui/base/button/private/button.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: Object.getOwnPropertyDescriptor(all, name).get
    });
}
_export(exports, {
    get BasicButton () {
        return BasicButton;
    },
    get BasicButtonLink () {
        return BasicButtonLink;
    },
    get Button () {
        return Button;
    },
    get ButtonLink () {
        return ButtonLink;
    },
    get CircleButton () {
        return CircleButton;
    },
    get LinkButton () {
        return LinkButton;
    },
    get PrimaryButton () {
        return PrimaryButton;
    },
    get SecondaryButton () {
        return SecondaryButton;
    },
    get TertiaryButton () {
        return TertiaryButton;
    }
});
const _jsxruntime = require("react/jsx-runtime");
const _react = _interop_require_wildcard(require("react"));
const _internal_button = require("./internal_button");
function _getRequireWildcardCache(nodeInterop) {
    if (typeof WeakMap !== "function") return null;
    var cacheBabelInterop = new WeakMap();
    var cacheNodeInterop = new WeakMap();
    return (_getRequireWildcardCache = function(nodeInterop) {
        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
    })(nodeInterop);
}
function _interop_require_wildcard(obj, nodeInterop) {
    if (!nodeInterop && obj && obj.__esModule) {
        return obj;
    }
    if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
        return {
            default: obj
        };
    }
    var cache = _getRequireWildcardCache(nodeInterop);
    if (cache && cache.has(obj)) {
        return cache.get(obj);
    }
    var newObj = {
        __proto__: null
    };
    var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
    for(var key in obj){
        if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
            if (desc && (desc.get || desc.set)) {
                Object.defineProperty(newObj, key, desc);
            } else {
                newObj[key] = obj[key];
            }
        }
    }
    newObj.default = obj;
    if (cache) {
        cache.set(obj, newObj);
    }
    return newObj;
}
const Button = _react.forwardRef((props, ref)=>{
    return (0, _jsxruntime.jsx)(_internal_button.InternalButton, {
        ...props,
        ref: ref
    });
});
const ButtonLink = _react.forwardRef((props, ref)=>{
    return (0, _jsxruntime.jsx)(_internal_button.InternalAnchor, {
        ...props,
        ref: ref
    });
});
const BasicButton = _react.forwardRef((props, ref)=>{
    return (0, _internal_button.renderButtonOrAnchor)({
        ...props,
        variant: 'basic'
    }, ref);
});
const BasicButtonLink = _react.forwardRef((props, ref)=>{
    return (0, _jsxruntime.jsx)(_internal_button.InternalAnchor, {
        ...props,
        variant: "basic",
        ref: ref
    });
});
const PrimaryButton = (props)=>(0, _internal_button.renderButtonOrAnchor)({
        ...props,
        variant: 'primary'
    });
const SecondaryButton = (props)=>(0, _internal_button.renderButtonOrAnchor)({
        ...props,
        variant: 'secondary'
    });
const TertiaryButton = (props)=>(0, _internal_button.renderButtonOrAnchor)({
        ...props,
        variant: 'tertiary'
    });
const LinkButton = (props)=>(0, _internal_button.renderButtonOrAnchor)({
        ...props,
        variant: 'linkButton'
    });
function isCircleButtonWithChildren(props) {
    return typeof props.children !== 'undefined';
}
const CircleButton = _react.forwardRef((props, ref)=>{
    if (isCircleButtonWithChildren(props)) return (0, _jsxruntime.jsx)(_internal_button.CircleButtonWithLabel, {
        ref: ref,
        ...props
    });
    return (0, _jsxruntime.jsx)(_internal_button.CircleButtonWithoutLabel, {
        ref: ref,
        ...props
    });
});
