// @canva/app-ui-kit/dist/cjs/ui/apps/developing/ui_kit/styles/tokens.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "tokens", {
    enumerable: true,
    get: function() {
        return tokens;
    }
});
const _tokenscss = _interop_require_default(require("./tokens.css"));
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
const tokens = {
    space0: _tokenscss.default.uiKitSpace0,
    space050: _tokenscss.default.uiKitSpace050,
    space1: _tokenscss.default.uiKitSpace1,
    space150: _tokenscss.default.uiKitSpace150,
    space2: _tokenscss.default.uiKitSpace2,
    space3: _tokenscss.default.uiKitSpace3,
    space4: _tokenscss.default.uiKitSpace4,
    space6: _tokenscss.default.uiKitSpace6,
    space8: _tokenscss.default.uiKitSpace8,
    space12: _tokenscss.default.uiKitSpace12,
    baseUnit: _tokenscss.default.uiKitBaseUnit,
    borderRadius: _tokenscss.default.uiKitBorderRadius,
    minTouchableArea: _tokenscss.default.uiKitMinTouchableArea,
    shadowSurface: _tokenscss.default.uiKitShadowSurface,
    shadowSurfaceHigh: _tokenscss.default.uiKitShadowSurfaceHigh,
    /**  @deprecated use shadowSurface instead. */ boxShadowFaint: _tokenscss.default.uiKitBoxShadowFaint,
    /**  @deprecated use shadowSurface instead. */ boxShadowMedium: _tokenscss.default.uiKitBoxShadowMedium,
    /**  @deprecated use shadowSurfaceHigh instead. */ boxShadowHeavy: _tokenscss.default.uiKitBoxShadowHeavy,
    hoverTransition: _tokenscss.default.uiKitHoverTransition,
    fadeTransition: _tokenscss.default.uiKitFadeTransition,
    colorOverlay: _tokenscss.default.uiKitColorOverlay,
    colorPrimary: _tokenscss.default.uiKitColorPrimary,
    colorPrimaryHover: _tokenscss.default.uiKitColorPrimaryHover,
    colorPrimaryActive: _tokenscss.default.uiKitColorPrimaryActive,
    colorPrimaryDisabled: _tokenscss.default.uiKitColorPrimaryDisabled,
    colorPrimaryFore: _tokenscss.default.uiKitColorPrimaryFore,
    colorPrimaryForeDisabled: _tokenscss.default.uiKitColorPrimaryForeDisabled,
    colorSecondary: _tokenscss.default.uiKitColorSecondary,
    colorSecondaryHover: _tokenscss.default.uiKitColorSecondaryHover,
    colorSecondaryActive: _tokenscss.default.uiKitColorSecondaryActive,
    colorSecondaryDisabled: _tokenscss.default.uiKitColorSecondaryDisabled,
    colorSecondaryFore: _tokenscss.default.uiKitColorSecondaryFore,
    colorSecondaryForeDisabled: _tokenscss.default.uiKitColorSecondaryForeDisabled,
    colorTertiary: _tokenscss.default.uiKitColorTertiary,
    colorTertiaryHover: _tokenscss.default.uiKitColorTertiaryHover,
    colorTertiaryActive: _tokenscss.default.uiKitColorTertiaryActive,
    colorTertiaryDisabled: _tokenscss.default.uiKitColorTertiaryDisabled,
    colorTertiaryFore: _tokenscss.default.uiKitColorTertiaryFore,
    colorTertiaryForeDisabled: _tokenscss.default.uiKitColorTertiaryForeDisabled,
    colorContrast: _tokenscss.default.uiKitColorContrast,
    colorContrastHover: _tokenscss.default.uiKitColorContrastHover,
    colorContrastActive: _tokenscss.default.uiKitColorContrastActive,
    colorContrastDisabled: _tokenscss.default.uiKitColorContrastDisabled,
    colorContrastFore: _tokenscss.default.uiKitColorContrastFore,
    colorContrastForeDisabled: _tokenscss.default.uiKitColorContrastForeDisabled,
    colorNeutral: _tokenscss.default.uiKitColorNeutral,
    colorNeutralHover: _tokenscss.default.uiKitColorNeutralHover,
    colorNeutralActive: _tokenscss.default.uiKitColorNeutralActive,
    colorNeutralDisabled: _tokenscss.default.uiKitColorNeutralDisabled,
    colorNeutralLow: _tokenscss.default.uiKitColorNeutralLow,
    colorNeutralFore: _tokenscss.default.uiKitColorNeutralFore,
    colorNeutralForeLow: _tokenscss.default.uiKitColorNeutralForeLow,
    colorNeutralForeDisabled: _tokenscss.default.uiKitColorNeutralForeDisabled,
    colorPositive: _tokenscss.default.uiKitColorPositive,
    colorPositiveHover: _tokenscss.default.uiKitColorPositiveHover,
    colorPositiveActive: _tokenscss.default.uiKitColorPositiveActive,
    colorPositiveDisabled: _tokenscss.default.uiKitColorPositiveDisabled,
    colorPositiveLow: _tokenscss.default.uiKitColorPositiveLow,
    colorPositiveFore: _tokenscss.default.uiKitColorPositiveFore,
    colorPositiveForeLow: _tokenscss.default.uiKitColorPositiveForeLow,
    colorPositiveForeDisabled: _tokenscss.default.uiKitColorPositiveForeDisabled,
    colorInfo: _tokenscss.default.uiKitColorInfo,
    colorInfoHover: _tokenscss.default.uiKitColorInfoHover,
    colorInfoActive: _tokenscss.default.uiKitColorInfoActive,
    colorInfoDisabled: _tokenscss.default.uiKitColorInfoDisabled,
    colorInfoLow: _tokenscss.default.uiKitColorInfoLow,
    colorInfoFore: _tokenscss.default.uiKitColorInfoFore,
    colorInfoForeLow: _tokenscss.default.uiKitColorInfoForeLow,
    colorInfoForeDisabled: _tokenscss.default.uiKitColorInfoForeDisabled,
    colorWarn: _tokenscss.default.uiKitColorWarn,
    colorWarnHover: _tokenscss.default.uiKitColorWarnHover,
    colorWarnActive: _tokenscss.default.uiKitColorWarnActive,
    colorWarnDisabled: _tokenscss.default.uiKitColorWarnDisabled,
    colorWarnLow: _tokenscss.default.uiKitColorWarnLow,
    colorWarnFore: _tokenscss.default.uiKitColorWarnFore,
    colorWarnForeLow: _tokenscss.default.uiKitColorWarnForeLow,
    colorWarnForeDisabled: _tokenscss.default.uiKitColorWarnForeDisabled,
    colorCritical: _tokenscss.default.uiKitColorCritical,
    colorCriticalHover: _tokenscss.default.uiKitColorCriticalHover,
    colorCriticalActive: _tokenscss.default.uiKitColorCriticalActive,
    colorCriticalDisabled: _tokenscss.default.uiKitColorCriticalDisabled,
    colorCriticalLow: _tokenscss.default.uiKitColorCriticalLow,
    colorCriticalFore: _tokenscss.default.uiKitColorCriticalFore,
    colorCriticalForeLow: _tokenscss.default.uiKitColorCriticalForeLow,
    colorCriticalForeDisabled: _tokenscss.default.uiKitColorCriticalForeDisabled,
    colorCanvas: _tokenscss.default.uiKitColorCanvas,
    colorTabdock: _tokenscss.default.uiKitColorTabdock,
    colorPage: _tokenscss.default.uiKitColorPage,
    colorSurface: _tokenscss.default.uiKitColorSurface,
    colorTypographyPrimary: _tokenscss.default.uiKitColorTypographyPrimary,
    colorTypographySecondary: _tokenscss.default.uiKitColorTypographySecondary,
    colorTypographyTertiary: _tokenscss.default.uiKitColorTypographyTertiary,
    /**  @deprecated - use `colorTypographyPlaceholder` instead */ colorTypographyQuaternary: _tokenscss.default.uiKitColorTypographyQuaternary,
    colorTypographyPlaceholder: _tokenscss.default.uiKitColorTypographyPlaceholder,
    colorTypographyPositive: _tokenscss.default.uiKitColorTypographyPositive,
    colorTypographyCritical: _tokenscss.default.uiKitColorTypographyCritical,
    colorTypographyCriticalHover: _tokenscss.default.uiKitColorTypographyCriticalHover,
    colorTypographyCriticalActive: _tokenscss.default.uiKitColorTypographyCriticalActive,
    colorTypographyWarn: _tokenscss.default.uiKitColorTypographyWarn,
    colorTypographyInfo: _tokenscss.default.uiKitColorTypographyInfo,
    colorTypographyLink: _tokenscss.default.uiKitColorTypographyLink,
    colorTypographyLinkHover: _tokenscss.default.uiKitColorTypographyLinkHover,
    colorTypographyLinkActive: _tokenscss.default.uiKitColorTypographyLinkActive,
    colorBorder: _tokenscss.default.uiKitColorBorder,
    colorBorderHover: _tokenscss.default.uiKitColorBorderHover,
    colorBorderActive: _tokenscss.default.uiKitColorBorderActive,
    colorBorderDisabled: _tokenscss.default.uiKitColorBorderDisabled,
    colorBorderCritical: _tokenscss.default.uiKitColorBorderCritical,
    colorBorderLow: _tokenscss.default.uiKitColorBorderLow,
    colorBorderStrong: _tokenscss.default.uiKitColorBorderStrong
};
