// @canva/app-ui-kit/dist/esm/ui/apps/developing/ui_kit/components/button/button.js
import { jsx as _jsx } from "react/jsx-runtime";
import { Button as EaselButton } from '../../../../../base/button/button';
/** 
 * Triggers an action, such as the submission of a form.
 */ export function Button(props) {
    const commonProps = {
        ariaLabel: props.ariaLabel,
        children: props.children,
        disabled: props.disabled,
        selected: props.selected,
        icon: props.icon,
        iconPosition: props.iconPosition,
        alignment: props.alignment,
        loading: props.loading,
        onClick: props.onClick,
        tooltipLabel: props.tooltipLabel,
        stretch: props.stretch,
        type: props.type || 'button',
        variant: props.variant || 'tertiary'
    };
    const shouldIncludeSize = ()=>{
        return !!(props.icon && (!props.variant || props.variant === 'tertiary' || props.variant === 'contrast') && !props.children);
    };
    const shouldIncludePressed = ()=>{
        return props.variant !== 'primary';
    };
    return _jsx(EaselButton, {
        ...commonProps,
        size: shouldIncludeSize() ? props.size || 'medium' : undefined,
        pressed: shouldIncludePressed() ? props.pressed : undefined
    });
}
