// @canva/app-ui-kit/dist/cjs/ui/base/layout/private/columns/columns.js
"use strict"
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: Object.getOwnPropertyDescriptor(all, name).get
    });
}
_export(exports, {
    get Column () {
        return Column;
    },
    get Columns () {
        return Columns;
    },
    get alignXs () {
        return alignXs;
    },
    get alignYs () {
        return alignYs;
    },
    get breakpoints () {
        return breakpoints;
    },
    get heights () {
        return heights;
    }
});
const _jsxruntime = require("react/jsx-runtime");
const _preconditions = require('../../../../../base/preconditions');
const _classnames = _interop_require_default(require("classnames"));
const _react = _interop_require_wildcard(require("react"));
const _common_props = require('../../../box/common_props/common_props');
const _metrics = require('../../../metrics/metrics');
const _responsive = require('../../../responsive/responsive');
const _columnscss = _interop_require_wildcard(require("./columns.css"));
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function _getRequireWildcardCache(nodeInterop) {
    if (typeof WeakMap !== "function") return null;
    var cacheBabelInterop = new WeakMap();
    var cacheNodeInterop = new WeakMap();
    return (_getRequireWildcardCache = function(nodeInterop) {
        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
    })(nodeInterop);
}
function _interop_require_wildcard(obj, nodeInterop) {
    if (!nodeInterop && obj && obj.__esModule) {
        return obj;
    }
    if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
        return {
            default: obj
        };
    }
    var cache = _getRequireWildcardCache(nodeInterop);
    if (cache && cache.has(obj)) {
        return cache.get(obj);
    }
    var newObj = {
        __proto__: null
    };
    var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
    for(var key in obj){
        if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
            if (desc && (desc.get || desc.set)) {
                Object.defineProperty(newObj, key, desc);
            } else {
                newObj[key] = obj[key];
            }
        }
    }
    newObj.default = obj;
    if (cache) {
        cache.set(obj, newObj);
    }
    return newObj;
}
const alignXs = [
    'start',
    'center',
    'end',
    'spaceBetween',
    'spaceAround'
];
const alignYs = [
    'start',
    'center',
    'end',
    'stretch',
    'baseline'
];
const heights = [
    'unset',
    'full'
];
const breakpoints = [
    'small',
    'medium',
    'large',
    'xLarge'
];
const Column = _react.forwardRef(({ width = 'fluid', children, tagName: TagName = 'div', ...rest }, ref)=>{
    return (0, _jsxruntime.jsx)(TagName, {
        className: (0, _classnames.default)(_columnscss.default.column, getWidthClassName(width)),
        style: getWidthStyle(width),
        ref: ref,
        ...(0, _common_props.commonPropsToDOMProps)(rest),
        children: children
    });
});
const spacingPropertyMap = {
    default: _columnscss.customProperties.spacing,
    smallUp: _columnscss.customProperties.smallUpSpacing,
    mediumUp: _columnscss.customProperties.mediumUpSpacing,
    largeUp: _columnscss.customProperties.largeUpSpacing,
    xLargeUp: _columnscss.customProperties.xLargeUpSpacing
};
const alignPropertyMap = {
    default: _columnscss.customProperties.align,
    smallUp: _columnscss.customProperties.smallUpAlign,
    mediumUp: _columnscss.customProperties.mediumUpAlign,
    largeUp: _columnscss.customProperties.largeUpAlign,
    xLargeUp: _columnscss.customProperties.xLargeUpAlign
};
const alignYPropertyMap = {
    default: _columnscss.customProperties.alignY,
    smallUp: _columnscss.customProperties.smallUpAlignY,
    mediumUp: _columnscss.customProperties.mediumUpAlignY,
    largeUp: _columnscss.customProperties.largeUpAlignY,
    xLargeUp: _columnscss.customProperties.xLargeUpAlignY
};
const heightPropertyMap = {
    default: _columnscss.customProperties.heightValue,
    smallUp: _columnscss.customProperties.smallUpHeight,
    mediumUp: _columnscss.customProperties.mediumUpHeight,
    largeUp: _columnscss.customProperties.largeUpHeight,
    xLargeUp: _columnscss.customProperties.xLargeUpHeight
};
const getAlignValue = (align)=>{
    switch(align){
        case 'start':
            return 'flex-start';
        case 'center':
            return 'center';
        case 'end':
            return 'flex-end';
        case 'spaceBetween':
            return 'space-between';
        case 'spaceAround':
            return 'space-around';
        default:
            throw new _preconditions.UnreachableError(align);
    }
};
const getAlignYValue = (alignY)=>alignY === 'start' || alignY === 'end' ? `flex-${alignY}` : alignY;
const getHeightValue = (height)=>{
    switch(height){
        case 'full':
            return '100%';
        case 'unset':
            return 'unset';
        default:
            throw new _preconditions.UnreachableError(height);
    }
};
const Columns = _react.forwardRef(
    ({ children, spacing, align, alignY, height, tagName: TagName = 'div', reversed, collapseBelow, collapsed, ...rest }, ref)=>{
        return ((0, _jsxruntime.jsx)(TagName, {
                    className: (0, _classnames.default)(_columnscss.default.columns, reversed && (collapsed || collapseBelow) && _columnscss.default.rowReverse, collapseBelow && collapseBelowClassNames[collapseBelow], collapsed && _columnscss.default.collapsed),
                    style: {
                        ...(spacing && (0, _responsive.generateStyle)(spacingPropertyMap, spacing, _metrics.getSpaceValue)),
                        ...(align && (0, _responsive.generateStyle)(alignPropertyMap, align, getAlignValue)),
                        ...(alignY && (0, _responsive.generateStyle)(alignYPropertyMap, alignY, getAlignYValue)),
                        ...(height && (0, _responsive.generateStyle)(heightPropertyMap, height, getHeightValue))
                    },
                    ref: ref,
                    ...(0, _common_props.commonPropsToDOMProps)(rest),
                    children: reversed && !collapsed && !collapseBelow ? _react.Children.toArray(children).reverse() : children
                }));
    }
);
const collapseBelowClassNames = {
    ['small']: _columnscss.default.collapseBelowSmall,
    ['medium']: _columnscss.default.collapseBelowMedium,
    ['large']: _columnscss.default.collapseBelowLarge,
    ['xLarge']: _columnscss.default.collapseBelowXLarge
};
function getWidthClassName(width) {
    if (isFixedWidth(width)) return _columnscss.default.widthFixed;
    switch(width){
        case 'fluid':
            return null;
        case 'content':
            return _columnscss.default.widthContent;
        case 'containedContent':
            return _columnscss.default.widthContainedContent;
        default:
            throw new _preconditions.UnreachableError(width);
    }
}
function getWidthStyle(width) {
    if (isFixedWidth(width)) return {
        [_columnscss.customProperties.columnWidth]: width
    };
    return null;
}
function isFixedWidth(width) {
    if (width !== 'fluid' && width !== 'content' && width !== 'containedContent') return true;
    return false;
}
