// @canva/app-ui-kit/dist/cjs/ui/base/button/base_button/private/base_button.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: Object.getOwnPropertyDescriptor(all, name).get
    });
}
_export(exports, {
    get BaseAnchor () {
        return BaseAnchor;
    },
    get BaseButton () {
        return BaseButton;
    }
});
const _jsxruntime = require("react/jsx-runtime");
const _classnames = _interop_require_default(require("classnames"));
const _react = _interop_require_wildcard(require("react"));
const _access_mode_state = require('../../../a11y/access_mode_state/access_mode_state');
const _button_aria_attributes = require('../../../a11y/button_aria_attributes/button_aria_attributes');
const _box = require('../../../box/box');
const _handle = require('../../../handle/handle');
const _base_buttoncss = _interop_require_default(require("./base_button.css"));
const _button_tones = require("./button_tones");
const _with_tooltip = require("./with_tooltip");
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function _getRequireWildcardCache(nodeInterop) {
    if (typeof WeakMap !== "function") return null;
    var cacheBabelInterop = new WeakMap();
    var cacheNodeInterop = new WeakMap();
    return (_getRequireWildcardCache = function(nodeInterop) {
        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
    })(nodeInterop);
}
function _interop_require_wildcard(obj, nodeInterop) {
    if (!nodeInterop && obj && obj.__esModule) {
        return obj;
    }
    if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
        return {
            default: obj
        };
    }
    var cache = _getRequireWildcardCache(nodeInterop);
    if (cache && cache.has(obj)) {
        return cache.get(obj);
    }
    var newObj = {
        __proto__: null
    };
    var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
    for(var key in obj){
        if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
            if (desc && (desc.get || desc.set)) {
                Object.defineProperty(newObj, key, desc);
            } else {
                newObj[key] = obj[key];
            }
        }
    }
    newObj.default = obj;
    if (cache) {
        cache.set(obj, newObj);
    }
    return newObj;
}
const InternalBaseButton = _react.forwardRef(function InternalBaseButton(props, ref) {
    var _props_display;
    const boxAttributes = (0, _box.useBox)({
        className: props.className,
        display: (_props_display = props.display) !== null && _props_display !== void 0 ? _props_display : 'inline-flex',
        border: props.border,
        borderTop: props.borderTop,
        borderBottom: props.borderBottom,
        borderStart: props.borderStart,
        borderEnd: props.borderEnd,
        borderRadius: props.borderRadius,
        shadow: props.shadow,
        padding: props.padding,
        paddingX: props.paddingX,
        paddingY: props.paddingY,
        paddingTop: props.paddingTop,
        paddingBottom: props.paddingBottom,
        paddingStart: props.paddingStart,
        paddingEnd: props.paddingEnd,
        width: props.width,
        height: props.height,
        minWidth: props.minWidth,
        minHeight: props.minHeight,
        flexDirection: props.flexDirection,
        flexWrap: props.flexWrap,
        justifyContent: props.justifyContent,
        alignItems: props.alignItems,
        flex: props.flex,
        alignSelfInFlex: props.alignSelfInFlex,
        alignSelfInGrid: props.alignSelfInGrid,
        position: props.position,
        inset: props.inset,
        start: props.start,
        end: props.end,
        top: props.top,
        bottom: props.bottom
    });
    const ariaAttributes = (0, _button_aria_attributes.getButtonAriaAttributes)({
        tagName: 'button',
        role: props.role,
        active: props.active,
        selected: props.selected,
        pressed: props.pressed,
        disabled: props.disabled,
        disclosure: props.disclosure,
        tabIndex: props.tabIndex,
        ariaLabel: props.ariaLabel,
        ariaLabelledBy: props.ariaLabelledBy,
        ariaControls: props.ariaControls,
        ariaOwns: props.ariaOwns,
        ariaHasPopup: props.ariaHasPopup,
        ariaDescribedBy: props.ariaDescribedBy,
        ariaRoleDescription: props.ariaRoleDescription,
        ariaHidden: props.ariaHidden,
        ariaActiveDescendant: props.ariaActiveDescendant,
        ariaCurrent: props.ariaCurrent
    });
    const disabledAttributes = getDisabledProps(props.disabled);
    const disableSubmit = useShouldDisableSubmit(props.type === 'submit');
    const stateAttributes = (0, _button_tones.useStateMaps)(props);
    const toneClassNames = (0, _button_tones.useToneClassNames)(props);
    const className = (0, _classnames.default)(_base_buttoncss.default.button, boxAttributes.className, toneClassNames, {
        [_base_buttoncss.default.hideOutline]: (0, _access_mode_state.useIsMouseAccessMode)()
    });
    const buttonRef = _react.useRef(null);
    _react.useImperativeHandle(ref, ()=>(0, _handle.createBasicHTMLHandle)(buttonRef));
    var _props_type;
    return (0, _jsxruntime.jsx)("button", {
        ref: buttonRef,
        id: props.id,
        type: (_props_type = props.type) !== null && _props_type !== void 0 ? _props_type : 'button',
        form: props.form,
        draggable: props.draggable,
        onClick: props.onClick,
        onMouseDown: props.onMouseDown,
        onMouseUp: props.onMouseUp,
        onMouseEnter: props.onMouseEnter,
        onMouseLeave: props.onMouseLeave,
        onPointerDown: props.onPointerDown,
        onPointerUp: props.onPointerUp,
        onPointerEnter: props.onPointerEnter,
        onPointerLeave: props.onPointerLeave,
        onPointerMove: props.onPointerMove,
        onPointerCancel: props.onPointerCancel,
        onFocus: props.onFocus,
        onBlur: props.onBlur,
        onDragStart: props.onDragStart,
        onDragEnd: props.onDragEnd,
        onContextMenu: props.onContextMenu,
        ...boxAttributes,
        ...ariaAttributes,
        ...disabledAttributes,
        disabled: disableSubmit || undefined,
        className: className,
        style: {
            ...boxAttributes.style,
            ...stateAttributes.style
        },
        children: props.children
    });
});
const InternalBaseAnchor = _react.forwardRef(function InternalBaseAnchor(props, ref) {
    var _props_display;
    const boxAttributes = (0, _box.useBox)({
        className: props.className,
        display: (_props_display = props.display) !== null && _props_display !== void 0 ? _props_display : 'inline-flex',
        border: props.border,
        borderTop: props.borderTop,
        borderBottom: props.borderBottom,
        borderStart: props.borderStart,
        borderEnd: props.borderEnd,
        borderRadius: props.borderRadius,
        shadow: props.shadow,
        padding: props.padding,
        paddingX: props.paddingX,
        paddingY: props.paddingY,
        paddingTop: props.paddingTop,
        paddingBottom: props.paddingBottom,
        paddingStart: props.paddingStart,
        paddingEnd: props.paddingEnd,
        width: props.width,
        height: props.height,
        minWidth: props.minWidth,
        minHeight: props.minHeight,
        flexDirection: props.flexDirection,
        flexWrap: props.flexWrap,
        justifyContent: props.justifyContent,
        alignItems: props.alignItems,
        flex: props.flex,
        alignSelfInFlex: props.alignSelfInFlex,
        alignSelfInGrid: props.alignSelfInGrid,
        position: props.position,
        inset: props.inset,
        start: props.start,
        end: props.end,
        top: props.top,
        bottom: props.bottom
    });
    const ariaAttributes = (0, _button_aria_attributes.getButtonAriaAttributes)({
        tagName: 'a',
        role: props.role,
        active: props.active,
        selected: props.selected,
        pressed: props.pressed,
        disabled: props.disabled,
        disclosure: props.disclosure,
        tabIndex: props.tabIndex,
        ariaLabel: props.ariaLabel,
        ariaLabelledBy: props.ariaLabelledBy,
        ariaControls: props.ariaControls,
        ariaOwns: props.ariaOwns,
        ariaHasPopup: props.ariaHasPopup,
        ariaDescribedBy: props.ariaDescribedBy,
        ariaRoleDescription: props.ariaRoleDescription,
        ariaHidden: props.ariaHidden,
        ariaActiveDescendant: props.ariaActiveDescendant,
        ariaCurrent: props.ariaCurrent
    });
    const disabledAttributes = getDisabledProps(props.disabled);
    const stateAttributes = (0, _button_tones.useStateMaps)(props);
    const toneClassNames = (0, _button_tones.useToneClassNames)(props);
    const className = (0, _classnames.default)(_base_buttoncss.default.button, boxAttributes.className, toneClassNames, {
        [_base_buttoncss.default.hideOutline]: (0, _access_mode_state.useIsMouseAccessMode)()
    });
    const anchorRef = _react.useRef(null);
    _react.useImperativeHandle(ref, ()=>(0, _handle.createBasicHTMLHandle)(anchorRef));
    return (0, _jsxruntime.jsx)("a", {
        ref: anchorRef,
        id: props.id,
        href: props.href,
        rel: props.rel,
        target: props.target,
        download: props.download,
        onClick: props.onClick,
        onMouseDown: props.onMouseDown,
        onMouseUp: props.onMouseUp,
        onMouseEnter: props.onMouseEnter,
        onMouseLeave: props.onMouseLeave,
        onPointerDown: props.onPointerDown,
        onPointerUp: props.onPointerUp,
        onPointerEnter: props.onPointerEnter,
        onPointerLeave: props.onPointerLeave,
        onPointerMove: props.onPointerMove,
        onPointerCancel: props.onPointerCancel,
        onFocus: props.onFocus,
        onBlur: props.onBlur,
        onDragStart: props.onDragStart,
        onDragEnd: props.onDragEnd,
        onContextMenu: props.onContextMenu,
        ...boxAttributes,
        ...ariaAttributes,
        ...disabledAttributes,
        className: className,
        style: {
            ...boxAttributes.style,
            ...stateAttributes.style
        },
        children: props.children
    });
});
function getDisabledProps(disabled) {
    if (!disabled) return {};
    return {
        onClick: stopClickPropagation,
        onClickCapture: stopClickPropagation,
        onKeyDownCapture: stopClickPropagation,
        onMouseDownCapture: stopClickPropagation,
        onMouseUpCapture: stopClickPropagation,
        ['aria-disabled']: true
    };
}
function stopClickPropagation(e) {
    const IS_CLICK = [
        'click',
        'mousedown',
        'mouseup',
        'submit'
    ].includes(e.type);
    const IS_ENTER_OR_SPACE = 'code' in e && [
        'Enter',
        'Space'
    ].includes(e.code);
    if (IS_CLICK || IS_ENTER_OR_SPACE) {
        e.stopPropagation();
        e.preventDefault();
    }
}
function useShouldDisableSubmit(initialShouldDisableSubmit) {
    const [disableSubmit, setDisableSubmit] = _react.useState(initialShouldDisableSubmit);
    _react.useEffect(()=>{
        setDisableSubmit(false);
    }, []);
    return disableSubmit;
}
const BaseButton = (0, _with_tooltip.withTooltip)(InternalBaseButton);
const BaseAnchor = (0, _with_tooltip.withTooltip)(InternalBaseAnchor);
