// @canva/app-ui-kit/dist/cjs/ui/base/button/base_button/private/button_tones.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: Object.getOwnPropertyDescriptor(all, name).get
    });
}
_export(exports, {
    get buttonTones () {
        return buttonTones;
    },
    get useStateMaps () {
        return useStateMaps;
    },
    get useToneClassNames () {
        return useToneClassNames;
    }
});
const _classnames = _interop_require_default(require("classnames"));
const _pointer_input_state = require('../../../a11y/pointer_input_state/pointer_input_state');
const _button_tonescss = _interop_require_wildcard(require("./button_tones.css"));
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function _getRequireWildcardCache(nodeInterop) {
    if (typeof WeakMap !== "function") return null;
    var cacheBabelInterop = new WeakMap();
    var cacheNodeInterop = new WeakMap();
    return (_getRequireWildcardCache = function(nodeInterop) {
        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
    })(nodeInterop);
}
function _interop_require_wildcard(obj, nodeInterop) {
    if (!nodeInterop && obj && obj.__esModule) {
        return obj;
    }
    if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
        return {
            default: obj
        };
    }
    var cache = _getRequireWildcardCache(nodeInterop);
    if (cache && cache.has(obj)) {
        return cache.get(obj);
    }
    var newObj = {
        __proto__: null
    };
    var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
    for(var key in obj){
        if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
            if (desc && (desc.get || desc.set)) {
                Object.defineProperty(newObj, key, desc);
            } else {
                newObj[key] = obj[key];
            }
        }
    }
    newObj.default = obj;
    if (cache) {
        cache.set(obj, newObj);
    }
    return newObj;
}
const buttonTones = [
    'primary',
    'secondary',
    'tertiary',
    'contrast',
    'critical'
];
function isStateMap(value) {
    return typeof value === 'object' && value != null;
}
function isSelectedTone(tone, selected) {
    return (tone === 'secondary' || tone === 'tertiary') && selected;
}
function useToneClassNames({ tone, pressed, selected, disabled, border, color, backgroundColor, borderColor }) {
    const hasCustomColors = !!(tone || color || backgroundColor || borderColor);
    const toneClassName = isSelectedTone(tone, selected) ? _button_tonescss.default.selected : tone != null && (0, _button_tonescss.getStyle)(tone);
    const borderClassName = border && border !== 'none' && (0, _button_tonescss.getStyle)(`${border}Border`);
    return (0, _classnames.default)(toneClassName, borderClassName, {
        [_button_tonescss.default.tone]: hasCustomColors || borderClassName,
        [_button_tonescss.default.pressed]: pressed,
        [_button_tonescss.default.disabled]: disabled,
        [_button_tonescss.default.hoverDisabled]: !(0, _pointer_input_state.useIsHoveringInput)()
    });
}
function useStateMaps(props) {
    const style = {
        ...applyStateMap(props.backgroundColor, _button_tonescss.customProperties.defaultBackgroundColor, _button_tonescss.customProperties.hoveredBackgroundColor, _button_tonescss.customProperties.pressedBackgroundColor, _button_tonescss.customProperties.disabledBackgroundColor),
        ...applyStateMap(props.color, _button_tonescss.customProperties.defaultColor, _button_tonescss.customProperties.hoveredColor, _button_tonescss.customProperties.pressedColor, _button_tonescss.customProperties.disabledColor),
        ...applyStateMap(props.borderColor, _button_tonescss.customProperties.defaultBorderColor, _button_tonescss.customProperties.hoveredBorderColor, _button_tonescss.customProperties.pressedBorderColor, _button_tonescss.customProperties.disabledBorderColor)
    };
    return {
        style
    };
}
function applyStateMap(value, defaultCustomProperty, hoveredCustomProperty, pressedCustomProperty, disabledCustomProperty) {
    const style = {};
    if (isStateMap(value)) {
        if (value.default) style[defaultCustomProperty] = value.default;
        if (value.hovered) style[hoveredCustomProperty] = value.hovered;
        if (value.pressed) style[pressedCustomProperty] = value.pressed;
        if (value.disabled) style[disabledCustomProperty] = value.disabled;
    } else if (value) {
        style[defaultCustomProperty] = value;
        style[disabledCustomProperty] = value;
    }
    return style;
}
