// @canva/app-ui-kit/dist/esm/ui/apps/developing/ui_kit/components/color_picker/color_selector.js
import { jsx as _jsx } from "react/jsx-runtime";
import 'react';
import { ColorPickerFlyout as EaselColorPickerFlyout } from '../../../../../base/color_picker/color_picker';
import { AddColorButton as EaselAddColorButton, Swatch as EaselSwatch } from '../../../../../base/swatch/swatch';
/** 
 * A square that displays a color and, when clicked, opens a flyout for selecting a color.
 */ export function ColorSelector(props) {
    const { color, disabled, onChange, onDeleteColor, onOpen, onClose, triggerMode } = props;
    return _jsx(EaselColorPickerFlyout, {
        onChange: onChange,
        color: color,
        onDeleteColor: onDeleteColor,
        onOpen: onOpen,
        onClose: onClose,
        trigger: (props)=>{
            if (triggerMode === 'addColorButton') return _jsx(EaselAddColorButton, {
                ...props,
                size: "small",
                disabled: disabled
            });
            return _jsx(EaselSwatch, {
                ...props,
                fill: [
                    color
                ],
                size: "small",
                disabled: disabled
            });
        }
    });
}
