// @canva/app-ui-kit/dist/esm/ui/base/layout/private/rows/rows.js
import { jsx as _jsx } from "react/jsx-runtime";
import * as React from 'react';
import { commonPropsToDOMProps } from '../../../box/common_props/common_props';
import { getSpaceValue } from '../../../metrics/metrics';
import { generateStyle } from '../../../responsive/responsive';
import styles, { customProperties } from './rows.css';
export const aligns = [
    'start',
    'center',
    'end',
    'stretch'
];
const spacingPropertyMap = {
    default: customProperties.spacing,
    smallUp: customProperties.smallUpSpacing,
    mediumUp: customProperties.mediumUpSpacing,
    largeUp: customProperties.largeUpSpacing,
    xLargeUp: customProperties.xLargeUpSpacing
};
const alignPropertyMap = {
    default: customProperties.align,
    smallUp: customProperties.smallUpAlign,
    mediumUp: customProperties.mediumUpAlign,
    largeUp: customProperties.largeUpAlign,
    xLargeUp: customProperties.xLargeUpAlign
};
export const Rows = React.forwardRef(
    function Rows({ children, spacing, align, tagName: TagName = 'div', ...rest }, ref) {
        return (_jsx(TagName, {
                    className: styles.rows,
                    style: {
                        ...(spacing && generateStyle(spacingPropertyMap, spacing, getSpaceValue)),
                        ...(align && generateStyle(alignPropertyMap, align))
                    },
                    ref: ref,
                    ...commonPropsToDOMProps(rest),
                    children: children
                }));
    }
);
Rows.displayName = 'Rows';
