// @canva/app-ui-kit/dist/esm/ui/base/typography/internal/typography.js
import { jsx as _jsx } from "react/jsx-runtime";
import { UnreachableError } from '../../../../base/preconditions';
import classNames from 'classnames';
import * as React from 'react';
import { useIsMouseAccessMode } from '../../a11y/access_mode_state/access_mode_state';
import { createBasicHTMLHandle } from '../../handle/handle';
import styles, { getStyle } from './typography.css';
import { getTypographyAriaAttributes } from './typography_aria_attributes';
export const InheritColor = Symbol('inherit');
export const typographyTones = [
    'primary',
    'secondary',
    'tertiary',
    'critical',
    'info',
    'success',
    'warn',
    InheritColor
];
export const typographySizes = [
    'xxlarge',
    'xlarge',
    'large',
    'medium',
    'small',
    'xsmall',
    'xxsmall'
];
export const Text = React.memo(React.forwardRef((props, ref)=>{
    var _props_tagName;
    const tagName = (_props_tagName = props.tagName) !== null && _props_tagName !== void 0 ? _props_tagName : 'p';
    var _props_weight;
    const variant = (_props_weight = props.weight) !== null && _props_weight !== void 0 ? _props_weight : 'regular';
    return _jsx(TextComponent, {
        ...props,
        variant: variant,
        tagName: tagName,
        ref: ref
    });
}));
export const Title = React.memo(React.forwardRef((props, ref)=>{
    var _props_tagName;
    const tagName = (_props_tagName = props.tagName) !== null && _props_tagName !== void 0 ? _props_tagName : getDefaultHeadingTagName(props.size);
    return _jsx(TextComponent, {
        ...props,
        variant: "title",
        tagName: tagName,
        ref: ref
    });
}));
const TextComponent = React.forwardRef((props, ref)=>{
    const { id, alignment = 'start', tone = 'primary', margins = 'none', capitalization = 'default', size = 'medium', variant, tagName: TagName, lineClamp, wrapStyle, children, allowUserSelect, className, elementTiming, ...aria } = props;
    const hideOutline = useIsMouseAccessMode() && props.tabIndex === -1;
    const innerRef = React.useRef(null);
    React.useImperativeHandle(ref, ()=>createBasicHTMLHandle(innerRef));
    if (children === '' || children == null) return null;
    const textWrapClassName = getTextWrapClassName({
        lineClamp,
        wrapStyle
    });
    const lineClampStyles = lineClamp && lineClamp > 1 ? {
        WebkitLineClamp: lineClamp
    } : {};
    const finalClassName = classNames(getTextClassName({
        variant,
        size
    }), margins === 'legacy' && styles.margin, getStyle(alignment), tone === InheritColor ? styles.inheritColor : tone && getStyle(tone), textWrapClassName, allowUserSelect && styles.allowUserSelect, capitalization === 'uppercase' && styles.uppercase, hideOutline && styles.hideOutline, className);
    return _jsx(TagName, {
        id: id,
        className: finalClassName,
        style: lineClampStyles,
        tabIndex: props.tabIndex,
        elementtiming: elementTiming,
        ref: innerRef,
        ...getTypographyAriaAttributes(aria),
        children: children
    });
});
function getTextWrapClassName({ lineClamp, wrapStyle }) {
    if (lineClamp == null && wrapStyle == null)
        return null;
    if (lineClamp === 1)
    return styles.ellipsis;
    return {
        [styles.lineClamp]: lineClamp && lineClamp > 1,
        [styles.wrapStyleBalance]: wrapStyle === 'balance',
        [styles.wrapStylePretty]: wrapStyle === 'pretty'
    };
}
function getVariantClassNamePart(variant) {
    switch(variant){
        case 'regular':
            return '';
        case 'title':
            return 'Title';
        case 'bold':
            return 'Bold';
        default:
            throw new UnreachableError(variant);
    }
}
function getSizeClassNamePart(size) {
    switch(size){
        case 'xxlarge':
            return 'ExtraExtraLarge';
        case 'xlarge':
            return 'ExtraLarge';
        case 'large':
            return 'Large';
        case 'medium':
            return 'Medium';
        case 'small':
            return 'Small';
        case 'xsmall':
            return 'ExtraSmall';
        case 'xxsmall':
            return 'ExtraExtraSmall';
        default:
            throw new UnreachableError(size);
    }
}
function getTextClassName({ variant, size }) {
    return getStyle(`text${getVariantClassNamePart(variant)}${getSizeClassNamePart(size)}`);
}
export function getDefaultHeadingTagName(size = 'medium') {
    switch(size){
        case 'xxlarge':
            return 'h1';
        case 'xlarge':
            return 'h1';
        case 'large':
            return 'h2';
        case 'medium':
            return 'h3';
        case 'small':
            return 'h4';
        case 'xsmall':
            return 'h5';
        case 'xxsmall':
            return 'h6';
        default:
            throw new UnreachableError(size);
    }
}
