// @canva/app-ui-kit/dist/cjs/ui/base/form/select/select.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: Object.getOwnPropertyDescriptor(all, name).get
    });
}
_export(exports, {
    get BaseComboSelect () {
        return _base_combo_select.BaseComboSelect;
    },
    get BaseMultiSelect () {
        return _base_multi_select.BaseMultiSelect;
    },
    get BaseMultiSelectStore () {
        return _base_multi_select.BaseMultiSelectStore;
    },
    get BaseSelect () {
        return _base_select.BaseSelect;
    },
    get BaseSelectItem () {
        return _stateless_base_select.BaseSelectItem;
    },
    get BaseSelectMenu () {
        return _stateless_base_select.BaseSelectMenu;
    },
    get BaseSelectPresenter () {
        return _base_select_presenter.BaseSelectPresenter;
    },
    get BaseSelectStore () {
        return _base_select_presenter.BaseSelectStore;
    },
    get BaseSelectTrigger () {
        return _stateless_base_select.BaseSelectTrigger;
    },
    get Select () {
        return _select.Select;
    },
    get StatelessBaseSelect () {
        return _stateless_base_select.StatelessBaseSelect;
    },
    get WindowedSelectMenu () {
        return _stateless_base_select.WindowedSelectMenu;
    },
    get defaultAreEqual () {
        return _base_select_util.defaultAreEqual;
    },
    get defaultFilterFn () {
        return _base_select_util.defaultFilterFn;
    },
    get defaultGetLabel () {
        return _base_select_util.defaultGetLabel;
    },
    get defaultIsSelected () {
        return _base_select_util.defaultIsSelected;
    },
    get deriveIdentifier () {
        return _base_select_util.deriveIdentifier;
    },
    get flattenOptions () {
        return _base_select_util.flattenOptions;
    },
    get joinLabels () {
        return _base_select_util.joinLabels;
    },
    get normalizeOptionGroups () {
        return _base_select_util.normalizeOptionGroups;
    }
});
const _select = require("./private/select");
const _base_combo_select = require("./private/base_combo_select");
const _base_multi_select = require("./private/base_multi_select");
const _base_select = require("./private/base_select");
const _base_select_presenter = require("./private/base_select_presenter");
const _base_select_util = require("./private/base_select_util");
const _stateless_base_select = require("./private/stateless_base_select");
