// @canva/app-ui-kit/dist/esm/ui/base/form/multiline_input/private/multiline_input_with_footer.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import classNames from 'classnames';
import * as React from 'react';
import { MultilineInput } from './multiline_input';
import styles from './multiline_input.css';
export const MultilineInputWithFooter = React.forwardRef(function MultilineInputWithFooter(props, ref) {
    const { EndContent, StartContent, ...multilineProps } = props;
    return _jsx(MultilineInput, {
        ...multilineProps,
        ref: ref,
        footer: _jsxs("div", {
            className: classNames(styles.footer, {
                [styles.footerDisabled]: props.disabled,
                [styles.footerReadOnly]: props.readOnly,
                [styles.footerFadein]: !props.borderless
            }),
            role: "none",
            children: [
                _jsx("span", {
                    children: StartContent
                }),
                _jsx("span", {
                    children: EndContent
                })
            ]
        })
    });
});
