// @canva/app-ui-kit/dist/cjs/ui/base/date_time/utils/utils.js
"use strict"
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: Object.getOwnPropertyDescriptor(all, name).get
    });
}
_export(exports, {
    get DAY () {
        return _utils.DAY;
    },
    get DayOfWeek () {
        return _utils.DayOfWeek;
    },
    get HOUR () {
        return _utils.HOUR;
    },
    get MINUTE () {
        return _utils.MINUTE;
    },
    get MONDAY_DATE () {
        return _utils.MONDAY_DATE;
    },
    get Month () {
        return _utils.Month;
    },
    get addDays () {
        return _utils.addDays;
    },
    get addMonths () {
        return _utils.addMonths;
    },
    get addYears () {
        return _utils.addYears;
    },
    get checkMonth () {
        return _utils.checkMonth;
    },
    get compareDates () {
        return _utils.compareDates;
    },
    get compareDays () {
        return _utils.compareDays;
    },
    get compareMonthAndYear () {
        return _utils.compareMonthAndYear;
    },
    get countDaysInMonth () {
        return _utils.countDaysInMonth;
    },
    get dateObjToDateString () {
        return _utils.dateObjToDateString;
    },
    get dayHours () {
        return _utils.dayHours;
    },
    get dayMinutes () {
        return _utils.dayMinutes;
    },
    get dayOfWeek () {
        return _utils.dayOfWeek;
    },
    get dayOfWeekStartOfMonth () {
        return _utils.dayOfWeekStartOfMonth;
    },
    get dayPeriod () {
        return _utils.dayPeriod;
    },
    get differenceInDays () {
        return _utils.differenceInDays;
    },
    get format () {
        return _utils.format;
    },
    get formatSecondsToTime () {
        return _utils.formatSecondsToTime;
    },
    get fromTimezoneToUTC () {
        return _utils.fromTimezoneToUTC;
    },
    get fromUTCToTimezone () {
        return _utils.fromUTCToTimezone;
    },
    get getDate () {
        return _utils.getDate;
    },
    get getMonth () {
        return _utils.getMonth;
    },
    get getTimeModeByLocale () {
        return _utils.getTimeModeByLocale;
    },
    get getTimeOfDay () {
        return _utils.getTimeOfDay;
    },
    get getTimestamp () {
        return _utils.getTimestamp;
    },
    get getTimezone () {
        return _utils.getTimezone;
    },
    get getTimezoneOffset () {
        return _utils.getTimezoneOffset;
    },
    get getYear () {
        return _utils.getYear;
    },
    get isEqualDates () {
        return _utils.isEqualDates;
    },
    get key () {
        return _utils.key;
    },
    get max () {
        return _utils.max;
    },
    get maxInputHour () {
        return _utils.maxInputHour;
    },
    get maxInputMinutes () {
        return _utils.maxInputMinutes;
    },
    get min () {
        return _utils.min;
    },
    get minInputHour () {
        return _utils.minInputHour;
    },
    get monthDays () {
        return _utils.monthDays;
    },
    get msToTimeString () {
        return _utils.msToTimeString;
    },
    get nextDay () {
        return _utils.nextDay;
    },
    get nextHour () {
        return _utils.nextHour;
    },
    get nextMonth () {
        return _utils.nextMonth;
    },
    get nextWeek () {
        return _utils.nextWeek;
    },
    get numberOfCompleteWeeks () {
        return _utils.numberOfCompleteWeeks;
    },
    get padTime () {
        return _utils.padTime;
    },
    get prevDay () {
        return _utils.prevDay;
    },
    get prevMinute () {
        return _utils.prevMinute;
    },
    get prevMonth () {
        return _utils.prevMonth;
    },
    get prevWeek () {
        return _utils.prevWeek;
    },
    get setLocalTimeToDate () {
        return _utils.setLocalTimeToDate;
    },
    get setTimeToDate () {
        return _utils.setTimeToDate;
    },
    get startOfDay () {
        return _utils.startOfDay;
    },
    get timeStringToMs () {
        return _utils.timeStringToMs;
    },
    get toDate () {
        return _utils.toDate;
    },
    get toDateObject () {
        return _utils.toDateObject;
    },
    get toISO () {
        return _utils.toISO;
    },
    get toLocalDateTimeObject () {
        return _utils.toLocalDateTimeObject;
    },
    get toTime () {
        return _utils.toTime;
    },
    get toUTCDateTimeObject () {
        return _utils.toUTCDateTimeObject;
    }
});
const _utils = require("./private/utils");
