// @canva/app-ui-kit/dist/esm/ui/base/date_time/date_time_picker/months_navigation/private/months_navigation.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { Preconditions } from '../../../../../../base/preconditions';
import classNames from 'classnames';
import * as React from 'react';
import { Button, CircleButton } from '../../../../button/button';
import { compareMonthAndYear, format, nextMonth, prevMonth, toLocalDateTimeObject, toUTCDateTimeObject } from '../../../utils/utils';
import { ChevronLeftIcon } from '../../../../icons/chevron_left/icon';
import { ChevronRightIcon } from '../../../../icons/chevron_right/icon';
import styles from './months_navigation.css';
import { MonthsNavigationMessages } from './months_navigation.messages';
export class MonthsNavigation extends React.PureComponent {
    componentDidUpdate(prevProps) {
        if (this.props.animate) {
            this.prevDate = this.date;
            setTimeout(()=>{
                requestAnimationFrame(()=>{
                    if (this.titleRef) this.titleRef.classList.add(styles.animate);
                });
            }, 0);
            this.animateKey = !this.animateKey;
        }
    }
    get today() {
        return Preconditions.checkExists(this.props.today);
    }
    get date() {
        return {
            month: this.props.month,
            year: this.props.year
        };
    }
    getTitleText(date) {
        return format(date, this.props.locale, 'MMMM YYYY');
    }
    isCurrentMonth() {
        return compareMonthAndYear(this.today, this.date) === 0;
    }
    get disabled() {
        return this.props.disabled === 'both';
    }
    get prevDisabled() {
        return this.disabled || this.props.disabled === 'prev';
    }
    get nextDisabled() {
        return this.disabled || this.props.disabled === 'next';
    }
    renderTitle() {
        if (this.props.animate && this.prevDate) return this.renderAnimatedTitle();
        return _jsx(Button, {
            variant: "subtleLinkButton",
            className: classNames(styles.reset, {
                [styles.disabled]: this.disabled || this.isCurrentMonth()
            }),
            onClick: !this.disabled && !this.isCurrentMonth() ? this.onReset : undefined,
            ariaLabel: MonthsNavigationMessages.reset(),
            disabled: this.disabled || this.isCurrentMonth(),
            children: _jsx("span", {
                className: styles.title,
                "aria-live": "polite",
                id: this.props.titleId,
                children: this.getTitleText(this.date)
            })
        });
    }
    renderAnimatedTitle() {
        return _jsx(Button, {
            variant: "subtleLinkButton",
            className: classNames(styles.reset, {
                [styles.disabled]: this.disabled || this.isCurrentMonth()
            }),
            onClick: !this.disabled && !this.isCurrentMonth() ? this.onReset : undefined,
            ariaLabel: MonthsNavigationMessages.reset(),
            disabled: this.disabled || this.isCurrentMonth(),
            children: _jsxs("span", {
                ref: this.onTitleRef,
                className: styles.animationWrap,
                children: [
                    _jsx("span", {
                        className: classNames(styles.title, styles.prevTitle),
                        children: this.getTitleText(this.prevDate)
                    }),
                    _jsx("span", {
                        className: classNames(styles.title, styles.newTitle),
                        "aria-live": "polite",
                        id: this.props.titleId,
                        children: this.getTitleText(this.date)
                    })
                ]
            })
        }, `title${this.animateKey ? 1 : 0}`);
    }
    render() {
        return _jsxs("div", {
            className: styles.monthsNavigation,
            children: [
                _jsx(CircleButton, {
                    variant: "tertiary",
                    className: styles.prevMonth,
                    icon: ChevronLeftIcon,
                    iconSize: "medium",
                    size: "small",
                    onClick: !this.prevDisabled ? this.onPreviousMonth : undefined,
                    disabled: this.prevDisabled,
                    ariaLabel: MonthsNavigationMessages.prev()
                }),
                this.renderTitle(),
                _jsx(CircleButton, {
                    variant: "tertiary",
                    className: styles.nextMonth,
                    icon: ChevronRightIcon,
                    iconSize: "medium",
                    size: "small",
                    onClick: !this.nextDisabled ? this.onNextMonth : undefined,
                    disabled: this.nextDisabled,
                    ariaLabel: MonthsNavigationMessages.next()
                })
            ]
        });
    }
    constructor(...args){
        super(...args), this.prevDate = this.date, this.animateKey = false, this.onTitleRef = (ref)=>{
            this.titleRef = ref;
        }, this.onReset = ()=>{
            this.props.onChange(this.today);
        }, this.onPreviousMonth = ()=>{
            this.props.onChange(toUTCDateTimeObject(prevMonth(this.date)));
        }, this.onNextMonth = ()=>{
            this.props.onChange(toUTCDateTimeObject(nextMonth(this.date)));
        };
    }
}
MonthsNavigation.defaultProps = {
    today: toLocalDateTimeObject(new Date()),
    animate: false
};
