// @canva/app-ui-kit/dist/cjs/ui/base/form/date_time_input/private/date_time_input.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "DateTimeInput", {
    enumerable: true,
    get: function() {
        return DateTimeInput;
    }
});
const _jsxruntime = require("react/jsx-runtime");
const _classnames = _interop_require_default(require("classnames"));
const _react = _interop_require_wildcard(require("react"));
const _utils = require('../../../date_time/utils/utils');
const _device_capabilities = require('../../../device_capabilities/device_capabilities');
const _base_input = require('../../base_input/base_input');
const _date_time_inputcss = _interop_require_default(require("./date_time_input.css"));
const _utils1 = require("./utils");
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function _getRequireWildcardCache(nodeInterop) {
    if (typeof WeakMap !== "function") return null;
    var cacheBabelInterop = new WeakMap();
    var cacheNodeInterop = new WeakMap();
    return (_getRequireWildcardCache = function(nodeInterop) {
        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
    })(nodeInterop);
}
function _interop_require_wildcard(obj, nodeInterop) {
    if (!nodeInterop && obj && obj.__esModule) {
        return obj;
    }
    if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
        return {
            default: obj
        };
    }
    var cache = _getRequireWildcardCache(nodeInterop);
    if (cache && cache.has(obj)) {
        return cache.get(obj);
    }
    var newObj = {
        __proto__: null
    };
    var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
    for(var key in obj){
        if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
            if (desc && (desc.get || desc.set)) {
                Object.defineProperty(newObj, key, desc);
            } else {
                newObj[key] = obj[key];
            }
        }
    }
    newObj.default = obj;
    if (cache) {
        cache.set(obj, newObj);
    }
    return newObj;
}
const globalNavigator = typeof navigator !== 'undefined' ? navigator : undefined;
const DateTimeInput = _react.memo(_react.forwardRef(
    function DateTimeInput({ value: valueProp, onClick: onClickProp, onChange: onChangeProp, onChangeComplete: onChangeCompleteProp, min: minProp, max: maxProp, className, end, timezone: timezoneProp, mode, ...passThroughProps }, ref) {
        const disableNativeControls = !(0, _device_capabilities.isVirtualKeyboard)();
        const isFirefox = useIsFirefox();
        const onClick = _react.useCallback((e)=>{
            if (disableNativeControls) e.preventDefault();
            onClickProp === null || onClickProp === void 0 ? void 0 : onClickProp(e);
        }, [
            disableNativeControls,
            onClickProp
        ]);
        const onChange = _react.useCallback((text)=>{
            const dateInTimezone = text === '' ? undefined : text;
            if (dateInTimezone == null) return onChangeProp === null || onChangeProp === void 0 ? void 0 : onChangeProp(dateInTimezone);
            if (mode === 'date') return onChangeProp === null || onChangeProp === void 0 ? void 0 : onChangeProp((0, _utils1.toDateObj)(dateInTimezone));
            var _timezoneProp_offset;
            const timezoneOffset = (_timezoneProp_offset = timezoneProp === null || timezoneProp === void 0 ? void 0 : timezoneProp.offset) !== null && _timezoneProp_offset !== void 0 ? _timezoneProp_offset : (0, _utils.getTimezoneOffset)(dateInTimezone);
            const utcValue = (0, _utils1.toUTCDateTimeObj)(dateInTimezone, timezoneOffset);
            onChangeProp === null || onChangeProp === void 0 ? void 0 : onChangeProp(utcValue);
        }, [
            mode,
            onChangeProp,
            timezoneProp === null || timezoneProp === void 0 ? void 0 : timezoneProp.offset
        ]);
        const onChangeComplete = _react.useCallback((text)=>{
            const dateInTimezone = text === '' ? undefined : text;
            if (dateInTimezone == null) return onChangeCompleteProp === null || onChangeCompleteProp === void 0 ? void 0 : onChangeCompleteProp(dateInTimezone);
            if (mode === 'date') return onChangeCompleteProp === null || onChangeCompleteProp === void 0 ? void 0 : onChangeCompleteProp((0, _utils1.toDateObj)(dateInTimezone));
            var _timezoneProp_offset;
            const timezoneOffset = (_timezoneProp_offset = timezoneProp === null || timezoneProp === void 0 ? void 0 : timezoneProp.offset) !== null && _timezoneProp_offset !== void 0 ? _timezoneProp_offset : (0, _utils.getTimezoneOffset)(dateInTimezone);
            const utcValue = (0, _utils1.toUTCDateTimeObj)(dateInTimezone, timezoneOffset);
            onChangeCompleteProp === null || onChangeCompleteProp === void 0 ? void 0 : onChangeCompleteProp(utcValue);
        }, [
            mode,
            onChangeCompleteProp,
            timezoneProp === null || timezoneProp === void 0 ? void 0 : timezoneProp.offset
        ]);
        const pickerOnChange = _react.useCallback((utcDateTime)=>{
            if (utcDateTime == null) return onChangeProp === null || onChangeProp === void 0 ? void 0 : onChangeProp(utcDateTime);
            const dateTimeObj = (0, _utils.toUTCDateTimeObject)(utcDateTime);
            if (mode === 'date') {
                const { year, month, day } = dateTimeObj;
                onChangeProp === null || onChangeProp === void 0 ? void 0 : onChangeProp({
                    year,
                    month,
                    day
                });
            } else onChangeProp === null || onChangeProp === void 0 ? void 0 : onChangeProp(dateTimeObj);
        }, [
            mode,
            onChangeProp
        ]);
        const pickerOnChangeComplete = _react.useCallback((utcDateTime)=>{
            if (utcDateTime == null) return onChangeCompleteProp === null || onChangeCompleteProp === void 0 ? void 0 : onChangeCompleteProp(utcDateTime);
            const dateTimeObj = (0, _utils.toUTCDateTimeObject)(utcDateTime);
            if (mode === 'date') {
                const { year, month, day } = dateTimeObj;
                onChangeCompleteProp === null || onChangeCompleteProp === void 0 ? void 0 : onChangeCompleteProp({
                    year,
                    month,
                    day
                });
            } else onChangeCompleteProp === null || onChangeCompleteProp === void 0 ? void 0 : onChangeCompleteProp(dateTimeObj);
        }, [
            mode,
            onChangeCompleteProp
        ]);
        const now = new Date();
        const timezone = mode === 'date' ? {
            offset: 0,
            label: 'GMT'
        } : timezoneProp;
        var _timezone_offset;
        const timezoneOffset = (_timezone_offset = timezone === null || timezone === void 0 ? void 0 : timezone.offset) !== null && _timezone_offset !== void 0 ? _timezone_offset : (0, _utils.getTimezoneOffset)(valueProp !== null && valueProp !== void 0 ? valueProp : now);
        let value, min, max, utcValue, utcMin, utcMax;
        if (mode === 'date') {
            const minDate = minProp === 'today' ? (0, _utils.toLocalDateTimeObject)(now) : minProp;
            value = utcValue = (0, _utils1.toDateString)(valueProp);
            min = (0, _utils1.toDateString)(minDate);
            utcMin = minProp === 'today' ? minProp : min;
            max = utcMax = (0, _utils1.toDateString)(maxProp);
        } else {
            const minDate = minProp === 'today' ? {
                ...(0, _utils.toLocalDateTimeObject)(now),
                hours: 0,
                minutes: 0
            } : minProp;
            value = (0, _utils1.toTimezoneDateTimeString)(valueProp, timezoneOffset);
            min = (0, _utils1.toTimezoneDateTimeString)(minDate, timezoneOffset);
            max = (0, _utils1.toTimezoneDateTimeString)(maxProp, timezoneOffset);
            utcValue = (0, _utils1.toTimezoneDateTimeString)(valueProp, 0);
            utcMin = minProp === 'today' ? minProp : (0, _utils1.toTimezoneDateTimeString)(minDate, 0);
            utcMax = (0, _utils1.toTimezoneDateTimeString)(maxProp, 0);
        }
        const { year, month, day } = (0, _utils.toLocalDateTimeObject)(now);
        const todayLocalDate = (0, _utils.dateObjToDateString)({
            year,
            month,
            day
        });
        const dateTimePickerDecoratorProps = {
            mode,
            timezone,
            animate: true,
            date: utcValue,
            today: mode === 'date' ? todayLocalDate : undefined,
            minSelectableDate: utcMin,
            maxSelectableDate: utcMax,
            onChange: pickerOnChange,
            onChangeComplete: pickerOnChangeComplete,
            disabled: passThroughProps.disabled || passThroughProps.readOnly
        };
        const renderEnd = ()=>{
            return typeof end === 'function' ? end(dateTimePickerDecoratorProps) : end;
        };
        return (0, _jsxruntime.jsx)(_base_input.BaseInput, {
            ...passThroughProps,
            type: mode === 'date' ? 'date' : 'datetime-local',
            ref: ref,
            value: value !== null && value !== void 0 ? value : '',
            onChange: onChange,
            onChangeComplete: onChangeComplete,
            min: min,
            max: max,
            onClick: onClick,
            end: disableNativeControls ? renderEnd() : undefined,
            className: (0, _classnames.default)(_date_time_inputcss.default.dateTimeInput, className),
            inputClassName: (0, _classnames.default)(isFirefox && _date_time_inputcss.default.clipNativeControls),
            placeholder: mode === 'date' ? 'YYYY-MM-DD' : 'YYYY-MM-DDTHH:mm'
        });
    }
));
DateTimeInput.displayName = 'DateTimeInput';
function useIsFirefox() {
    const [isFirefox, setIsFirefox] = _react.useState(false);
    _react.useLayoutEffect(()=>{
        setIsFirefox(globalNavigator != null && /Firefox/i.test(globalNavigator.userAgent));
    }, []);
    return isFirefox;
}
