// @canva/app-ui-kit/dist/cjs/ui/base/provider/private/provider.js
"use strict"
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: Object.getOwnPropertyDescriptor(all, name).get
    });
}
_export(exports, {
    get EaselProvider () {
        return EaselProvider;
    },
    get useAutoplayVideos () {
        return useAutoplayVideos;
    },
    get useDirection () {
        return useDirection;
    },
    get useDisableDialogBlur () {
        return useDisableDialogBlur;
    },
    get useDisableFocusTraps () {
        return useDisableFocusTraps;
    },
    get useEnableAnimations () {
        return useEnableAnimations;
    },
    get useEnableHighColorContrast () {
        return useEnableHighColorContrast;
    },
    get useEnableUserSelection () {
        return useEnableUserSelection;
    },
    get useExpectedPixelsPerRem () {
        return useExpectedPixelsPerRem;
    },
    get useTheme () {
        return useTheme;
    }
});
const _jsxruntime = require("react/jsx-runtime");
const _preconditions = require('../../../../base/preconditions');
const _react = _interop_require_wildcard(require("react"));
const _interaction_manager = require('../../a11y/interaction_manager/interaction_manager');
function _getRequireWildcardCache(nodeInterop) {
    if (typeof WeakMap !== "function") return null;
    var cacheBabelInterop = new WeakMap();
    var cacheNodeInterop = new WeakMap();
    return (_getRequireWildcardCache = function(nodeInterop) {
        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
    })(nodeInterop);
}
function _interop_require_wildcard(obj, nodeInterop) {
    if (!nodeInterop && obj && obj.__esModule) {
        return obj;
    }
    if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
        return {
            default: obj
        };
    }
    var cache = _getRequireWildcardCache(nodeInterop);
    if (cache && cache.has(obj)) {
        return cache.get(obj);
    }
    var newObj = {
        __proto__: null
    };
    var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
    for(var key in obj){
        if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
            if (desc && (desc.get || desc.set)) {
                Object.defineProperty(newObj, key, desc);
            } else {
                newObj[key] = obj[key];
            }
        }
    }
    newObj.default = obj;
    if (cache) {
        cache.set(obj, newObj);
    }
    return newObj;
}
const EaselContext = _react.createContext(null);
function EaselProvider({ enableAnimations = true, direction, disableFocusTraps, enableUserSelection, theme, enableHighColorContrast, autoplayVideos, disableDialogBlur, expectedPixelsPerRem, children }) {
    return (0, _jsxruntime.jsx)(_interaction_manager.InteractionManager, {
        children: (0, _jsxruntime.jsx)(EaselContext.Provider, {
            value: {
                enableAnimations,
                direction,
                disableFocusTraps,
                enableUserSelection,
                theme,
                enableHighColorContrast,
                autoplayVideos,
                disableDialogBlur,
                expectedPixelsPerRem
            },
            children: children
        })
    });
}
function useEaselContext() {
    return _preconditions.Preconditions.checkExists(_react.useContext(EaselContext), 'Did you forget to render a provider?');
}
function useEnableAnimations() {
    return useEaselContext().enableAnimations;
}
function useDirection() {
    return useEaselContext().direction;
}
function useDisableFocusTraps() {
    return useEaselContext().disableFocusTraps;
}
function useEnableUserSelection() {
    return useEaselContext().enableUserSelection;
}
function useTheme() {
    return useEaselContext().theme;
}
function useEnableHighColorContrast() {
    return useEaselContext().enableHighColorContrast;
}
function useAutoplayVideos() {
    return useEaselContext().autoplayVideos;
}
function useDisableDialogBlur() {
    return useEaselContext().disableDialogBlur;
}
function useExpectedPixelsPerRem() {
    return useEaselContext().expectedPixelsPerRem;
}
