// @canva/app-ui-kit/dist/cjs/ui/base/form/switch/private/switch.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "Switch", {
    enumerable: true,
    get: function() {
        return Switch;
    }
});
const _jsxruntime = require("react/jsx-runtime");
const _classnames = _interop_require_default(require("classnames"));
const _react = _interop_require_wildcard(require("react"));
const _base_button = require('../../../button/base_button/base_button');
const _controllable_value = require('../../../controllable_value/controllable_value');
const _icon = require('../../../icons/check/icon');
const _provider = require('../../../provider/provider');
const _typography = require('../../../typography/typography');
const _switchcss = _interop_require_default(require("./switch.css"));
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function _getRequireWildcardCache(nodeInterop) {
    if (typeof WeakMap !== "function") return null;
    var cacheBabelInterop = new WeakMap();
    var cacheNodeInterop = new WeakMap();
    return (_getRequireWildcardCache = function(nodeInterop) {
        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
    })(nodeInterop);
}
function _interop_require_wildcard(obj, nodeInterop) {
    if (!nodeInterop && obj && obj.__esModule) {
        return obj;
    }
    if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
        return {
            default: obj
        };
    }
    var cache = _getRequireWildcardCache(nodeInterop);
    if (cache && cache.has(obj)) {
        return cache.get(obj);
    }
    var newObj = {
        __proto__: null
    };
    var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
    for(var key in obj){
        if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
            if (desc && (desc.get || desc.set)) {
                Object.defineProperty(newObj, key, desc);
            } else {
                newObj[key] = obj[key];
            }
        }
    }
    newObj.default = obj;
    if (cache) {
        cache.set(obj, newObj);
    }
    return newObj;
}
const Switch = _react.forwardRef((props, ref)=>{
    const { id, value: valueProp, defaultValue: defaultValueProp, onChange: onChangeProp, disabled, label, labelWeight = 'regular', description, ariaLabel, ariaLabelledBy, ariaDescribedBy, onFocus, onBlur } = props;
    const [value, setValue] = (0, _controllable_value.useControllableValue)({
        value: valueProp,
        defaultValue: defaultValueProp !== null && defaultValueProp !== void 0 ? defaultValueProp : false
    });
    const labelId = _react.useId();
    const descriptionId = _react.useId();
    const generatedButtonId = _react.useId();
    const buttonId = id || generatedButtonId;
    const switchClassNames = (0, _classnames.default)(_switchcss.default.switch, {
        [_switchcss.default.switchOn]: value,
        [_switchcss.default.disabled]: disabled,
        [_switchcss.default.animated]: (0, _provider.useEnableAnimations)()
    });
    const onChange = _react.useCallback((value, e)=>{
        setValue(value);
        onChangeProp === null || onChangeProp === void 0 ? void 0 : onChangeProp(value, e);
    }, [
        onChangeProp,
        setValue
    ]);
    const getButton = (buttonRef, buttonId, labelId, descriptionId)=>(0, _jsxruntime.jsx)(_base_button.BaseButton, {
            ref: buttonRef,
            role: "switch",
            onClick: (event)=>onChange(!value, event),
            className: switchClassNames,
            disabled: disabled,
            active: value,
            ariaLabel: ariaLabel,
            ariaLabelledBy: ariaLabelledBy !== null && ariaLabelledBy !== void 0 ? ariaLabelledBy : labelId,
            ariaDescribedBy: ariaDescribedBy !== null && ariaDescribedBy !== void 0 ? ariaDescribedBy : description ? descriptionId : undefined,
            id: buttonId,
            onFocus: onFocus,
            onBlur: onBlur,
            display: "flex",
            border: "none",
            borderRadius: "elementRound",
            alignItems: "center",
            position: "relative",
            backgroundColor: {
                default: value ? _switchcss.default.switchOnColor : _switchcss.default.switchOffColor,
                hovered: value ? _switchcss.default.switchOnColorHover : _switchcss.default.switchOffColorHover,
                pressed: value ? _switchcss.default.switchOnColorHover : _switchcss.default.switchOffColorHover,
                disabled: _switchcss.default.switchDisabledColor
            },
            children: (0, _jsxruntime.jsx)("span", {
                className: _switchcss.default.animationWrapper,
                children: (0, _jsxruntime.jsx)("span", {
                    className: _switchcss.default.thumb,
                    children: (0, _jsxruntime.jsx)(_icon.CheckIcon, {
                        size: "tiny",
                        className: _switchcss.default.tick
                    })
                })
            })
        });
    if (label == null) return getButton(ref, id);
    return (0, _jsxruntime.jsxs)("div", {
        className: _switchcss.default.switchWrapper,
        children: [
            (0, _jsxruntime.jsx)("label", {
                htmlFor: buttonId,
                className: _switchcss.default.switchText,
                id: labelId,
                children: (0, _jsxruntime.jsx)(SwitchLabelAndDescription, {
                    label: label,
                    description: description,
                    disabled: disabled,
                    descriptionId: descriptionId,
                    labelWeight: labelWeight
                })
            }),
            getButton(ref, buttonId, labelId, descriptionId)
        ]
    });
});
function SwitchLabelAndDescription({ label, description, disabled, descriptionId, labelWeight }) {
    const labelElement = typeof label === 'string' ? (0, _jsxruntime.jsx)(_typography.Text, {
        tagName: "span",
        margins: "none",
        tone: disabled ? 'tertiary' : 'primary',
        weight: labelWeight,
        className: _switchcss.default.switchLabel,
        children: label
    }) : label;
    if (!description) return labelElement;
    return (0, _jsxruntime.jsxs)(_jsxruntime.Fragment, {
        children: [
            labelElement,
            (0, _jsxruntime.jsx)("div", {
                "aria-hidden": true,
                children: (0, _jsxruntime.jsx)(_typography.Text, {
                    className: _switchcss.default.switchDescription,
                    id: descriptionId,
                    tone: "tertiary",
                    size: "small",
                    children: description
                })
            })
        ]
    });
}
