// @canva/app-ui-kit/dist/esm/ui/base/scrollable/private/base_scrollable.js
import { jsx as _jsx } from "react/jsx-runtime";
import { Preconditions } from '../../../../base/preconditions';
import classNames from 'classnames';
import * as React from 'react';
import { useDirection } from '../../provider/provider';
import { getContextualizedScrollLeft, getNormalizedScrollLeft } from '../../scroll_controls/normalized_scroll';
import styles from './base_scrollable.css';
const ROUNDING_ERROR = 1;
export const BaseScrollable = React.forwardRef((props, forwardedRef)=>{
    const { children, direction, onScroll, className, ariaLabel, role = 'region' } = props;
    const isRtl = useDirection() === 'RTL';
    const ref = React.useRef(null);
    React.useImperativeHandle(forwardedRef, ()=>{
        const scrollContainer = Preconditions.checkExists(ref.current);
        return {
            getScrollState: ()=>getScrollState(scrollContainer, direction, isRtl),
            scrollBy: (options)=>scrollContainer.scrollBy(options),
            scrollTo: (options)=>{
                const { position, behavior } = options;
                if (!position) {
                    scrollContainer.scrollTo(options);
                    return;
                }
                const state = getScrollState(scrollContainer, direction, isRtl);
                if (state.scrollable) {
                    const { raw } = state;
                    if (direction === 'vertical') {
                        const top = position === 'start' ? 0 : raw.scroll.height - raw.client.height;
                        scrollContainer.scrollTo({
                            top,
                            behavior
                        });
                    } else {
                        const startPosition = isRtl ? raw.scroll.width - raw.client.width : 0;
                        const endPosition = isRtl ? 0 : raw.scroll.width - raw.client.width;
                        const left = position === 'start' ? startPosition : endPosition;
                        scrollContainer.scrollTo({
                            left: getContextualizedScrollLeft(scrollContainer, left, isRtl),
                            behavior: options === null || options === void 0 ? void 0 : options.behavior
                        });
                    }
                }
            }
        };
    }, [
        direction,
        isRtl
    ]);
    const handleScroll = React.useCallback(()=>{
        if (onScroll && ref.current) {
            const scrollContainer = ref.current;
            onScroll(()=>getScrollState(scrollContainer, direction, isRtl));
        }
    }, [
        onScroll,
        direction,
        isRtl
    ]);
    return _jsx("div", {
        tabIndex: 0,
        "aria-label": ariaLabel,
        role: ariaLabel ? role : undefined,
        className: classNames(styles.scrollContainer, className, {
            [styles.vertical]: direction === 'vertical',
            [styles.horizontal]: direction === 'horizontal'
        }),
        ref: ref,
        onScroll: handleScroll,
        children: children
    });
});
function getScrollState(scrollContainer, direction, isRtl) {
    const scrollable = direction === 'vertical' ? scrollContainer.scrollHeight > scrollContainer.clientHeight : scrollContainer.scrollWidth > scrollContainer.clientWidth;
    if (!scrollable) return {
        scrollable: false
    };
    const raw = {
        scroll: {
            top: scrollContainer.scrollTop,
            height: scrollContainer.scrollHeight,
            left: getNormalizedScrollLeft(scrollContainer, isRtl),
            width: scrollContainer.scrollWidth
        },
        client: {
            top: scrollContainer.clientTop,
            height: scrollContainer.clientHeight,
            left: scrollContainer.clientLeft,
            width: scrollContainer.clientWidth
        }
    };
    let atStart, atEnd;
    if (direction === 'vertical') {
        atStart = raw.scroll.top <= 0;
        atEnd = raw.scroll.height - raw.scroll.top - raw.client.height < ROUNDING_ERROR;
    } else {
        const atLeftMost = raw.scroll.left <= 0;
        const atRightMost = raw.scroll.width - raw.scroll.left - raw.client.width < ROUNDING_ERROR;
        atStart = isRtl ? atRightMost : atLeftMost;
        atEnd = isRtl ? atLeftMost : atRightMost;
    }
    return {
        scrollable: true,
        raw,
        atStart,
        atEnd
    };
}
