// @canva/app-ui-kit/dist/esm/ui/base/scrollable/private/scrollable_with_indicator.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import composeRefs from '@seznam/compose-react-refs';
import { debounce } from '../../../../base/debounce';
import { UnreachableError } from '../../../../base/preconditions';
import classNames from 'classnames';
import * as React from 'react';
import Measure from 'react-measure';
import { useDirection, useEnableAnimations } from '../../provider/provider';
import { BaseScrollable } from './base_scrollable';
import styles from './scrollable_with_indicator.css';
const UPDATE_INDICATORS_DEBOUNCE_MS = 16;
const isWebkit = ()=>typeof window !== 'undefined' ? /AppleWebKit\//i.test(window.navigator.userAgent) : undefined;
const runningInWebkit = isWebkit();
export const ScrollableWithIndicator = React.forwardRef((props, forwardedRef)=>{
    const [showStartIndicator, setShowStartIndicator] = React.useState(false);
    const [showEndIndicator, setShowEndIndicator] = React.useState(false);
    const [widths, setWidths] = React.useState(null);
    const scrollableRef = React.useRef(null);
    const isRtl = useDirection() === 'RTL';
    const { children, direction, onScroll, ariaLabel, role, type, background } = props;
    const applyWebkitRtlFix = runningInWebkit && isRtl && direction === 'horizontal';
    const enableAnimations = useEnableAnimations();
    const updateIndicators = React.useMemo(()=>{
        return debounce((getScrollState)=>{
            const scrollState = getScrollState();
            if (!scrollState.scrollable) {
                setShowStartIndicator(false);
                setShowEndIndicator(false);
            } else {
                setShowStartIndicator(!scrollState.atStart);
                setShowEndIndicator(!scrollState.atEnd);
            }
        }, UPDATE_INDICATORS_DEBOUNCE_MS, {
            leading: true
        });
    }, []);
    const updateWidths = React.useMemo(()=>{
        return debounce((getScrollState)=>{
            const scrollState = getScrollState();
            if (scrollState.scrollable) {
                const { raw } = scrollState;
                setWidths({
                    client: raw.client.width,
                    scroll: raw.scroll.width
                });
            }
        }, UPDATE_INDICATORS_DEBOUNCE_MS, {
            leading: true
        });
    }, []);
    const handleScroll = React.useCallback((getScrollState)=>{
        updateIndicators(getScrollState);
        onScroll && onScroll(getScrollState);
    }, [
        updateIndicators,
        onScroll
    ]);
    const handleChildrenResize = React.useCallback(()=>{
        const ref = scrollableRef.current;
        if (!ref) return;
        updateIndicators(ref.getScrollState);
        if (applyWebkitRtlFix)
        updateWidths(ref.getScrollState);
    }, [
        updateIndicators,
        updateWidths,
        applyWebkitRtlFix
    ]);
    const indicatorWrapperSharedStyles = classNames(styles.indicatorWrapper, getDirectionClassName(direction));
    const indicatorSharedStyles = classNames(styles.indicator, getBackgroundClassName(background), getIndicatorTypeClassName(type), getDirectionClassName(direction), {
        [styles.animate]: enableAnimations
    });
    return _jsxs(BaseScrollable, {
        direction: direction,
        onScroll: handleScroll,
        ariaLabel: ariaLabel,
        role: role,
        ref: composeRefs(forwardedRef, scrollableRef),
        className: classNames(styles.scrollContainer, getDirectionClassName(direction)),
        children: [
            _jsx("div", {
                className: classNames(indicatorWrapperSharedStyles, styles.start),
                style: {
                    right: applyWebkitRtlFix && !!widths ? `${widths.scroll - widths.client}px` : undefined
                },
                children: _jsx("div", {
                    className: classNames(indicatorSharedStyles, styles.start, {
                        [styles.isVisible]: showStartIndicator
                    })
                })
            }),
            _jsx("div", {
                className: classNames(indicatorWrapperSharedStyles, styles.end),
                style: {
                    right: applyWebkitRtlFix && !!widths ? `${widths.scroll}px` : undefined
                },
                children: _jsx("div", {
                    className: classNames(indicatorSharedStyles, styles.end, {
                        [styles.isVisible]: showEndIndicator
                    })
                })
            }),
            _jsx(Measure, {
                onResize: handleChildrenResize,
                children: ({ measureRef })=>_jsx("div", {
                        ref: measureRef,
                        className: styles.childrenWrapper,
                        children: children
                    })
            })
        ]
    });
});
function getBackgroundClassName(background) {
    switch(background){
        case 'canvas':
            return styles.canvas;
        case 'tabdock':
            return styles.tabdock;
        case 'surface':
            return styles.surface;
        case 'page':
            return styles.page;
        default:
            throw new UnreachableError(background);
    }
}
function getIndicatorTypeClassName(type) {
    switch(type){
        case 'shadow':
            return styles.shadow;
        case 'fade':
            return styles.fade;
        default:
            throw new UnreachableError(type);
    }
}
function getDirectionClassName(direction) {
    switch(direction){
        case 'vertical':
            return styles.vertical;
        case 'horizontal':
            return styles.horizontal;
        default:
            throw new UnreachableError(direction);
    }
}
