// @canva/app-ui-kit/dist/cjs/ui/base/surface/internal/content_container.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "ContentContainer", {
    enumerable: true,
    get: function() {
        return ContentContainer;
    }
});
const _jsxruntime = require("react/jsx-runtime");
const _react = _interop_require_wildcard(require("react"));
const _focus_lock = require('../../a11y/focus_lock/focus_lock');
function _getRequireWildcardCache(nodeInterop) {
    if (typeof WeakMap !== "function") return null;
    var cacheBabelInterop = new WeakMap();
    var cacheNodeInterop = new WeakMap();
    return (_getRequireWildcardCache = function(nodeInterop) {
        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
    })(nodeInterop);
}
function _interop_require_wildcard(obj, nodeInterop) {
    if (!nodeInterop && obj && obj.__esModule) {
        return obj;
    }
    if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
        return {
            default: obj
        };
    }
    var cache = _getRequireWildcardCache(nodeInterop);
    if (cache && cache.has(obj)) {
        return cache.get(obj);
    }
    var newObj = {
        __proto__: null
    };
    var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
    for(var key in obj){
        if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
            if (desc && (desc.get || desc.set)) {
                Object.defineProperty(newObj, key, desc);
            } else {
                newObj[key] = obj[key];
            }
        }
    }
    newObj.default = obj;
    if (cache) {
        cache.set(obj, newObj);
    }
    return newObj;
}
const ContentContainer = _react.forwardRef(
    function ContentContainer({ ariaLabel, ariaLabelledBy, ariaDescribedBy, captureFocus = true, autoFocus = true, onRequestEscape, children,
    className, ...rest }, ref) {
        const divProps = {
            ['aria-label']: ariaLabel,
            ['aria-labelledby']: ariaLabelledBy,
            ['aria-describedby']: ariaDescribedBy,
            ...rest
        };
        return (0, _jsxruntime.jsx)(_focus_lock.FocusLock, {
            returnFocus: captureFocus,
            preventScroll: "on-autofocus",
            disabled: !captureFocus,
            onRequestEscape: onRequestEscape,
            autoFocus: autoFocus,
            ref: ref,
            lockProps: divProps,
            className: className,
            makeContainerFocusable: true,
            children: children
        });
    }
);
