// @canva/app-ui-kit/dist/cjs/ui/base/metrics/metrics.js
"use strict"

Object.defineProperty(exports, "__esModule", {
  value: true
});
function _export(target, all) {
  for (var name in all) Object.defineProperty(target, name, {
    enumerable: true,
    get: Object.getOwnPropertyDescriptor(all, name).get
  });
}
_export(exports, {
  get Breakpoint() {
    return Breakpoint;
  },
  get DEFAULT_ROOT_FONT_SIZE() {
    return DEFAULT_ROOT_FONT_SIZE;
  },
  get aboveLarge() {
    return aboveLarge;
  },
  get aboveMedium() {
    return aboveMedium;
  },
  get aboveSmall() {
    return aboveSmall;
  },
  get aboveXLarge() {
    return aboveXLarge;
  },
  get baseUnit() {
    return baseUnit;
  },
  get belowLarge() {
    return belowLarge;
  },
  get belowMedium() {
    return belowMedium;
  },
  get belowSmall() {
    return belowSmall;
  },
  get belowXLarge() {
    return belowXLarge;
  },
  get breakpointFromScreenWidth() {
    return breakpointFromScreenWidth;
  },
  get getRootFontSize() {
    return getRootFontSize;
  },
  get getSpaceValue() {
    return getSpaceValue;
  },
  get largeBreakpoint() {
    return largeBreakpoint;
  },
  get mediumBreakpoint() {
    return mediumBreakpoint;
  },
  get resolveUnitOrNumberSize() {
    return resolveUnitOrNumberSize;
  },
  get smallBreakpoint() {
    return smallBreakpoint;
  },
  get unitSizeToNumber() {
    return unitSizeToNumber;
  },
  get unitSpaces() {
    return unitSpaces;
  },
  get useBaseUnitInRem() {
    return useBaseUnitInRem;
  },
  get useTextZoomBaseUnit() {
    return useTextZoomBaseUnit;
  },
  get xLargeBreakpoint() {
    return xLargeBreakpoint;
  }
});
const _provider = require('../provider/provider');
const _metricscss = _interop_require_default(require("./metrics.css"));
function _interop_require_default(obj) {
  return obj && obj.__esModule ? obj : {
    default: obj
  };
}
const baseUnit = parseInt(_metricscss.default.baseUnit, 10) || 8;
const Breakpoint = {
  DEFAULT: 0,
  SMALL: 1,
  MEDIUM: 2,
  LARGE: 3,
  XLARGE: 4
};
const smallBreakpoint = parseInt(_metricscss.default.smallBreakpoint, 10) || 600;
const mediumBreakpoint = parseInt(_metricscss.default.mediumBreakpoint, 10) || 900;
const largeBreakpoint = parseInt(_metricscss.default.largeBreakpoint, 10) || 1200;
const xLargeBreakpoint = parseInt(_metricscss.default.xLargeBreakpoint, 10) || 1650;
const breakpointFromScreenWidth = width => {
  if (width >= xLargeBreakpoint) return 4;else if (width >= largeBreakpoint) return 3;else if (width >= mediumBreakpoint) return 2;else if (width >= smallBreakpoint) return 1;else return 0;
};
const aboveSmall = breakpoint => breakpoint >= 1;
const belowSmall = breakpoint => breakpoint < 1;
const aboveMedium = breakpoint => breakpoint >= 2;
const belowMedium = breakpoint => breakpoint < 2;
const aboveLarge = breakpoint => breakpoint >= 3;
const belowLarge = breakpoint => breakpoint < 3;
const aboveXLarge = breakpoint => breakpoint >= 4;
const belowXLarge = breakpoint => breakpoint < 4;
const unitSpaces = ['0', '0.25u', '0.5u', '1u', '1.5u', '2u', '3u', '4u', '6u', '8u', '12u'];
const getSpaceValue = space => {
  return spaceValues[space];
};
const spaceValues = {
  ['0']: _metricscss.default.space0,
  ['0.25u']: _metricscss.default.space025,
  ['0.5u']: _metricscss.default.space050,
  ['1u']: _metricscss.default.space1,
  ['1.5u']: _metricscss.default.space150,
  ['2u']: _metricscss.default.space2,
  ['3u']: _metricscss.default.space3,
  ['4u']: _metricscss.default.space4,
  ['6u']: _metricscss.default.space6,
  ['8u']: _metricscss.default.space8,
  ['12u']: _metricscss.default.space12
};
const unitSizeToNumber = unitSize => {
  const numeric = unitSize.slice(0, -1);
  return Number(numeric) * baseUnit;
};
const resolveUnitOrNumberSize = size => {
  if (size == null || typeof size === 'number') return size;
  return unitSizeToNumber(size);
};
const DEFAULT_ROOT_FONT_SIZE = 10;
const useBaseUnitInRem = () => {
  const rootFontSize = (0, _provider.useExpectedPixelsPerRem)() || DEFAULT_ROOT_FONT_SIZE;
  return baseUnit / rootFontSize;
};
const getRootFontSize = () => {
  if (window == null) return DEFAULT_ROOT_FONT_SIZE;
  return parseFloat(getComputedStyle(window.document.documentElement).fontSize) || DEFAULT_ROOT_FONT_SIZE;
};
const useTextZoomBaseUnit = (currentRootFontSize = getRootFontSize()) => {
  const originalRootFontSize = (0, _provider.useExpectedPixelsPerRem)() || DEFAULT_ROOT_FONT_SIZE;
  const textZoomLevel = currentRootFontSize / originalRootFontSize;
  return baseUnit * textZoomLevel;
};
