// @canva/app-ui-kit/dist/esm/ui/base/surface/flyout/private/flyout.js
import { jsx as _jsx, jsxs as _jsxs, Fragment as _Fragment } from "react/jsx-runtime";
import { UnreachableError } from '../../../../../base/preconditions';
import * as React from 'react';
import { isPortraitMobileScreen } from '../../../device_capabilities/device_capabilities';
import { SkipRenderOnClient } from '../../../hydration/skip_render_on_client';
import { isReferenceObject, Popover } from '../../popover/popover';
import { Sheet } from '../../sheet/sheet';
const modes = [
    'sheet',
    'popover'
];
export function Flyout(
    { open, onRequestClose, onCloseComplete, blockOutsidePointerEvents, trigger, width = '32u', dangerouslySetWidthMode = 'fixed', header, footer, children, title, description, headerStart, headerEnd, headerDivider = 'always', headerAlignment = 'auto', ...contentContainerProps }
) {
    const mode = useMode();
    const resolveContent = (content)=>typeof content === 'function' ? modes.map((m, i)=>_jsx(SkipRenderOnClient, {
                shouldRenderOnClient: ()=>mode === m,
                initiallyVisible: i === 0,
                children: content({
                    mode: m
                })
            }, m)) : content;
    const resolvedHeader = resolveContent(header);
    const resolvedFooter = resolveContent(footer);
    const endDecorator = resolveContent(headerEnd);
    const startDecorator = resolveContent(headerStart);
    const headerAndFooterProps = {
        footer: hasContent(resolvedFooter) ? resolvedFooter : null,
        header: hasContent(resolvedHeader) ? resolvedHeader : null,
        title,
        description,
        headerStart: startDecorator,
        headerEnd: endDecorator,
        headerDivider,
        headerAlignment: headerAlignment === 'auto' ? mode === 'sheet' ? 'center' : 'start' : headerAlignment
    };
    return _jsx("div", {
        children: [
            ...modes.map((m, i)=>{
                switch(m){
                    case 'popover':
                        const popoverWidth = width === 'trigger' ? 'reference' : width;
                        return _jsx(SkipRenderOnClient, {
                            shouldRenderOnClient: ()=>mode === 'popover',
                            initiallyVisible: i === 0,
                            children: _jsx(Popover, {
                                open: open,
                                onRequestClose: onRequestClose,
                                onCloseComplete: onCloseComplete,
                                blockOutsidePointerEvents: blockOutsidePointerEvents,
                                reference: trigger,
                                width: popoverWidth,
                                widthMode: dangerouslySetWidthMode,
                                ...headerAndFooterProps,
                                ...contentContainerProps,
                                children: resolveContent(children)
                            })
                        }, m);
                    case 'sheet':
                        return _jsx(SkipRenderOnClient, {
                            shouldRenderOnClient: ()=>mode === 'sheet',
                            initiallyVisible: i === 0,
                            children: _jsx(FlyoutSheet, {
                                open: open,
                                onRequestClose: onRequestClose,
                                onCloseComplete: onCloseComplete,
                                trigger: trigger,
                                ...headerAndFooterProps,
                                ...contentContainerProps,
                                children: resolveContent(children)
                            })
                        }, m);
                    default:
                        throw new UnreachableError(m);
                }
            })
        ]
    });
}
function useMode() {
    return isPortraitMobileScreen() ? 'sheet' : 'popover';
}
const hasContent = (node)=>React.Children.count(node) > 0;
function FlyoutSheet({ id: idProp, open, onRequestClose, onCloseComplete, onScroll, trigger, handleSafeAreaInsetBottom = true, children, ...contentContainerProps }) {
    const defaultId = React.useId();
    const id = idProp || defaultId;
    return _jsxs(_Fragment, {
        children: [
            typeof trigger === 'function' ? trigger({
                ariaControls: open ? id : undefined,
                active: open,
                disclosure: true,
                ariaHasPopup: contentContainerProps.role
            }) : !isReferenceObject(trigger) && trigger,
            _jsx(Sheet, {
                open: open,
                onRequestClose: onRequestClose,
                onCloseComplete: onCloseComplete,
                onScroll: onScroll,
                id: id,
                handleSafeAreaInsetBottom: handleSafeAreaInsetBottom,
                ...contentContainerProps,
                children: children
            })
        ]
    });
}
