// @canva/app-ui-kit/dist/esm/ui/apps/developing/ui_kit/components/card/horizontal/horizontal.js
import { jsx as _jsx, jsxs as _jsxs, Fragment as _Fragment } from "react/jsx-runtime";
import 'react';
import { Box } from '../../../../../../base/box/box';
import { Card as EaselHorizontalCard, CardDecorator, CardIconThumbnail, CardImageThumbnail } from '../../../../../../base/card/card';
import { AnimatedSpinnerIcon } from '../../../../../../base/icons/animated_spinner/icon';
import { TemplateIcon } from '../../../../../../base/icons/template/icon';
import { CARD_THUMBNAIL_SIDE_LENGTH_PX } from '../internal_card';
import styles from './horizontal.css';
export const HorizontalCard = ({ title, ariaLabel, description, thumbnail, onClick, bottomEnd, bottomEndVisibility, disabled, loading })=>{
    return _jsx(EaselHorizontalCard, {
        layout: "horizontal",
        hoverBoundary: "card",
        padding: "1u",
        hoverEffect: "bgColorTransition",
        title: title,
        description: description,
        ariaLabel: ariaLabel,
        onClick: disabled !== true && loading !== true ? onClick : undefined,
        disabled: disabled,
        decorators: _jsxs(_Fragment, {
            children: [
                bottomEnd && _jsx(CardDecorator, {
                    location: "bottom-end",
                    visibility: bottomEndVisibility,
                    isInteractive: true,
                    children: bottomEnd
                }),
                loading && _jsx(Box, {
                    className: styles.loading,
                    display: "flex",
                    justifyContent: "center",
                    alignItems: "center",
                    width: "full",
                    height: "full",
                    children: _jsx(AnimatedSpinnerIcon, {
                        size: "medium",
                        className: styles.spinner
                    })
                })
            ]
        }),
        thumbnail: _jsxs(_Fragment, {
            children: [
                thumbnail && ('url' in thumbnail ? _jsx(CardImageThumbnail, {
                    src: thumbnail.url,
                    alt: thumbnail.alt,
                    width: CARD_THUMBNAIL_SIDE_LENGTH_PX,
                    height: CARD_THUMBNAIL_SIDE_LENGTH_PX,
                    borderRadius: "elementSmall"
                }) : _jsx(CardIconThumbnail, {
                    Icon: thumbnail.icon,
                    width: CARD_THUMBNAIL_SIDE_LENGTH_PX,
                    height: CARD_THUMBNAIL_SIDE_LENGTH_PX,
                    borderRadius: "elementSmall",
                    border: "none"
                })),
                !thumbnail && _jsx(CardIconThumbnail, {
                    Icon: TemplateIcon,
                    width: CARD_THUMBNAIL_SIDE_LENGTH_PX,
                    height: CARD_THUMBNAIL_SIDE_LENGTH_PX,
                    borderRadius: "elementSmall",
                    border: "none"
                })
            ]
        })
    });
};
