// @canva/app-ui-kit/dist/cjs/ui/base/mobile_event_handler/mobile_event_handler.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
function _export(target, all) {
  for (var name in all) Object.defineProperty(target, name, {
    enumerable: true,
    get: Object.getOwnPropertyDescriptor(all, name).get
  });
}
_export(exports, {
  get MobileEventHandler() {
    return MobileEventHandler;
  },
  get MobileEvents() {
    return MobileEvents;
  }
});
const _jsxruntime = require("react/jsx-runtime");
const _react = _interop_require_wildcard(require("react"));
function _getRequireWildcardCache(nodeInterop) {
  if (typeof WeakMap !== "function") return null;
  var cacheBabelInterop = new WeakMap();
  var cacheNodeInterop = new WeakMap();
  return (_getRequireWildcardCache = function (nodeInterop) {
    return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
  })(nodeInterop);
}
function _interop_require_wildcard(obj, nodeInterop) {
  if (!nodeInterop && obj && obj.__esModule) {
    return obj;
  }
  if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
    return {
      default: obj
    };
  }
  var cache = _getRequireWildcardCache(nodeInterop);
  if (cache && cache.has(obj)) {
    return cache.get(obj);
  }
  var newObj = {
    __proto__: null
  };
  var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
  for (var key in obj) {
    if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
      var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
      if (desc && (desc.get || desc.set)) {
        Object.defineProperty(newObj, key, desc);
      } else {
        newObj[key] = obj[key];
      }
    }
  }
  newObj.default = obj;
  if (cache) {
    cache.set(obj, newObj);
  }
  return newObj;
}
const MobileEventHandlerContext = _react.createContext(undefined);
const MobileEvents = {
  BACK: "backbutton"
};
const MobileEventHandler = function MobileEventHandler({
  onBackButton,
  children
}) {
  const [nodeContext] = _react.useState(() => ({
    handler: onBackButton,
    childContexts: []
  }));
  _react.useEffect(() => {
    nodeContext.handler = onBackButton;
  }, [nodeContext, onBackButton]);
  const parentContext = _react.useContext(MobileEventHandlerContext);
  _react.useEffect(() => {
    if (parentContext) {
      parentContext.childContexts.unshift(nodeContext);
      return () => {
        const {
          childContexts
        } = parentContext;
        const index = childContexts.indexOf(nodeContext);
        if (index > -1) childContexts.splice(index, 1);
      };
    } else {
      const processHandlerTree = () => processContext(nodeContext);
      return backButtonTrigger(processHandlerTree);
    }
  },
  []);
  return (0, _jsxruntime.jsx)(MobileEventHandlerContext.Provider, {
    value: nodeContext,
    children: children
  });
};
const backButtonTrigger = onTrigger => {
  try {
    var _self_Android_getBackButtonEventName, _self_Android;
    var _self_Android_getBackButtonEventName1;
    const backButtonEventName = (_self_Android_getBackButtonEventName1 = (_self_Android = self.Android) === null || _self_Android === void 0 ? void 0 : (_self_Android_getBackButtonEventName = _self_Android.getBackButtonEventName) === null || _self_Android_getBackButtonEventName === void 0 ? void 0 : _self_Android_getBackButtonEventName.call(_self_Android)) !== null && _self_Android_getBackButtonEventName1 !== void 0 ? _self_Android_getBackButtonEventName1 : "backbutton";
    document.addEventListener(backButtonEventName, onTrigger);
    return () => document.removeEventListener(backButtonEventName, onTrigger);
  } catch (e) {
    if (window.self === window.top)
      throw e;
  }
};
const processContext = ({
  childContexts,
  handler
}) => {
  let propagationStopped = false;
  for (const ctx of childContexts) {
    propagationStopped = processContext(ctx);
    if (propagationStopped) break;
  }
  if (!propagationStopped)
    handler({
      stopPropagation: () => propagationStopped = true
    });
  return propagationStopped;
};
