// @canva/app-ui-kit/dist/esm/ui/base/color_picker/sat_val_picker/private/sat_val_picker.js
import { jsx as _jsx } from "react/jsx-runtime";
import * as React from 'react';
import { HsvColor } from '../../hsv_color/hsv_color';
import styles from './sat_val_picker.css';
export function SatValPicker(
    { color, Draggable, onChange: onChangeProp, onChangeStart, onChangeComplete }
) {
    const pickerRef = React.useRef(null);
    const setPickerRef = React.useCallback((el)=>{
        pickerRef.current = el;
        el === null || el === void 0 ? void 0 : el.addEventListener('touchmove', (event)=>event.stopPropagation());
    }, []);
    const [browserWindow, setBrowserWindow] = React.useState();
    React.useLayoutEffect(()=>{
        var _pickerRef_current;
        var _pickerRef_current_ownerDocument_defaultView;
        setBrowserWindow((_pickerRef_current_ownerDocument_defaultView = (_pickerRef_current = pickerRef.current) === null || _pickerRef_current === void 0 ? void 0 : _pickerRef_current.ownerDocument.defaultView) !== null && _pickerRef_current_ownerDocument_defaultView !== void 0 ? _pickerRef_current_ownerDocument_defaultView : window);
    }, []);
    const [changing, setChanging] = React.useState(false);
    const onChange = ({ s, v })=>{
        if (!changing) {
            setChanging(true);
            onChangeStart && onChangeStart();
        }
        if (s !== color.s || v !== color.v) onChangeProp === null || onChangeProp === void 0 ? void 0 : onChangeProp(new HsvColor(color.h, s, v, color.alpha));
    };
    const moveKnob = (knob)=>{
        onChange({
            s: knob.leftPercent,
            v: 1 - knob.topPercent
        });
    };
    const stopChanging = React.useCallback(()=>{
        setChanging(false);
        if (onChangeComplete) onChangeComplete();
    }, [
        onChangeComplete
    ]);
    const onKeyDown = (event)=>{
        const INCREMENT = 0.01;
        let { s, v } = color;
        switch(event.key){
            case 'ArrowUp':
                v += INCREMENT;
                break;
            case 'ArrowDown':
                v -= INCREMENT;
                break;
            case 'ArrowLeft':
                s -= INCREMENT;
                break;
            case 'ArrowRight':
                s += INCREMENT;
                break;
            default:
                return;
        }
        event.preventDefault();
        const clamp = (n)=>Math.min(Math.max(0, n), 1);
        onChange({
            s: clamp(s),
            v: clamp(v)
        });
    };
    const onKeyUp = React.useCallback(()=>{
        stopChanging();
    }, [
        stopChanging
    ]);
    const onMouseUp = React.useCallback(()=>{
        stopChanging();
    }, [
        stopChanging
    ]);
    const onMouseDown = (event)=>{
        var _pickerRef_current;
        if (event.button !== 0)
            return;
        (_pickerRef_current = pickerRef.current) === null || _pickerRef_current === void 0 ? void 0 : _pickerRef_current.focus();
        moveKnob(mouseToKnobCoords(pickerRef.current.getBoundingClientRect(), event));
    };
    const onDragMove = (event)=>moveKnob(mouseToKnobCoords(pickerRef.current.getBoundingClientRect(), event));
    const onDragEnd = React.useCallback((event)=>{
        stopChanging();
    }, [
        stopChanging
    ]);
    const knobPosition = {
        width: `${color.s * 100}%`,
        height: `${100 - color.v * 100}%`
    };
    return _jsx(Draggable, {
        onMouseDown: onMouseDown,
        onDragMove: onDragMove,
        onDragEnd: onDragEnd,
        browserWindow: browserWindow,
        raf: browserWindow === null || browserWindow === void 0 ? void 0 : browserWindow.requestAnimationFrame,
        caf: browserWindow === null || browserWindow === void 0 ? void 0 : browserWindow.cancelAnimationFrame,
        children: ({
            onMouseDown,
            onTouchStart
        })=>
            _jsx("div", {
                className: styles.root,
                onMouseDown: onMouseDown,
                onTouchStart: onTouchStart,
                onMouseUp: onMouseUp,
                onKeyDown: onKeyDown,
                onKeyUp: onKeyUp,
                ref: setPickerRef,
                tabIndex: 0,
                style: {
                    backgroundColor: `hsl(${color.h * 360}, 100%, 50%)`
                },
                children: _jsx("div", {
                    className: styles.whiteGradient,
                    children: _jsx("div", {
                        className: styles.blackGradient,
                        children: _jsx("div", {
                            className: styles.knobPositioner,
                            style: knobPosition,
                            children: _jsx("div", {
                                className: styles.knob,
                                style: {
                                    backgroundColor: color.toRgb().toHexString()
                                }
                            })
                        })
                    })
                })
            })
    });
}
export function mouseToKnobCoords(pickerRect, { clientX, clientY }) {
    const clamp = (val)=>Math.min(1, Math.max(0, val));
    return {
        leftPercent: clamp((clientX - pickerRect.left) / pickerRect.width),
        topPercent: clamp((clientY - pickerRect.top) / pickerRect.height)
    };
}
