// @canva/app-ui-kit/dist/cjs/ui/base/form/checkbox/private/checkbox.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: Object.getOwnPropertyDescriptor(all, name).get
    });
}
_export(exports, {
    get Checkbox () {
        return Checkbox;
    },
    get FakeCheckbox () {
        return FakeCheckbox;
    },
    get InheritColor () {
        return InheritColor;
    }
});
const _jsxruntime = require("react/jsx-runtime");
const _preconditions = require('../../../../../base/preconditions');
const _classnames = _interop_require_default(require("classnames"));
const _mobxreactlite = require("mobx-react-lite");
const _react = _interop_require_wildcard(require("react"));
const _access_mode_state = require('../../../a11y/access_mode_state/access_mode_state');
const _pointer_input_state = require('../../../a11y/pointer_input_state/pointer_input_state');
const _controllable_value = require('../../../controllable_value/controllable_value');
const _icon = require('../../../icons/check/icon');
const _icon1 = require('../../../icons/minus/icon');
const _layout = require('../../../layout/layout');
const _typography = require('../../../typography/typography');
const _checkboxcss = _interop_require_wildcard(require("./checkbox.css"));
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function _getRequireWildcardCache(nodeInterop) {
    if (typeof WeakMap !== "function") return null;
    var cacheBabelInterop = new WeakMap();
    var cacheNodeInterop = new WeakMap();
    return (_getRequireWildcardCache = function(nodeInterop) {
        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
    })(nodeInterop);
}
function _interop_require_wildcard(obj, nodeInterop) {
    if (!nodeInterop && obj && obj.__esModule) {
        return obj;
    }
    if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
        return {
            default: obj
        };
    }
    var cache = _getRequireWildcardCache(nodeInterop);
    if (cache && cache.has(obj)) {
        return cache.get(obj);
    }
    var newObj = {
        __proto__: null
    };
    var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
    for(var key in obj){
        if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
            if (desc && (desc.get || desc.set)) {
                Object.defineProperty(newObj, key, desc);
            } else {
                newObj[key] = obj[key];
            }
        }
    }
    newObj.default = obj;
    if (cache) {
        cache.set(obj, newObj);
    }
    return newObj;
}
const InheritColor = Symbol('inherit');
const Checkbox = (0, _mobxreactlite.observer)(function Checkbox(props) {
    const { label, description, checked: checkedProp, defaultChecked, disabled, autoComplete, value = '', labelClassName, className, size = 'small', id, ariaLabel, ariaLabelledBy, ariaDescribedBy, tone = 'primary', focusRef, onFocus, onBlur, onChange: onChangeProp } = props;
    const [checked, setChecked] = (0, _controllable_value.useControllableValue)({
        value: checkedProp === 'indeterminate' ? false : checkedProp,
        defaultValue: defaultChecked !== null && defaultChecked !== void 0 ? defaultChecked : false
    });
    if (description != null) _preconditions.Preconditions.checkArgument(label != null, 'Provide label to display description in the checkbox');
    const descriptionId = _react.useId();
    const formattedValue = _react.useMemo(()=>{
        if (value === undefined) return undefined;
        if (typeof value === 'string' || typeof value === 'number') return value;
        return JSON.stringify(value);
    }, [
        value
    ]);
    const onChange = _react.useCallback((event)=>{
        setChecked(event.target.checked);
        onChangeProp === null || onChangeProp === void 0 ? void 0 : onChangeProp(value, event.target.checked);
    }, [
        setChecked,
        onChangeProp,
        value
    ]);
    const hoverSupported = (0, _pointer_input_state.useIsHoveringInput)();
    const containerClassName = (0, _classnames.default)(_checkboxcss.default.container, {
        [_checkboxcss.default.disabled]: !!disabled,
        [_checkboxcss.default.hoverSupported]: hoverSupported,
        [_checkboxcss.default.labelContainer]: label != null
    }, className);
    const labelContent = (0, _jsxruntime.jsx)(_typography.Text, {
        className: (0, _classnames.default)(_checkboxcss.default.checkboxLabel, labelClassName),
        tagName: typeof label === 'string' ? 'p' : 'div',
        tone: disabled ? 'tertiary' : 'primary',
        children: label
    });
    return (0, _jsxruntime.jsxs)("label", {
        className: containerClassName,
        children: [
            (0, _jsxruntime.jsxs)("div", {
                className: (0, _classnames.default)(_checkboxcss.default.checkboxWrapper, {
                    [_checkboxcss.default.medium]: size === 'medium'
                }),
                children: [
                    (0, _jsxruntime.jsx)("input", {
                        type: "checkbox",
                        className: _checkboxcss.default.checkbox,
                        onChange: onChange,
                        onFocus: onFocus,
                        onBlur: onBlur,
                        checked: checked,
                        disabled: disabled,
                        autoComplete: autoComplete,
                        value: formattedValue,
                        id: id,
                        "aria-label": ariaLabel,
                        "aria-describedby": ariaDescribedBy !== null && ariaDescribedBy !== void 0 ? ariaDescribedBy : description ? descriptionId : undefined,
                        "aria-labelledby": ariaLabelledBy,
                        "aria-checked": checkedProp === 'indeterminate' ? 'mixed' : undefined,
                        ref: focusRef
                    }),
                    (0, _jsxruntime.jsx)(FakeCheckbox, {
                        checked: checkedProp === 'indeterminate' ? checkedProp : checked,
                        disabled: disabled,
                        size: size,
                        tone: tone,
                        className: _access_mode_state.accessModeState.isMouseMode ? _checkboxcss.default.hideOutline : undefined
                    })
                ]
            }),
            label != null && (0, _jsxruntime.jsx)(_layout.Spacer, {
                direction: "horizontal",
                size: "0.5u"
            }),
            label != null && (description != null ? (0, _jsxruntime.jsxs)("div", {
                className: _checkboxcss.default.checkboxText,
                children: [
                    labelContent,
                    (0, _jsxruntime.jsx)("div", {
                        "aria-hidden": true,
                        children: (0, _jsxruntime.jsx)(_typography.Text, {
                            className: _checkboxcss.default.checkboxDescription,
                            id: descriptionId,
                            tone: "tertiary",
                            size: "small",
                            children: description
                        })
                    })
                ]
            }) : labelContent)
        ]
    });
});
const FakeCheckbox = ({ checked, disabled, size = 'small', tone = 'primary', className })=>{
    return (0, _jsxruntime.jsx)("span", {
        "aria-hidden": true,
        className: (0, _classnames.default)(_checkboxcss.default.uiCheckbox, size === 'medium' && _checkboxcss.default.medium, tone === InheritColor ? _checkboxcss.default.inherit : (0, _checkboxcss.getStyle)(tone), checked === true && _checkboxcss.default.checked, checked === 'indeterminate' && _checkboxcss.default.indeterminate, disabled && _checkboxcss.default.disabled, className),
        children: checked === 'indeterminate' ? (0, _jsxruntime.jsx)(_icon1.MinusIcon, {
            size: size,
            className: _checkboxcss.default.minus
        }) : (0, _jsxruntime.jsx)(_icon.CheckIcon, {
            size: size,
            className: _checkboxcss.default.tick
        })
    });
};
