// @canva/app-ui-kit/dist/cjs/ui/base/menu/private/placeholder.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "MenuItemPlaceholder", {
    enumerable: true,
    get: function() {
        return MenuItemPlaceholder;
    }
});
const _jsxruntime = require("react/jsx-runtime");
const _react = _interop_require_wildcard(require("react"));
const _box = require('../../box/box');
const _layout = require('../../layout/layout');
const _placeholder = require('../../placeholder/placeholder');
const _placeholder1 = require('../../typography/placeholder/placeholder');
const _placeholdercss = _interop_require_default(require("./placeholder.css"));
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function _getRequireWildcardCache(nodeInterop) {
    if (typeof WeakMap !== "function") return null;
    var cacheBabelInterop = new WeakMap();
    var cacheNodeInterop = new WeakMap();
    return (_getRequireWildcardCache = function(nodeInterop) {
        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
    })(nodeInterop);
}
function _interop_require_wildcard(obj, nodeInterop) {
    if (!nodeInterop && obj && obj.__esModule) {
        return obj;
    }
    if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
        return {
            default: obj
        };
    }
    var cache = _getRequireWildcardCache(nodeInterop);
    if (cache && cache.has(obj)) {
        return cache.get(obj);
    }
    var newObj = {
        __proto__: null
    };
    var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
    for(var key in obj){
        if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
            if (desc && (desc.get || desc.set)) {
                Object.defineProperty(newObj, key, desc);
            } else {
                newObj[key] = obj[key];
            }
        }
    }
    newObj.default = obj;
    if (cache) {
        cache.set(obj, newObj);
    }
    return newObj;
}
const MenuItemPlaceholder = _react.memo(
    function MenuItemPlaceholder({ index, startIcon = false, width: _width }) {
        const width = _width || getPlaceholderWidth(index || 0);
        return (0, _jsxruntime.jsx)(_box.Box, {
            tagName: "li",
            role: "none",
            paddingX: "2u",
            className: _placeholdercss.default.menuItemPlaceholder,
            children: (0, _jsxruntime.jsxs)(_layout.Columns, {
                spacing: "1u",
                align: "start",
                alignY: "center",
                children: [
                    startIcon && (0, _jsxruntime.jsx)(_layout.Column, {
                        width: "content",
                        children: (0, _jsxruntime.jsx)(_box.Box, {
                            className: _placeholdercss.default.startIcon,
                            children: (0, _jsxruntime.jsx)(_placeholder.Placeholder, {
                                shape: "square",
                                index: index
                            })
                        })
                    }),
                    (0, _jsxruntime.jsx)(_layout.Column, {
                        width: width,
                        children: (0, _jsxruntime.jsx)(_placeholder1.TextPlaceholder, {
                            size: "medium",
                            index: index
                        })
                    })
                ]
            })
        });
    }
);
MenuItemPlaceholder.displayName = 'MenuItemPlaceholder';
const placeholderWidthDefaultValues = [
    '2/3',
    '1/2'
];
const getPlaceholderWidth = (index)=>{
    return index != null ? placeholderWidthDefaultValues[index % 2] : '2/3';
};
