// @canva/app-ui-kit/dist/esm/ui/base/avatar/private/acronym_avatar.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { RgbaColor } from '../../../../base/color/color';
import { CssColors } from '../../../../base/color/css_colors';
import { UnreachableError } from '../../../../base/preconditions';
import classNames from 'classnames';
import * as React from 'react';
import { getContrastingForegroundColor } from '../../colors/colors';
import { getInitials } from '../../initials/initials';
import styles, { customProperties } from './acronym_avatar.css';
import { DEFAULT_BACKGROUND_COLOR, generateAvatarBackgroundColor } from './utils';
export const AcronymAvatar = React.memo((props)=>{
    const { name, borderColor, backgroundColor, backgroundSeed, role, ariaLabel, className, style, shape = 'circle', ...tooltipTriggerProps } = props;
    let normalizedColor;
    if (backgroundColor) {
        const cssColor = CssColors.fromString(backgroundColor);
        if (cssColor instanceof RgbaColor) normalizedColor = CssColors.toRgbaString(cssColor);
        else normalizedColor = CssColors.toHexString(cssColor);
    } else if (backgroundSeed) normalizedColor = generateAvatarBackgroundColor(backgroundSeed);
    else normalizedColor = DEFAULT_BACKGROUND_COLOR;
    const clipId = React.useId();
    let defs;
    switch(shape){
        case 'circle':
            defs = _jsx("defs", {
                children: _jsx("clipPath", {
                    id: clipId,
                    children: _jsx("circle", {
                        id: `${clipId}-path`,
                        cx: "50%",
                        cy: "50%",
                        r: "50%"
                    })
                })
            });
            break;
        case 'square':
            defs = _jsx("defs", {
                children: _jsx("clipPath", {
                    id: clipId,
                    children: _jsx("rect", {
                        id: `${clipId}-path`,
                        width: "100%",
                        height: "100%"
                    })
                })
            });
            break;
        default:
            throw new UnreachableError(shape);
    }
    let fill;
    switch(shape){
        case 'circle':
            fill = _jsx("circle", {
                cx: "50%",
                cy: "50%",
                fill: normalizedColor,
                shapeRendering: "geometricPrecision",
                clipPath: `url(#${clipId})`,
                stroke: borderColor,
                className: classNames(styles.circle, {
                    [styles.withStroke]: borderColor
                })
            });
            break;
        case 'square':
            fill = _jsx("rect", {
                width: "100%",
                height: "100%",
                fill: normalizedColor,
                shapeRendering: "geometricPrecision",
                clipPath: `url(#${clipId})`,
                stroke: borderColor,
                className: classNames(styles.square, {
                    [styles.withStroke]: borderColor
                })
            });
            break;
        default:
            throw new UnreachableError(shape);
    }
    const avatarTextColor = getContrastingForegroundColor(normalizedColor);
    return _jsx("span", {
        role: role,
        "aria-label": ariaLabel,
        className: classNames(styles.container, {
            [styles.square]: shape === 'square'
        }),
        style: {
            [customProperties.avatarTextColor]: avatarTextColor
        },
        ...tooltipTriggerProps,
        children: _jsxs("svg", {
            className: className,
            style: style,
            children: [
                defs,
                fill,
                _jsx("text", {
                    x: "50%",
                    y: "50%",
                    "aria-hidden": "true",
                    textAnchor: "middle",
                    fill: avatarTextColor,
                    fontWeight: "500",
                    fontSize: "50%",
                    dominantBaseline: "central",
                    letterSpacing: "0.01rem",
                    dy: "-0.04em",
                    children: getInitials(name).toUpperCase()
                })
            ]
        })
    });
});
