// @canva/app-ui-kit/dist/esm/ui/base/avatar/private/avatar_group_utils.js
import { jsx as _jsx } from "react/jsx-runtime";
import classnames from 'classnames';
import * as React from 'react';
import { useIsHoveringInput } from '../../a11y/pointer_input_state/pointer_input_state';
import { BaseButton } from '../../button/base_button/base_button';
import { NumberAbbreviationPresenter } from '../../number_abbreviation/number_abbreviation_presenter';
import { Tooltip } from '../../tooltip/tooltip';
import avatarStyles from './avatar.css';
import styles from './avatar_group.css';
import { AvatarGroupMessages } from './avatar_group.messages';
import { useAvatarSize, useAvatarSizeStyle } from './avatar_size';
import sizedAvatarStyles from './avatar_size.css';
import internalAvatarStyles from './internal_avatar.css';
export const OverflowAvatar = React.forwardRef(
    ({ count, size = 'medium', role = 'img', onClick, active, pressed, disclosure, ariaHasPopup, id, ariaOwns, tooltipLabel, ariaLabel }, ref)=>{
        const hoverSupported = useIsHoveringInput();
        const label = tooltipLabel !== null && tooltipLabel !== void 0 ? tooltipLabel : AvatarGroupMessages.more(count);
        const avatarSize = useAvatarSize(size);
        const sizeStyle = useAvatarSizeStyle(size);
        const abbreviationPresenter = React.useMemo(()=>new NumberAbbreviationPresenter('en-US', {
                maximumSignificantDigits: 2
            }), []);
        const formattedCount = `+${abbreviationPresenter.abbreviateNumber(count)}`;
        const wrapWithContainer = (_children)=>{
            return _jsx("li", {
                className: classnames(styles.avatarGroupItem, {
                    [styles.overflowAvatarSmall]: avatarSize ? parseFloat(avatarSize) < 3.2 : false,
                    [styles.overflowAvatarThreeChars]: formattedCount.length === 3,
                    [styles.overflowAvatarFourChars]: formattedCount.length === 4,
                    [styles.overflowAvatarFiveChars]: formattedCount.length === 5
                }),
                style: sizeStyle,
                children: _children
            });
        };
        const children = _jsx("span", {
            className: classnames(internalAvatarStyles.avatar, sizedAvatarStyles.internalSizedContainer, styles.overflowAvatar),
            role: role,
            "aria-label": ariaLabel !== null && ariaLabel !== void 0 ? ariaLabel : label,
            children: formattedCount
        });
        if (onClick) return wrapWithContainer(_jsx(BaseButton, {
            onClick: onClick,
            className: classnames(avatarStyles.avatarButton, {
                [styles.hoverSupported]: hoverSupported
            }),
            display: "block",
            borderRadius: "elementRound",
            position: "relative",
            pressed: pressed !== null && pressed !== void 0 ? pressed : active,
            disclosure: disclosure,
            ariaHasPopup: ariaHasPopup,
            id: id,
            ref: ref,
            ariaOwns: ariaOwns,
            ariaLabel: ariaLabel !== null && ariaLabel !== void 0 ? ariaLabel : label,
            tooltipLabel: label,
            children: children
        }));
        if (label && role !== 'presentation') return wrapWithContainer(_jsx(Tooltip, {
            label: label,
            children: ({ tooltipId, ...rest })=>_jsx("div", {
                    className: styles.overflowAvatarContainer,
                    ...rest,
                    children: children
                })
        }));
        return wrapWithContainer(children);
    }
);
