// @canva/app-ui-kit/dist/cjs/ui/base/thumbnail/thumbnail.js
"use strict"
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: Object.getOwnPropertyDescriptor(all, name).get
    });
}
_export(exports, {
    get IconThumbnail () {
        return _icon_thumbnail.IconThumbnail;
    },
    get ImageComponent () {
        return _image.ImageComponent;
    },
    get ImagePlaceholder () {
        return _image.ImagePlaceholder;
    },
    get ImageThumbnail () {
        return _image_thumbnail.ImageThumbnail;
    },
    get ImageThumbnailPlaceholder () {
        return _image_thumbnail.ImageThumbnailPlaceholder;
    },
    get StaticImage () {
        return _image.StaticImage;
    },
    get Thumbnail () {
        return _thumbnail.Thumbnail;
    },
    get ThumbnailContainer () {
        return _thumbnail_container.ThumbnailContainer;
    },
    get VideoThumbnail () {
        return _video_thumbnail.VideoThumbnail;
    },
    get useInThumbnailContainerContext () {
        return _thumbnail_container.useInThumbnailContainerContext;
    }
});
const _icon_thumbnail = require("./private/icon_thumbnail");
const _image = require("./private/image");
const _image_thumbnail = require("./private/image_thumbnail");
const _thumbnail = require("./private/thumbnail");
const _thumbnail_container = require("./private/thumbnail_container");
const _video_thumbnail = require("./private/video_thumbnail");
