// @canva/app-ui-kit/dist/cjs/ui/base/box/private/box.js
"use strict"
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: Object.getOwnPropertyDescriptor(all, name).get
    });
}
_export(exports, {
    get Box () {
        return Box;
    },
    get useBox () {
        return useBox;
    }
});
const _jsxruntime = require("react/jsx-runtime");
const _classnames = _interop_require_default(require("classnames"));
const _react = _interop_require_wildcard(require("react"));
const _common_props = require('../common_props/common_props');
const _atoms = require("./atoms");
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function _getRequireWildcardCache(nodeInterop) {
    if (typeof WeakMap !== "function") return null;
    var cacheBabelInterop = new WeakMap();
    var cacheNodeInterop = new WeakMap();
    return (_getRequireWildcardCache = function(nodeInterop) {
        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
    })(nodeInterop);
}
function _interop_require_wildcard(obj, nodeInterop) {
    if (!nodeInterop && obj && obj.__esModule) {
        return obj;
    }
    if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
        return {
            default: obj
        };
    }
    var cache = _getRequireWildcardCache(nodeInterop);
    if (cache && cache.has(obj)) {
        return cache.get(obj);
    }
    var newObj = {
        __proto__: null
    };
    var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
    for(var key in obj){
        if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
            if (desc && (desc.get || desc.set)) {
                Object.defineProperty(newObj, key, desc);
            } else {
                newObj[key] = obj[key];
            }
        }
    }
    newObj.default = obj;
    if (cache) {
        cache.set(obj, newObj);
    }
    return newObj;
}
function useBox(options) {
    const { className, customProperties, display, background, border, borderTop, borderBottom, borderStart, borderEnd, borderRadius, shadow, padding, paddingX, paddingY, paddingTop, paddingBottom, paddingStart, paddingEnd, width, height, minWidth, minHeight, flexDirection, flexWrap, justifyContent, alignItems, flex, alignSelfInFlex, alignSelfInGrid, position, inset, start, end, top, bottom } = options;
    const cls = (0, _classnames.default)(className, (0, _atoms.resetAtom)(true), (0, _atoms.backgroundAtom)(background), (0, _atoms.borderAtom)(border), (0, _atoms.borderPositionAtom)({
        top: borderTop,
        bottom: borderBottom,
        start: borderStart,
        end: borderEnd
    }), (0, _atoms.borderRadiusAtom)(borderRadius), (0, _atoms.shadowAtom)(shadow), (0, _atoms.widthAtom)(width), (0, _atoms.heightAtom)(height), (0, _atoms.minWidthAtom)(minWidth), (0, _atoms.minHeightAtom)(minHeight), (0, _atoms.getPaddingClassNames)(padding, paddingX, paddingY, paddingTop, paddingBottom, paddingStart, paddingEnd), (0, _atoms.getFlexItemBehaviorClassNames)(flex), (0, _atoms.getDisplayClassNames)(display), (0, _atoms.getFlexDirectionClassNames)(flexDirection), (0, _atoms.flexWrapAtom)(flexWrap), (0, _atoms.justifyContentAtom)(justifyContent, display), (0, _atoms.alignItemsAtom)(alignItems, display), (0, _atoms.alignSelfAtom)('flex', alignSelfInFlex), (0, _atoms.alignSelfAtom)('grid', alignSelfInGrid), (0, _atoms.positionAtom)(position), (0, _atoms.positionInsetAtom)(inset), (0, _atoms.positionStartAtom)(start), (0, _atoms.positionEndAtom)(end), (0, _atoms.positionTopAtom)(top), (0, _atoms.positionBottomAtom)(bottom));
    const style = {
        ...(0, _atoms.getDisplayStyles)(display),
        ...(0, _atoms.getPaddingStyles)(padding, paddingX, paddingY, paddingTop, paddingBottom, paddingStart, paddingEnd),
        ...(0, _atoms.getFlexDirectionStyles)(flexDirection),
        ...customProperties
    };
    return {
        className: cls,
        style
    };
}
const Box = _react.forwardRef(function Box(props, ref) {
    const { tagName: TagName = 'div', children, tabIndex, testId, ...rest } = props;
    return (0, _jsxruntime.jsx)(TagName, {
        ref: ref,
        ...useBox(rest),
        ...(0, _common_props.commonPropsToDOMProps)(rest),
        tabIndex: tabIndex,
        "data-testid": testId,
        children: children
    });
});
