// @canva/app-ui-kit/dist/cjs/ui/apps/developing/ui_kit/components/form/pills_input/input_pill/input_pill.js
"use strict"
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "InputPill", {
    enumerable: true,
    get: function() {
        return InputPill;
    }
});
const _jsxruntime = require("react/jsx-runtime");
const _preconditions = require('../../../../../../../../base/preconditions');
const _pills_input = require('../../../../../../../base/form/pills_input/pills_input');
function InputPill({ truncateText, ...props }) {
    return (0, _jsxruntime.jsx)(_pills_input.InputPill, {
        text: props.text,
        tone: props.tone,
        disabled: props.disabled,
        maxWidth: mapMaxWidth(truncateText),
        onRemoveClick: props.onRemoveClick
    });
}
function mapMaxWidth(truncateText) {
    switch(truncateText){
        case undefined:
            return undefined;
        case true:
            return '25u';
        case false:
            return '100%';
        default:
            throw new _preconditions.UnreachableError(truncateText);
    }
}
