// @canva/app-ui-kit/dist/cjs/ui/base/tooltip/private/tooltip.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: Object.getOwnPropertyDescriptor(all, name).get
    });
}
_export(exports, {
    get FloatingTooltip () {
        return FloatingTooltip;
    },
    get Tooltip () {
        return Tooltip;
    },
    get TriggeredTooltip () {
        return TriggeredTooltip;
    }
});
const _jsxruntime = require("react/jsx-runtime");
const _mobx = _interop_require_wildcard(require("mobx"));
const _mobxreactlite = require("mobx-react-lite");
const _react = _interop_require_wildcard(require("react"));
const _metrics = require('../../metrics/metrics');
const _internal_tooltip = require('../internal/internal_tooltip');
const _tooltip_presenter = require("./tooltip_presenter");
function _getRequireWildcardCache(nodeInterop) {
    if (typeof WeakMap !== "function") return null;
    var cacheBabelInterop = new WeakMap();
    var cacheNodeInterop = new WeakMap();
    return (_getRequireWildcardCache = function(nodeInterop) {
        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
    })(nodeInterop);
}
function _interop_require_wildcard(obj, nodeInterop) {
    if (!nodeInterop && obj && obj.__esModule) {
        return obj;
    }
    if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
        return {
            default: obj
        };
    }
    var cache = _getRequireWildcardCache(nodeInterop);
    if (cache && cache.has(obj)) {
        return cache.get(obj);
    }
    var newObj = {
        __proto__: null
    };
    var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
    for(var key in obj){
        if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
            if (desc && (desc.get || desc.set)) {
                Object.defineProperty(newObj, key, desc);
            } else {
                newObj[key] = obj[key];
            }
        }
    }
    newObj.default = obj;
    if (cache) {
        cache.set(obj, newObj);
    }
    return newObj;
}
const FADE_OUT_DELAY = 2000;
const Tooltip = _react.forwardRef((props, ref)=>(0, _jsxruntime.jsx)(_internal_tooltip.InternalTooltip, {
        ...props,
        forwardedRef: ref
    }));
const TriggeredTooltip = (0, _mobxreactlite.observer)((props)=>{
    const [presenter] = _react.useState(()=>new _tooltip_presenter.TooltipPresenter());
    const { autoclose = true, state, ...rest } = props;
    _react.useEffect(()=>{
        const disposer = _mobx.autorun(()=>{
            if (autoclose && state.open) presenter.hideTooltip(props, FADE_OUT_DELAY);
        });
        return ()=>disposer();
    }, [
        autoclose,
        state,
        presenter,
        rest,
        props
    ]);
    return (0, _jsxruntime.jsx)(_internal_tooltip.StatelessTooltip, {
        ...rest,
        open: state.open,
        mode: "tooltip"
    });
});
const FloatingTooltip = (0, _mobxreactlite.observer)((props)=>{
    const refObj = _react.useMemo(()=>new FloatingTooltipReferenceObject({
            info: props
        }), [
        props
    ]);
    const [info, setInfo] = _react.useState(props);
    const { arrow, open, label, placement = 'right-end', direction, x, y } = props;
    _react.useEffect(()=>{
        if (open)
        setInfo({
            label,
            placement,
            direction,
            x,
            y
        });
    }, [
        open,
        label,
        placement,
        direction,
        x,
        y
    ]);
    return (0, _jsxruntime.jsx)(_internal_tooltip.StatelessTooltip, {
        label: info.label,
        open: open,
        arrow: !!arrow,
        refObj: refObj,
        placement: info.placement,
        direction: info.direction,
        mode: "tooltip"
    });
});
class FloatingTooltipReferenceObject {
    get left() {
        return this.tooltip.info.x - FloatingTooltipReferenceObject.horizontalOffset;
    }
    get top() {
        return this.tooltip.info.y - FloatingTooltipReferenceObject.verticalOffset;
    }
    get width() {
        return 2 * FloatingTooltipReferenceObject.horizontalOffset;
    }
    get height() {
        return 2 * FloatingTooltipReferenceObject.verticalOffset;
    }
    get right() {
        return this.left + this.width;
    }
    get bottom() {
        return this.top + this.height;
    }
    get clientWidth() {
        return this.width;
    }
    get clientHeight() {
        return this.height;
    }
    get x() {
        return this.tooltip.info.x;
    }
    get y() {
        return this.tooltip.info.y;
    }
    toJSON() {
        const { bottom, height, left, right, top, width, x, y } = this;
        return {
            bottom,
            height,
            left,
            right,
            top,
            width,
            x,
            y
        };
    }
    getBoundingClientRect() {
        return this;
    }
    constructor(tooltip){
        this.tooltip = tooltip;
    }
}
FloatingTooltipReferenceObject.horizontalOffset = 3 * _metrics.baseUnit;
FloatingTooltipReferenceObject.verticalOffset = 8 * _metrics.baseUnit;
