// @canva/app-ui-kit/dist/esm/ui/base/gestures/tap_recognizer.js
import { getCommonEventProperties } from './recognizer';
export class TapRecognizer {
    getBrowserHandledTouchActions() {
        return [
            'pan-x',
            'pan-y',
            'pinch-zoom'
        ];
    }
    onPointerDown(pointer, allPointers) {
        var _this_opts_disabled;
        if (pointer.start.button != null && pointer.start.button !== 0)
            return;
        if ((_this_opts_disabled = this.opts.disabled) === null || _this_opts_disabled === void 0 ? void 0 : _this_opts_disabled.get()) return;
        if (allPointers.size === 1) this.state = pointer;
        else
        this.state = undefined;
    }
    onPointerUp(pointer) {
        var _this_opts_onTap, _this_opts;
        if (!this.state) return;
        const start = this.state.start;
        const current = pointer.current;
        const timeDelta = current.timestamp - start.timestamp;
        const distanceSqrd = (current.x - start.x) ** 2 + (current.y - start.y) ** 2;
        if (timeDelta <= 250 && distanceSqrd <= 81) (_this_opts_onTap = (_this_opts = this.opts).onTap) === null || _this_opts_onTap === void 0 ? void 0 : _this_opts_onTap.call(_this_opts, getCommonEventProperties(start));
        this.state = undefined;
    }
    onPointerCancel(pointer) {
        this.state = undefined;
    }
    constructor(opts){
        this.opts = opts;
    }
}
