// @canva/app-ui-kit/dist/esm/ui/base/button/base_button/private/base_button.js
import { jsx as _jsx } from "react/jsx-runtime";
import classNames from 'classnames';
import * as React from 'react';
import { useIsMouseAccessMode } from '../../../a11y/access_mode_state/access_mode_state';
import { getButtonAriaAttributes } from '../../../a11y/button_aria_attributes/button_aria_attributes';
import { useBox } from '../../../box/box';
import { createBasicHTMLHandle } from '../../../handle/handle';
import styles from './base_button.css';
import { useStateMaps, useToneClassNames } from './button_tones';
import { withTooltip } from './with_tooltip';
const InternalBaseButton = React.forwardRef(function InternalBaseButton(props, ref) {
    var _props_display;
    const boxAttributes = useBox({
        className: props.className,
        display: (_props_display = props.display) !== null && _props_display !== void 0 ? _props_display : 'inline-flex',
        border: props.border,
        borderTop: props.borderTop,
        borderBottom: props.borderBottom,
        borderStart: props.borderStart,
        borderEnd: props.borderEnd,
        borderRadius: props.borderRadius,
        shadow: props.shadow,
        padding: props.padding,
        paddingX: props.paddingX,
        paddingY: props.paddingY,
        paddingTop: props.paddingTop,
        paddingBottom: props.paddingBottom,
        paddingStart: props.paddingStart,
        paddingEnd: props.paddingEnd,
        width: props.width,
        height: props.height,
        minWidth: props.minWidth,
        minHeight: props.minHeight,
        flexDirection: props.flexDirection,
        flexWrap: props.flexWrap,
        justifyContent: props.justifyContent,
        alignItems: props.alignItems,
        flex: props.flex,
        alignSelfInFlex: props.alignSelfInFlex,
        alignSelfInGrid: props.alignSelfInGrid,
        position: props.position,
        inset: props.inset,
        start: props.start,
        end: props.end,
        top: props.top,
        bottom: props.bottom
    });
    const ariaAttributes = getButtonAriaAttributes({
        tagName: 'button',
        role: props.role,
        active: props.active,
        selected: props.selected,
        pressed: props.pressed,
        disabled: props.disabled,
        disclosure: props.disclosure,
        tabIndex: props.tabIndex,
        ariaLabel: props.ariaLabel,
        ariaLabelledBy: props.ariaLabelledBy,
        ariaControls: props.ariaControls,
        ariaOwns: props.ariaOwns,
        ariaHasPopup: props.ariaHasPopup,
        ariaDescribedBy: props.ariaDescribedBy,
        ariaRoleDescription: props.ariaRoleDescription,
        ariaHidden: props.ariaHidden,
        ariaActiveDescendant: props.ariaActiveDescendant,
        ariaCurrent: props.ariaCurrent
    });
    const disabledAttributes = getDisabledProps(props.disabled);
    const disableSubmit = useShouldDisableSubmit(props.type === 'submit');
    const stateAttributes = useStateMaps(props);
    const toneClassNames = useToneClassNames(props);
    const className = classNames(styles.button, boxAttributes.className, toneClassNames, {
        [styles.hideOutline]: useIsMouseAccessMode()
    });
    const buttonRef = React.useRef(null);
    React.useImperativeHandle(ref, ()=>createBasicHTMLHandle(buttonRef));
    var _props_type;
    return _jsx("button", {
        ref: buttonRef,
        id: props.id,
        type: (_props_type = props.type) !== null && _props_type !== void 0 ? _props_type : 'button',
        form: props.form,
        draggable: props.draggable,
        onClick: props.onClick,
        onMouseDown: props.onMouseDown,
        onMouseUp: props.onMouseUp,
        onMouseEnter: props.onMouseEnter,
        onMouseLeave: props.onMouseLeave,
        onPointerDown: props.onPointerDown,
        onPointerUp: props.onPointerUp,
        onPointerEnter: props.onPointerEnter,
        onPointerLeave: props.onPointerLeave,
        onPointerMove: props.onPointerMove,
        onPointerCancel: props.onPointerCancel,
        onFocus: props.onFocus,
        onBlur: props.onBlur,
        onDragStart: props.onDragStart,
        onDragEnd: props.onDragEnd,
        onContextMenu: props.onContextMenu,
        ...boxAttributes,
        ...ariaAttributes,
        ...disabledAttributes,
        disabled: disableSubmit || undefined,
        className: className,
        style: {
            ...boxAttributes.style,
            ...stateAttributes.style
        },
        children: props.children
    });
});
const InternalBaseAnchor = React.forwardRef(function InternalBaseAnchor(props, ref) {
    var _props_display;
    const boxAttributes = useBox({
        className: props.className,
        display: (_props_display = props.display) !== null && _props_display !== void 0 ? _props_display : 'inline-flex',
        border: props.border,
        borderTop: props.borderTop,
        borderBottom: props.borderBottom,
        borderStart: props.borderStart,
        borderEnd: props.borderEnd,
        borderRadius: props.borderRadius,
        shadow: props.shadow,
        padding: props.padding,
        paddingX: props.paddingX,
        paddingY: props.paddingY,
        paddingTop: props.paddingTop,
        paddingBottom: props.paddingBottom,
        paddingStart: props.paddingStart,
        paddingEnd: props.paddingEnd,
        width: props.width,
        height: props.height,
        minWidth: props.minWidth,
        minHeight: props.minHeight,
        flexDirection: props.flexDirection,
        flexWrap: props.flexWrap,
        justifyContent: props.justifyContent,
        alignItems: props.alignItems,
        flex: props.flex,
        alignSelfInFlex: props.alignSelfInFlex,
        alignSelfInGrid: props.alignSelfInGrid,
        position: props.position,
        inset: props.inset,
        start: props.start,
        end: props.end,
        top: props.top,
        bottom: props.bottom
    });
    const ariaAttributes = getButtonAriaAttributes({
        tagName: 'a',
        role: props.role,
        active: props.active,
        selected: props.selected,
        pressed: props.pressed,
        disabled: props.disabled,
        disclosure: props.disclosure,
        tabIndex: props.tabIndex,
        ariaLabel: props.ariaLabel,
        ariaLabelledBy: props.ariaLabelledBy,
        ariaControls: props.ariaControls,
        ariaOwns: props.ariaOwns,
        ariaHasPopup: props.ariaHasPopup,
        ariaDescribedBy: props.ariaDescribedBy,
        ariaRoleDescription: props.ariaRoleDescription,
        ariaHidden: props.ariaHidden,
        ariaActiveDescendant: props.ariaActiveDescendant,
        ariaCurrent: props.ariaCurrent
    });
    const disabledAttributes = getDisabledProps(props.disabled);
    const stateAttributes = useStateMaps(props);
    const toneClassNames = useToneClassNames(props);
    const className = classNames(styles.button, boxAttributes.className, toneClassNames, {
        [styles.hideOutline]: useIsMouseAccessMode()
    });
    const anchorRef = React.useRef(null);
    React.useImperativeHandle(ref, ()=>createBasicHTMLHandle(anchorRef));
    return _jsx("a", {
        ref: anchorRef,
        id: props.id,
        href: props.href,
        rel: props.rel,
        target: props.target,
        download: props.download,
        onClick: props.onClick,
        onMouseDown: props.onMouseDown,
        onMouseUp: props.onMouseUp,
        onMouseEnter: props.onMouseEnter,
        onMouseLeave: props.onMouseLeave,
        onPointerDown: props.onPointerDown,
        onPointerUp: props.onPointerUp,
        onPointerEnter: props.onPointerEnter,
        onPointerLeave: props.onPointerLeave,
        onPointerMove: props.onPointerMove,
        onPointerCancel: props.onPointerCancel,
        onFocus: props.onFocus,
        onBlur: props.onBlur,
        onDragStart: props.onDragStart,
        onDragEnd: props.onDragEnd,
        onContextMenu: props.onContextMenu,
        ...boxAttributes,
        ...ariaAttributes,
        ...disabledAttributes,
        className: className,
        style: {
            ...boxAttributes.style,
            ...stateAttributes.style
        },
        children: props.children
    });
});
function getDisabledProps(disabled) {
    if (!disabled) return {};
    return {
        onClick: stopClickPropagation,
        onClickCapture: stopClickPropagation,
        onKeyDownCapture: stopClickPropagation,
        onMouseDownCapture: stopClickPropagation,
        onMouseUpCapture: stopClickPropagation,
        ['aria-disabled']: true
    };
}
function stopClickPropagation(e) {
    const IS_CLICK = [
        'click',
        'mousedown',
        'mouseup',
        'submit'
    ].includes(e.type);
    const IS_ENTER_OR_SPACE = 'code' in e && [
        'Enter',
        'Space'
    ].includes(e.code);
    if (IS_CLICK || IS_ENTER_OR_SPACE) {
        e.stopPropagation();
        e.preventDefault();
    }
}
function useShouldDisableSubmit(initialShouldDisableSubmit) {
    const [disableSubmit, setDisableSubmit] = React.useState(initialShouldDisableSubmit);
    React.useEffect(()=>{
        setDisableSubmit(false);
    }, []);
    return disableSubmit;
}
export const BaseButton = withTooltip(InternalBaseButton);
export const BaseAnchor = withTooltip(InternalBaseAnchor);
