// @canva/app-ui-kit/dist/esm/ui/base/button/base_button/private/button_tones.js
import classNames from 'classnames';
import { useIsHoveringInput } from '../../../a11y/pointer_input_state/pointer_input_state';
import tones, { customProperties, getStyle } from './button_tones.css';
export const buttonTones = [
    'primary',
    'secondary',
    'tertiary',
    'contrast',
    'critical'
];
function isStateMap(value) {
    return typeof value === 'object' && value != null;
}
function isSelectedTone(tone, selected) {
    return (tone === 'secondary' || tone === 'tertiary') && selected;
}
export function useToneClassNames(
    { tone, pressed, selected, disabled, border, color, backgroundColor, borderColor }
) {
    const hasCustomColors = !!(tone || color || backgroundColor || borderColor);
    const toneClassName = isSelectedTone(tone, selected) ? tones.selected : tone != null && getStyle(tone);
    const borderClassName = border && border !== 'none' && getStyle(`${border}Border`);
    return classNames(toneClassName, borderClassName, {
        [tones.tone]: hasCustomColors || borderClassName,
        [tones.pressed]: pressed,
        [tones.disabled]: disabled,
        [tones.hoverDisabled]: !useIsHoveringInput()
    });
}
export function useStateMaps(props) {
    const style = {
        ...applyStateMap(props.backgroundColor, customProperties.defaultBackgroundColor, customProperties.hoveredBackgroundColor, customProperties.pressedBackgroundColor, customProperties.disabledBackgroundColor),
        ...applyStateMap(props.color, customProperties.defaultColor, customProperties.hoveredColor, customProperties.pressedColor, customProperties.disabledColor),
        ...applyStateMap(props.borderColor, customProperties.defaultBorderColor, customProperties.hoveredBorderColor, customProperties.pressedBorderColor, customProperties.disabledBorderColor)
    };
    return {
        style
    };
}
function applyStateMap(value, defaultCustomProperty, hoveredCustomProperty, pressedCustomProperty, disabledCustomProperty) {
    const style = {};
    if (isStateMap(value)) {
        if (value.default) style[defaultCustomProperty] = value.default;
        if (value.hovered) style[hoveredCustomProperty] = value.hovered;
        if (value.pressed) style[pressedCustomProperty] = value.pressed;
        if (value.disabled) style[disabledCustomProperty] = value.disabled;
    } else if (value) {
        style[defaultCustomProperty] = value;
        style[disabledCustomProperty] = value;
    }
    return style;
}
