// @canva/app-ui-kit/dist/esm/ui/base/button/base_button/private/with_tooltip.js
import { jsx as _jsx } from "react/jsx-runtime";
import * as React from 'react';
import { Tooltip } from '../../../tooltip/tooltip';
export function getComponentAriaAttributesConsideringTooltip(ariaAttrsProps, tooltipProps) {
    const ariaAttrs = {
        ...ariaAttrsProps
    };
    if ((ariaAttrsProps.children == null || typeof ariaAttrsProps.children !== 'string') && ariaAttrsProps.ariaLabel == null && ariaAttrsProps.ariaLabelledBy == null)
    ariaAttrs.ariaLabel = tooltipProps.tooltipDescription != null ? `${tooltipProps.tooltipLabel}: ${tooltipProps.tooltipDescription}` : tooltipProps.tooltipLabel;
    else {
        const hasRedundantTooltip = typeof ariaAttrsProps.children === 'string' && ariaAttrsProps.children === tooltipProps.tooltipLabel || ariaAttrsProps.ariaLabel === tooltipProps.tooltipLabel;
        if (!hasRedundantTooltip && ariaAttrs.ariaDescribedBy == null) ariaAttrs.ariaDescribedBy = tooltipProps.tooltipId;
    }
    return ariaAttrs;
}
export const withTooltip = (Component)=>{
    return React.forwardRef((props, ref)=>{
        if (props.tooltipLabel == null) return _jsx(Component, {
            ...props,
            ref: ref
        });
        return _jsx(Tooltip, {
            label: props.tooltipLabel,
            description: props.tooltipDescription,
            shortcut: props.tooltipShortcut,
            disabled: props.tooltipDisabled,
            placement: props.tooltipPlacement,
            lineClamp: props.tooltipLineClamp,
            closeOnClick: props.tooltipCloseOnClick,
            children: (triggerProps)=>{
                var _props_tooltipLabel;
                const ariaAttributesConsideringTooltip = getComponentAriaAttributesConsideringTooltip(props, {
                    tooltipId: triggerProps.tooltipId,
                    tooltipLabel: (_props_tooltipLabel = props.tooltipLabel) !== null && _props_tooltipLabel !== void 0 ? _props_tooltipLabel : '',
                    tooltipDescription: props.tooltipDescription
                });
                return _jsx(InnerMergedTooltipEvents, {
                    componentEventProps: props,
                    tooltipTriggerProps: triggerProps,
                    children: (mergedEvents)=>{
                        return _jsx(Component, {
                            ...props,
                            ...ariaAttributesConsideringTooltip,
                            ...mergedEvents,
                            ref: ref
                        });
                    }
                });
            }
        });
    });
};
const InnerMergedTooltipEvents = ({ componentEventProps, tooltipTriggerProps, children })=>{
    const { onMouseEnter: componentOnMouseEnter, onMouseLeave: componentOnMouseLeave, onMouseDown: componentOnMouseDown, onFocus: componentOnFocus, onBlur: componentOnBlur } = componentEventProps;
    const { onMouseEnter: tooltipTriggerOnMouseEnter, onMouseLeave: tooltipTriggerMouseLeave, onMouseDown: tooltipTriggerMouseDown, onFocus: tooltipTriggerFocus, onBlur: tooltipTriggerOnBlur } = tooltipTriggerProps;
    const onMouseEnter = React.useCallback((event)=>{
        componentOnMouseEnter === null || componentOnMouseEnter === void 0 ? void 0 : componentOnMouseEnter(event);
        tooltipTriggerOnMouseEnter === null || tooltipTriggerOnMouseEnter === void 0 ? void 0 : tooltipTriggerOnMouseEnter();
    }, [
        componentOnMouseEnter,
        tooltipTriggerOnMouseEnter
    ]);
    const onMouseLeave = React.useCallback((event)=>{
        componentOnMouseLeave === null || componentOnMouseLeave === void 0 ? void 0 : componentOnMouseLeave(event);
        tooltipTriggerMouseLeave === null || tooltipTriggerMouseLeave === void 0 ? void 0 : tooltipTriggerMouseLeave();
    }, [
        componentOnMouseLeave,
        tooltipTriggerMouseLeave
    ]);
    const onMouseDown = React.useCallback((event)=>{
        componentOnMouseDown === null || componentOnMouseDown === void 0 ? void 0 : componentOnMouseDown(event);
        tooltipTriggerMouseDown === null || tooltipTriggerMouseDown === void 0 ? void 0 : tooltipTriggerMouseDown();
    }, [
        componentOnMouseDown,
        tooltipTriggerMouseDown
    ]);
    const onFocus = React.useCallback((event)=>{
        componentOnFocus === null || componentOnFocus === void 0 ? void 0 : componentOnFocus(event);
        tooltipTriggerFocus === null || tooltipTriggerFocus === void 0 ? void 0 : tooltipTriggerFocus();
    }, [
        componentOnFocus,
        tooltipTriggerFocus
    ]);
    const onBlur = React.useCallback((event)=>{
        componentOnBlur === null || componentOnBlur === void 0 ? void 0 : componentOnBlur(event);
        tooltipTriggerOnBlur === null || tooltipTriggerOnBlur === void 0 ? void 0 : tooltipTriggerOnBlur();
    }, [
        componentOnBlur,
        tooltipTriggerOnBlur
    ]);
    return React.useMemo(()=>{
        return children({
            onMouseEnter,
            onMouseLeave,
            onMouseDown,
            onFocus,
            onBlur
        });
    }, [
        children,
        onMouseEnter,
        onMouseLeave,
        onMouseDown,
        onFocus,
        onBlur
    ]);
};
