// @canva/app-ui-kit/dist/cjs/ui/apps/developing/ui_kit/components/card/typography/typography.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "TypographyCard", {
    enumerable: true,
    get: function() {
        return TypographyCard;
    }
});
const _jsxruntime = require("react/jsx-runtime");
const _react = _interop_require_wildcard(require("react"));
const _box = require('../../../../../../base/box/box');
const _button = require('../../../../../../base/button/button');
const _icon = require('../../../../../../base/icons/animated_spinner/icon');
const _typographycss = _interop_require_default(require("./typography.css"));
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function _getRequireWildcardCache(nodeInterop) {
    if (typeof WeakMap !== "function") return null;
    var cacheBabelInterop = new WeakMap();
    var cacheNodeInterop = new WeakMap();
    return (_getRequireWildcardCache = function(nodeInterop) {
        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
    })(nodeInterop);
}
function _interop_require_wildcard(obj, nodeInterop) {
    if (!nodeInterop && obj && obj.__esModule) {
        return obj;
    }
    if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
        return {
            default: obj
        };
    }
    var cache = _getRequireWildcardCache(nodeInterop);
    if (cache && cache.has(obj)) {
        return cache.get(obj);
    }
    var newObj = {
        __proto__: null
    };
    var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
    for(var key in obj){
        if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
            if (desc && (desc.get || desc.set)) {
                Object.defineProperty(newObj, key, desc);
            } else {
                newObj[key] = obj[key];
            }
        }
    }
    newObj.default = obj;
    if (cache) {
        cache.set(obj, newObj);
    }
    return newObj;
}
const TypographyCard = ({ ariaLabel, onClick, onDragStart, children, loading })=>{
    const [isDragging, setIsDragging] = _react.useState(false);
    const [isHovering, setIsHovering] = _react.useState(false);
    const boxBackground = isHovering && !loading ? 'neutral' : 'surface';
    return (0, _jsxruntime.jsx)(_button.BasicButton, {
        ariaLabel: ariaLabel,
        onClick: loading ? undefined : onClick,
        draggable: loading !== true && onDragStart !== undefined,
        onDragStart: (event)=>{
            if (onDragStart === undefined || loading === true) return;
            setIsDragging(true);
            onDragStart(event);
        },
        onDragEnd: ()=>setIsDragging(false),
        onMouseEnter: ()=>setIsHovering(true),
        onMouseLeave: ()=>setIsHovering(false),
        noChildWrapper: true,
        children: (0, _jsxruntime.jsxs)(_box.Box, {
            padding: "2u",
            borderRadius: "legacyStandard",
            background: boxBackground,
            border: "standard",
            className: loading ? _typographycss.default.wrapperLoading : _typographycss.default.wrapper,
            children: [
                (0, _jsxruntime.jsx)("span", {
                    style: {
                        opacity: isDragging ? 0 : 1
                    },
                    children: children
                }),
                loading && (0, _jsxruntime.jsx)(_box.Box, {
                    className: _typographycss.default.loading,
                    display: "flex",
                    justifyContent: "center",
                    alignItems: "center",
                    width: "full",
                    height: "full",
                    borderRadius: "legacyStandard",
                    children: (0, _jsxruntime.jsx)(_icon.AnimatedSpinnerIcon, {
                        size: "medium",
                        className: _typographycss.default.spinner
                    })
                })
            ]
        })
    });
};
