// @canva/app-ui-kit/dist/esm/ui/base/card/private/card_context.js
import { jsx as _jsx } from "react/jsx-runtime";
import { makeObservable } from '../../../../base/make_observable/make_observable';
import * as mobx from 'mobx';
import * as React from 'react';
import { pointerInputState } from '../../a11y/pointer_input_state/pointer_input_state';
export const useCardConfiguration = ()=>useCardConfigurationStore().configuration;
export function useCardConfigurationSetter(configuration) {
    const { setConfiguration } = useCardConfigurationStore();
    const willMount = React.useRef(true);
    if (willMount.current) setConfiguration(configuration);
    willMount.current = false;
    React.useEffect(()=>{
        setConfiguration(configuration);
    }, [
        setConfiguration,
        configuration
    ]);
}
export function useCardHover(args) {
    var _args_track;
    const track = (_args_track = args === null || args === void 0 ? void 0 : args.track) !== null && _args_track !== void 0 ? _args_track : true;
    const interactionStore = useCardInteractionStore();
    React.useEffect(()=>{
        if (track) mobx.runInAction(()=>interactionStore.trackHover = true);
    }, [
        interactionStore,
        track
    ]);
    return interactionStore.isHovered;
}
export function useCardFocus() {
    const interactionStore = useCardInteractionStore();
    React.useEffect(()=>{
        mobx.runInAction(()=>interactionStore.trackFocus = true);
    }, [
        interactionStore
    ]);
    return interactionStore.isFocused;
}
export function useCardMouseDown() {
    const interactionStore = useCardInteractionStore();
    React.useEffect(()=>{
        mobx.runInAction(()=>interactionStore.trackMouseDown = true);
    }, [
        interactionStore
    ]);
    return interactionStore.isMouseDown;
}
export const useCardEventHandlers = ()=>useCardInteractionStore().eventHandlers;
export function withCardContextProvider(Component) {
    const WithProvider = (props)=>{
        const [context] = React.useState(createContext);
        return _jsx(CardContext.Provider, {
            value: context,
            children: _jsx(Component, {
                ...props
            })
        });
    };
    var _Component_displayName;
    WithProvider.displayName = `withCardContext(${(_Component_displayName = Component.displayName) !== null && _Component_displayName !== void 0 ? _Component_displayName : Component.name})`;
    return WithProvider;
}
class CardInteractionStore {
    static _makeObservable(instance) {
        makeObservable(instance, {
            isHovered: mobx.observable.ref,
            isFocused: mobx.observable.ref,
            isMouseDown: mobx.observable.ref,
            trackHover: mobx.observable.ref,
            trackFocus: mobx.observable.ref,
            trackMouseDown: mobx.observable.ref,
            eventHandlers: mobx.computed,
            setHovered: mobx.action.bound,
            setFocused: mobx.action.bound,
            setMouseDown: mobx.action.bound
        });
    }
    get eventHandlers() {
        const trackHover = pointerInputState.isHoveringInput && this.trackHover;
        return {
            onMouseEnter: trackHover ? ()=>this.setHovered(true) : undefined,
            onMouseLeave: trackHover ? ()=>this.setHovered(false) : undefined,
            onFocus: this.trackFocus ? ()=>this.setFocused(true) : undefined,
            onBlur: this.trackFocus ? ()=>this.setFocused(false) : undefined,
            onMouseDown: this.trackMouseDown ? ()=>this.setMouseDown(true) : undefined,
            onMouseUp: this.trackMouseDown ? ()=>this.setMouseDown(false) : undefined
        };
    }
    setHovered(isHovered) {
        this.isHovered = isHovered;
    }
    setFocused(isFocused) {
        this.isFocused = isFocused;
    }
    setMouseDown(isMouseDown) {
        this.isMouseDown = isMouseDown;
    }
    constructor(){
        this.isHovered = (CardInteractionStore._makeObservable(this), false);
        this.isFocused = false;
        this.isMouseDown = false;
        this.trackHover = false;
        this.trackFocus = false;
        this.trackMouseDown = false;
    }
}
class CardConfigurationStore {
    static _makeObservable(instance) {
        makeObservable(instance, {
            configuration: mobx.observable.struct,
            setConfiguration: mobx.action.bound
        });
    }
    setConfiguration(configuration) {
        this.configuration = configuration;
    }
    constructor(){
        this.configuration = (CardConfigurationStore._makeObservable(this), {
            disabled: false,
            border: 'none',
            borderRadius: 'element',
            layout: 'vertical',
            selected: false,
            selectableMode: 'none',
            selectableBoundary: 'thumbnail',
            hoverEffect: 'none',
            hoverBoundary: 'thumbnail'
        });
    }
}
function createContext() {
    return {
        interactionStore: new CardInteractionStore(),
        configurationStore: new CardConfigurationStore()
    };
}
const CardContext = React.createContext(undefined);
const useCardContext = ()=>{
    var _React_useContext;
    return (_React_useContext = React.useContext(CardContext)) !== null && _React_useContext !== void 0 ? _React_useContext : createContext();
};
const useCardConfigurationStore = ()=>useCardContext().configurationStore;
const useCardInteractionStore = ()=>useCardContext().interactionStore;
