// @canva/app-ui-kit/dist/esm/ui/base/card/private/card_thumbnail.js
import { jsx as _jsx } from "react/jsx-runtime";
import classNames from 'classnames';
import { action, autorun } from 'mobx';
import { observer, useLocalObservable } from 'mobx-react-lite';
import * as React from 'react';
import { usePrefersReducedMotion } from '../../animation/supports_animation';
import { useAutoplayVideos } from '../../provider/provider';
import { IconThumbnail, ImageThumbnail, ImageThumbnailPlaceholder, StaticImage, Thumbnail, ThumbnailContainer, VideoThumbnail } from '../../thumbnail/thumbnail';
import { SelectableWrapper } from './card';
import styles from './card.css';
import { useCardConfiguration, useCardFocus, useCardHover } from './card_context';
export const CardThumbnailContainer = observer(({ border, borderRadius: borderRadiusProp, background: backgroundProp = 'secondary', aspectRatio, padding, children, className, width, height, fit })=>{
    const shouldBgColorChangeOnHover = useShouldBgColorChangeOnHover();
    const background = useCardThumbnailBackground(backgroundProp, !!shouldBgColorChangeOnHover);
    const borderRadius = useBorderRadius(borderRadiusProp);
    return _jsx(SelectableWrapper, {
        boundary: "thumbnail",
        borderRadius: borderRadius,
        children: _jsx(ThumbnailContainer, {
            border: border,
            borderRadius: borderRadius,
            padding: padding,
            background: background,
            width: width,
            height: height,
            aspectRatio: aspectRatio,
            fit: fit,
            className: classNames(className, {
                [styles.bgColorTransition]: shouldBgColorChangeOnHover
            }),
            children: children
        })
    });
});
export const CardThumbnail = observer(({ border = 'low', borderRadius: borderRadiusProp, children, ...props })=>{
    const borderRadius = useBorderRadius(borderRadiusProp);
    const thumbnail = _jsx(Thumbnail, {
        ...props,
        border: border,
        borderRadius: borderRadius,
        children: children
    });
    return _jsx(SelectableWrapper, {
        boundary: "thumbnail",
        borderRadius: borderRadius,
        children: thumbnail
    });
});
export const CardPlaceholder = observer(({ border = 'low', borderRadius: borderRadiusProp, ...cardPlaceholderProps })=>{
    const borderRadius = useBorderRadius(borderRadiusProp);
    const placeholderThumbnail = _jsx(ImageThumbnailPlaceholder, {
        ...cardPlaceholderProps,
        border: border,
        borderRadius: borderRadius
    });
    return _jsx(SelectableWrapper, {
        boundary: "thumbnail",
        borderRadius: borderRadius,
        children: placeholderThumbnail
    });
});
export const CardIconThumbnail = observer(({ border = 'low', borderRadius: borderRadiusProp, ...iconThumbnailProps })=>{
    const borderRadius = useBorderRadius(borderRadiusProp);
    const iconThumbnail = _jsx(IconThumbnail, {
        ...iconThumbnailProps,
        border: border,
        borderRadius: borderRadius
    });
    return _jsx(SelectableWrapper, {
        boundary: "thumbnail",
        borderRadius: borderRadius,
        children: iconThumbnail
    });
});
export const CardImageThumbnail = observer(({ renderImage = StaticImage, border = 'low', borderRadius: borderRadiusProp, ...props })=>{
    const borderRadius = useBorderRadius(borderRadiusProp);
    const imageThumbnail = _jsx(ImageThumbnail, {
        ...props,
        border: border,
        borderRadius: borderRadius,
        src: props.src,
        renderImage: ({ ...imageProps })=>renderImage({
                ...imageProps,
                className: classNames(props.className, styles.image),
                crossOrigin: props.crossOrigin,
                src: props.src,
                alt: props.alt || '',
                elementTiming: props.elementTiming
            })
    });
    return _jsx(SelectableWrapper, {
        boundary: "thumbnail",
        borderRadius: borderRadius,
        children: imageThumbnail
    });
});
export const CardVideoThumbnail = observer(React.forwardRef((props, ref)=>{
    const { border = 'low', borderRadius: borderRadiusProp, className, videoSrc, imageSrc, imageLoadingState, imageFallback, imagePlaceholder, imageAlt, videoLoadingState, VideoComponent, renderImage, loadVideo, onVideoLoad, onImageLoad, autoplay } = props;
    const isFocused = useCardFocus();
    const borderRadius = useBorderRadius(borderRadiusProp);
    const autoplayVideos = useAutoplayVideos() || 'ADAPTIVE';
    const prefersReducedMotion = usePrefersReducedMotion();
    const enableAutoplay = !prefersReducedMotion && autoplayVideos === 'ADAPTIVE' || autoplayVideos === 'AUTOPLAY';
    const whenToPlay = imageSrc == null ? 'always' : props.whenToPlay === 'always' ? enableAutoplay ? 'always' : 'on-hover' : props.whenToPlay;
    const isHovered = useCardHover({
        track: whenToPlay === 'on-hover'
    });
    const store = useLocalObservable(()=>({
            playState: 'pause',
            setPlayState: action((value)=>{
                store.playState = value;
            })
        }));
    const { setPlayState } = store;
    React.useEffect(()=>autorun(()=>{
            if ((whenToPlay === 'always' || whenToPlay === 'now') && store.playState !== 'play') setPlayState('play');
            if (whenToPlay === 'never' && store.playState === 'play') setPlayState('pause');
        }), [
        whenToPlay,
        store.playState,
        setPlayState
    ]);
    React.useEffect(()=>autorun(()=>{
            if (whenToPlay === 'on-hover') {
                if ((isHovered || isFocused) && store.playState !== 'play') setPlayState('play');
                else if (!isHovered && !isFocused && store.playState === 'play') setPlayState('pause');
            }
        }), [
        whenToPlay,
        isHovered,
        isFocused,
        store.playState,
        setPlayState
    ]);
    const videoThumbnail = _jsx(VideoThumbnail, {
        ...props,
        ref: ref,
        border: border,
        borderRadius: borderRadius,
        videoSrc: videoSrc,
        imageSrc: imageSrc,
        playState: store.playState,
        autoplay: autoplay,
        delayVideoLoad: whenToPlay === 'on-hover' ? 150 : 0,
        VideoComponent: VideoComponent,
        videoClassName: classNames(styles.video, className),
        videoLoadingState: videoLoadingState,
        loadVideo: loadVideo,
        onVideoLoad: onVideoLoad,
        imageAlt: imageAlt,
        renderImage: renderImage,
        imageClassName: classNames(className, styles.image),
        imagePlaceholder: imagePlaceholder,
        imageLoadingState: imageLoadingState,
        imageFallback: imageFallback,
        onImageLoad: onImageLoad
    });
    return _jsx(SelectableWrapper, {
        boundary: "thumbnail",
        borderRadius: borderRadius,
        children: videoThumbnail
    });
}));
const backgroundMapping = {
    ['contrastOnLight']: {
        hover: 'contrastOnLightHover',
        default: 'contrastOnLight'
    },
    ['contrastOnDark']: {
        hover: 'contrastOnDarkHover',
        default: 'contrastOnDark'
    },
    ['secondary']: {
        hover: 'secondaryHover',
        default: 'secondary'
    },
    ['none']: {
        hover: 'secondaryHover',
        default: 'none'
    }
};
function getBackground(background, isHover) {
    return backgroundMapping[background][isHover ? 'hover' : 'default'];
}
const useShouldBgColorChangeOnHover = ()=>{
    const { hoverEffect, hoverBoundary } = useCardConfiguration();
    return hoverBoundary === 'thumbnail' && hoverEffect === 'bgColorTransition';
};
function useBorderRadius(borderRadius) {
    const { borderRadius: cardBorderRadius } = useCardConfiguration();
    return borderRadius !== null && borderRadius !== void 0 ? borderRadius : cardBorderRadius;
}
function useCardThumbnailBackground(background, shouldBgColorChangeOnHover) {
    const isHovered = useCardHover({
        track: shouldBgColorChangeOnHover
    });
    return shouldBgColorChangeOnHover ? getBackground(background, isHovered) : background;
}
