// @canva/app-ui-kit/dist/esm/ui/base/card/private/card_content.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import classNames from 'classnames';
import 'react';
import { Rows } from '../../layout/layout';
import { Text } from '../../typography/typography';
import styles from './card.css';
export const CardContent = function CardContent({ titleId, descriptionId, title, description, preTitle, content }) {
    const hasStandardContent = !!(title || description || preTitle);
    const customContent = typeof content === 'function' ? content({
        titleId,
        descriptionId
    }) : content;
    if (hasStandardContent) return _jsxs(Rows, {
        spacing: "0",
        children: [
            typeof preTitle !== 'string' ? preTitle : _jsx(CardPreTitle, {
                children: preTitle
            }),
            title != null && typeof title !== 'string' ? _jsx("div", {
                className: styles.title,
                id: titleId,
                children: title
            }) : _jsx(CardTitle, {
                id: titleId,
                lineClamp: description ? 1 : 2,
                children: title
            }),
            description != null && typeof description !== 'string' ? _jsx("div", {
                className: styles.description,
                id: descriptionId,
                children: description
            }) : _jsx(CardDescription, {
                id: descriptionId,
                children: description
            }),
            customContent
        ]
    });
    return customContent;
};
export const CardTitle = function CardTitle({ className, weight = 'bold', lineClamp = 2, ...rest }) {
    return _jsx(Text, {
        className: classNames(styles.title, className),
        weight: weight,
        lineClamp: lineClamp,
        ...rest
    });
};
export const CardDescription = function CardDescription({ size = 'small', tone = 'secondary', lineClamp = 1, ...rest }) {
    return _jsx(Text, {
        size: size,
        tone: tone,
        lineClamp: lineClamp,
        className: styles.description,
        ...rest
    });
};
export const CardPreTitle = function CardPreTitle({ size = 'xsmall', tone = 'secondary', capitalization = 'uppercase', lineClamp = 1, ...rest }) {
    return _jsx(Text, {
        size: size,
        tone: tone,
        capitalization: capitalization,
        lineClamp: lineClamp,
        ...rest
    });
};
