// @canva/app-ui-kit/dist/esm/ui/base/card/private/card.js
import { jsx as _jsx, jsxs as _jsxs, Fragment as _Fragment } from "react/jsx-runtime";
import classNames from 'classnames';
import { observer } from 'mobx-react-lite';
import * as React from 'react';
import { accessModeState } from '../../a11y/access_mode_state/access_mode_state';
import { BadgeContainer } from '../../badge/badge';
import { ExtendedBadgeVisibilityTrigger } from '../../badge/temporary_exports';
import { Spacer } from '../../layout/layout';
import { useEnableAnimations } from '../../provider/provider';
import { useInThumbnailContainerContext } from '../../thumbnail/thumbnail';
import styles from './card.css';
import { CardContent } from './card_content';
import { useCardConfiguration, useCardConfigurationSetter, useCardEventHandlers, useCardFocus, useCardHover, useCardMouseDown, withCardContextProvider } from './card_context';
import { CardImageThumbnail } from './card_thumbnail';
export const Card = withCardContextProvider(observer(function Card({ title, description, preTitle, content, ariaControls, ariaHasPopup, ariaExpanded, ariaLabel, ariaLabelledBy, buttonAriaLabel, buttonAriaLabelledBy, href, target, draggable, buttonId, buttonRef, onClick, onDragStart, onDragEnd, onContextMenu, tabIndex, disabled = false, border = 'none', borderRadius = 'element', bleedX, padding = '0', background = 'none', layout = 'vertical', thumbnail: thumbnailProp, stretchThumbnail, decorators, decoratorsInset = '1u', decoratorsBoundary: decoratorsBoundaryProp, hoverEffect = 'none', hoverBoundary: hoverBoundaryProp, selectableMode = 'none', selected = false, selectableBoundary: selectableBoundaryProp, onPointerDown, onPointerUp, onPointerEnter, onPointerLeave, onPointerMove, onPointerCancel }) {
    const { decoratorsBoundary, hoverBoundary, selectableBoundary } = getBoundaries({
        decoratorsBoundary: decoratorsBoundaryProp,
        hoverBoundary: hoverBoundaryProp,
        selectableBoundary: selectableBoundaryProp
    });
    useCardConfigurationSetter({
        disabled,
        layout,
        border,
        borderRadius,
        hoverEffect,
        hoverBoundary,
        selected,
        selectableMode,
        selectableBoundary
    });
    const enableAnimations = useEnableAnimations();
    const { onMouseEnter, onMouseLeave, onMouseUp, onMouseDown } = useCardEventHandlers();
    const titleId = React.useId();
    const descriptionId = React.useId();
    const cardAriaProps = getCardAriaLabel({
        ariaLabelledBy,
        ariaLabel,
        titleId
    });
    const buttonAriaProps = getButtonAriaLabel({
        buttonAriaLabelledBy,
        buttonAriaLabel,
        ariaLabelledBy,
        ariaLabel,
        titleId
    });
    const gutterSize = {
        ['horizontal']: '1.5u',
        ['vertical']: '1u'
    };
    const hasContent = !!(title || description || preTitle || content);
    const isThumbnailOnlyCard = decoratorsBoundary === 'thumbnail' && !hasContent && padding === '0';
    const cardButton = _jsx(InternalCardButton, {
        href: href,
        target: target,
        onClick: onClick,
        onPointerDown: onPointerDown,
        onPointerUp: onPointerUp,
        onPointerEnter: onPointerEnter,
        onPointerLeave: onPointerLeave,
        onPointerMove: onPointerMove,
        onPointerCancel: onPointerCancel,
        buttonAriaLabel: buttonAriaProps.buttonAriaLabel,
        buttonAriaLabelledBy: buttonAriaProps.buttonAriaLabelledBy,
        ariaPressed: selectableMode === 'primary' ? selected || false : undefined,
        ariaControls: ariaControls,
        ariaHasPopup: ariaHasPopup,
        ariaExpanded: ariaExpanded,
        buttonId: buttonId,
        buttonRef: buttonRef,
        disabled: disabled,
        draggable: draggable,
        onDragStart: onDragStart,
        onDragEnd: onDragEnd,
        onContextMenu: onContextMenu,
        tabIndex: tabIndex
    });
    const cardContainerClassName = classNames(styles.cardContainer, {
        [styles.disabled]: disabled,
        [styles.animated]: enableAnimations,
        [styles.cardHoverTrigger]: selectableMode === 'primary',
        [styles.scaleOnHover]: hoverEffect === 'scaleThumbnail',
        [styles.bgTransitionOnHover]: hoverEffect === 'bgColorTransition' && hoverBoundary === 'card'
    });
    const cardContainerProps = {
        ariaLabel: cardAriaProps.ariaLabel,
        ariaLabelledBy: cardAriaProps.ariaLabelledBy,
        onMouseEnter,
        onMouseLeave,
        onMouseUp,
        onMouseDown,
        role: 'group'
    };
    const badgeContainerProps = {
        paddingStyle: decoratorsInset === '1u' ? 'standard' : 'compact'
    };
    const thumbnail = typeof thumbnailProp === 'string' ? _jsx(CardImageThumbnail, {
        src: thumbnailProp
    }) : thumbnailProp;
    const decoratedThumbnail = decoratorsBoundary === 'thumbnail' ? _jsxs(BadgeContainer, {
        ...(isThumbnailOnlyCard ? cardContainerProps : {}),
        ...badgeContainerProps,
        className: classNames(styles.thumbnailArea, isThumbnailOnlyCard && cardContainerClassName, {
            [styles.stretchThumbnail]: stretchThumbnail
        }),
        children: [
            isThumbnailOnlyCard && cardButton,
            thumbnail,
            decorators
        ]
    }) : layout === 'horizontal' ? _jsx("div", {
        className: styles.thumbnailArea,
        children: thumbnail
    }) : thumbnail;
    if (isThumbnailOnlyCard) return decoratedThumbnail;
    const CardWithContentContainer = decoratorsBoundary === 'card' ? BadgeContainer : ExtendedBadgeVisibilityTrigger;
    return (_jsx(SelectableWrapper, {
                boundary: "card",
                borderRadius: borderRadius,
                className: classNames({
                    [styles.horizontalWrapper]: layout === 'horizontal'
                }),
                children: _jsxs(CardWithContentContainer, {
                    ...cardContainerProps,
                    ...(decoratorsBoundary === 'card' ? badgeContainerProps : {}),
                    ariaDescribedBy: descriptionId,
                    className: classNames(cardContainerClassName, {
                        [styles.backgroundNeutralLow]: background === 'neutralLow',
                        [styles.backgroundNone]: background === 'none',
                        [styles.borderLow]: border === 'low',
                        [styles.borderRadiusElement]: borderRadius === 'element',
                        [styles.borderRadiusElementSmall]: borderRadius === 'elementSmall',
                        [styles.borderRadiusElementRelaxed]: borderRadius === 'elementRelaxed',
                        [styles.paddingOneUnit]: padding === '1u',
                        [styles.paddingTwoUnit]: padding === '2u',
                        [styles.bleedXOneUnit]: bleedX && padding === '1u',
                        [styles.bleedXTwoUnit]: bleedX && padding === '2u',
                        [styles.vertical]: layout === 'vertical',
                        [styles.horizontal]: layout === 'horizontal',
                        [styles.selected]: selected
                    }),
                    children: [
                        cardButton,
                        thumbnail != null && _jsxs(_Fragment, {
                            children: [
                                decoratedThumbnail,
                                _jsx(Spacer, {
                                    size: gutterSize[layout],
                                    direction: layout
                                })
                            ]
                        }),
                        _jsx(CardContent, {
                            title: title,
                            description: description,
                            preTitle: preTitle,
                            titleId: titleId,
                            descriptionId: descriptionId,
                            content: content
                        }),
                        decoratorsBoundary === 'card' && decorators
                    ]
                })
            }));
}));
export const WithCardState = observer(function WithCardState({ children }) {
    const state = {
        isHovered: useCardHover(),
        isFocused: useCardFocus(),
        isMouseDown: useCardMouseDown()
    };
    return children(state);
});
function getBoundaries({ decoratorsBoundary, selectableBoundary, hoverBoundary }) {
    var _ref, _ref1;
    const fallbackBoundary = (_ref1 = (_ref = decoratorsBoundary !== null && decoratorsBoundary !== void 0 ? decoratorsBoundary : selectableBoundary) !== null && _ref !== void 0 ? _ref : hoverBoundary) !== null && _ref1 !== void 0 ? _ref1 : 'thumbnail';
    return {
        decoratorsBoundary: decoratorsBoundary !== null && decoratorsBoundary !== void 0 ? decoratorsBoundary : fallbackBoundary,
        selectableBoundary: selectableBoundary !== null && selectableBoundary !== void 0 ? selectableBoundary : fallbackBoundary,
        hoverBoundary: hoverBoundary !== null && hoverBoundary !== void 0 ? hoverBoundary : fallbackBoundary
    };
}
const getCardAriaLabel = ({ ariaLabelledBy, ariaLabel, titleId })=>{
    if (ariaLabelledBy != null) return {
        ariaLabelledBy
    };
    if (ariaLabel != null) return {
        ariaLabel
    };
    if (titleId != null) return {
        ariaLabelledBy: titleId
    };
    return {};
};
const getButtonAriaLabel = ({ ariaLabelledBy, ariaLabel, titleId, buttonAriaLabelledBy, buttonAriaLabel })=>{
    if (buttonAriaLabelledBy != null) return {
        buttonAriaLabelledBy
    };
    if (buttonAriaLabel != null) return {
        buttonAriaLabel
    };
    if (ariaLabelledBy != null) return {
        buttonAriaLabelledBy: ariaLabelledBy
    };
    if (ariaLabel != null) return {
        buttonAriaLabel: ariaLabel
    };
    if (titleId != null) return {
        buttonAriaLabelledBy: titleId
    };
    return {};
};
const InternalCardButton = observer(function InternalCardButton({ ariaControls, ariaHasPopup, ariaExpanded, ariaPressed, buttonAriaLabelledBy, href, target, buttonAriaLabel, draggable, disabled, onDragStart, onDragEnd, buttonId, buttonRef, onContextMenu, tabIndex, onClick, onPointerDown, onPointerUp, onPointerEnter, onPointerLeave, onPointerMove, onPointerCancel }) {
    const { onFocus, onBlur } = useCardEventHandlers();
    const preventDefault = React.useCallback((e)=>e.preventDefault(), []);
    const doNothing = React.useCallback(()=>{}, []);
    if (!(href || onClick)) return null;
    if (href) return (_jsx("a", {
            id: buttonId,
            href: href,
            target: target,
            className: classNames(styles.cardButton, {
                [styles.hideOutline]: accessModeState.isMouseMode
            }),
            onClick: disabled ? preventDefault : onClick,
            onPointerDown: disabled ? preventDefault : onPointerDown,
            onPointerUp: disabled ? preventDefault : onPointerUp,
            onPointerEnter: onPointerEnter,
            onPointerLeave: onPointerLeave,
            onPointerMove: onPointerMove,
            onPointerCancel: onPointerCancel,
            "aria-controls": ariaControls,
            "aria-haspopup": ariaHasPopup,
            "aria-labelledby": buttonAriaLabelledBy,
            "aria-label": buttonAriaLabelledBy ? undefined : buttonAriaLabel,
            "aria-disabled": disabled || undefined,
            "aria-expanded": ariaHasPopup && ariaExpanded,
            draggable: draggable,
            onFocus: onFocus,
            onBlur: onBlur,
            onDragStart: onDragStart,
            onDragEnd: onDragEnd,
            onContextMenu: onContextMenu,
            ref: buttonRef,
            tabIndex: tabIndex
        }));
    const onKeyUp = (e)=>{
        if (!disabled && e.key === ' ') e.currentTarget.click();
    };
    const onKeyDown = (e)=>{
        if (!disabled) {
            if (e.key === ' ') e.preventDefault();
            else if (e.key === 'Enter') {
                e.preventDefault();
                e.currentTarget.click();
            }
        }
    };
    return _jsx("div", {
        role: "button",
        "aria-disabled": disabled || undefined,
        "aria-controls": ariaControls,
        "aria-haspopup": ariaHasPopup,
        "aria-expanded": ariaExpanded,
        "aria-pressed": ariaPressed,
        "aria-label": buttonAriaLabelledBy ? undefined : buttonAriaLabel,
        "aria-labelledby": buttonAriaLabelledBy,
        className: classNames(styles.divButton, styles.cardButton, {
            [styles.hideOutline]: accessModeState.isMouseMode
        }),
        draggable: draggable,
        id: buttonId,
        onClick: disabled ? doNothing : onClick,
        onPointerDown: disabled ? doNothing : onPointerDown,
        onPointerUp: disabled ? doNothing : onPointerUp,
        onPointerEnter: onPointerEnter,
        onPointerLeave: onPointerLeave,
        onPointerMove: onPointerMove,
        onPointerCancel: onPointerCancel,
        onContextMenu: onContextMenu,
        onFocus: onFocus,
        onBlur: onBlur,
        onKeyUp: onKeyUp,
        onKeyDown: onKeyDown,
        onDragStart: onDragStart,
        onDragEnd: onDragEnd,
        ref: buttonRef,
        tabIndex: tabIndex !== null && tabIndex !== void 0 ? tabIndex : 0
    });
});
export const SelectableWrapper = observer(({ children, boundary, borderRadius, className })=>{
    const { selected, selectableMode, selectableBoundary } = useCardConfiguration();
    const enableAnimations = useEnableAnimations();
    const isInThumbnailContainerContext = useInThumbnailContainerContext();
    if (!isInThumbnailContainerContext && selectableBoundary === boundary && selectableMode !== 'none') return _jsx("div", {
        className: classNames(styles.selectableWrapper, className, {
            [styles.selected]: selected,
            [styles.selectableOnHover]: selectableMode === 'primary',
            [styles.animated]: enableAnimations,
            [styles.borderRadiusElement]: borderRadius === 'element',
            [styles.borderRadiusElementSmall]: borderRadius === 'elementSmall',
            [styles.borderRadiusElementRelaxed]: borderRadius === 'elementRelaxed'
        }),
        children: children
    });
    return children;
});
