// @canva/app-ui-kit/dist/cjs/ui/base/layout/layout.js
"use strict"
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: Object.getOwnPropertyDescriptor(all, name).get
    });
}
_export(exports, {
    get Bleed () {
        return _bleed.Bleed;
    },
    get Column () {
        return _columns.Column;
    },
    get Columns () {
        return _columns.Columns;
    },
    get Container () {
        return _container.Container;
    },
    get Grid () {
        return _grid.Grid;
    },
    get Hide () {
        return _hide.Hide;
    },
    get Inline () {
        return _inline.Inline;
    },
    get Rows () {
        return _rows.Rows;
    },
    get Spacer () {
        return _spacer.Spacer;
    }
});
const _bleed = require("./private/bleed/bleed");
const _columns = require("./private/columns/columns");
const _container = require("./private/container/container");
const _grid = require("./private/grid/grid");
const _hide = require("./private/hide/hide");
const _inline = require("./private/inline/inline");
const _rows = require("./private/rows/rows");
const _spacer = require("./private/spacer/spacer");
