// @canva/app-ui-kit/dist/cjs/ui/base/swatch/private/internal_swatch.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: Object.getOwnPropertyDescriptor(all, name).get
    });
}
_export(exports, {
    get ClickableSwatch () {
        return ClickableSwatch;
    },
    get DeletableSwatch () {
        return DeletableSwatch;
    },
    get StaticSwatch () {
        return StaticSwatch;
    },
    get getBorderRadiusStyles () {
        return getBorderRadiusStyles;
    }
});
const _jsxruntime = require("react/jsx-runtime");
const _preconditions = require('../../../../base/preconditions');
const _classnames = _interop_require_default(require("classnames"));
const _mobxreactlite = require("mobx-react-lite");
const _react = _interop_require_wildcard(require("react"));
const _pointer_input_state = require('../../a11y/pointer_input_state/pointer_input_state');
const _badge = require('../../badge/badge');
const _base_button = require('../../button/base_button/base_button');
const _button = require('../../button/button');
const _handle = require('../../handle/handle');
const _icon = require('../../icons/x/icon');
const _swatchcss = _interop_require_default(require("./swatch.css"));
const _swatchmessages = require("./swatch.messages");
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function _getRequireWildcardCache(nodeInterop) {
    if (typeof WeakMap !== "function") return null;
    var cacheBabelInterop = new WeakMap();
    var cacheNodeInterop = new WeakMap();
    return (_getRequireWildcardCache = function(nodeInterop) {
        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
    })(nodeInterop);
}
function _interop_require_wildcard(obj, nodeInterop) {
    if (!nodeInterop && obj && obj.__esModule) {
        return obj;
    }
    if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
        return {
            default: obj
        };
    }
    var cache = _getRequireWildcardCache(nodeInterop);
    if (cache && cache.has(obj)) {
        return cache.get(obj);
    }
    var newObj = {
        __proto__: null
    };
    var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
    for(var key in obj){
        if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
            if (desc && (desc.get || desc.set)) {
                Object.defineProperty(newObj, key, desc);
            } else {
                newObj[key] = obj[key];
            }
        }
    }
    newObj.default = obj;
    if (cache) {
        cache.set(obj, newObj);
    }
    return newObj;
}
function getBorderRadiusStyles(shape, size) {
    if (shape === 'circle') return undefined;
    switch(size){
        case 'xxsmall':
        case 'xsmall':
            return _swatchcss.default.borderRadiusElementSmall;
        case 'small':
        case 'medium':
        case 'large':
            return _swatchcss.default.borderRadiusElement;
        default:
            throw new _preconditions.UnreachableError(size);
    }
}
const StaticSwatch = _react.forwardRef(
    ({ className, children, ariaLabel, ariaDescribedBy, ariaLabelledBy, id, isRoot }, ref)=>{
        const spanRef = (0, _handle.useHandleRef)(ref, _handle.createFocusHandle);
        return (0, _jsxruntime.jsx)("span", {
            ref: spanRef,
            "aria-label": ariaLabelledBy ? undefined : ariaLabel,
            "aria-labelledby": ariaLabelledBy,
            "aria-describedby": ariaDescribedBy,
            id: id,
            className: (0, _classnames.default)(className, isRoot && _swatchcss.default.swatchRoot),
            role: "img",
            children: children
        });
    }
);
const ClickableSwatch = (0, _mobxreactlite.observer)(_react.forwardRef(
    function _ClickableSwatch({ size = 'medium', shape = 'circle', onClick, active, selected, pressed, disabled, ariaLabel, disclosure, role, tabIndex, stretch, aspectRatio, children, ariaHasPopup, ariaControls, tooltipLabel, tooltipDescription, disableTooltip, isRoot }, ref) {
        const buttonRef = (0, _handle.useHandleRef)(ref, _handle.createFocusHandle);
        const borderRadiusStyles = getBorderRadiusStyles(shape, size);
        const hoverSupported = (0, _pointer_input_state.useIsHoveringInput)();
        return (0, _jsxruntime.jsx)(_base_button.BaseButton, {
            onClick: onClick,
            active: active,
            selected: selected,
            pressed: pressed,
            disclosure: disclosure,
            disabled: disabled,
            role: role,
            ref: buttonRef,
            tabIndex: tabIndex,
            ariaLabel: ariaLabel,
            ariaHasPopup: ariaHasPopup,
            ariaControls: ariaControls,
            tooltipLabel: tooltipLabel,
            tooltipDescription: tooltipDescription,
            tooltipDisabled: disableTooltip,
            position: "relative",
            width: stretch ? 'full' : undefined,
            height: stretch && aspectRatio ? 'full' : undefined,
            className: (0, _classnames.default)(_swatchcss.default.swatchButton, borderRadiusStyles, {
                [_swatchcss.default.swatchRoot]: isRoot,
                [_swatchcss.default.active]: selected || pressed || active,
                [_swatchcss.default.disabled]: disabled,
                [_swatchcss.default.fixed]: !stretch,
                [_swatchcss.default.hoverSupported]: hoverSupported
            }),
            children: children
        });
    }
));
const DeletableSwatch = (0, _mobxreactlite.observer)(_react.forwardRef(
    function deletableSwatch({ onDelete, deleteButtonVisibility, stretch, aspectRatio, disableTooltip, children, isRoot }, ref) {
        const buttonRef = (0, _handle.useHandleRef)(ref, _handle.createFocusHandle);
        return (0, _jsxruntime.jsxs)(_badge.BadgeContainer, {
            className: (0, _classnames.default)(_swatchcss.default.swatchWrapper, {
                [_swatchcss.default.swatchRoot]: isRoot,
                [_swatchcss.default.stretch]: stretch,
                [_swatchcss.default.fixed]: !stretch,
                [_swatchcss.default.aspectRatio]: aspectRatio
            }),
            children: [
                children,
                (0, _jsxruntime.jsx)(_badge.BadgeGroup, {
                    location: "top-end",
                    visibility: !_pointer_input_state.pointerInputState.isMouseInput ? 'always' : deleteButtonVisibility,
                    children: (0, _jsxruntime.jsx)(_button.CircleButton, {
                        ref: buttonRef,
                        icon: _icon.XIcon,
                        variant: "contrast",
                        onClick: onDelete,
                        tooltipLabel: _swatchmessages.SwatchMessages.deleteTooltip(),
                        tooltipDisabled: disableTooltip,
                        size: "tiny",
                        iconSize: "tiny",
                        className: _swatchcss.default.deleteButton
                    })
                })
            ]
        });
    }
));
