// @canva/app-ui-kit/dist/esm/ui/base/button/repeat_click_on_hold/repeat_click_on_hold.js
import * as React from 'react';
export const SLOW_REPEAT_DELAY = 450;
export const FAST_REPEAT_DELAY = 350;
export const SLOW_REPEAT_INTERVAL = 100;
export const FAST_REPEAT_INTERVAL = 40;
export function WithRepeatClickOnHold(props) {
    return props.children(useRepeatClickOnHold(props));
}
export function useRepeatClickOnHold({ speed = 'none', disabled, onClick }) {
    const repeatDelayTimeoutIdRef = React.useRef(undefined);
    const repeatIntervalIdRef = React.useRef(undefined);
    const isPointerDownRef = React.useRef(false);
    const onClickRef = React.useRef(onClick);
    React.useLayoutEffect(()=>{
        onClickRef.current = onClick;
    });
    const clearTimers = React.useCallback(()=>{
        if (repeatDelayTimeoutIdRef.current) {
            window.clearTimeout(repeatDelayTimeoutIdRef.current);
            repeatDelayTimeoutIdRef.current = undefined;
        }
        if (repeatIntervalIdRef.current) {
            window.clearInterval(repeatIntervalIdRef.current);
            repeatIntervalIdRef.current = undefined;
        }
    }, []);
    React.useEffect(()=>{
        return ()=>{
            clearTimers();
        };
    }, [
        clearTimers
    ]);
    const maybeClick = React.useCallback((e)=>{
        var _onClickRef_current;
        if (!disabled) (_onClickRef_current = onClickRef.current) === null || _onClickRef_current === void 0 ? void 0 : _onClickRef_current.call(onClickRef, e);
    }, [
        disabled
    ]);
    const onPointerDown = React.useCallback((e)=>{
        isPointerDownRef.current = true;
        const repeatInterval = speed === 'fast' ? FAST_REPEAT_INTERVAL : SLOW_REPEAT_INTERVAL;
        const repeatDelay = speed === 'fast' ? FAST_REPEAT_DELAY : SLOW_REPEAT_DELAY;
        clearTimers();
        repeatDelayTimeoutIdRef.current = window.setTimeout(()=>{
            maybeClick(e);
            repeatIntervalIdRef.current = window.setInterval(()=>{
                maybeClick(e);
            }, repeatInterval);
        }, repeatDelay);
    }, [
        speed,
        maybeClick,
        clearTimers
    ]);
    const onPointerUp = React.useCallback((e)=>{
        if (isPointerDownRef.current) {
            isPointerDownRef.current = false;
            if (!repeatIntervalIdRef.current) maybeClick(e);
            clearTimers();
        }
        e.preventDefault();
    }, [
        clearTimers,
        maybeClick
    ]);
    const onPointerLeave = React.useCallback((e)=>{
        isPointerDownRef.current = false;
        clearTimers();
        e.preventDefault();
    }, [
        clearTimers
    ]);
    const onContextMenu = React.useCallback((e)=>{
        e.preventDefault();
        return false;
    }, []);
    return speed === 'none' ? {
        onClick
    } : {
        onPointerDown,
        onPointerUp,
        onPointerLeave,
        onContextMenu,
        onClick: undefined
    };
}
